<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>ideal_genom_qc.AncestryQC &mdash; IDEAL-GENOM-QC v0.1.0</title>
      <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../../_static/css/theme.css?v=19f00094" />

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../_static/jquery.js?v=5d32c60e"></script>
        <script src="../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../../_static/documentation_options.js?v=01f34227"></script>
        <script src="../../_static/doctools.js?v=9a2dae69"></script>
        <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #2980B9" >

          
          
          <a href="../../index.html" class="icon icon-home">
            IDEAL-GENOM-QC
          </a>
              <div class="version">
                0.1.0
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">User Guide</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../getting_started.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../configuration.html">Configuration Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../examples.html">Examples</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../SampleQC.html">SampleQC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../AncestryQC.html">AncestryQC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../PopStructure.html">Population Structure</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../VariantQC.html">VariantQC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../Helpers.html">Helpers</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../get_references.html">Get References</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Additional Resources</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../faq.html">Frequently Asked Questions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../troubleshooting.html">Troubleshooting Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../contributing.html">Contributing Guide</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #2980B9" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">IDEAL-GENOM-QC</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../index.html">Module code</a></li>
      <li class="breadcrumb-item active">ideal_genom_qc.AncestryQC</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for ideal_genom_qc.AncestryQC</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">psutil</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">logging</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">seaborn</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sns</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">typing</span><span class="w"> </span><span class="kn">import</span> <span class="n">Union</span><span class="p">,</span> <span class="n">Literal</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">ideal_genom_qc.Helpers</span><span class="w"> </span><span class="kn">import</span> <span class="n">shell_do</span><span class="p">,</span> <span class="n">delete_temp_files</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">ideal_genom_qc.get_references</span><span class="w"> </span><span class="kn">import</span> <span class="n">Fetcher1000Genome</span><span class="p">,</span> <span class="n">FetcherLDRegions</span>

<span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">%(asctime)s</span><span class="s2"> - </span><span class="si">%(levelname)s</span><span class="s2"> - </span><span class="si">%(message)s</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span>

<div class="viewcode-block" id="ReferenceGenomicMerger">
<a class="viewcode-back" href="../../AncestryQC.html#ideal_genom_qc.AncestryQC.ReferenceGenomicMerger">[docs]</a>
<span class="k">class</span><span class="w"> </span><span class="nc">ReferenceGenomicMerger</span><span class="p">():</span>

<div class="viewcode-block" id="ReferenceGenomicMerger.__init__">
<a class="viewcode-back" href="../../AncestryQC.html#ideal_genom_qc.AncestryQC.ReferenceGenomicMerger.__init__">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">input_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">input_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">output_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">output_name</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span> <span class="n">high_ld_regions</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">reference_files</span><span class="p">:</span> <span class="nb">dict</span><span class="p">,</span> <span class="n">built</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;38&#39;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initialize AncestryQC class.</span>
<span class="sd">        This class performs ancestry quality control on genetic data by comparing study samples against reference populations.</span>
<span class="sd">        </span>
<span class="sd">        Parameters:</span>
<span class="sd">        -----------</span>
<span class="sd">        input_path: Path </span>
<span class="sd">            Path to directory containing input files</span>
<span class="sd">        input_name: str </span>
<span class="sd">            Name of input file without extension</span>
<span class="sd">        output_path: Path </span>
<span class="sd">            Path to directory for output files</span>
<span class="sd">        output_name: str </span>
<span class="sd">            Name for output files without extension</span>
<span class="sd">        high_ld_regions: Path </span>
<span class="sd">            Path to file containing high LD regions to exclude</span>
<span class="sd">        reference_files: dict </span>
<span class="sd">            Dictionary containing paths to reference population files</span>
<span class="sd">        built: str (optional) </span>
<span class="sd">            Genome build version (&#39;37&#39; or &#39;38&#39;). Defaults to &#39;38&#39;</span>
<span class="sd">        </span>
<span class="sd">        Raises:</span>
<span class="sd">        -------</span>
<span class="sd">        TypeError: </span>
<span class="sd">            If input arguments are not of correct type</span>
<span class="sd">        ValueError: </span>
<span class="sd">            If genome build version is not &#39;37&#39; or &#39;38&#39;</span>
<span class="sd">        FileNotFoundError: </span>
<span class="sd">            If required input files/directories do not exist</span>
<span class="sd">        </span>
<span class="sd">        Attributes:</span>
<span class="sd">        -----------</span>
<span class="sd">            reference_AC_GT_filtered: Filtered reference allele counts and genotypes</span>
<span class="sd">            study_AC_GT_filtered: Filtered study allele counts and genotypes</span>
<span class="sd">            pruned_reference: LD-pruned reference data</span>
<span class="sd">            pruned_study: LD-pruned study data</span>
<span class="sd">            reference_fixed_chr: Reference data with fixed chromosomes</span>
<span class="sd">            reference_fixed_pos: Reference data with fixed positions</span>
<span class="sd">            reference_flipped: Reference data with flipped alleles</span>
<span class="sd">            reference_cleaned: Final cleaned reference data</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">input_path</span><span class="p">,</span> <span class="n">Path</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;input_path should be a Path object&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="n">Path</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;output_path should be a Path object&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">high_ld_regions</span><span class="p">,</span> <span class="n">Path</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;high_ld_regions should be a Path object&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">reference_files</span><span class="p">,</span> <span class="nb">dict</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;reference_files should be a dictionary&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">input_name</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;input_name should be a string&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">output_name</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;output_name should be a string&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">built</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;built should be a string&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">built</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;37&#39;</span><span class="p">,</span> <span class="s1">&#39;38&#39;</span><span class="p">]:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;built should be either &#39;37&#39; or &#39;38&#39;&quot;</span><span class="p">)</span>
        
        <span class="k">if</span> <span class="ow">not</span> <span class="n">input_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;input_path does not exist: </span><span class="si">{</span><span class="n">input_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">output_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="s2">&quot;output_path does not exist&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">high_ld_regions</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="s2">&quot;high_ld_regions does not exist&quot;</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">input_path</span> <span class="o">=</span> <span class="n">input_path</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">input_name</span> <span class="o">=</span> <span class="n">input_name</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">output_path</span><span class="o">=</span> <span class="n">output_path</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">=</span> <span class="n">output_name</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">high_ld_regions</span> <span class="o">=</span> <span class="n">high_ld_regions</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reference_files</span> <span class="o">=</span> <span class="n">reference_files</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">reference_AC_GT_filtered</span><span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">study_AC_GT_filtered</span>    <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pruned_reference</span>        <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pruned_study</span>            <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reference_fixed_chr</span>     <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reference_fixed_pos</span>     <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reference_flipped</span>       <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reference_cleaned</span>       <span class="o">=</span> <span class="kc">None</span>

        <span class="k">pass</span></div>


<div class="viewcode-block" id="ReferenceGenomicMerger.execute_rename_snpid">
<a class="viewcode-back" href="../../AncestryQC.html#ideal_genom_qc.AncestryQC.ReferenceGenomicMerger.execute_rename_snpid">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">execute_rename_snpid</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Executes the SNP ID renaming process using PLINK2.</span>
<span class="sd">        This method renames SNP IDs in the PLINK binary files to a standardized format of &#39;chr:pos:a1:a2&#39;.</span>
<span class="sd">        The renaming is performed using PLINK2&#39;s --set-all-var-ids parameter.</span>
<span class="sd">        </span>
<span class="sd">        Parameter:</span>
<span class="sd">        ----------</span>
<span class="sd">        rename (bool, optional): Flag to control whether SNP renaming should be performed. </span>
<span class="sd">            Defaults to True.</span>

<span class="sd">        Returns:</span>
<span class="sd">        --------</span>
<span class="sd">            None</span>

<span class="sd">        Raises:</span>
<span class="sd">        -------</span>
<span class="sd">            TypeError: If rename parameter is not a boolean.</span>

<span class="sd">        Notes:</span>
<span class="sd">        ------</span>
<span class="sd">            - The renamed files will be saved with &#39;-renamed&#39; suffix</span>
<span class="sd">            - Thread count is optimized based on available CPU cores</span>
<span class="sd">            - The new SNP ID format will be: chromosome:position:allele1:allele2</span>
<span class="sd">            - Sets self.renamed_snps to True if renaming is performed</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;STEP: Renaming SNP IDs in the study data using PLINK2&quot;</span><span class="p">)</span>

        <span class="n">cpu_count</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">cpu_count</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">max_threads</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">cpu_count</span> <span class="o">-</span> <span class="mi">2</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Dynamically calculate fallback as half of available cores or default to 2</span>
            <span class="n">max_threads</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="p">(</span><span class="n">psutil</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">(</span><span class="n">logical</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="ow">or</span> <span class="mi">2</span><span class="p">)</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span>

        <span class="n">plink2_cmd</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink2 --bfile </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">input_path</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="si">}</span><span class="s2"> --set-all-var-ids @:#:$r:$a --threads </span><span class="si">{</span><span class="n">max_threads</span><span class="si">}</span><span class="s2"> --make-bed --out </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">output_path</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="o">+</span><span class="w"> </span><span class="s1">&#39;-renamed&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="c1"># Execute PLINK2 command</span>
        <span class="n">shell_do</span><span class="p">(</span><span class="n">plink2_cmd</span><span class="p">,</span> <span class="n">log</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="k">return</span></div>


<div class="viewcode-block" id="ReferenceGenomicMerger.execute_filter_prob_snps">
<a class="viewcode-back" href="../../AncestryQC.html#ideal_genom_qc.AncestryQC.ReferenceGenomicMerger.execute_filter_prob_snps">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">execute_filter_prob_snps</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span><span class="o">-&gt;</span><span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Executes the filtering of problematic SNPs (A-&gt;T and C-&gt;G) from both study and reference data.</span>
<span class="sd">        This method performs the following operations:</span>
<span class="sd">        1. Identifies and filters A-&gt;T and C-&gt;G SNPs from study data</span>
<span class="sd">        2. Identifies and filters A-&gt;T and C-&gt;G SNPs from reference data</span>
<span class="sd">        3. Creates new PLINK binary files excluding the identified problematic SNPs</span>
<span class="sd">        4. Uses maximum available CPU threads (total cores - 2) and 2/3 of available memory</span>
<span class="sd">        The method handles both renamed and original SNP scenarios, determined by self.renamed_snps.</span>
<span class="sd">        </span>
<span class="sd">        Returns:</span>
<span class="sd">        --------</span>
<span class="sd">            None</span>

<span class="sd">        Side Effects:</span>
<span class="sd">        -------------</span>
<span class="sd">            - Creates filtered SNP list files in the output directory</span>
<span class="sd">            - Creates new PLINK binary files (.bed, .bim, .fam) in the output directory</span>
<span class="sd">            - Sets self.reference_AC_GT_filtered and self.study_AC_GT_filtered paths</span>
<span class="sd">            - Logs progress and statistics of filtering operations</span>

<span class="sd">        Requires:</span>
<span class="sd">        ---------</span>
<span class="sd">            - Valid PLINK binary files for both study and reference data</span>
<span class="sd">            - Proper initialization of input_path, output_path, and reference_files</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;STEP: Filtering A-&gt;T and C-&gt;G SNPs from study and reference data.&quot;</span><span class="p">)</span>

        <span class="n">cpu_count</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">cpu_count</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">max_threads</span> <span class="o">=</span> <span class="n">cpu_count</span><span class="o">-</span><span class="mi">2</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">max_threads</span> <span class="o">=</span> <span class="mi">10</span>
        
        <span class="c1"># Get the virtual memory details</span>
        <span class="n">memory_info</span> <span class="o">=</span> <span class="n">psutil</span><span class="o">.</span><span class="n">virtual_memory</span><span class="p">()</span>
        <span class="n">available_memory_mb</span> <span class="o">=</span> <span class="n">memory_info</span><span class="o">.</span><span class="n">available</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1024</span> <span class="o">*</span> <span class="mi">1024</span><span class="p">)</span>
        <span class="n">memory</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="mi">2</span><span class="o">*</span><span class="n">available_memory_mb</span><span class="o">/</span><span class="mi">3</span><span class="p">,</span><span class="mi">0</span><span class="p">)</span>


        <span class="c1"># find A-&gt;T and C-&gt;G SNPs in study data</span>
        <span class="n">filtered_study</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_filter_non_AT_or_GC_snps</span><span class="p">(</span><span class="n">target_bim</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">output_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="si">}</span><span class="s2">-renamed.bim&quot;</span><span class="p">,</span> <span class="n">output_filename</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;STEP: Filtering problematic SNPs from the study data: filtered study data&quot;</span><span class="p">)</span>

        <span class="c1"># find A-&gt;T and C-&gt;G SNPs in reference data</span>
        <span class="n">filtered_reference</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_filter_non_AT_or_GC_snps</span><span class="p">(</span><span class="n">target_bim</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_files</span><span class="p">[</span><span class="s1">&#39;bim&#39;</span><span class="p">],</span> <span class="n">output_filename</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_files</span><span class="p">[</span><span class="s1">&#39;bim&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">stem</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;STEP: Filtering problematic SNPs from the study data: filtered reference data&quot;</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">reference_AC_GT_filtered</span><span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_files</span><span class="p">[</span><span class="s1">&#39;bim&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">stem</span><span class="si">}</span><span class="s2">-no_ac_gt_snps&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">study_AC_GT_filtered</span>    <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="si">}</span><span class="s2">-no_ac_gt_snps&quot;</span>

        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filtered_study</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;STEP: Filtering problematic SNPs from the study data: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">readlines</span><span class="p">())</span><span class="si">}</span><span class="s2"> SNPs filtered&quot;</span><span class="p">)</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filtered_reference</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;STEP: Filtering problematic SNPs from the reference data: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">readlines</span><span class="p">())</span><span class="si">}</span><span class="s2"> SNPs filtered&quot;</span><span class="p">)</span>


        <span class="c1"># PLINK command: generate cleaned study data files</span>
        <span class="n">plink_cmd1</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink --bfile  </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">output_path</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="o">+</span><span class="s1">&#39;-renamed&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2"> --chr 1-22 --exclude </span><span class="si">{</span><span class="n">filtered_study</span><span class="si">}</span><span class="s2"> --keep-allele-order --threads </span><span class="si">{</span><span class="n">max_threads</span><span class="si">}</span><span class="s2"> --make-bed --out </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">study_AC_GT_filtered</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="c1"># Make sure the reference bim path is valid and extract the base filename</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_files</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;bim&#39;</span><span class="p">)</span> <span class="ow">or</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_files</span><span class="p">[</span><span class="s1">&#39;bim&#39;</span><span class="p">],</span> <span class="n">Path</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;reference_files dictionary must contain a valid &#39;bim&#39; Path&quot;</span><span class="p">)</span>
        
        <span class="n">reference_base</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_files</span><span class="p">[</span><span class="s1">&#39;bim&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
        
        <span class="c1"># PLINK command: generate cleaned reference data files</span>
        <span class="n">plink_cmd2</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink --bfile </span><span class="si">{</span><span class="n">reference_base</span><span class="si">}</span><span class="s2"> --biallelic-only strict --chr 1-22 --exclude </span><span class="si">{</span><span class="n">filtered_reference</span><span class="si">}</span><span class="s2"> --keep-allele-order --allow-extra-chr --memory </span><span class="si">{</span><span class="n">memory</span><span class="si">}</span><span class="s2"> --threads </span><span class="si">{</span><span class="n">max_threads</span><span class="si">}</span><span class="s2"> --make-bed --out </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_AC_GT_filtered</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="c1"># execute PLINK commands</span>
        <span class="n">cmds</span> <span class="o">=</span> <span class="p">[</span><span class="n">plink_cmd1</span><span class="p">,</span> <span class="n">plink_cmd2</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">cmd</span> <span class="ow">in</span> <span class="n">cmds</span><span class="p">:</span>
            <span class="n">shell_do</span><span class="p">(</span><span class="n">cmd</span><span class="p">,</span> <span class="n">log</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="k">return</span></div>

    
<div class="viewcode-block" id="ReferenceGenomicMerger.execute_ld_pruning">
<a class="viewcode-back" href="../../AncestryQC.html#ideal_genom_qc.AncestryQC.ReferenceGenomicMerger.execute_ld_pruning">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">execute_ld_pruning</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ind_pair</span><span class="p">:</span><span class="nb">list</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Execute linkage disequilibrium (LD) pruning on study and reference data.</span>
<span class="sd">        </span>
<span class="sd">        This method performs LD-based pruning using PLINK to remove highly correlated SNPs </span>
<span class="sd">        from both study and reference datasets. The pruning is done using a sliding window </span>
<span class="sd">        approach where SNPs are removed based on their pairwise correlation (r²).</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ind_pair : list</span>
<span class="sd">            A list containing three elements:</span>
<span class="sd">            </span>
<span class="sd">            - ind_pair[0] (int): Window size in SNPs  </span>
<span class="sd">            - ind_pair[1] (int): Number of SNPs to shift the window at each step  </span>
<span class="sd">            - ind_pair[2] (float): r² threshold for pruning</span>
<span class="sd">        </span>
<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        TypeError</span>
<span class="sd">            If ind_pair is not a list.</span>
<span class="sd">        TypeError</span>
<span class="sd">            If first two elements of ind_pair are not integers.</span>
<span class="sd">        TypeError</span>
<span class="sd">            If third element of ind_pair is not a float.</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None</span>
<span class="sd">        </span>
<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        - Uses PLINK&#39;s `--indep-pairwise` command for pruning.</span>
<span class="sd">        - Excludes high LD regions specified in `self.high_ld_regions`.</span>
<span class="sd">        - Creates pruned datasets for both study and reference data.</span>
<span class="sd">        - Updates `self.pruned_reference` and `self.pruned_study` with paths to pruned files.</span>
<span class="sd">        - Uses all available CPU threads except 2 for processing.</span>
<span class="sd">        &quot;&quot;&quot;</span>


        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ind_pair</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;ind_pair should be a list&quot;</span><span class="p">)</span>
        
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ind_pair</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="nb">int</span><span class="p">)</span> <span class="ow">or</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ind_pair</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="nb">int</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;The first two elements in ind_pair values should be integers (windows size and step size)&quot;</span><span class="p">)</span>
        
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ind_pair</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="nb">float</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;The third element in ind_pair should be a float (r^2 threshold)&quot;</span><span class="p">)</span>
        
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;STEP: LD-based pruning of study and reference data&quot;</span><span class="p">)</span>

        <span class="n">cpu_count</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">cpu_count</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">max_threads</span> <span class="o">=</span> <span class="n">cpu_count</span><span class="o">-</span><span class="mi">2</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">max_threads</span> <span class="o">=</span> <span class="mi">10</span>

        <span class="c1"># PLINK command: generates prune.in and prune.out files from study data</span>
        <span class="n">plink_cmd1</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink --bfile </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">study_AC_GT_filtered</span><span class="p">)</span><span class="si">}</span><span class="s2"> --exclude range </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">high_ld_regions</span><span class="si">}</span><span class="s2"> --keep-allele-order --indep-pairwise </span><span class="si">{</span><span class="n">ind_pair</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">ind_pair</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">ind_pair</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="si">}</span><span class="s2"> --threads </span><span class="si">{</span><span class="n">max_threads</span><span class="si">}</span><span class="s2"> --out </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_path</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="c1"># PLINK command: prune study data and creates a filtered binary file</span>
        <span class="n">plink_cmd2</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink --bfile </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">study_AC_GT_filtered</span><span class="p">)</span><span class="si">}</span><span class="s2"> --extract </span><span class="si">{</span><span class="nb">str</span><span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">output_path</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="p">)</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s1">&#39;.prune.in&#39;</span><span class="p">))</span><span class="si">}</span><span class="s2"> --keep-allele-order --threads </span><span class="si">{</span><span class="n">max_threads</span><span class="si">}</span><span class="s2"> --make-bed --out </span><span class="si">{</span><span class="nb">str</span><span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">output_path</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="o">+</span><span class="s1">&#39;-pruned&#39;</span><span class="p">)))</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="c1"># PLINK command: generates a pruned reference data files</span>
        <span class="n">plink_cmd3</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink --bfile </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_AC_GT_filtered</span><span class="p">)</span><span class="si">}</span><span class="s2"> --extract </span><span class="si">{</span><span class="nb">str</span><span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">output_path</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="p">)</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s1">&#39;.prune.in&#39;</span><span class="p">))</span><span class="si">}</span><span class="s2"> --keep-allele-order --make-bed --threads </span><span class="si">{</span><span class="n">max_threads</span><span class="si">}</span><span class="s2"> --out </span><span class="si">{</span><span class="nb">str</span><span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">output_path</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_files</span><span class="p">[</span><span class="s1">&#39;bim&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">stem</span><span class="o">+</span><span class="s1">&#39;-pruned&#39;</span><span class="p">)))</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">pruned_reference</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_path</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_files</span><span class="p">[</span><span class="s1">&#39;bim&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">stem</span><span class="o">+</span><span class="s1">&#39;-pruned&#39;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pruned_study</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_path</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="o">+</span><span class="s1">&#39;-pruned&#39;</span><span class="p">)</span>

        <span class="c1"># execute PLINK commands</span>
        <span class="n">cmds</span> <span class="o">=</span> <span class="p">[</span><span class="n">plink_cmd1</span><span class="p">,</span> <span class="n">plink_cmd2</span><span class="p">,</span> <span class="n">plink_cmd3</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">cmd</span> <span class="ow">in</span> <span class="n">cmds</span><span class="p">:</span>
            <span class="n">shell_do</span><span class="p">(</span><span class="n">cmd</span><span class="p">,</span> <span class="n">log</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="k">return</span></div>

    
<div class="viewcode-block" id="ReferenceGenomicMerger.execute_fix_chromosome_mismatch">
<a class="viewcode-back" href="../../AncestryQC.html#ideal_genom_qc.AncestryQC.ReferenceGenomicMerger.execute_fix_chromosome_mismatch">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">execute_fix_chromosome_mismatch</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Fix chromosome mismatch between study data and reference panel.</span>

<span class="sd">        This method executes PLINK commands to correct any chromosome mismatches between the study data</span>
<span class="sd">        and reference panel datasets. It identifies mismatches using internal methods and updates</span>
<span class="sd">        the chromosome assignments in the reference panel to match the study data.</span>

<span class="sd">        The method performs the following steps:</span>
<span class="sd">        1. Identifies chromosome mismatches between study and reference BIM files</span>
<span class="sd">        2. Creates an update file for chromosome reassignment</span>
<span class="sd">        3. Executes PLINK command to update chromosome assignments in reference panel</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        - Creates new PLINK binary files with updated chromosome assignments</span>
<span class="sd">        - The updated files are saved with &#39;-updateChr&#39; suffix</span>
<span class="sd">        </span>
<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        ValueError</span>
<span class="sd">            If pruned_study or pruned_reference is None, meaning execute_ld_pruning() was not called first</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;STEP: Fixing chromosome mismatch between study data and reference panel&quot;</span><span class="p">)</span>

        <span class="c1"># Check if pruned_study and pruned_reference have been set</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">pruned_study</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;pruned_study is not set. Make sure execute_ld_pruning() is called before this method and completed successfully.&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">pruned_reference</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;pruned_reference is not set. Make sure execute_ld_pruning() is called before this method and completed successfully.&quot;</span><span class="p">)</span>

        <span class="n">cpu_count</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">cpu_count</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">max_threads</span> <span class="o">=</span> <span class="n">cpu_count</span><span class="o">-</span><span class="mi">2</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">max_threads</span> <span class="o">=</span> <span class="mi">10</span>

        <span class="c1"># File paths - using with_suffix instead of with_name for more reliability</span>
        <span class="n">study_bim</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pruned_study</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s1">&#39;.bim&#39;</span><span class="p">)</span>
        <span class="n">reference_bim</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pruned_reference</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s1">&#39;.bim&#39;</span><span class="p">)</span>

        <span class="n">to_update_chr_file</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_find_chromosome_mismatch</span><span class="p">(</span><span class="n">study_bim</span><span class="p">,</span> <span class="n">reference_bim</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">reference_fixed_chr</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_files</span><span class="p">[</span><span class="s1">&#39;bim&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">stem</span><span class="si">}</span><span class="s2">-updateChr&quot;</span>

        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">to_update_chr_file</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;STEP: Fixing chromosome mismatch between study data and reference panel: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">readlines</span><span class="p">())</span><span class="si">}</span><span class="s2"> SNPs to update&quot;</span><span class="p">)</span>

        <span class="c1"># PLINK command</span>
        <span class="n">plink_cmd</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink --bfile </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">pruned_reference</span><span class="si">}</span><span class="s2"> --update-chr </span><span class="si">{</span><span class="n">to_update_chr_file</span><span class="si">}</span><span class="s2"> 1 2 --keep-allele-order --threads </span><span class="si">{</span><span class="n">max_threads</span><span class="si">}</span><span class="s2"> --make-bed --out </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_fixed_chr</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="c1"># Execute PLINK command</span>
        <span class="n">shell_do</span><span class="p">(</span><span class="n">plink_cmd</span><span class="p">,</span> <span class="n">log</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="k">return</span></div>

    
<div class="viewcode-block" id="ReferenceGenomicMerger.execute_fix_possition_mismatch">
<a class="viewcode-back" href="../../AncestryQC.html#ideal_genom_qc.AncestryQC.ReferenceGenomicMerger.execute_fix_possition_mismatch">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">execute_fix_possition_mismatch</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Fixes position mismatches between study data and reference panel.</span>

<span class="sd">        This method executes PLINK commands to update the positions of SNPs in the reference panel</span>
<span class="sd">        to match those in the study data. It processes previously identified position mismatches</span>
<span class="sd">        and creates new binary PLINK files with corrected positions.</span>

<span class="sd">        The method:</span>
<span class="sd">        1. Determines optimal thread count for processing</span>
<span class="sd">        2. Identifies position mismatches between study and reference BIM files</span>
<span class="sd">        3. Updates reference panel positions using PLINK</span>
<span class="sd">        4. Creates new binary files with corrected positions</span>

<span class="sd">        Returns:</span>
<span class="sd">        --------</span>
<span class="sd">            None</span>

<span class="sd">        Side Effects:</span>
<span class="sd">        -------------</span>
<span class="sd">            - Creates new PLINK binary files (.bed, .bim, .fam) with updated positions</span>
<span class="sd">            - Logs the number of SNPs being updated</span>
<span class="sd">            - Modifies self.reference_fixed_pos with path to updated files</span>

<span class="sd">        Dependencies:</span>
<span class="sd">        -------------</span>
<span class="sd">            - Requires PLINK to be installed and accessible</span>
<span class="sd">            - Expects pruned study and reference files to exist</span>
<span class="sd">            - Requires previous chromosome fixing step to be completed</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;STEP: Fixing position mismatch between study data and reference panel&quot;</span><span class="p">)</span>

        <span class="n">cpu_count</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">cpu_count</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">max_threads</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">cpu_count</span> <span class="o">-</span> <span class="mi">2</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Dynamically calculate fallback as half of available cores or default to 2</span>
            <span class="n">max_threads</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="p">(</span><span class="n">psutil</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">(</span><span class="n">logical</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="ow">or</span> <span class="mi">2</span><span class="p">)</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span>

        <span class="c1"># Check if pruned_study and reference_fixed_chr have been properly set</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">pruned_study</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;pruned_study is not set. Make sure execute_ld_pruning() is called before this method and completed successfully.&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_fixed_chr</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;reference_fixed_chr is not set. Make sure execute_fix_chromosome_mismatch() is called before this method.&quot;</span><span class="p">)</span>
            
        <span class="c1"># File paths - using with_suffix instead of with_name for more reliability</span>
        <span class="n">study_bim</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pruned_study</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s1">&#39;.bim&#39;</span><span class="p">)</span>
        <span class="n">reference_bim</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_fixed_chr</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s1">&#39;.bim&#39;</span><span class="p">)</span>

        <span class="n">to_update_pos_file</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_find_position_mismatch</span><span class="p">(</span><span class="n">study_bim</span><span class="p">,</span> <span class="n">reference_bim</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">reference_fixed_pos</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_files</span><span class="p">[</span><span class="s1">&#39;bim&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">stem</span><span class="si">}</span><span class="s2">-updatePos&quot;</span>

        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">to_update_pos_file</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;STEP: Fixing position mismatch between study data and reference panel: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">readlines</span><span class="p">())</span><span class="si">}</span><span class="s2"> SNPs to update&quot;</span><span class="p">)</span>

        <span class="c1"># PLINK command</span>
        <span class="n">plink_cmd</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink --bfile </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_fixed_chr</span><span class="si">}</span><span class="s2"> --update-map </span><span class="si">{</span><span class="n">to_update_pos_file</span><span class="si">}</span><span class="s2"> --keep-allele-order --threads </span><span class="si">{</span><span class="n">max_threads</span><span class="si">}</span><span class="s2"> --make-bed --out </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_fixed_pos</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="c1"># Execute PLINK command</span>
        <span class="n">shell_do</span><span class="p">(</span><span class="n">plink_cmd</span><span class="p">,</span> <span class="n">log</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="k">return</span></div>

    
<div class="viewcode-block" id="ReferenceGenomicMerger.execute_fix_allele_flip">
<a class="viewcode-back" href="../../AncestryQC.html#ideal_genom_qc.AncestryQC.ReferenceGenomicMerger.execute_fix_allele_flip">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">execute_fix_allele_flip</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">            </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">            Executes the allele flipping process between study data and reference panel.</span>
<span class="sd">    </span>
<span class="sd">            This method performs the following steps:</span>
<span class="sd">            1. Identifies SNPs requiring allele flipping between study and reference data</span>
<span class="sd">            2. Creates a list of SNPs to flip</span>
<span class="sd">            3. Generates a new reference panel with flipped alleles using PLINK</span>
<span class="sd">    </span>
<span class="sd">            The method uses multi-threading capabilities based on available CPU cores,</span>
<span class="sd">            reserving 2 cores for system processes when possible.</span>
<span class="sd">    </span>
<span class="sd">            Returns:</span>
<span class="sd">            --------</span>
<span class="sd">                None</span>
<span class="sd">    </span>
<span class="sd">            Side Effects:</span>
<span class="sd">            -------------</span>
<span class="sd">                - Creates a .toFlip file containing SNPs requiring allele flipping</span>
<span class="sd">                - Generates new PLINK binary files (.bed, .bim, .fam) with flipped alleles</span>
<span class="sd">                - Logs the number of SNPs requiring flipping</span>
<span class="sd">                - Updates self.reference_flipped with the path to new flipped reference files</span>
<span class="sd">    </span>
<span class="sd">            Dependencies:</span>
<span class="sd">            -------------</span>
<span class="sd">                - PLINK must be installed and accessible in system PATH</span>
<span class="sd">                - Requires valid PLINK binary files for both study and reference data</span>
<span class="sd">                - Requires write permissions in output directory</span>
<span class="sd">            &quot;&quot;&quot;</span>
    
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;STEP: Allele flipping between study data and reference panel&quot;</span><span class="p">)</span>
    
            <span class="n">cpu_count</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">cpu_count</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">max_threads</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">cpu_count</span> <span class="o">-</span> <span class="mi">2</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">max_threads</span> <span class="o">=</span> <span class="mi">10</span>
                
            <span class="c1"># Check if pruned_study and reference_fixed_pos have been properly set</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">pruned_study</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;pruned_study is not set. Make sure execute_ld_pruning() is called before this method and completed successfully.&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_fixed_pos</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;reference_fixed_pos is not set. Make sure execute_fix_possition_mismatch() is called before this method.&quot;</span><span class="p">)</span>
                
            <span class="c1"># File paths - using with_suffix for consistency and reliability</span>
            <span class="n">study_bim</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pruned_study</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s1">&#39;.bim&#39;</span><span class="p">)</span>
            <span class="n">reference_bim</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_fixed_pos</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s1">&#39;.bim&#39;</span><span class="p">)</span>
    
            <span class="n">to_flip_file</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_files</span><span class="p">[</span><span class="s1">&#39;bim&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">stem</span><span class="si">}</span><span class="s2">.toFlip&quot;</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_find_allele_flip</span><span class="p">(</span><span class="n">study_bim</span><span class="p">,</span> <span class="n">reference_bim</span><span class="p">,</span> <span class="n">to_flip_file</span><span class="p">)</span>
    
            <span class="bp">self</span><span class="o">.</span><span class="n">reference_flipped</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_files</span><span class="p">[</span><span class="s1">&#39;bim&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">stem</span><span class="si">}</span><span class="s2">-flipped&quot;</span>
    
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">to_flip_file</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;STEP: Allele flipping between study data and reference panel: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">readlines</span><span class="p">())</span><span class="si">}</span><span class="s2"> SNPs to flip&quot;</span><span class="p">)</span>
    
            <span class="c1"># plink command</span>
            <span class="n">plink_cmd</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink --bfile </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_fixed_pos</span><span class="si">}</span><span class="s2"> --flip </span><span class="si">{</span><span class="n">to_flip_file</span><span class="si">}</span><span class="s2"> --keep-allele-order --threads </span><span class="si">{</span><span class="n">max_threads</span><span class="si">}</span><span class="s2"> --make-bed --out </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_flipped</span><span class="si">}</span><span class="s2">&quot;</span>
    
            <span class="c1"># execute PLINK command</span>
            <span class="n">shell_do</span><span class="p">(</span><span class="n">plink_cmd</span><span class="p">,</span> <span class="n">log</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    
            <span class="k">return</span></div>


<div class="viewcode-block" id="ReferenceGenomicMerger.execute_remove_mismatches">
<a class="viewcode-back" href="../../AncestryQC.html#ideal_genom_qc.AncestryQC.ReferenceGenomicMerger.execute_remove_mismatches">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">execute_remove_mismatches</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Removes mismatched SNPs from the reference data based on allele comparisons between study and reference datasets.</span>

<span class="sd">        This method performs the following steps:</span>
<span class="sd">        1. Determines optimal thread count for processing</span>
<span class="sd">        2. Identifies allele mismatches between study and reference BIM files</span>
<span class="sd">        3. Creates a list of SNPs to remove</span>
<span class="sd">        4. Generates a cleaned reference dataset excluding mismatched SNPs</span>

<span class="sd">        The method utilizes PLINK to perform the actual SNP removal while maintaining allele order.</span>

<span class="sd">        Returns:</span>
<span class="sd">        --------</span>
<span class="sd">            None</span>

<span class="sd">        Side Effects:</span>
<span class="sd">        -------------</span>
<span class="sd">            - Creates a file listing SNPs to be removed at {output_path}/{reference_bim_stem}.toRemove</span>
<span class="sd">            - Generates cleaned reference files at {output_path}/{reference_bim_stem}-cleaned.bed/bim/fam</span>
<span class="sd">            - Logs the number of SNPs being removed</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;STEP: Removing mismatched SNPs from reference data&quot;</span><span class="p">)</span>

        <span class="n">cpu_count</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">cpu_count</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">max_threads</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">cpu_count</span> <span class="o">-</span> <span class="mi">2</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Dynamically calculate fallback as half of available cores or default to 2</span>
            <span class="n">max_threads</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="p">(</span><span class="n">psutil</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">(</span><span class="n">logical</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="ow">or</span> <span class="mi">2</span><span class="p">)</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">pruned_study</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;pruned_study is not set. Make sure execute_ld_pruning() is called before this method and completed successfully.&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">pruned_reference</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;pruned_reference is not set. Make sure execute_ld_pruning() is called before this method and completed successfully.&quot;</span><span class="p">)</span>

        <span class="c1"># File paths - using with_suffix instead of with_name for more reliability</span>
        <span class="n">study_bim</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pruned_study</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s1">&#39;.bim&#39;</span><span class="p">)</span>
        <span class="n">reference_bim</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pruned_reference</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s1">&#39;.bim&#39;</span><span class="p">)</span>

        <span class="n">mismatches_file</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_files</span><span class="p">[</span><span class="s1">&#39;bim&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">stem</span><span class="si">}</span><span class="s2">.toRemove&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_find_allele_flip</span><span class="p">(</span><span class="n">study_bim</span><span class="p">,</span> <span class="n">reference_bim</span><span class="p">,</span> <span class="n">mismatches_file</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">reference_cleaned</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_files</span><span class="p">[</span><span class="s1">&#39;bim&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">stem</span><span class="si">}</span><span class="s2">-cleaned&quot;</span>

        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">mismatches_file</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;STEP: Removing mismatched SNPs from reference data: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">readlines</span><span class="p">())</span><span class="si">}</span><span class="s2"> SNPs to remove&quot;</span><span class="p">)</span>

        <span class="c1"># plink command</span>
        <span class="n">plink_cmd</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink --bfile </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_flipped</span><span class="si">}</span><span class="s2"> --exclude </span><span class="si">{</span><span class="n">mismatches_file</span><span class="si">}</span><span class="s2"> --keep-allele-order --threads </span><span class="si">{</span><span class="n">max_threads</span><span class="si">}</span><span class="s2"> --make-bed --out </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_cleaned</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="c1"># execute PLINK command</span>
        <span class="n">shell_do</span><span class="p">(</span><span class="n">plink_cmd</span><span class="p">,</span> <span class="n">log</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="k">return</span></div>

    
<div class="viewcode-block" id="ReferenceGenomicMerger.execute_merge_data">
<a class="viewcode-back" href="../../AncestryQC.html#ideal_genom_qc.AncestryQC.ReferenceGenomicMerger.execute_merge_data">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">execute_merge_data</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Merge study and reference data using PLINK.</span>

<span class="sd">        This method merges the pruned study data with the cleaned reference data using PLINK&#39;s</span>
<span class="sd">        --bmerge functionality. It automatically determines the optimal number of threads to use</span>
<span class="sd">        based on available CPU cores.</span>

<span class="sd">        The method:</span>
<span class="sd">        1. Calculates optimal thread count (CPU count - 2 or half of available cores)</span>
<span class="sd">        2. Constructs PLINK command for merging datasets</span>
<span class="sd">        3. Executes the merge operation via shell command</span>

<span class="sd">        Returns:</span>
<span class="sd">        --------</span>
<span class="sd">            None</span>

<span class="sd">        Side effects:</span>
<span class="sd">        -------------</span>
<span class="sd">            - Creates merged PLINK binary files (.bed, .bim, .fam) in the output directory</span>
<span class="sd">            - Logs the merge operation</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;STEP: Merging study and reference data&quot;</span><span class="p">)</span>

        <span class="n">cpu_count</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">cpu_count</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">max_threads</span> <span class="o">=</span> <span class="n">cpu_count</span><span class="o">-</span><span class="mi">2</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">max_threads</span> <span class="o">=</span> <span class="mi">10</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_cleaned</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;reference_cleaned is not set. Make sure execute_remove_mismatches() is called before this method and completed successfully.&quot;</span><span class="p">)</span>

        <span class="c1"># plink command</span>
        <span class="n">plink_cmd</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink --bfile </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">pruned_study</span><span class="si">}</span><span class="s2"> --bmerge </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_cleaned</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s1">&#39;.bed&#39;</span><span class="p">))</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_cleaned</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s1">&#39;.bim&#39;</span><span class="p">))</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_cleaned</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s1">&#39;.fam&#39;</span><span class="p">))</span><span class="si">}</span><span class="s2"> --keep-allele-order --threads </span><span class="si">{</span><span class="n">max_threads</span><span class="si">}</span><span class="s2"> --make-bed --out </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">output_path</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-merged&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="c1"># execute PLINK command</span>
        <span class="n">shell_do</span><span class="p">(</span><span class="n">plink_cmd</span><span class="p">,</span> <span class="n">log</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="k">return</span></div>


    <span class="k">def</span><span class="w"> </span><span class="nf">_filter_non_AT_or_GC_snps</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">target_bim</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">output_filename</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Path</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Filter SNPs that are not A/T or G/C variants from a PLINK BIM file.</span>
<span class="sd">        This method reads a BIM file and identifies SNPs that are either A/T or G/C variants.</span>
<span class="sd">        These variants are known as strand-ambiguous SNPs because their complementary alleles </span>
<span class="sd">        are the same as their original alleles, making it impossible to determine the correct </span>
<span class="sd">        strand without additional information.</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        target_bim : Path</span>
<span class="sd">            Path to the input BIM file containing SNP information</span>
<span class="sd">        output_filename : str</span>
<span class="sd">            Base name for the output file (without extension)</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        Path</span>
<span class="sd">            Path to the output file containing filtered SNP IDs with .ac_get_snps extension</span>
<span class="sd">        </span>
<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        The input BIM file should be tab-delimited with standard PLINK BIM format.</span>
<span class="sd">        Only columns containing SNP ID (column 2) and alleles (columns 5 and 6) are used.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
            <span class="n">target_bim</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">usecols</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">5</span><span class="p">],</span> <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;SNP&quot;</span><span class="p">,</span> <span class="s2">&quot;A1&quot;</span><span class="p">,</span> <span class="s2">&quot;A2&quot;</span><span class="p">]</span>
        <span class="p">)</span>

        <span class="n">output_file</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">output_filename</span><span class="si">}</span><span class="s2">.ac_get_snps&quot;</span>

        <span class="n">filtered_snps</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[[</span><span class="s1">&#39;A1&#39;</span><span class="p">,</span> <span class="s1">&#39;A2&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">x</span><span class="p">))</span> <span class="ow">in</span> <span class="p">{</span><span class="s2">&quot;AT&quot;</span><span class="p">,</span> <span class="s2">&quot;TA&quot;</span><span class="p">,</span> <span class="s2">&quot;GC&quot;</span><span class="p">,</span> <span class="s2">&quot;CG&quot;</span><span class="p">},</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)]</span>
        
        <span class="n">filtered_snps</span><span class="p">[[</span><span class="s2">&quot;SNP&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_file</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">output_file</span>
    
    <span class="k">def</span><span class="w"> </span><span class="nf">_find_chromosome_mismatch</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">study_bim</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">reference_bim</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Path</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Find chromosome mismatches between study and reference BIM files.</span>

<span class="sd">        This function identifies SNPs where the chromosome assignment differs between</span>
<span class="sd">        the study dataset and the reference panel, despite having the same rsID.</span>
<span class="sd">        Sex chromosomes (X, Y) are excluded from the update list.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        study_bim : Path</span>
<span class="sd">            Path to the study BIM file to check for mismatches</span>
<span class="sd">        reference_bim : Path</span>
<span class="sd">            Path to the reference BIM file to compare against</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        Path</span>
<span class="sd">            Path to output file containing SNPs that need chromosome updates.</span>
<span class="sd">            File format is tab-separated with columns: chromosome, rsID</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">col_names</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span> <span class="s2">&quot;rsid&quot;</span><span class="p">,</span> <span class="s2">&quot;pos_cm&quot;</span><span class="p">,</span> <span class="s2">&quot;pos_bp&quot;</span><span class="p">,</span> <span class="s2">&quot;allele1&quot;</span><span class="p">,</span> <span class="s2">&quot;allele2&quot;</span><span class="p">]</span>
        <span class="n">study_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">study_bim</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="n">col_names</span><span class="p">)</span>
        <span class="n">reference_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">reference_bim</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="n">col_names</span><span class="p">)</span>

        <span class="c1"># Find mismatches where rsID is the same but chromosome differs</span>
        <span class="n">mismatch_df</span> <span class="o">=</span> <span class="n">reference_df</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">study_df</span><span class="p">[[</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span> <span class="s2">&quot;rsid&quot;</span><span class="p">]],</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;rsid&quot;</span><span class="p">,</span> <span class="n">suffixes</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;_ref&quot;</span><span class="p">,</span> <span class="s2">&quot;_study&quot;</span><span class="p">))</span>
        <span class="n">chromosome_mismatch_df</span> <span class="o">=</span> <span class="n">mismatch_df</span><span class="p">[</span><span class="n">mismatch_df</span><span class="p">[</span><span class="s2">&quot;chr_ref&quot;</span><span class="p">]</span> <span class="o">!=</span> <span class="n">mismatch_df</span><span class="p">[</span><span class="s2">&quot;chr_study&quot;</span><span class="p">]]</span>

        <span class="c1"># Exclude chromosomes X and Y from updates</span>
        <span class="n">mismatch_df</span> <span class="o">=</span> <span class="n">mismatch_df</span><span class="p">[</span><span class="o">~</span><span class="n">mismatch_df</span><span class="p">[</span><span class="s2">&quot;chr_study&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="s2">&quot;X&quot;</span><span class="p">,</span> <span class="s2">&quot;Y&quot;</span><span class="p">])]</span>

        <span class="n">to_update_chr_file</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_path</span> <span class="o">/</span> <span class="s2">&quot;all_phase3.toUpdateChr&quot;</span>

        <span class="c1"># Save the mismatch data to a file</span>
        <span class="n">chromosome_mismatch_df</span><span class="p">[[</span><span class="s2">&quot;chr_study&quot;</span><span class="p">,</span> <span class="s2">&quot;rsid&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">to_update_chr_file</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">to_update_chr_file</span>
    
    <span class="k">def</span><span class="w"> </span><span class="nf">_find_position_mismatch</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">study_bim</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">reference_bim</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Path</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Find SNPs with mismatched positions between study and reference datasets.</span>

<span class="sd">        This method compares the base pair positions of SNPs between a study dataset and a </span>
<span class="sd">        reference dataset to identify SNPs that have different positions despite having the </span>
<span class="sd">        same rsID.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        study_bim : Path</span>
<span class="sd">            Path to the PLINK .bim file of the study dataset.</span>
<span class="sd">        reference_bim : Path</span>
<span class="sd">            Path to the PLINK .bim file of the reference dataset.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        Path</span>
<span class="sd">            Path to the output file containing SNPs that need position updates.</span>
<span class="sd">            The output file contains two columns (rsID and new position) without headers.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        The output file format is compatible with PLINK&#39;s --update-map command for updating</span>
<span class="sd">        SNP positions in the study dataset.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">col_names</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span> <span class="s2">&quot;rsid&quot;</span><span class="p">,</span> <span class="s2">&quot;pos_cm&quot;</span><span class="p">,</span> <span class="s2">&quot;pos_bp&quot;</span><span class="p">,</span> <span class="s2">&quot;allele1&quot;</span><span class="p">,</span> <span class="s2">&quot;allele2&quot;</span><span class="p">]</span>
        <span class="n">study_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">study_bim</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="n">col_names</span><span class="p">)</span>
        <span class="n">reference_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">reference_bim</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="n">col_names</span><span class="p">)</span>

        <span class="c1"># Create a dictionary from file1 with column 2 as key and column 4 as value</span>
        <span class="n">a</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">study_df</span><span class="p">[</span><span class="s1">&#39;rsid&#39;</span><span class="p">],</span> <span class="n">study_df</span><span class="p">[</span><span class="s1">&#39;pos_bp&#39;</span><span class="p">]))</span>

        <span class="c1"># Filter rows in reference_df where column 2 exists in &#39;a&#39; and the values isn column 4 differ</span>
        <span class="n">filtered</span> <span class="o">=</span> <span class="n">reference_df</span><span class="p">[</span><span class="n">reference_df</span><span class="p">[</span><span class="s1">&#39;rsid&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">a</span><span class="p">)</span><span class="o">.</span><span class="n">notna</span><span class="p">()</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">reference_df</span><span class="p">[</span><span class="s1">&#39;pos_bp&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="n">reference_df</span><span class="p">[</span><span class="s1">&#39;rsid&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">a</span><span class="p">))]</span>

        <span class="c1"># Print the result to a file</span>
        <span class="n">to_update_pos_file</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_files</span><span class="p">[</span><span class="s1">&#39;bim&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">stem</span><span class="si">}</span><span class="s2">.toUpdatePos&quot;</span>
        <span class="n">filtered</span><span class="p">[[</span><span class="s1">&#39;rsid&#39;</span><span class="p">,</span> <span class="s1">&#39;pos_bp&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">to_update_pos_file</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">to_update_pos_file</span>
    
    <span class="k">def</span><span class="w"> </span><span class="nf">_find_allele_flip</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">study_bim</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">reference_bim</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">output_filename</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Find SNPs with allele flips between study and reference datasets.</span>

<span class="sd">        This method identifies Single Nucleotide Polymorphisms (SNPs) where the alleles are</span>
<span class="sd">        flipped between the study and reference datasets. A flip occurs when the allele</span>
<span class="sd">        pairs don&#39;t match in either order.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        study_bim : Path</span>
<span class="sd">            Path to the study .bim file containing SNP information</span>
<span class="sd">        reference_bim : Path</span>
<span class="sd">            Path to the reference .bim file containing SNP information</span>
<span class="sd">        output_filename : Path</span>
<span class="sd">            Path where the list of flipped SNPs will be saved</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None</span>
<span class="sd">            Writes rsids of flipped SNPs to the specified output file</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        The .bim files should be tab-separated with columns:</span>
<span class="sd">        chromosome, rsid, genetic_distance, base_pair_position, allele1, allele2</span>

<span class="sd">        The output file will contain one rsid per line for SNPs where alleles don&#39;t match</span>
<span class="sd">        between study and reference in either order (A1/A2 or A2/A1).</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">col_names</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span> <span class="s2">&quot;rsid&quot;</span><span class="p">,</span> <span class="s2">&quot;pos_cm&quot;</span><span class="p">,</span> <span class="s2">&quot;pos_bp&quot;</span><span class="p">,</span> <span class="s2">&quot;allele1&quot;</span><span class="p">,</span> <span class="s2">&quot;allele2&quot;</span><span class="p">]</span>
        <span class="n">study_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">study_bim</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="n">col_names</span><span class="p">)</span>
        <span class="n">reference_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">reference_bim</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="n">col_names</span><span class="p">)</span>

        <span class="c1"># Create a dictionary with the composite key from file1</span>
        <span class="n">a</span> <span class="o">=</span> <span class="p">{</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;chr&#39;</span><span class="p">]</span><span class="si">}{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;rsid&#39;</span><span class="p">]</span><span class="si">}{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;pos_bp&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;allele1&#39;</span><span class="p">]</span><span class="si">}{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;allele2&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">study_df</span><span class="o">.</span><span class="n">iterrows</span><span class="p">()}</span>

        <span class="c1"># Filtering the rows in file2 based on the conditions</span>
        <span class="n">filtered</span> <span class="o">=</span> <span class="n">reference_df</span><span class="p">[</span>
            <span class="n">reference_df</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
                <span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;chr&#39;</span><span class="p">]</span><span class="si">}{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;rsid&#39;</span><span class="p">]</span><span class="si">}{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;pos_bp&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span> <span class="ow">in</span> <span class="n">a</span> <span class="ow">and</span> 
                    <span class="n">a</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;chr&#39;</span><span class="p">]</span><span class="si">}{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;rsid&#39;</span><span class="p">]</span><span class="si">}{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;pos_bp&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">{</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;allele1&#39;</span><span class="p">]</span><span class="si">}{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;allele2&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;allele2&#39;</span><span class="p">]</span><span class="si">}{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;allele1&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">}</span>
                <span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span>
            <span class="p">)</span>
        <span class="p">]</span>

        <span class="c1"># Save the second column of filtered rows to a file</span>
        <span class="n">filtered</span><span class="p">[</span><span class="s1">&#39;rsid&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_filename</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="k">return</span></div>

    
<div class="viewcode-block" id="GenomicOutlierAnalyzer">
<a class="viewcode-back" href="../../AncestryQC.html#ideal_genom_qc.AncestryQC.GenomicOutlierAnalyzer">[docs]</a>
<span class="k">class</span><span class="w"> </span><span class="nc">GenomicOutlierAnalyzer</span><span class="p">:</span>

<div class="viewcode-block" id="GenomicOutlierAnalyzer.__init__">
<a class="viewcode-back" href="../../AncestryQC.html#ideal_genom_qc.AncestryQC.GenomicOutlierAnalyzer.__init__">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">input_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">input_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">merged_file</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">reference_tags</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">output_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">output_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initialize GenomicOutlierAnalyzer object with input and output parameters.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        input_path : Path</span>
<span class="sd">            Path to input directory containing files to process</span>
<span class="sd">        input_name : str</span>
<span class="sd">            Name of input file </span>
<span class="sd">        merged_file : Path</span>
<span class="sd">            Path to merged genotype file</span>
<span class="sd">        reference_tags : Path</span>
<span class="sd">            Path to file containing reference population tags</span>
<span class="sd">        output_path : Path</span>
<span class="sd">            Path to output directory</span>
<span class="sd">        output_name : str</span>
<span class="sd">            Name for output files</span>

<span class="sd">        Attributes</span>
<span class="sd">        ----------</span>
<span class="sd">        einvectors : numpy.ndarray, None</span>
<span class="sd">            Principal component eigenvectors, initialized as None</span>
<span class="sd">        eigenvalues : numpy.ndarray, None</span>
<span class="sd">            Principal component eigenvalues, initialized as None</span>
<span class="sd">        ancestry_fails : list, None</span>
<span class="sd">            List of samples failing ancestry QC, initialized as None</span>
<span class="sd">        population_tags : pandas.DataFrame, None</span>
<span class="sd">            DataFrame containing population reference tags, initialized as None</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">merged_file</span> <span class="o">=</span> <span class="n">merged_file</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reference_tags</span> <span class="o">=</span> <span class="n">reference_tags</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">output_path</span><span class="o">=</span> <span class="n">output_path</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">=</span> <span class="n">output_name</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">input_path</span> <span class="o">=</span> <span class="n">input_path</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">input_name</span> <span class="o">=</span> <span class="n">input_name</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">einvectors</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">eigenvalues</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ancestry_fails</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">population_tags</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="k">pass</span></div>


<div class="viewcode-block" id="GenomicOutlierAnalyzer.execute_pca">
<a class="viewcode-back" href="../../AncestryQC.html#ideal_genom_qc.AncestryQC.GenomicOutlierAnalyzer.execute_pca">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">execute_pca</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">pca</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span> <span class="n">maf</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.01</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Perform Principal Component Analysis (PCA) on the genetic data using PLINK.</span>

<span class="sd">        This method executes PCA on the merged genetic data file, calculating the specified</span>
<span class="sd">        number of principal components. It automatically determines the optimal number of</span>
<span class="sd">        threads and memory allocation based on system resources.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        pca : int, default=10</span>
<span class="sd">            Number of principal components to calculate.</span>
<span class="sd">            Must be a positive integer.</span>
<span class="sd">        maf : float, default=0.01</span>
<span class="sd">            Minor allele frequency threshold for filtering variants.</span>
<span class="sd">            Must be between 0 and 0.5.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None</span>

<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        TypeError</span>
<span class="sd">            If pca is not an integer or maf is not a float</span>
<span class="sd">        ValueError</span>
<span class="sd">            If pca is not positive or maf is not between 0 and 0.5</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        The method creates two output files:</span>
<span class="sd">        - {output_name}-pca.eigenvec: Contains the eigenvectors (PC loadings)</span>
<span class="sd">        - {output_name}-pca.eigenval: Contains the eigenvalues</span>

<span class="sd">        The results are stored in self.einvectors and self.eigenvalues attributes.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">pca</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;pca should be an integer&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">pca</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;pca should be a positive integer&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">maf</span><span class="p">,</span> <span class="nb">float</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;maf should be a float&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">maf</span> <span class="o">&lt;</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">maf</span> <span class="o">&gt;</span> <span class="mf">0.5</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;maf should be a float between 0 and 0.5&quot;</span><span class="p">)</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;STEP: Performing principal component decomposition&quot;</span><span class="p">)</span>

        <span class="n">cpu_count</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">cpu_count</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">max_threads</span> <span class="o">=</span> <span class="n">cpu_count</span><span class="o">-</span><span class="mi">2</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">max_threads</span> <span class="o">=</span> <span class="mi">10</span>

        <span class="c1"># Get the virtual memory details</span>
        <span class="n">memory_info</span> <span class="o">=</span> <span class="n">psutil</span><span class="o">.</span><span class="n">virtual_memory</span><span class="p">()</span>
        <span class="n">available_memory_mb</span> <span class="o">=</span> <span class="n">memory_info</span><span class="o">.</span><span class="n">available</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1024</span> <span class="o">*</span> <span class="mi">1024</span><span class="p">)</span>
        <span class="n">memory</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="mi">2</span><span class="o">*</span><span class="n">available_memory_mb</span><span class="o">/</span><span class="mi">3</span><span class="p">,</span><span class="mi">0</span><span class="p">)</span>

        <span class="c1"># PLINK command: generate PCA for reference data</span>
        <span class="n">plink_cmd</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink --bfile </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">merged_file</span><span class="p">)</span><span class="si">}</span><span class="s2"> --keep-allele-order --maf </span><span class="si">{</span><span class="n">maf</span><span class="si">}</span><span class="s2"> --out </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_path</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-pca&#39;</span><span class="p">))</span><span class="si">}</span><span class="s2"> --pca </span><span class="si">{</span><span class="n">pca</span><span class="si">}</span><span class="s2"> --memory </span><span class="si">{</span><span class="n">memory</span><span class="si">}</span><span class="s2"> --threads </span><span class="si">{</span><span class="n">max_threads</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="c1"># execute PLINK command</span>
        <span class="n">shell_do</span><span class="p">(</span><span class="n">plink_cmd</span><span class="p">,</span> <span class="n">log</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">einvectors</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_path</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-pca.eigenvec&#39;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">eigenvalues</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_path</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-pca.eigenval&#39;</span><span class="p">)</span>

        <span class="k">return</span></div>

    
<div class="viewcode-block" id="GenomicOutlierAnalyzer.find_ancestry_outliers">
<a class="viewcode-back" href="../../AncestryQC.html#ideal_genom_qc.AncestryQC.GenomicOutlierAnalyzer.find_ancestry_outliers">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">find_ancestry_outliers</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ref_threshold</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span> <span class="n">stu_threshold</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span> <span class="n">reference_pop</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">num_pcs</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">fails_dir</span><span class="p">:</span> <span class="n">Path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">())</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Identifies ancestry outliers in the dataset based on PCA analysis.</span>
<span class="sd">        This method analyzes population structure using principal component analysis (PCA) and identifies</span>
<span class="sd">        samples that are potential ancestry outliers based on their distance from reference populations.</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ref_threshold : float</span>
<span class="sd">            Distance threshold for reference population samples</span>
<span class="sd">        stu_threshold : float</span>
<span class="sd">            Distance threshold for study population samples</span>
<span class="sd">        reference_pop : str</span>
<span class="sd">            Name of the reference population to compare against</span>
<span class="sd">        num_pcs : int, optional</span>
<span class="sd">            Number of principal components to use in the analysis (default is 2)</span>
<span class="sd">        fails_dir : Path, optional</span>
<span class="sd">            Directory path to save failed samples information (default is empty Path)</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None</span>
<span class="sd">            Results are stored in the ancestry_fails attribute</span>

<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        TypeError</span>
<span class="sd">            If parameters are not of the expected type</span>
<span class="sd">        ValueError</span>
<span class="sd">            If num_pcs is not a positive integer</span>
<span class="sd">        </span>
<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        The method requires:</span>
<span class="sd">        - A reference tags file with population information</span>
<span class="sd">        - An eigenvectors file from PCA analysis</span>
<span class="sd">        - Both files should be previously set in the class instance</span>
<span class="sd">        The results are saved in:</span>
<span class="sd">        - population_tags: CSV file with population assignments</span>
<span class="sd">        - ancestry_fails: List of samples identified as ancestry outliers</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ref_threshold</span><span class="p">,</span> <span class="p">(</span><span class="nb">float</span><span class="p">,</span> <span class="nb">int</span><span class="p">)):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;ref_threshold should be a float&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">stu_threshold</span><span class="p">,</span> <span class="p">(</span><span class="nb">float</span><span class="p">,</span> <span class="nb">int</span><span class="p">)):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;stu_threshold should be a float&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">reference_pop</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;reference_pop should be a string&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">num_pcs</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;num_pcs should be an integer&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">num_pcs</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;num_pcs should be a positive integer&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">fails_dir</span><span class="p">,</span> <span class="n">Path</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;fails_dir should be a Path object&quot;</span><span class="p">)</span>
        
        <span class="k">if</span> <span class="ow">not</span> <span class="n">fails_dir</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;STEP: Identifying ancestry outliers: `fails_dir` does not exist.&quot;</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;STEP: Identifying ancestry outliers: ancestry outliers will be saved in </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">output_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">fails_dir</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_path</span>
        
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;STEP: Identifying ancestry outliers&quot;</span><span class="p">)</span>

        <span class="n">df_tags</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_tags</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">usecols</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;#IID&#39;</span><span class="p">,</span> <span class="s1">&#39;SuperPop&#39;</span><span class="p">])</span>
        <span class="n">df_tags</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;0&#39;</span>
        <span class="n">df_tags</span> <span class="o">=</span> <span class="n">df_tags</span><span class="p">[[</span><span class="s1">&#39;ID&#39;</span><span class="p">,</span> <span class="s1">&#39;#IID&#39;</span><span class="p">,</span> <span class="s1">&#39;SuperPop&#39;</span><span class="p">]]</span>
        <span class="n">df_tags</span> <span class="o">=</span> <span class="n">df_tags</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;ID&#39;</span><span class="p">:</span> <span class="s1">&#39;ID1&#39;</span><span class="p">,</span> <span class="s1">&#39;#IID&#39;</span><span class="p">:</span> <span class="s1">&#39;ID2&#39;</span><span class="p">,</span> <span class="s1">&#39;SuperPop&#39;</span><span class="p">:</span> <span class="s1">&#39;SuperPop&#39;</span><span class="p">})</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">einvectors</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;einvectors is not set. Make sure execute_pca() is called before this method and completed successfully.&quot;</span><span class="p">)</span>

        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">einvectors</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="sa">r</span><span class="s2">&quot;\s+&quot;</span><span class="p">,</span><span class="n">engine</span><span class="o">=</span><span class="s1">&#39;python&#39;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;STEP: Identifying ancestry outliers: read eigenvec file&quot;</span><span class="p">)</span>

        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]]</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">{</span><span class="mi">0</span><span class="p">:</span> <span class="s1">&#39;ID1&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">:</span><span class="s1">&#39;ID2&#39;</span><span class="p">})</span>

        <span class="n">df</span><span class="p">[</span><span class="s1">&#39;ID2&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;ID2&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
        <span class="n">df</span><span class="p">[</span><span class="s1">&#39;ID1&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;ID1&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>

        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">df_tags</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;ID1&#39;</span><span class="p">,</span> <span class="s1">&#39;ID2&#39;</span><span class="p">],</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;left&#39;</span><span class="p">)</span>
        <span class="n">df</span><span class="p">[</span><span class="s1">&#39;SuperPop&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;SuperPop&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="s1">&#39;StPop&#39;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">output_path</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span> <span class="o">+</span> <span class="s1">&#39;pop_tags.csv&#39;</span><span class="p">)),</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">population_tags</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_path</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span> <span class="o">+</span> <span class="s1">&#39;pop_tags.csv&#39;</span><span class="p">)</span>

        <span class="c1"># filter samples who are ethnicity outliers</span>
        <span class="n">ancestry_fails</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_find_pca_fails</span><span class="p">(</span>
            <span class="n">output_path</span>  <span class="o">=</span> <span class="n">fails_dir</span><span class="p">,</span>
            <span class="n">df_tags</span>      <span class="o">=</span> <span class="n">df</span><span class="p">,</span>
            <span class="n">ref_threshold</span><span class="o">=</span> <span class="n">ref_threshold</span><span class="p">,</span>
            <span class="n">stu_threshold</span><span class="o">=</span> <span class="n">stu_threshold</span><span class="p">,</span>
            <span class="n">reference_pop</span><span class="o">=</span> <span class="n">reference_pop</span><span class="p">,</span>
            <span class="n">num_pcs</span>      <span class="o">=</span> <span class="n">num_pcs</span>
        <span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">ancestry_fails</span> <span class="o">=</span> <span class="n">ancestry_fails</span>

        <span class="k">return</span></div>

    
<div class="viewcode-block" id="GenomicOutlierAnalyzer.execute_drop_ancestry_outliers">
<a class="viewcode-back" href="../../AncestryQC.html#ideal_genom_qc.AncestryQC.GenomicOutlierAnalyzer.execute_drop_ancestry_outliers">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">execute_drop_ancestry_outliers</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">output_dir</span><span class="p">:</span> <span class="n">Path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">())</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Drop ancestry outliers from the study data by removing samples identified as ancestry outliers</span>
<span class="sd">        using PLINK command line tool.</span>
<span class="sd">        This method reads a file containing samples identified as ancestry outliers and creates new</span>
<span class="sd">        binary PLINK files excluding these samples.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        output_dir : Path, optional</span>
<span class="sd">            Directory where the cleaned files will be saved. If not provided or doesn&#39;t exist,</span>
<span class="sd">            files will be saved in self.output_path.</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None</span>

<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        TypeError</span>
<span class="sd">            If output_dir is not a Path object.</span>
<span class="sd">        </span>
<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        The method creates new PLINK binary files (.bed, .bim, .fam) with the suffix &#39;-ancestry-cleaned&#39;</span>
<span class="sd">        excluding the samples listed in self.ancestry_fails file.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;STEP: Dropping ancestry outliers from the study data&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="n">Path</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;output_dir should be a Path object&quot;</span><span class="p">)</span>
        
        <span class="k">if</span> <span class="ow">not</span> <span class="n">output_dir</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;STEP: Dropping ancestry outliers from the study data: `output_dir` does not exist.&quot;</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;STEP: Dropping ancestry outliers from the study data: ancestry outliers will be saved in </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">output_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">output_dir</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_path</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">ancestry_fails</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;ancestry_fails is not set. Make sure find_ancestry_outliers() is called before this method and completed successfully.&quot;</span><span class="p">)</span>

        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ancestry_fails</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;STEP: Dropping ancestry outliers from the study data: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">readlines</span><span class="p">())</span><span class="si">}</span><span class="s2"> samples identified as ancestry outliers&quot;</span><span class="p">)</span>

        <span class="c1"># create cleaned binary files</span>
        <span class="n">plink_cmd2</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink --bfile </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">input_path</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="p">)</span><span class="si">}</span><span class="s2"> --allow-no-sex --remove </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ancestry_fails</span><span class="p">)</span><span class="si">}</span><span class="s2"> --make-bed --out </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">output_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-ancestry-cleaned&#39;</span><span class="p">))</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="c1"># execute PLINK command</span>
        <span class="n">shell_do</span><span class="p">(</span><span class="n">plink_cmd2</span><span class="p">,</span> <span class="n">log</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="k">return</span></div>

    
<div class="viewcode-block" id="GenomicOutlierAnalyzer.draw_pca_plot">
<a class="viewcode-back" href="../../AncestryQC.html#ideal_genom_qc.AncestryQC.GenomicOutlierAnalyzer.draw_pca_plot">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">draw_pca_plot</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">reference_pop</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">aspect_ratio</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">Literal</span><span class="p">[</span><span class="s1">&#39;auto&#39;</span><span class="p">,</span> <span class="s1">&#39;equal&#39;</span><span class="p">],</span> <span class="nb">float</span><span class="p">],</span> <span class="n">exclude_outliers</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">plot_dir</span><span class="p">:</span> <span class="n">Path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(),</span> <span class="n">plot_name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;pca_plot.svg&#39;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Generate 2D and 3D PCA plots from eigenvector data and population tags.</span>
<span class="sd">        This method creates two PCA visualization plots:</span>
<span class="sd">        - A 2D scatter plot showing PC1 vs PC2 colored by super-population</span>
<span class="sd">        - A 3D scatter plot showing PC1 vs PC2 vs PC3 colored by super-population</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        plot_dir : Path, optional</span>
<span class="sd">            Directory path where plots will be saved. Defaults to current directory.</span>
<span class="sd">            If directory doesn&#39;t exist, plots will be saved in self.output_path</span>
<span class="sd">        plot_name : str, optional</span>
<span class="sd">            Base name for the plot files. Defaults to &#39;pca_plot.jpeg&#39;.</span>
<span class="sd">            Final filenames will be prefixed with &#39;2D-&#39; and &#39;3D-&#39;</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None</span>
<span class="sd">        </span>
<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        TypeError</span>
<span class="sd">            If plot_dir is not a Path object</span>
<span class="sd">            If plot_name is not a string</span>
<span class="sd">        </span>
<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        Requires the following class attributes to be set:</span>
<span class="sd">        - self.population_tags : Path to population tags file (tab-separated)</span>
<span class="sd">        - self.einvectors : Path to eigenvectors file (space-separated)</span>
<span class="sd">        - self.output_path : Path to output directory (used if plot_dir doesn&#39;t exist)</span>
<span class="sd">        The population tags file should contain columns &#39;ID1&#39;, &#39;ID2&#39;, and &#39;SuperPop&#39;</span>
<span class="sd">        The eigenvectors file should contain the principal components data</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;STEP: Generating PCA plots&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">plot_dir</span><span class="p">,</span> <span class="n">Path</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;plot_dir should be a Path object&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">plot_name</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;plot_name should be a string&quot;</span><span class="p">)</span>
        
        <span class="k">if</span> <span class="ow">not</span> <span class="n">plot_dir</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;STEP: Generating PCA plots: `plot_dir` does not exist.&#39;</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;STEP: Generating PCA plots: pca plots will be saved in </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">output_path</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="n">plot_dir</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_path</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">population_tags</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;population_tags is not set. Make sure find_ancestry_outliers() is called before this method and completed successfully.&quot;</span><span class="p">)</span>

        <span class="c1"># add population tags to pca output</span>
        <span class="n">df_tags</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">population_tags</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="n">df_tags</span><span class="p">[</span><span class="s1">&#39;ID1&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_tags</span><span class="p">[</span><span class="s1">&#39;ID1&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">einvectors</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;einvectors is not set. Make sure execute_pca() is called before this method and completed successfully.&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">eigenvalues</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;eigenvalues is not set. Make sure execute_pca() is called before this method and completed successfully.&quot;</span><span class="p">)</span>
        
        <span class="c1"># load .eigenval file and calculate variance explained by the first two PCs</span>
        <span class="n">df_eigenval</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">eigenvalues</span><span class="p">,</span>
            <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="n">sep</span>   <span class="o">=</span><span class="sa">r</span><span class="s2">&quot;\s+&quot;</span><span class="p">,</span>
            <span class="n">engine</span><span class="o">=</span><span class="s1">&#39;python&#39;</span>
        <span class="p">)</span>

        <span class="n">total_variance</span> <span class="o">=</span> <span class="n">df_eigenval</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
        <span class="n">pc1_var</span> <span class="o">=</span> <span class="n">df_eigenval</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">pc2_var</span> <span class="o">=</span> <span class="n">df_eigenval</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>

        <span class="n">pc1_var_perc</span> <span class="o">=</span> <span class="nb">round</span><span class="p">((</span><span class="n">pc1_var</span> <span class="o">/</span> <span class="n">total_variance</span><span class="p">)</span> <span class="o">*</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
        <span class="n">pc2_var_perc</span> <span class="o">=</span> <span class="nb">round</span><span class="p">((</span><span class="n">pc2_var</span> <span class="o">/</span> <span class="n">total_variance</span><span class="p">)</span> <span class="o">*</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>

        <span class="c1"># load .eigenvec file and keep the first three principal components</span>
        <span class="n">df_eigenvec</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">einvectors</span><span class="p">,</span>
            <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="n">sep</span>   <span class="o">=</span><span class="sa">r</span><span class="s2">&quot;\s+&quot;</span><span class="p">,</span>
            <span class="n">engine</span><span class="o">=</span><span class="s1">&#39;python&#39;</span>
        <span class="p">)</span>
        <span class="n">df_eigenvec</span> <span class="o">=</span> <span class="n">df_eigenvec</span><span class="p">[</span><span class="n">df_eigenvec</span><span class="o">.</span><span class="n">columns</span><span class="p">[:</span><span class="mi">5</span><span class="p">]]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">df_eigenvec</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ID1&#39;</span><span class="p">,</span> <span class="s1">&#39;ID2&#39;</span><span class="p">,</span> <span class="s1">&#39;pc_1&#39;</span><span class="p">,</span> <span class="s1">&#39;pc_2&#39;</span><span class="p">,</span> <span class="s1">&#39;pc_3&#39;</span><span class="p">]</span>
        <span class="n">df_eigenvec</span><span class="p">[</span><span class="s1">&#39;ID1&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_eigenvec</span><span class="p">[</span><span class="s1">&#39;ID1&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">exclude_outliers</span><span class="p">:</span>
            <span class="c1"># load ancestry outliers</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">ancestry_fails</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;ancestry_fails is not set. Make sure find_ancestry_outliers() is called before this method and completed successfully.&quot;</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;STEP: Generating PCA plots: excluding ancestry outliers&quot;</span><span class="p">)</span>

            <span class="n">df_outliers</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ancestry_fails</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="sa">r</span><span class="s1">&#39;\s+&#39;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">engine</span><span class="o">=</span><span class="s1">&#39;python&#39;</span><span class="p">)</span>
            <span class="n">df_outliers</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ID1&#39;</span><span class="p">,</span> <span class="s1">&#39;ID2&#39;</span><span class="p">]</span>
            <span class="n">df_outliers</span><span class="p">[</span><span class="s1">&#39;ID1&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_outliers</span><span class="p">[</span><span class="s1">&#39;ID1&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
            <span class="n">df_outliers</span><span class="p">[</span><span class="s1">&#39;ID2&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_outliers</span><span class="p">[</span><span class="s1">&#39;ID2&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>

            <span class="n">df_eigenvec</span> <span class="o">=</span> <span class="n">df_eigenvec</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">df_outliers</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;ID1&#39;</span><span class="p">,</span> <span class="s1">&#39;ID2&#39;</span><span class="p">],</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;left&#39;</span><span class="p">,</span> <span class="n">indicator</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">df_eigenvec</span> <span class="o">=</span> <span class="n">df_eigenvec</span><span class="p">[</span><span class="n">df_eigenvec</span><span class="p">[</span><span class="s1">&#39;_merge&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;left_only&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;_merge&#39;</span><span class="p">])</span>

            <span class="n">plot_name</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;no-outliers-</span><span class="si">{</span><span class="n">plot_name</span><span class="si">}</span><span class="s1">&#39;</span>

        <span class="c1"># merge to get data with tagged populations</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">df_eigenvec</span><span class="p">,</span> <span class="n">df_tags</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;ID1&#39;</span><span class="p">,</span> <span class="s1">&#39;ID2&#39;</span><span class="p">])</span>

        <span class="c1"># generates a 2D scatter plot</span>
        <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span><span class="mi">10</span><span class="p">))</span>
        <span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">df</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s1">&#39;pc_1&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;pc_2&#39;</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="s1">&#39;SuperPop&#39;</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">70</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="n">aspect_ratio</span><span class="p">,</span> <span class="n">adjustable</span><span class="o">=</span><span class="s1">&#39;datalim&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;PC_1 (</span><span class="si">{</span><span class="n">pc1_var_perc</span><span class="si">}</span><span class="s1">%)&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;PC_2 (</span><span class="si">{</span><span class="n">pc2_var_perc</span><span class="si">}</span><span class="s1">%)&#39;</span><span class="p">)</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">plot_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;2D-aspect-</span><span class="si">{</span><span class="n">aspect_ratio</span><span class="si">}</span><span class="s1">-</span><span class="si">{</span><span class="n">plot_name</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">400</span><span class="p">)</span>

        <span class="n">fig</span><span class="o">.</span><span class="n">clf</span><span class="p">()</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

        <span class="n">fig3</span><span class="p">,</span> <span class="n">ax3</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span><span class="mi">10</span><span class="p">))</span>
        <span class="n">df_zoom</span> <span class="o">=</span> <span class="n">df</span><span class="p">[(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;SuperPop&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;StPop&#39;</span><span class="p">)</span> <span class="o">|</span> <span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;SuperPop&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">reference_pop</span><span class="p">)]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">df_zoom</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s1">&#39;pc_1&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;pc_2&#39;</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="s1">&#39;SuperPop&#39;</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax3</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">70</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="n">aspect_ratio</span><span class="p">,</span> <span class="n">adjustable</span><span class="o">=</span><span class="s1">&#39;datalim&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;PC_1 (</span><span class="si">{</span><span class="n">pc1_var_perc</span><span class="si">}</span><span class="s1">%)&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;PC_2 (</span><span class="si">{</span><span class="n">pc2_var_perc</span><span class="si">}</span><span class="s1">%)&#39;</span><span class="p">)</span>
        <span class="n">fig3</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="n">fig3</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">plot_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;2D-zoom-aspect-</span><span class="si">{</span><span class="n">aspect_ratio</span><span class="si">}</span><span class="s1">-</span><span class="si">{</span><span class="n">plot_name</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">400</span><span class="p">)</span>

        <span class="c1"># generates a 3D scatter plot</span>
        <span class="n">fig2</span><span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span>
        <span class="n">ax</span>  <span class="o">=</span> <span class="n">fig2</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">111</span><span class="p">,</span> <span class="n">projection</span><span class="o">=</span><span class="s1">&#39;3d&#39;</span><span class="p">)</span>

        <span class="n">grouped</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;SuperPop&#39;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">s</span><span class="p">,</span> <span class="n">group</span> <span class="ow">in</span> <span class="n">grouped</span><span class="p">:</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span>
                <span class="n">group</span><span class="p">[</span><span class="s1">&#39;pc_1&#39;</span><span class="p">],</span>
                <span class="n">group</span><span class="p">[</span><span class="s1">&#39;pc_2&#39;</span><span class="p">],</span>
                <span class="n">group</span><span class="p">[</span><span class="s1">&#39;pc_3&#39;</span><span class="p">],</span>
                <span class="n">label</span><span class="o">=</span><span class="n">s</span>
            <span class="p">)</span>

        <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">plot_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;3D-</span><span class="si">{</span><span class="n">plot_name</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">400</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

        <span class="k">return</span></div>

    
    <span class="k">def</span><span class="w"> </span><span class="nf">_set_population_tags</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">psam_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">study_fam_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Sets population tags for genetic data by combining information from a PSAM file and a study FAM file.</span>

<span class="sd">        This method processes population information from reference data (PSAM file) and study data (FAM file), </span>
<span class="sd">        combining them into a single DataFrame with consistent column naming and structure.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        psam_path : Path</span>
<span class="sd">            Path to the PSAM file containing reference population information.</span>
<span class="sd">        study_fam_path : Path</span>
<span class="sd">            Path to the FAM file containing study individual IDs.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        pd.DataFrame</span>
<span class="sd">            Combined DataFrame containing:</span>
<span class="sd">                - ID1: Family or group identifier (0 for reference data)</span>
<span class="sd">                - ID2: Individual identifier</span>
<span class="sd">                - SuperPop: Population tag (&#39;StPop&#39; for study individuals, actual population for reference data)</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        The PSAM file should contain at least &#39;#IID&#39; and &#39;SuperPop&#39; columns.</span>
<span class="sd">        The FAM file should be space-separated with no header.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Read population information from the .psam file</span>
        <span class="n">df_psam</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
            <span class="n">psam_path</span><span class="p">,</span>
            <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span>
            <span class="n">usecols</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;#IID&#39;</span><span class="p">,</span> <span class="s1">&#39;SuperPop&#39;</span><span class="p">]</span>
        <span class="p">)</span>

        <span class="c1"># Set an ID column and rename columns for consistency</span>
        <span class="n">df_psam</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">df_psam</span> <span class="o">=</span> <span class="n">df_psam</span><span class="p">[[</span><span class="s1">&#39;ID&#39;</span><span class="p">,</span> <span class="s1">&#39;#IID&#39;</span><span class="p">,</span> <span class="s1">&#39;SuperPop&#39;</span><span class="p">]]</span>
        <span class="n">df_psam</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ID1&#39;</span><span class="p">,</span> <span class="s1">&#39;ID2&#39;</span><span class="p">,</span> <span class="s1">&#39;SuperPop&#39;</span><span class="p">]</span>

        <span class="c1"># read individual IDs from the study .fam file</span>
        <span class="n">df_fam</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
            <span class="n">study_fam_path</span><span class="p">,</span>
            <span class="n">sep</span><span class="o">=</span><span class="s1">&#39; &#39;</span><span class="p">,</span>
            <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="n">index_col</span><span class="o">=</span><span class="kc">False</span>
        <span class="p">)</span>

        <span class="c1"># select relevant columns, assign a placeholder population tag, and rename columns</span>
        <span class="n">df_fam</span> <span class="o">=</span> <span class="n">df_fam</span><span class="p">[</span><span class="n">df_fam</span><span class="o">.</span><span class="n">columns</span><span class="p">[:</span><span class="mi">2</span><span class="p">]]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">df_fam</span><span class="p">[</span><span class="s1">&#39;SuperPop&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;StPop&#39;</span>
        <span class="n">df_fam</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ID1&#39;</span><span class="p">,</span> <span class="s1">&#39;ID2&#39;</span><span class="p">,</span> <span class="s1">&#39;SuperPop&#39;</span><span class="p">]</span>

        <span class="c1"># concatenate the two DataFrames to merge the information</span>
        <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">df_fam</span><span class="p">,</span> <span class="n">df_psam</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    
    <span class="k">def</span><span class="w"> </span><span class="nf">_find_pca_fails</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">output_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">df_tags</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">ref_threshold</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span> <span class="n">stu_threshold</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span> <span class="n">reference_pop</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">num_pcs</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">2</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Path</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Identifies ancestry outliers based on PCA results using two thresholds:</span>
<span class="sd">        one for reference population and another for study population.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        output_path : Path</span>
<span class="sd">            Path where the output file will be saved</span>
<span class="sd">        df_tags : pd.DataFrame</span>
<span class="sd">            DataFrame containing subject IDs and population tags</span>
<span class="sd">        ref_threshold : int</span>
<span class="sd">            Number of standard deviations from reference population mean to consider a subject as outlier</span>
<span class="sd">        stu_threshold : int</span>
<span class="sd">            Number of standard deviations from study population mean to consider a subject as outlier</span>
<span class="sd">        reference_pop : str</span>
<span class="sd">            Reference population name as it appears in df_tags</span>
<span class="sd">        num_pcs : int, optional</span>
<span class="sd">            Number of principal components to use in the analysis (default is 2)</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        str</span>
<span class="sd">            Path to the output file containing the IDs of subjects identified as ancestry outliers</span>

<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        TypeError</span>
<span class="sd">            If ref_threshold, stu_threshold are not numeric</span>
<span class="sd">            If reference_pop is not a string</span>
<span class="sd">            If num_pcs is not an integer</span>
<span class="sd">        ValueError</span>
<span class="sd">            If ref_threshold, stu_threshold are not positive</span>
<span class="sd">            If num_pcs is less than 1</span>
<span class="sd">            If num_pcs is greater than available PCs in eigenvec file</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        The method identifies outliers that deviate significantly from both:</span>
<span class="sd">        1. The reference population mean (by ref_threshold standard deviations)</span>
<span class="sd">        2. The study population mean (by stu_threshold standard deviations)</span>
<span class="sd">        Only subjects that are outliers in both criteria are included in the final output.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ref_threshold</span><span class="p">,</span> <span class="p">(</span><span class="nb">float</span><span class="p">,</span> <span class="nb">int</span><span class="p">)):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;ref_threshold should be an integer or float value&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">stu_threshold</span><span class="p">,</span> <span class="p">(</span><span class="nb">float</span><span class="p">,</span> <span class="nb">int</span><span class="p">)):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;stu_threshold should be an integer or float value&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">stu_threshold</span><span class="o">&lt;=</span><span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;stu_threshold should be a positive value&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">ref_threshold</span><span class="o">&lt;=</span><span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;ref_threshold should be a positive value&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">reference_pop</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;reference_pop should be a string&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">num_pcs</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;num_pcs should be an integer value&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">num_pcs</span><span class="o">&lt;</span><span class="mi">1</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;num_pcs should be a positive integer&quot;</span><span class="p">)</span>

        <span class="c1"># filters reference subjects</span>
        <span class="n">mask1</span> <span class="o">=</span> <span class="p">(</span><span class="n">df_tags</span><span class="p">[</span><span class="s1">&#39;SuperPop&#39;</span><span class="p">]</span><span class="o">==</span><span class="n">reference_pop</span><span class="p">)</span>
        <span class="c1"># filters subjects from study data</span>
        <span class="n">mask2</span> <span class="o">=</span> <span class="p">(</span><span class="n">df_tags</span><span class="p">[</span><span class="s1">&#39;SuperPop&#39;</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;StPop&#39;</span><span class="p">)</span>

        <span class="c1"># generates two data frames with filtered subjects</span>
        <span class="n">df_ref</span> <span class="o">=</span> <span class="n">df_tags</span><span class="p">[</span><span class="n">mask1</span><span class="p">]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">df_stu</span> <span class="o">=</span> <span class="n">df_tags</span><span class="p">[</span><span class="n">mask2</span><span class="p">]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">einvectors</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;einvectors is not set. Make sure execute_pca() is called before this method and completed successfully.&quot;</span><span class="p">)</span>

        <span class="c1"># read .eigenvec file</span>
        <span class="n">df_eigenvec</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">einvectors</span><span class="p">,</span>
            <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="n">sep</span>   <span class="o">=</span><span class="sa">r</span><span class="s2">&quot;\s+&quot;</span><span class="p">,</span>
            <span class="n">engine</span><span class="o">=</span><span class="s1">&#39;python&#39;</span>
        <span class="p">)</span>

        <span class="k">if</span> <span class="n">num_pcs</span><span class="o">&gt;</span><span class="n">df_eigenvec</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">-</span><span class="mi">2</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;num_pcs should be less than or equal to the number of principal components in the .eigenvec file&quot;</span><span class="p">)</span>
        
        <span class="n">df_eigenvec</span> <span class="o">=</span> <span class="n">df_eigenvec</span><span class="p">[</span><span class="n">df_eigenvec</span><span class="o">.</span><span class="n">columns</span><span class="p">[:</span><span class="mi">2</span><span class="o">+</span><span class="n">num_pcs</span><span class="p">]]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="c1"># renames columns for consistency</span>
        <span class="n">new_col_names</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">2</span><span class="o">+</span><span class="n">num_pcs</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">k</span><span class="o">&lt;</span><span class="mi">2</span><span class="p">:</span>
                <span class="n">new_col_names</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;ID</span><span class="si">{</span><span class="n">k</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">new_col_names</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;pc_</span><span class="si">{</span><span class="n">k</span><span class="o">-</span><span class="mi">1</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">df_eigenvec</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="n">new_col_names</span>

        <span class="n">df_eigenvec</span><span class="p">[</span><span class="s1">&#39;ID1&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_eigenvec</span><span class="p">[</span><span class="s1">&#39;ID1&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
        <span class="n">df_eigenvec</span><span class="p">[</span><span class="s1">&#39;ID2&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_eigenvec</span><span class="p">[</span><span class="s1">&#39;ID2&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>

        <span class="c1"># merge filtered subjects with its principal components</span>
        <span class="n">df_ref</span> <span class="o">=</span> <span class="n">df_ref</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">df_eigenvec</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;ID1&#39;</span><span class="p">,</span> <span class="s1">&#39;ID2&#39;</span><span class="p">])</span>\
            <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;SuperPop&#39;</span><span class="p">],</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">df_stu</span> <span class="o">=</span> <span class="n">df_stu</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">df_eigenvec</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;ID1&#39;</span><span class="p">,</span> <span class="s1">&#39;ID2&#39;</span><span class="p">])</span>\
            <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;SuperPop&#39;</span><span class="p">],</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="c1"># computes mean and standard deviation by columns in reference data</span>
        <span class="n">mean_ref</span><span class="o">=</span> <span class="n">df_ref</span><span class="p">[</span><span class="n">df_ref</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">2</span><span class="p">:]]</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
        <span class="n">std_ref</span> <span class="o">=</span> <span class="n">df_ref</span><span class="p">[</span><span class="n">df_ref</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">2</span><span class="p">:]]</span><span class="o">.</span><span class="n">std</span><span class="p">()</span>

        <span class="c1"># creates empty data frame</span>
        <span class="n">outliers_1</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="n">df_ref</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
        <span class="n">outliers_1</span><span class="p">[</span><span class="n">df_stu</span><span class="o">.</span><span class="n">columns</span><span class="p">[:</span><span class="mi">2</span><span class="p">]]</span> <span class="o">=</span> <span class="n">df_stu</span><span class="p">[</span><span class="n">df_stu</span><span class="o">.</span><span class="n">columns</span><span class="p">[:</span><span class="mi">2</span><span class="p">]]</span>

        <span class="c1"># identifies subjects with more than `ref_threshold` std deviations from the reference mean</span>
        <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">outliers_1</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">2</span><span class="p">:]:</span>
            <span class="n">outliers_1</span><span class="p">[</span><span class="n">col</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">df_stu</span><span class="p">[</span><span class="n">col</span><span class="p">]</span> <span class="o">-</span> <span class="n">mean_ref</span><span class="p">[</span><span class="n">col</span><span class="p">])</span> <span class="o">&gt;</span> <span class="n">ref_threshold</span><span class="o">*</span><span class="n">std_ref</span><span class="p">[</span><span class="n">col</span><span class="p">])</span>

        <span class="n">outliers_1</span><span class="p">[</span><span class="s1">&#39;is_out&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">outliers_1</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span><span class="mi">2</span><span class="p">:],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span> <span class="o">&gt;</span><span class="mi">0</span><span class="p">)</span>

        <span class="n">df_1</span> <span class="o">=</span> <span class="n">outliers_1</span><span class="p">[</span><span class="n">outliers_1</span><span class="p">[</span><span class="s1">&#39;is_out&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)[[</span><span class="s1">&#39;ID1&#39;</span><span class="p">,</span> <span class="s1">&#39;ID2&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="c1"># computes mean and standard deviation by columns in study data</span>
        <span class="n">mean_stu</span><span class="o">=</span> <span class="n">df_stu</span><span class="p">[</span><span class="n">df_stu</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">2</span><span class="p">:]]</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
        <span class="n">std_stu</span> <span class="o">=</span> <span class="n">df_stu</span><span class="p">[</span><span class="n">df_stu</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">2</span><span class="p">:]]</span><span class="o">.</span><span class="n">std</span><span class="p">()</span>

        <span class="c1"># creates empty data frame</span>
        <span class="n">outliers_2</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="n">df_ref</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
        <span class="n">outliers_2</span><span class="p">[</span><span class="n">df_stu</span><span class="o">.</span><span class="n">columns</span><span class="p">[:</span><span class="mi">2</span><span class="p">]]</span> <span class="o">=</span> <span class="n">df_stu</span><span class="p">[</span><span class="n">df_stu</span><span class="o">.</span><span class="n">columns</span><span class="p">[:</span><span class="mi">2</span><span class="p">]]</span>

        <span class="c1"># identifies subjects with more than `stu_threshold` std deviation from the study mean</span>
        <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">outliers_2</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">2</span><span class="p">:]:</span>
            <span class="n">outliers_2</span><span class="p">[</span><span class="n">col</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">df_stu</span><span class="p">[</span><span class="n">col</span><span class="p">]</span> <span class="o">-</span> <span class="n">mean_stu</span><span class="p">[</span><span class="n">col</span><span class="p">])</span> <span class="o">&gt;</span> <span class="n">stu_threshold</span><span class="o">*</span><span class="n">std_stu</span><span class="p">[</span><span class="n">col</span><span class="p">])</span>

        <span class="n">outliers_2</span><span class="p">[</span><span class="s1">&#39;is_out&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">outliers_2</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span><span class="mi">2</span><span class="p">:],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span> <span class="o">&gt;</span><span class="mi">0</span><span class="p">)</span>

        <span class="n">df_2</span> <span class="o">=</span> <span class="n">outliers_2</span><span class="p">[</span><span class="n">outliers_2</span><span class="p">[</span><span class="s1">&#39;is_out&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)[[</span><span class="s1">&#39;ID1&#39;</span><span class="p">,</span> <span class="s1">&#39;ID2&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">df_1</span><span class="p">,</span> <span class="n">df_2</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;ID1&#39;</span><span class="p">,</span> <span class="s1">&#39;ID2&#39;</span><span class="p">])</span>

        <span class="n">ancestry_fails</span> <span class="o">=</span> <span class="n">output_path</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span> <span class="o">+</span> <span class="s1">&#39;_fail-ancestry-qc.txt&#39;</span><span class="p">)</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;STEP: Identifying ancestry outliers: </span><span class="si">{</span><span class="n">df</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2"> samples identified as ancestry outliers&quot;</span><span class="p">)</span>

        <span class="c1"># save samples considered as ethnicity outliers</span>
        <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span>
            <span class="n">ancestry_fails</span><span class="p">,</span>
            <span class="n">header</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="n">index</span> <span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="n">sep</span>   <span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">ancestry_fails</span></div>


<div class="viewcode-block" id="AncestryQC">
<a class="viewcode-back" href="../../AncestryQC.html#ideal_genom_qc.AncestryQC.AncestryQC">[docs]</a>
<span class="k">class</span><span class="w"> </span><span class="nc">AncestryQC</span><span class="p">:</span>

<div class="viewcode-block" id="AncestryQC.__init__">
<a class="viewcode-back" href="../../AncestryQC.html#ideal_genom_qc.AncestryQC.AncestryQC.__init__">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">input_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">input_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">output_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">output_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">high_ld_file</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">reference_files</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(),</span> <span class="n">recompute_merge</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">built</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;38&#39;</span><span class="p">,</span> <span class="n">rename_snps</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initialize AncestryQC class.</span>
<span class="sd">        This class performs ancestry quality control analysis on genetic data by merging it with 1000 Genomes reference data</span>
<span class="sd">        and running principal component analysis.</span>

<span class="sd">        Parameters:</span>
<span class="sd">        -----------</span>
<span class="sd">        input_path: Path </span>
<span class="sd">            Path to directory containing input files</span>
<span class="sd">        input_name: str </span>
<span class="sd">            Base name of input files (without extension) </span>
<span class="sd">        output_path: Path </span>
<span class="sd">            Path to directory where output files will be saved</span>
<span class="sd">        output_name: str </span>
<span class="sd">            Base name for output files</span>
<span class="sd">        high_ld_file: Path </span>
<span class="sd">            Path to file containing high LD regions to exclude</span>
<span class="sd">        reference_files: dict (optional) </span>
<span class="sd">            Dictionary with paths to reference files. Must contain &#39;bim&#39;, &#39;bed&#39;, &#39;fam&#39; and &#39;psam&#39; keys. </span>
<span class="sd">            If not provided, will download 1000 Genomes reference files. Defaults to empty dict.</span>
<span class="sd">        recompute_merge: bool (optional): </span>
<span class="sd">            Whether to recompute merge with reference even if merged files exist. Defaults to True.</span>
<span class="sd">        built: str (optional) </span>
<span class="sd">            Genome build version, either &#39;37&#39; or &#39;38&#39;. Defaults to &#39;38&#39;.</span>
<span class="sd">        rename_snps: bool (optional): </span>
<span class="sd">            Whether to rename SNPs to avoid duplicates during merge. Defaults to False.</span>
<span class="sd">        </span>
<span class="sd">        Raises:</span>
<span class="sd">        -------</span>
<span class="sd">            TypeError: If input arguments are not of expected types</span>
<span class="sd">            ValueError: If built is not &#39;37&#39; or &#39;38&#39;</span>
<span class="sd">            FileNotFoundError: If input_path or output_path do not exist</span>
<span class="sd">        </span>
<span class="sd">        Note:</span>
<span class="sd">        -----</span>
<span class="sd">            Creates the following directory structure under output_path:</span>
<span class="sd">            - ancestry_qc_results/</span>
<span class="sd">                - merging/</span>
<span class="sd">                - plots/ </span>
<span class="sd">                - fail_samples/</span>
<span class="sd">                - clean_files/</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">input_path</span><span class="p">,</span> <span class="n">Path</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;input_path should be a Path object&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="n">Path</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;output_path should be a Path object&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">high_ld_file</span><span class="p">,</span> <span class="n">Path</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;high_ld_regions should be a Path object&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">reference_files</span><span class="p">,</span> <span class="nb">dict</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;reference_files should be a dictionary&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">input_name</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span> 
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;input_name should be a string&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">output_name</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;output_name should be a string&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">recompute_merge</span><span class="p">,</span> <span class="nb">bool</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;recompute_merge should be a boolean&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">built</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;built should be a string&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">built</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;37&#39;</span><span class="p">,</span> <span class="s1">&#39;38&#39;</span><span class="p">]:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;built should be either &#39;37&#39; or &#39;38&#39;&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">rename_snps</span><span class="p">,</span> <span class="nb">bool</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;rename_snps should be a boolean&quot;</span><span class="p">)</span>
        
        <span class="k">if</span> <span class="ow">not</span> <span class="n">input_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="s2">&quot;input_path does not exist&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">output_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="s2">&quot;output_path does not exist&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">high_ld_file</span><span class="o">.</span><span class="n">is_file</span><span class="p">():</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;High LD file not found at </span><span class="si">{</span><span class="n">high_ld_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;High LD file will be fetched from the package&#39;</span><span class="p">)</span>
            
            <span class="n">ld_fetcher</span> <span class="o">=</span> <span class="n">FetcherLDRegions</span><span class="p">(</span><span class="n">built</span><span class="o">=</span><span class="n">built</span><span class="p">)</span>
            <span class="n">ld_fetcher</span><span class="o">.</span><span class="n">get_ld_regions</span><span class="p">()</span>

            <span class="n">ld_regions</span> <span class="o">=</span> <span class="n">ld_fetcher</span><span class="o">.</span><span class="n">ld_regions</span>
            <span class="k">if</span> <span class="n">ld_regions</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Failed to fetch high LD regions file&quot;</span><span class="p">)</span>
            <span class="n">high_ld_file</span> <span class="o">=</span> <span class="n">ld_regions</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;High LD file fetched from the package and saved at </span><span class="si">{</span><span class="n">high_ld_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        
        <span class="bp">self</span><span class="o">.</span><span class="n">input_path</span> <span class="o">=</span> <span class="n">input_path</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">input_name</span> <span class="o">=</span> <span class="n">input_name</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">output_path</span><span class="o">=</span> <span class="n">output_path</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">=</span> <span class="n">output_name</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reference_files</span> <span class="o">=</span> <span class="n">reference_files</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">high_ld_regions</span> <span class="o">=</span> <span class="n">high_ld_file</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">recompute_merge</span> <span class="o">=</span> <span class="n">recompute_merge</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">built</span> <span class="o">=</span> <span class="n">built</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">rename_snps</span> <span class="o">=</span> <span class="n">rename_snps</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">reference_files</span><span class="p">:</span>

            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No reference files provided. Fetching 1000 Genomes reference data for built </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">built</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="n">fetcher</span> <span class="o">=</span> <span class="n">Fetcher1000Genome</span><span class="p">(</span><span class="n">built</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">built</span><span class="p">)</span>
            <span class="n">fetcher</span><span class="o">.</span><span class="n">get_1000genomes</span><span class="p">()</span>
            <span class="n">fetcher</span><span class="o">.</span><span class="n">get_1000genomes_binaries</span><span class="p">()</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">reference_files</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s1">&#39;bim&#39;</span><span class="p">:</span> <span class="n">fetcher</span><span class="o">.</span><span class="n">bim_file</span><span class="p">,</span>
                <span class="s1">&#39;bed&#39;</span><span class="p">:</span> <span class="n">fetcher</span><span class="o">.</span><span class="n">bed_file</span><span class="p">,</span>
                <span class="s1">&#39;fam&#39;</span><span class="p">:</span> <span class="n">fetcher</span><span class="o">.</span><span class="n">fam_file</span><span class="p">,</span>
                <span class="s1">&#39;psam&#39;</span><span class="p">:</span> <span class="n">fetcher</span><span class="o">.</span><span class="n">psam_file</span>
            <span class="p">}</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_path</span> <span class="o">/</span> <span class="s1">&#39;ancestry_qc_results&#39;</span> 
        <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">merging_dir</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="s1">&#39;merging&#39;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">merging_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">plots_dir</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="s1">&#39;plots&#39;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">plots_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">fail_samples_dir</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="s1">&#39;fail_samples&#39;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">fail_samples_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">clean_files</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="s1">&#39;clean_files&#39;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">clean_files</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="k">pass</span></div>


<div class="viewcode-block" id="AncestryQC.merge_reference_study">
<a class="viewcode-back" href="../../AncestryQC.html#ideal_genom_qc.AncestryQC.AncestryQC.merge_reference_study">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">merge_reference_study</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ind_pair</span><span class="p">:</span> <span class="nb">list</span> <span class="o">=</span> <span class="p">[</span><span class="mi">50</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mf">0.2</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Merge reference and study data by applying quality control filters and merging steps.</span>
<span class="sd">        This method performs a series of quality control steps to merge study data with reference data:</span>
<span class="sd">        1. Filters problematic SNPs</span>
<span class="sd">        2. Performs LD pruning</span>
<span class="sd">        3. Fixes chromosome mismatches</span>
<span class="sd">        4. Fixes position mismatches  </span>
<span class="sd">        5. Fixes allele flips</span>
<span class="sd">        6. Removes remaining mismatches</span>
<span class="sd">        7. Merges the datasets</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ind_pair : list, default [50, 5, 0.2]</span>
<span class="sd">            Parameters for LD pruning: [window size, step size, r2 threshold]</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None</span>
<span class="sd">        </span>
<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        If recompute_merge is False, the method will skip the merging process and expect</span>
<span class="sd">        merged data to already exist in the merging directory.</span>
<span class="sd">        </span>
<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        TypeError</span>
<span class="sd">            If ind_pair is not a list</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ind_pair</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;ind_pair should be a list&quot;</span><span class="p">)</span>
        
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">recompute_merge</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;STEP: Merging study and reference data: recompute_merge is set to False. Skipping merging step&quot;</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;STEP: Merging study and reference data: merged data is expected to be in </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">merging_dir</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">return</span>

        <span class="n">rgm</span> <span class="o">=</span> <span class="n">ReferenceGenomicMerger</span><span class="p">(</span>
            <span class="n">input_path</span><span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">input_path</span><span class="p">,</span>
            <span class="n">input_name</span><span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="p">,</span>
            <span class="n">output_path</span><span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">merging_dir</span><span class="p">,</span> 
            <span class="n">output_name</span><span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="p">,</span>
            <span class="n">high_ld_regions</span> <span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">high_ld_regions</span><span class="p">,</span> 
            <span class="n">reference_files</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_files</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="n">rgm</span><span class="o">.</span><span class="n">execute_rename_snpid</span><span class="p">()</span>
        <span class="n">rgm</span><span class="o">.</span><span class="n">execute_filter_prob_snps</span><span class="p">()</span>
        <span class="n">rgm</span><span class="o">.</span><span class="n">execute_ld_pruning</span><span class="p">(</span><span class="n">ind_pair</span><span class="o">=</span><span class="n">ind_pair</span><span class="p">)</span>
        <span class="n">rgm</span><span class="o">.</span><span class="n">execute_fix_chromosome_mismatch</span><span class="p">()</span>
        <span class="n">rgm</span><span class="o">.</span><span class="n">execute_fix_possition_mismatch</span><span class="p">()</span>
        <span class="n">rgm</span><span class="o">.</span><span class="n">execute_fix_allele_flip</span><span class="p">()</span>
        <span class="n">rgm</span><span class="o">.</span><span class="n">execute_remove_mismatches</span><span class="p">()</span>
        <span class="n">rgm</span><span class="o">.</span><span class="n">execute_merge_data</span><span class="p">()</span>

        <span class="k">return</span></div>

    
    <span class="k">def</span><span class="w"> </span><span class="nf">_clean_merging_dir</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Cleans up the merging directory by removing unnecessary files.</span>
<span class="sd">        This method removes all files in the merging directory except:</span>
<span class="sd">        - Files containing &#39;-merged&#39; in their name</span>
<span class="sd">        - Log files with &#39;.log&#39; extension</span>
<span class="sd">        The cleanup helps manage disk space and removes intermediate files that are no longer needed</span>
<span class="sd">        after the merging process is complete.</span>
<span class="sd">        </span>
<span class="sd">        Returns:</span>
<span class="sd">        --------</span>

<span class="sd">            None</span>
<span class="sd">        &quot;&quot;&quot;</span>
        

        <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">merging_dir</span><span class="o">.</span><span class="n">iterdir</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">file</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="ow">and</span> <span class="s1">&#39;-merged&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">file</span><span class="o">.</span><span class="n">name</span> <span class="ow">and</span> <span class="n">file</span><span class="o">.</span><span class="n">suffix</span> <span class="o">!=</span> <span class="s1">&#39;.log&#39;</span><span class="p">:</span>
                <span class="n">file</span><span class="o">.</span><span class="n">unlink</span><span class="p">()</span>

        <span class="k">return</span>
    
<div class="viewcode-block" id="AncestryQC.run_pca">
<a class="viewcode-back" href="../../AncestryQC.html#ideal_genom_qc.AncestryQC.AncestryQC.run_pca">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">run_pca</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ref_population</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">pca</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span> <span class="n">maf</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.01</span><span class="p">,</span> <span class="n">num_pca</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span> <span class="n">ref_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">4</span><span class="p">,</span> <span class="n">stu_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">4</span><span class="p">,</span> <span class="n">aspect_ratio</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">Literal</span><span class="p">[</span><span class="s1">&#39;auto&#39;</span><span class="p">,</span> <span class="s1">&#39;equal&#39;</span><span class="p">],</span> <span class="nb">float</span><span class="p">]</span><span class="o">=</span><span class="s1">&#39;equal&#39;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Performs Principal Component Analysis (PCA) on genetic data and identifies ancestry outliers.</span>

<span class="sd">        This method executes a complete PCA workflow including:</span>
<span class="sd">        1. Running the PCA analysis</span>
<span class="sd">        2. Identifying ancestry outliers</span>
<span class="sd">        3. Removing identified outliers</span>
<span class="sd">        4. Generating PCA plots</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ref_population : str</span>
<span class="sd">            Reference population identifier for ancestry comparison</span>
<span class="sd">        pca : int, optional</span>
<span class="sd">            Number of principal components to calculate (default=10)</span>
<span class="sd">        maf : float, optional</span>
<span class="sd">            Minor allele frequency threshold for filtering (default=0.01)</span>
<span class="sd">        num_pca : int, optional</span>
<span class="sd">            Number of principal components to use in outlier detection (default=10)</span>
<span class="sd">        ref_threshold : float, optional</span>
<span class="sd">            Threshold for reference population outlier detection (default=4)</span>
<span class="sd">        stu_threshold : float, optional</span>
<span class="sd">            Threshold for study population outlier detection (default=4)</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None</span>
<span class="sd">            Results are saved to specified output directories</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        The method uses the GenomicOutlierAnalyzer class to perform the analysis and </span>
<span class="sd">        saves results in the directories specified during class initialization.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Make sure the reference tag path is valid before creating the analyzer</span>
        <span class="k">if</span> <span class="s1">&#39;psam&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_files</span> <span class="ow">or</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_files</span><span class="p">[</span><span class="s1">&#39;psam&#39;</span><span class="p">],</span> <span class="n">Path</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Reference files dictionary must contain a valid &#39;psam&#39; Path&quot;</span><span class="p">)</span>
        
        <span class="n">goa</span> <span class="o">=</span> <span class="n">GenomicOutlierAnalyzer</span><span class="p">(</span>
            <span class="n">input_path</span><span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">input_path</span><span class="p">,</span> 
            <span class="n">input_name</span><span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="p">,</span>
            <span class="n">merged_file</span><span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">merging_dir</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span> <span class="o">+</span> <span class="s1">&#39;-merged&#39;</span><span class="p">),</span>
            <span class="n">reference_tags</span><span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_files</span><span class="p">[</span><span class="s1">&#39;psam&#39;</span><span class="p">],</span>
            <span class="n">output_path</span><span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="p">,</span> 
            <span class="n">output_name</span><span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_name</span>
        <span class="p">)</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;STEP: Running PCA analysis: `ref_population` = </span><span class="si">{</span><span class="n">ref_population</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;STEP: Running PCA analysis: `pca` = </span><span class="si">{</span><span class="n">pca</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;STEP: Running PCA analysis: `maf` = </span><span class="si">{</span><span class="n">maf</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;STEP: Running PCA analysis: `num_pca` = </span><span class="si">{</span><span class="n">num_pca</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;STEP: Running PCA analysis: `ref_threshold` = </span><span class="si">{</span><span class="n">ref_threshold</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;STEP: Running PCA analysis: `stu_threshold` = </span><span class="si">{</span><span class="n">stu_threshold</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;STEP: Running PCA analysis: `psam_file` = </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_files</span><span class="p">[</span><span class="s1">&#39;psam&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="n">goa</span><span class="o">.</span><span class="n">execute_pca</span><span class="p">(</span><span class="n">pca</span><span class="o">=</span><span class="n">pca</span><span class="p">,</span> <span class="n">maf</span><span class="o">=</span><span class="n">maf</span><span class="p">)</span>
        <span class="n">goa</span><span class="o">.</span><span class="n">find_ancestry_outliers</span><span class="p">(</span>
            <span class="n">ref_threshold</span><span class="o">=</span><span class="n">ref_threshold</span><span class="p">,</span> 
            <span class="n">stu_threshold</span><span class="o">=</span><span class="n">stu_threshold</span><span class="p">,</span> 
            <span class="n">reference_pop</span><span class="o">=</span><span class="n">ref_population</span><span class="p">,</span> 
            <span class="n">num_pcs</span>      <span class="o">=</span><span class="n">num_pca</span><span class="p">,</span> 
            <span class="n">fails_dir</span>    <span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">fail_samples_dir</span>
        <span class="p">)</span>
        <span class="n">goa</span><span class="o">.</span><span class="n">execute_drop_ancestry_outliers</span><span class="p">(</span><span class="n">output_dir</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">clean_files</span><span class="p">)</span>
        <span class="n">goa</span><span class="o">.</span><span class="n">draw_pca_plot</span><span class="p">(</span><span class="n">plot_dir</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">plots_dir</span><span class="p">,</span> <span class="n">reference_pop</span><span class="o">=</span><span class="n">ref_population</span><span class="p">,</span> <span class="n">aspect_ratio</span><span class="o">=</span><span class="n">aspect_ratio</span><span class="p">)</span>
        <span class="n">goa</span><span class="o">.</span><span class="n">draw_pca_plot</span><span class="p">(</span><span class="n">plot_dir</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">plots_dir</span><span class="p">,</span> <span class="n">reference_pop</span><span class="o">=</span><span class="n">ref_population</span><span class="p">,</span> <span class="n">aspect_ratio</span><span class="o">=</span><span class="n">aspect_ratio</span><span class="p">,</span> <span class="n">exclude_outliers</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="k">return</span></div>
</div>

</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, Luis Giraldo Gonzalez Ricardo, Amabel Tenghe, and Ashwin Ashok Kumar Sreelatha.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>