<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>ideal_genom_qc.get_references &mdash; IDEAL-GENOM-QC v0.1.0</title>
      <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../../_static/css/theme.css?v=19f00094" />

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../_static/jquery.js?v=5d32c60e"></script>
        <script src="../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../../_static/documentation_options.js?v=01f34227"></script>
        <script src="../../_static/doctools.js?v=9a2dae69"></script>
        <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #2980B9" >

          
          
          <a href="../../index.html" class="icon icon-home">
            IDEAL-GENOM-QC
          </a>
              <div class="version">
                0.1.0
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">User Guide</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../getting_started.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../configuration.html">Configuration Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../examples.html">Examples</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../SampleQC.html">SampleQC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../AncestryQC.html">AncestryQC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../PopStructure.html">Population Structure</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../VariantQC.html">VariantQC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../Helpers.html">Helpers</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../get_references.html">Get References</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Additional Resources</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../faq.html">Frequently Asked Questions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../troubleshooting.html">Troubleshooting Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../contributing.html">Contributing Guide</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #2980B9" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">IDEAL-GENOM-QC</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../index.html">Module code</a></li>
      <li class="breadcrumb-item active">ideal_genom_qc.get_references</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for ideal_genom_qc.get_references</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span><span class="w"> </span><span class="nn">requests</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">logging</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">psutil</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">typing</span><span class="w"> </span><span class="kn">import</span> <span class="n">Optional</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">ideal_genom_qc.Helpers</span><span class="w"> </span><span class="kn">import</span> <span class="n">shell_do</span>

<span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">%(asctime)s</span><span class="s2"> - </span><span class="si">%(levelname)s</span><span class="s2"> - </span><span class="si">%(message)s</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span>

<div class="viewcode-block" id="Fetcher1000Genome">
<a class="viewcode-back" href="../../get_references.html#ideal_genom_qc.get_references.Fetcher1000Genome">[docs]</a>
<span class="k">class</span><span class="w"> </span><span class="nc">Fetcher1000Genome</span><span class="p">:</span>

<div class="viewcode-block" id="Fetcher1000Genome.__init__">
<a class="viewcode-back" href="../../get_references.html#ideal_genom_qc.get_references.Fetcher1000Genome.__init__">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">destination</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Path</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">built</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;38&#39;</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initialize a reference data handler.</span>
<span class="sd">        This class manages reference data files from 1000 Genomes Project.</span>
<span class="sd">        </span>
<span class="sd">        Paramters:</span>
<span class="sd">        ----------</span>
<span class="sd">        destination: Path (optional) </span>
<span class="sd">            Path where reference files will be stored. If not provided, defaults to &#39;../data/1000genomes_built_{built}&#39;.</span>
<span class="sd">        built: str (optional): </span>
<span class="sd">            Human genome build version. Defaults to &#39;38&#39;.</span>
<span class="sd">        </span>
<span class="sd">        Attributes:</span>
<span class="sd">        -----------</span>
<span class="sd">        destination: Path </span>
<span class="sd">            Directory path where reference files are stored</span>
<span class="sd">        built: str </span>
<span class="sd">            Human genome build version being used</span>
<span class="sd">        pgen_file: Path </span>
<span class="sd">            Path to PGEN format file</span>
<span class="sd">        pvar_file: Path </span>
<span class="sd">            Path to PVAR format file </span>
<span class="sd">        psam_file: Path </span>
<span class="sd">            Path to PSAM format file</span>
<span class="sd">        bed_file: Path </span>
<span class="sd">            Path to BED format file</span>
<span class="sd">        bim_file: Path </span>
<span class="sd">            Path to BIM format file </span>
<span class="sd">        fam_file: Path </span>
<span class="sd">            Path to FAM format file</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">built</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;Built must be a string representing the genome build version (e.g., &#39;37&#39; or &#39;38&#39;).&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">built</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;37&#39;</span><span class="p">,</span> <span class="s1">&#39;38&#39;</span><span class="p">]:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Built must be either &#39;37&#39; or &#39;38&#39;.&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">destination</span><span class="p">:</span>
            <span class="n">destination</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="vm">__file__</span><span class="p">)</span><span class="o">.</span><span class="n">resolve</span><span class="p">()</span><span class="o">.</span><span class="n">parent</span><span class="o">.</span><span class="n">parent</span> <span class="o">/</span> <span class="s2">&quot;data&quot;</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;1000genomes_built_</span><span class="si">{</span><span class="n">built</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Destination folder: </span><span class="si">{</span><span class="n">destination</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        
        <span class="bp">self</span><span class="o">.</span><span class="n">destination</span> <span class="o">=</span> <span class="n">destination</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">built</span> <span class="o">=</span> <span class="n">built</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">pgen_file</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pvar_file</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">psam_file</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">bed_file</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">bim_file</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">fam_file</span> <span class="o">=</span> <span class="kc">None</span>
        
        <span class="k">pass</span></div>


<div class="viewcode-block" id="Fetcher1000Genome.get_1000genomes">
<a class="viewcode-back" href="../../get_references.html#ideal_genom_qc.get_references.Fetcher1000Genome.get_1000genomes">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">get_1000genomes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">url_pgen</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">url_pvar</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">url_psam</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span><span class="o">-&gt;</span> <span class="n">Path</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Download and decompress 1000 Genomes reference data.</span>
<span class="sd">        This method downloads the PLINK2 binary files (.pgen, .pvar, .psam) for the 1000 Genomes </span>
<span class="sd">        reference dataset, corresponding to the specified genome build (37 or 38). If the files </span>
<span class="sd">        already exist in the destination directory, the download is skipped.</span>
<span class="sd">        </span>
<span class="sd">        Parameters:</span>
<span class="sd">        -----------</span>
<span class="sd">        url_pgen (str, optional): Custom URL for downloading the .pgen file. </span>
<span class="sd">            If None, uses default URL based on genome build.</span>
<span class="sd">        url_pvar (str, optional): Custom URL for downloading the .pvar file.</span>
<span class="sd">            If None, uses default URL based on genome build.</span>
<span class="sd">        url_psam (str, optional): Custom URL for downloading the .psam file.</span>
<span class="sd">            If None, uses default URL based on genome build.</span>
<span class="sd">        </span>
<span class="sd">        Returns:</span>
<span class="sd">        --------</span>
<span class="sd">            Path: Path object pointing to the decompressed .pgen file location.</span>
<span class="sd">        </span>
<span class="sd">        Note:</span>
<span class="sd">        -----</span>
<span class="sd">            The method requires plink2 to be installed and accessible in the system path</span>
<span class="sd">            for decompressing the .pgen file.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">destination</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">built</span> <span class="o">==</span> <span class="s1">&#39;38&#39;</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">url_pgen</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">url_pgen</span> <span class="o">=</span> <span class="sa">r</span><span class="s2">&quot;https://www.dropbox.com/s/j72j6uciq5zuzii/all_hg38.pgen.zst?dl=1&quot;</span>
            <span class="k">if</span> <span class="n">url_pvar</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">url_pvar</span> <span class="o">=</span> <span class="sa">r</span><span class="s2">&quot;https://www.dropbox.com/scl/fi/fn0bcm5oseyuawxfvkcpb/all_hg38_rs.pvar.zst?rlkey=przncwb78rhz4g4ukovocdxaz&amp;dl=1&quot;</span>
            <span class="k">if</span> <span class="n">url_psam</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">url_psam</span> <span class="o">=</span> <span class="sa">r</span><span class="s2">&quot;https://www.dropbox.com/scl/fi/u5udzzaibgyvxzfnjcvjc/hg38_corrected.psam?rlkey=oecjnk4vmbhc8b1p202l0ih4x&amp;dl=1&quot;</span>
        
        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">built</span> <span class="o">==</span> <span class="s1">&#39;37&#39;</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">url_pgen</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">url_pgen</span> <span class="o">=</span> <span class="sa">r</span><span class="s2">&quot;https://www.dropbox.com/s/y6ytfoybz48dc0u/all_phase3.pgen.zst?dl=1&quot;</span>
            <span class="k">if</span> <span class="n">url_pvar</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">url_pvar</span> <span class="o">=</span> <span class="sa">r</span><span class="s2">&quot;https://www.dropbox.com/s/odlexvo8fummcvt/all_phase3.pvar.zst?dl=1&quot;</span>
            <span class="k">if</span> <span class="n">url_psam</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">url_psam</span> <span class="o">=</span> <span class="sa">r</span><span class="s2">&quot;https://www.dropbox.com/scl/fi/haqvrumpuzfutklstazwk/phase3_corrected.psam?rlkey=0yyifzj2fb863ddbmsv4jkeq6&amp;dl=1&quot;</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_check_if_binaries_exist</span><span class="p">():</span>

            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;1000 Genomes binaries already exist. Skipping download.&quot;</span><span class="p">)</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">pvar_file</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">destination</span> <span class="o">/</span> <span class="s2">&quot;all_phase3.pvar.zst&quot;</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">psam_file</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">destination</span> <span class="o">/</span> <span class="s2">&quot;all_phase3.psam&quot;</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">pgen_decompressed</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">destination</span> <span class="o">/</span> <span class="s2">&quot;all_phase3.pgen&quot;</span>

            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">pgen_decompressed</span>
        
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Downloading 1000 Genomes data...&quot;</span><span class="p">)</span>
        
        <span class="k">if</span> <span class="n">url_pgen</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">pgen_file</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_download_file</span><span class="p">(</span><span class="n">url_pgen</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">destination</span> <span class="o">/</span> <span class="s2">&quot;all_phase3.pgen.zst&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">url_pvar</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">pvar_file</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_download_file</span><span class="p">(</span><span class="n">url_pvar</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">destination</span> <span class="o">/</span> <span class="s2">&quot;all_phase3.pvar.zst&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">url_psam</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">psam_file</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_download_file</span><span class="p">(</span><span class="n">url_psam</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">destination</span> <span class="o">/</span> <span class="s2">&quot;all_phase3.psam&quot;</span><span class="p">)</span>

        <span class="n">pgen_file</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">destination</span> <span class="o">/</span> <span class="s2">&quot;all_phase3.pgen.zst&quot;</span>
        <span class="n">pgen_decompressed</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">destination</span> <span class="o">/</span> <span class="s2">&quot;all_phase3.pgen&quot;</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Decompressing pgen file from 1000 Genomes data...&quot;</span><span class="p">)</span>

        <span class="c1"># plink2 command</span>
        <span class="n">plink2_cmd</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink2 --zst-decompress </span><span class="si">{</span><span class="n">pgen_file</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">pgen_decompressed</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="c1"># execute plink2 command</span>
        <span class="n">shell_do</span><span class="p">(</span><span class="n">plink2_cmd</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">pgen_file</span> <span class="o">=</span> <span class="n">pgen_decompressed</span>

        <span class="k">return</span> <span class="n">pgen_decompressed</span></div>


    <span class="k">def</span><span class="w"> </span><span class="nf">_download_file</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">url</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">destination</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Path</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Downloads a file from a given URL and saves it to the specified destination.</span>

<span class="sd">        Parameters:</span>
<span class="sd">        -----------</span>
<span class="sd">        url: str </span>
<span class="sd">            The URL of the file to download.</span>
<span class="sd">        destination: Path </span>
<span class="sd">            The path where the downloaded file will be saved.</span>

<span class="sd">        Returns:</span>
<span class="sd">        --------</span>
<span class="sd">        Path: </span>
<span class="sd">            The path to the downloaded file if successful, None if download fails.</span>

<span class="sd">        Raises:</span>
<span class="sd">        -------</span>
<span class="sd">            None: Exceptions are caught and logged internally.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">response</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">url</span><span class="p">,</span> <span class="n">stream</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">response</span><span class="o">.</span><span class="n">raise_for_status</span><span class="p">()</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">destination</span><span class="p">,</span> <span class="s1">&#39;wb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">out_file</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">chunk</span> <span class="ow">in</span> <span class="n">response</span><span class="o">.</span><span class="n">iter_content</span><span class="p">(</span><span class="n">chunk_size</span><span class="o">=</span><span class="mi">1024</span><span class="p">):</span>
                    <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">chunk</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">destination</span>
        <span class="k">except</span> <span class="n">requests</span><span class="o">.</span><span class="n">RequestException</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Failed to download </span><span class="si">{</span><span class="n">url</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">Path</span><span class="p">()</span>

<div class="viewcode-block" id="Fetcher1000Genome.get_1000genomes_binaries">
<a class="viewcode-back" href="../../get_references.html#ideal_genom_qc.get_references.Fetcher1000Genome.get_1000genomes_binaries">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">get_1000genomes_binaries</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Path</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Convert downloaded 1000 Genomes data into PLINK binary files (.bed, .bim, .fam).</span>
<span class="sd">        This method processes the downloaded 1000 Genomes data files and converts them into PLINK binary format.</span>
<span class="sd">        If the binary files already exist, it skips the conversion process. The method handles file cleanup</span>
<span class="sd">        and proper renaming of output files.</span>
<span class="sd">        The conversion is done in two steps:</span>
<span class="sd">        1. Convert pfile to binary format including only SNPs from chromosomes 1-22,X,Y,MT</span>
<span class="sd">        2. Update variant IDs and create final binary files</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        Path</span>
<span class="sd">            Path object pointing to the generated binary files (without extension)</span>
<span class="sd">            The actual files created will be .bed, .bim, .fam and .psam with the same prefix</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_check_if_binaries_exist</span><span class="p">():</span>

            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;1000 Genomes binaries already exist. Skipping conversion into bfiles...&quot;</span><span class="p">)</span>

            <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">destination</span> <span class="o">/</span> <span class="s2">&quot;all_phase3.pgen&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">unlink</span><span class="p">(</span><span class="n">missing_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">destination</span> <span class="o">/</span> <span class="s2">&quot;all_phase3.pgen.zst&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">unlink</span><span class="p">(</span><span class="n">missing_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">destination</span> <span class="o">/</span> <span class="s2">&quot;all_phase3.pvar.zst&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">unlink</span><span class="p">(</span><span class="n">missing_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">bed_file</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">destination</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;1kG_phase3_GRCh</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">built</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s1">&#39;.bed&#39;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">bim_file</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">destination</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;1kG_phase3_GRCh</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">built</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s1">&#39;.bim&#39;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">fam_file</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">destination</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;1kG_phase3_GRCh</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">built</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s1">&#39;.fam&#39;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">psam_file</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">destination</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;1kG_phase3_GRCh</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">built</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s1">&#39;.psam&#39;</span><span class="p">)</span>

            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">destination</span> <span class="o">/</span> <span class="s2">&quot;all_phase3&quot;</span>
        
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Converting 1000 Genomes data into bfiles...&quot;</span><span class="p">)</span>

        <span class="n">memory_info</span> <span class="o">=</span> <span class="n">psutil</span><span class="o">.</span><span class="n">virtual_memory</span><span class="p">()</span>
        <span class="n">available_memory_mb</span> <span class="o">=</span> <span class="n">memory_info</span><span class="o">.</span><span class="n">available</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1024</span> <span class="o">*</span> <span class="mi">1024</span><span class="p">)</span>
        <span class="n">memory</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="mi">2</span><span class="o">*</span><span class="n">available_memory_mb</span><span class="o">/</span><span class="mi">3</span><span class="p">,</span><span class="mi">0</span><span class="p">)</span>

        <span class="c1"># plink2 command</span>
        <span class="n">plink2_cmd</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink2 --pfile </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">destination</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="s1">&#39;all_phase3&#39;</span><span class="si">}</span><span class="s2"> vzs --chr 1-22,X,Y,MT --snps-only --max-alleles 2 --memory 10000 --make-bed --out </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">destination</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="s1">&#39;all_phase3&#39;</span><span class="si">}</span><span class="s2">&quot;</span>
        
        <span class="c1"># execute plink2 command</span>
        <span class="n">shell_do</span><span class="p">(</span><span class="n">plink2_cmd</span><span class="p">)</span>

        <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">destination</span> <span class="o">/</span> <span class="s2">&quot;all_phase3.pgen&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">unlink</span><span class="p">(</span><span class="n">missing_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">destination</span> <span class="o">/</span> <span class="s2">&quot;all_phase3.pgen.zst&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">unlink</span><span class="p">(</span><span class="n">missing_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">destination</span> <span class="o">/</span> <span class="s2">&quot;all_phase3.pvar.zst&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">unlink</span><span class="p">(</span><span class="n">missing_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Downloaded 1000 Genomes data deleted.&quot;</span><span class="p">)</span>

        <span class="c1"># PLINK2 command</span>
        <span class="n">plink2_cmd</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink2 --bfile </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">destination</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="s1">&#39;all_phase3&#39;</span><span class="si">}</span><span class="s2"> --set-all-var-ids @:#:$r:$a --memory 1000 --make-bed --out </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">destination</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="sa">f</span><span class="s1">&#39;1kG_phase3_GRCh</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">built</span><span class="si">}</span><span class="s1">&#39;</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="n">shell_do</span><span class="p">(</span><span class="n">plink2_cmd</span><span class="p">,</span> <span class="n">log</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">destination</span> <span class="o">/</span> <span class="s2">&quot;all_phase3.bed&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">unlink</span><span class="p">(</span><span class="n">missing_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">destination</span> <span class="o">/</span> <span class="s2">&quot;all_phase3.bim&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">unlink</span><span class="p">(</span><span class="n">missing_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">destination</span> <span class="o">/</span> <span class="s2">&quot;all_phase3.fam&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">unlink</span><span class="p">(</span><span class="n">missing_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">bed_file</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">destination</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;1kG_phase3_GRCh</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">built</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s1">&#39;.bed&#39;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">bim_file</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">destination</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;1kG_phase3_GRCh</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">built</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s1">&#39;.bim&#39;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">fam_file</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">destination</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;1kG_phase3_GRCh</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">built</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s1">&#39;.fam&#39;</span><span class="p">)</span>

        <span class="n">psam_renamed</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">destination</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;1kG_phase3_GRCh</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">built</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s1">&#39;.psam&#39;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">psam_file</span><span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">destination</span> <span class="o">/</span> <span class="s2">&quot;all_phase3.psam&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">psam_renamed</span><span class="p">)</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">destination</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;1kG_phase3_GRCh</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">built</span><span class="si">}</span><span class="s1">&#39;</span></div>

    
    <span class="k">def</span><span class="w"> </span><span class="nf">_check_if_binaries_exist</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Checks if all required binary files exist in the destination directory.</span>

<span class="sd">        This method verifies the existence of .bed, .bim, .fam, and .psam files</span>
<span class="sd">        for the 1000 Genomes Phase 3 reference panel in the specified genome build.</span>

<span class="sd">        Returns:</span>
<span class="sd">        --------</span>
<span class="sd">            bool: True if all required files exist, False otherwise.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">check_bed</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">destination</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;1kG_phase3_GRCh</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">built</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s1">&#39;.bed&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">exists</span><span class="p">()</span>
        <span class="n">check_bim</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">destination</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;1kG_phase3_GRCh</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">built</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s1">&#39;.bim&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">exists</span><span class="p">()</span>
        <span class="n">check_fam</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">destination</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;1kG_phase3_GRCh</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">built</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s1">&#39;.fam&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">exists</span><span class="p">()</span>
        <span class="n">check_psam</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">destination</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;1kG_phase3_GRCh</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">built</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s1">&#39;.psam&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">exists</span><span class="p">()</span>

        <span class="k">return</span> <span class="n">check_bed</span> <span class="ow">and</span> <span class="n">check_bim</span> <span class="ow">and</span> <span class="n">check_fam</span> <span class="ow">and</span> <span class="n">check_psam</span></div>

    
<div class="viewcode-block" id="FetcherLDRegions">
<a class="viewcode-back" href="../../get_references.html#ideal_genom_qc.get_references.FetcherLDRegions">[docs]</a>
<span class="k">class</span><span class="w"> </span><span class="nc">FetcherLDRegions</span><span class="p">:</span>

<div class="viewcode-block" id="FetcherLDRegions.__init__">
<a class="viewcode-back" href="../../get_references.html#ideal_genom_qc.get_references.FetcherLDRegions.__init__">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">destination</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Path</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">built</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;38&#39;</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initialize LDRegions object.</span>
<span class="sd">        This initializer sets up the destination path for LD regions files and the genome build version.</span>
<span class="sd">        If no destination is provided, it defaults to a &#39;data/ld_regions_files&#39; directory relative to</span>
<span class="sd">        the parent directory of the current file.</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        destination : Path, optional</span>
<span class="sd">            Path where LD region files will be stored. If None, uses default path.</span>
<span class="sd">        built : str, optional</span>
<span class="sd">            Genome build version, defaults to &#39;38&#39;.</span>
<span class="sd">        </span>
<span class="sd">        Attributes</span>
<span class="sd">        ----------</span>
<span class="sd">        destination : Path</span>
<span class="sd">            Directory path where LD region files are stored</span>
<span class="sd">        built : str</span>
<span class="sd">            Genome build version being used</span>
<span class="sd">        ld_regions : None</span>
<span class="sd">            Placeholder for LD regions data, initially set to None</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">destination</span><span class="p">:</span>
            <span class="n">destination</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="vm">__file__</span><span class="p">)</span><span class="o">.</span><span class="n">resolve</span><span class="p">()</span><span class="o">.</span><span class="n">parent</span><span class="o">.</span><span class="n">parent</span> <span class="o">/</span> <span class="s2">&quot;data&quot;</span> <span class="o">/</span> <span class="s2">&quot;ld_regions_files&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">destination</span> <span class="o">=</span> <span class="n">destination</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">built</span> <span class="o">=</span> <span class="n">built</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">ld_regions</span> <span class="o">=</span> <span class="kc">None</span>
        
        <span class="k">pass</span></div>


<div class="viewcode-block" id="FetcherLDRegions.get_ld_regions">
<a class="viewcode-back" href="../../get_references.html#ideal_genom_qc.get_references.FetcherLDRegions.get_ld_regions">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">get_ld_regions</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span><span class="o">-&gt;</span> <span class="n">Path</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Downloads or creates high LD regions file based on genome build version.</span>
<span class="sd">        This method handles the retrieval of high Linkage Disequilibrium (LD) regions for</span>
<span class="sd">        different genome builds (37 or 38). For build 37, it downloads the regions from a</span>
<span class="sd">        GitHub repository. For build 38, it creates the file from predefined coordinates.</span>
<span class="sd">        </span>
<span class="sd">        Returns:</span>
<span class="sd">        --------</span>
<span class="sd">        Path: Path to the created/downloaded LD regions file. Returns empty Path if</span>
<span class="sd">              download fails for build 37.</span>
<span class="sd">        </span>
<span class="sd">        Raises:</span>
<span class="sd">        -------</span>
<span class="sd">            None explicitly, but may raise standard I/O related exceptions.</span>
<span class="sd">        </span>
<span class="sd">        Notes:</span>
<span class="sd">        -----</span>
<span class="sd">            - For build 37: Downloads from genepi-freiburg/gwas repository</span>
<span class="sd">            - For build 38: Creates file from hardcoded coordinates from GWAS-pipeline</span>
<span class="sd">            - Files are named as &#39;high-LD-regions_GRCh{build}.txt&#39;</span>
<span class="sd">            - Creates destination directory if it doesn&#39;t exist</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">destination</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="n">out_dir</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">destination</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">built</span> <span class="o">==</span> <span class="s1">&#39;37&#39;</span><span class="p">:</span>
            <span class="n">url_ld_regions</span> <span class="o">=</span> <span class="sa">r</span><span class="s2">&quot;https://raw.githubusercontent.com/genepi-freiburg/gwas/refs/heads/master/single-pca/high-LD-regions.txt&quot;</span>
        
            <span class="n">ld</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">url_ld_regions</span><span class="p">)</span>


            <span class="k">if</span> <span class="n">ld</span><span class="o">.</span><span class="n">status_code</span> <span class="o">==</span> <span class="mi">200</span><span class="p">:</span>
                <span class="k">with</span> <span class="nb">open</span><span class="p">((</span><span class="n">out_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;high-LD-regions_GRCh</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">built</span><span class="si">}</span><span class="s2">.txt&quot;</span><span class="p">),</span> <span class="s2">&quot;wb&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">ld</span><span class="o">.</span><span class="n">content</span><span class="p">)</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;LD regions file for built </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">built</span><span class="si">}</span><span class="s2"> downloaded successfully to </span><span class="si">{</span><span class="n">out_dir</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

                <span class="bp">self</span><span class="o">.</span><span class="n">ld_regions</span> <span class="o">=</span> <span class="n">out_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;high-LD-regions_GRCh</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">built</span><span class="si">}</span><span class="s2">.txt&quot;</span>
                <span class="k">return</span> <span class="n">out_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;high-LD-regions_GRCh</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">built</span><span class="si">}</span><span class="s2">.txt&quot;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Failed to download .bim file: </span><span class="si">{</span><span class="n">ld</span><span class="o">.</span><span class="n">status_code</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

                <span class="k">return</span> <span class="n">Path</span><span class="p">()</span>

        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">built</span> <span class="o">==</span> <span class="s1">&#39;38&#39;</span><span class="p">:</span>
            <span class="c1"># extracted from</span>
            <span class="c1"># https://github.com/neurogenetics/GWAS-pipeline</span>
            <span class="n">data</span> <span class="o">=</span> <span class="p">[</span>
                <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">47534328</span><span class="p">,</span> <span class="mi">51534328</span><span class="p">,</span> <span class="s2">&quot;r1&quot;</span><span class="p">),</span>
                <span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">133742429</span><span class="p">,</span> <span class="mi">137242430</span><span class="p">,</span> <span class="s2">&quot;r2&quot;</span><span class="p">),</span>
                <span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">182135273</span><span class="p">,</span> <span class="mi">189135274</span><span class="p">,</span> <span class="s2">&quot;r3&quot;</span><span class="p">),</span>
                <span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">47458510</span><span class="p">,</span> <span class="mi">49962567</span><span class="p">,</span> <span class="s2">&quot;r4&quot;</span><span class="p">),</span>
                <span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">83450849</span><span class="p">,</span> <span class="mi">86950850</span><span class="p">,</span> <span class="s2">&quot;r5&quot;</span><span class="p">),</span>
                <span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">98664296</span><span class="p">,</span> <span class="mi">101164296</span><span class="p">,</span> <span class="s2">&quot;r6&quot;</span><span class="p">),</span>
                <span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">129664307</span><span class="p">,</span> <span class="mi">132664308</span><span class="p">,</span> <span class="s2">&quot;r7&quot;</span><span class="p">),</span>
                <span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">136164311</span><span class="p">,</span> <span class="mi">139164311</span><span class="p">,</span> <span class="s2">&quot;r8&quot;</span><span class="p">),</span>
                <span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">24999772</span><span class="p">,</span> <span class="mi">35032223</span><span class="p">,</span> <span class="s2">&quot;r9&quot;</span><span class="p">),</span>
                <span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">139678863</span><span class="p">,</span> <span class="mi">142178863</span><span class="p">,</span> <span class="s2">&quot;r10&quot;</span><span class="p">),</span>
                <span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">7142478</span><span class="p">,</span> <span class="mi">13142491</span><span class="p">,</span> <span class="s2">&quot;r11&quot;</span><span class="p">),</span>
                <span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">110987771</span><span class="p">,</span> <span class="mi">113987771</span><span class="p">,</span> <span class="s2">&quot;r12&quot;</span><span class="p">),</span>
                <span class="p">(</span><span class="mi">11</span><span class="p">,</span> <span class="mi">87789108</span><span class="p">,</span> <span class="mi">90766832</span><span class="p">,</span> <span class="s2">&quot;r13&quot;</span><span class="p">),</span>
                <span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">109062195</span><span class="p">,</span> <span class="mi">111562196</span><span class="p">,</span> <span class="s2">&quot;r14&quot;</span><span class="p">),</span>
                <span class="p">(</span><span class="mi">20</span><span class="p">,</span> <span class="mi">33412194</span><span class="p">,</span> <span class="mi">35912078</span><span class="p">,</span> <span class="s2">&quot;r15&quot;</span><span class="p">)</span>
            <span class="p">]</span>

            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">out_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;high-LD-regions_GRCH</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">built</span><span class="si">}</span><span class="s1">.txt&#39;</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">data</span><span class="p">:</span>
                    <span class="n">file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">line</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">line</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">line</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">line</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">ld_regions</span> <span class="o">=</span> <span class="n">out_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;high-LD-regions_GRCH</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">built</span><span class="si">}</span><span class="s2">.txt&quot;</span>
            <span class="k">return</span> <span class="n">out_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;high-LD-regions_GRCH</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">built</span><span class="si">}</span><span class="s1">.txt&#39;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unsupported genome build: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">built</span><span class="si">}</span><span class="s2">. Only &#39;37&#39; and &#39;38&#39; are supported.&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">Path</span><span class="p">()</span></div>
</div>

</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, Luis Giraldo Gonzalez Ricardo, Amabel Tenghe, and Ashwin Ashok Kumar Sreelatha.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>