<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>ideal_genom_qc.PopStructure &mdash; IDEAL-GENOM-QC v0.1.0</title>
      <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../../_static/css/theme.css?v=19f00094" />

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../_static/jquery.js?v=5d32c60e"></script>
        <script src="../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../../_static/documentation_options.js?v=01f34227"></script>
        <script src="../../_static/doctools.js?v=9a2dae69"></script>
        <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #2980B9" >

          
          
          <a href="../../index.html" class="icon icon-home">
            IDEAL-GENOM-QC
          </a>
              <div class="version">
                0.1.0
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">User Guide</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../getting_started.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../configuration.html">Configuration Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../examples.html">Examples</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../SampleQC.html">SampleQC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../AncestryQC.html">AncestryQC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../PopStructure.html">Population Structure</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../VariantQC.html">VariantQC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../Helpers.html">Helpers</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../get_references.html">Get References</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Additional Resources</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../faq.html">Frequently Asked Questions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../troubleshooting.html">Troubleshooting Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../contributing.html">Contributing Guide</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #2980B9" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">IDEAL-GENOM-QC</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../index.html">Module code</a></li>
      <li class="breadcrumb-item active">ideal_genom_qc.PopStructure</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for ideal_genom_qc.PopStructure</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Module to draw plots based on UMAP dimension reduction</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">umap</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">warnings</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">logging</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">psutil</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">typing</span><span class="w"> </span><span class="kn">import</span> <span class="n">Optional</span><span class="p">,</span> <span class="n">Union</span><span class="p">,</span> <span class="n">Literal</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">seaborn</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sns</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">ideal_genom_qc.Helpers</span><span class="w"> </span><span class="kn">import</span> <span class="n">shell_do</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">ideal_genom_qc.get_references</span><span class="w"> </span><span class="kn">import</span> <span class="n">FetcherLDRegions</span><span class="p">,</span> <span class="n">Fetcher1000Genome</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">ideal_genom_qc.AncestryQC</span><span class="w"> </span><span class="kn">import</span> <span class="n">ReferenceGenomicMerger</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.model_selection</span><span class="w"> </span><span class="kn">import</span> <span class="n">ParameterGrid</span>

<span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">%(asctime)s</span><span class="s2"> - </span><span class="si">%(levelname)s</span><span class="s2"> - </span><span class="si">%(message)s</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span>

<div class="viewcode-block" id="UMAPplot">
<a class="viewcode-back" href="../../PopStructure.html#ideal_genom_qc.PopStructure.UMAPplot">[docs]</a>
<span class="k">class</span><span class="w"> </span><span class="nc">UMAPplot</span><span class="p">:</span>

<div class="viewcode-block" id="UMAPplot.__init__">
<a class="viewcode-back" href="../../PopStructure.html#ideal_genom_qc.PopStructure.UMAPplot.__init__">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">input_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">input_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">output_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">high_ld_file</span><span class="p">:</span> <span class="n">Path</span><span class="o">=</span><span class="n">Path</span><span class="p">(),</span> <span class="n">built</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;38&#39;</span><span class="p">,</span> <span class="n">recompute_pca</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initialize UMAPplot object for population structure analysis.</span>
<span class="sd">        This class handles the creation of UMAP plots for genetic data, managing input/output paths</span>
<span class="sd">        and configuration for the analysis.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        input_path : Path</span>
<span class="sd">            Path to the directory containing input files</span>
<span class="sd">        input_name : str</span>
<span class="sd">            Name of the input file</span>
<span class="sd">        output_path : Path</span>
<span class="sd">            Path to the directory where results will be saved</span>
<span class="sd">        output_name : str</span>
<span class="sd">            Name for the output files</span>
<span class="sd">        high_ld_file : Path</span>
<span class="sd">            Path to the file containing high LD regions</span>
<span class="sd">        built : str, optional</span>
<span class="sd">            Genome build version, either &#39;37&#39; or &#39;38&#39; (default is &#39;38&#39;)</span>
<span class="sd">        recompute_pca : bool, optional</span>
<span class="sd">            Whether to recompute PCA analysis (default is True)</span>

<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        TypeError</span>
<span class="sd">            If input types are incorrect for any parameter</span>
<span class="sd">        ValueError</span>
<span class="sd">            If built is not &#39;37&#39; or &#39;38&#39;</span>
<span class="sd">        FileNotFoundError</span>
<span class="sd">            If input_path or output_path do not exist</span>
<span class="sd">        </span>
<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        If high_ld_file is not found, it will be automatically fetched from the package.</span>
<span class="sd">        Creates &#39;umap_results&#39; and &#39;plots&#39; directories in the output path.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">input_path</span><span class="p">,</span> <span class="n">Path</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;input_path should be a Path object&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="n">Path</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;output_path should be a Path object&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">high_ld_file</span><span class="p">,</span> <span class="n">Path</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;high_ld_regions should be a Path object&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">input_name</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span> 
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;input_name should be a string&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">recompute_pca</span><span class="p">,</span> <span class="nb">bool</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;recompute_merge should be a boolean&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">built</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;built should be a string&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">built</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;37&#39;</span><span class="p">,</span> <span class="s1">&#39;38&#39;</span><span class="p">]:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;built should be either &#39;37&#39; or &#39;38&#39;&quot;</span><span class="p">)</span>        
        <span class="k">if</span> <span class="ow">not</span> <span class="n">input_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="s2">&quot;input_path does not exist&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">output_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="s2">&quot;output_path does not exist&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">high_ld_file</span><span class="o">.</span><span class="n">is_file</span><span class="p">():</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;High LD file not found at </span><span class="si">{</span><span class="n">high_ld_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;High LD file will be fetched from the package&#39;</span><span class="p">)</span>
            
            <span class="n">ld_fetcher</span> <span class="o">=</span> <span class="n">FetcherLDRegions</span><span class="p">()</span>
            <span class="n">ld_fetcher</span><span class="o">.</span><span class="n">get_ld_regions</span><span class="p">()</span>

            <span class="k">if</span> <span class="n">ld_fetcher</span><span class="o">.</span><span class="n">ld_regions</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="s2">&quot;Could not fetch LD regions file.&quot;</span><span class="p">)</span>
                
            <span class="n">high_ld_file</span> <span class="o">=</span> <span class="n">ld_fetcher</span><span class="o">.</span><span class="n">ld_regions</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;High LD file fetched from the package and saved at </span><span class="si">{</span><span class="n">high_ld_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">input_path</span> <span class="o">=</span> <span class="n">input_path</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">input_name</span> <span class="o">=</span> <span class="n">input_name</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">output_path</span><span class="o">=</span> <span class="n">output_path</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">high_ld_regions</span> <span class="o">=</span> <span class="n">high_ld_file</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">recompute_pca</span> <span class="o">=</span> <span class="n">recompute_pca</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">built</span> <span class="o">=</span> <span class="n">built</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">files_to_keep</span><span class="o">=</span> <span class="p">[]</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_path</span> <span class="o">/</span> <span class="s1">&#39;umap_results&#39;</span> 
        <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">plots_dir</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="s1">&#39;plots&#39;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">plots_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="k">pass</span></div>


<div class="viewcode-block" id="UMAPplot.ld_pruning">
<a class="viewcode-back" href="../../PopStructure.html#ideal_genom_qc.PopStructure.UMAPplot.ld_pruning">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">ld_pruning</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">maf</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.001</span><span class="p">,</span> <span class="n">geno</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.1</span><span class="p">,</span> <span class="n">mind</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.2</span><span class="p">,</span> <span class="n">hwe</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">5e-8</span><span class="p">,</span> <span class="n">ind_pair</span><span class="p">:</span> <span class="nb">list</span> <span class="o">=</span> <span class="p">[</span><span class="mi">50</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mf">0.2</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Perform Linkage Disequilibrium (LD) pruning on genetic data using PLINK.</span>
<span class="sd">        This method filters SNPs based on specified thresholds for various quality control metrics</span>
<span class="sd">        and performs LD-based pruning to remove highly correlated variants.</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        maf : float, default=0.001</span>
<span class="sd">            Minor allele frequency threshold. Variants with MAF below this value are removed.</span>
<span class="sd">            Must be between 0 and 0.5.</span>
<span class="sd">        geno : float, default=0.1</span>
<span class="sd">            Maximum per-SNP missing rate. Variants with missing rate above this are removed.</span>
<span class="sd">            Must be between 0 and 1.</span>
<span class="sd">        mind : float, default=0.2</span>
<span class="sd">            Maximum per-individual missing rate. Samples with missing rate above this are removed.</span>
<span class="sd">            Must be between 0 and 1. Recommended range is 0.02 to 0.1.</span>
<span class="sd">        hwe : float, default=5e-8</span>
<span class="sd">            Hardy-Weinberg equilibrium exact test p-value threshold.</span>
<span class="sd">            Variants with p-value below this are removed. Must be between 0 and 1.</span>
<span class="sd">        ind_pair : list, default=[50, 5, 0.2]</span>
<span class="sd">            Parameters for pairwise LD pruning: [window size, step size, rÂ² threshold].</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None</span>
<span class="sd">            Creates pruned PLINK binary files in the results directory.</span>
<span class="sd">        </span>
<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        - Skips processing if recompute_pca is False</span>
<span class="sd">        - Uses multithreading with optimal thread count based on system CPU</span>
<span class="sd">        - Generates intermediate files: .prune.in and .prune.out</span>
<span class="sd">        - Creates final LD-pruned dataset with &#39;-LDpruned&#39; suffix</span>
<span class="sd">        </span>
<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        TypeError</span>
<span class="sd">            If input parameters are not of type float</span>
<span class="sd">        ValueError</span>
<span class="sd">            If input parameters are outside their valid ranges</span>
<span class="sd">        UserWarning</span>
<span class="sd">            If mind parameter is outside recommended range</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">recompute_pca</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;`recompuite_pca` is set to </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">recompute_pca</span><span class="si">}</span><span class="s2">. LD pruning will be skipped.&quot;</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;LD pruning already performed. Skipping this step.&quot;</span><span class="p">)</span>
            <span class="k">return</span>

        <span class="c1"># Check type of maf</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">maf</span><span class="p">,</span> <span class="nb">float</span><span class="p">):</span>
             <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;maf should be of type float.&quot;</span><span class="p">)</span>

        <span class="c1"># Check type of geno</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">geno</span><span class="p">,</span> <span class="nb">float</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;geno should be of type float.&quot;</span><span class="p">)</span>

        <span class="c1"># Check type of mind</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">mind</span><span class="p">,</span> <span class="nb">float</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;mind should be of type float.&quot;</span><span class="p">)</span>
        
        <span class="c1"># Check type of hwe</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">hwe</span><span class="p">,</span> <span class="nb">float</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;hwe should be of type float.&quot;</span><span class="p">)</span>
        
        <span class="c1"># Check if maf is in range</span>
        <span class="k">if</span> <span class="n">maf</span> <span class="o">&lt;=</span> <span class="mf">0.0</span> <span class="ow">or</span> <span class="n">maf</span> <span class="o">&gt;=</span> <span class="mf">0.5</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;maf should be between 0 and 0.5&quot;</span><span class="p">)</span>
        
        <span class="c1"># Check if geno is in range</span>
        <span class="k">if</span> <span class="n">geno</span> <span class="o">&lt;=</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">geno</span> <span class="o">&gt;=</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;geno should be between 0 and 1&quot;</span><span class="p">)</span>
        
        <span class="c1"># Check if mind is in range</span>
        <span class="k">if</span> <span class="n">mind</span> <span class="o">&lt;</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">mind</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;mind should be between 0 and 1&quot;</span><span class="p">)</span>
        
        <span class="c1"># Check if mind is around typical values</span>
        <span class="k">if</span> <span class="n">mind</span> <span class="o">&lt;=</span> <span class="mf">0.02</span> <span class="ow">and</span> <span class="n">mind</span> <span class="o">&gt;=</span> <span class="mf">0.1</span><span class="p">:</span>
            <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;The &#39;mind&#39; value </span><span class="si">{</span><span class="n">mind</span><span class="si">}</span><span class="s2"> is outside the recommended range of 0.02 to 0.1.&quot;</span><span class="p">,</span> <span class="ne">UserWarning</span><span class="p">)</span>

        <span class="c1"># Check if hwe is in range</span>
        <span class="k">if</span> <span class="n">hwe</span> <span class="o">&lt;</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">hwe</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;hwe should be between 0 and 1&quot;</span><span class="p">)</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Executing LD pruning with the following parameters:&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;LD pruning parameters: maf=</span><span class="si">{</span><span class="n">maf</span><span class="si">}</span><span class="s2">, geno=</span><span class="si">{</span><span class="n">geno</span><span class="si">}</span><span class="s2">, mind=</span><span class="si">{</span><span class="n">mind</span><span class="si">}</span><span class="s2">, hwe=</span><span class="si">{</span><span class="n">hwe</span><span class="si">}</span><span class="s2">, ind_pair=</span><span class="si">{</span><span class="n">ind_pair</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="n">cpu_count</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">cpu_count</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">max_threads</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">cpu_count</span> <span class="o">-</span> <span class="mi">2</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Dynamically calculate fallback as half of available cores or default to 2</span>
            <span class="n">max_threads</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="p">(</span><span class="n">psutil</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">(</span><span class="n">logical</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="ow">or</span> <span class="mi">2</span><span class="p">)</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span>

        <span class="c1"># Get the virtual memory details</span>
        <span class="n">memory_info</span> <span class="o">=</span> <span class="n">psutil</span><span class="o">.</span><span class="n">virtual_memory</span><span class="p">()</span>
        <span class="n">available_memory_mb</span> <span class="o">=</span> <span class="n">memory_info</span><span class="o">.</span><span class="n">available</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1024</span> <span class="o">*</span> <span class="mi">1024</span><span class="p">)</span>
        <span class="n">memory</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="mi">2</span><span class="o">*</span><span class="n">available_memory_mb</span><span class="o">/</span><span class="mi">3</span><span class="p">,</span><span class="mi">0</span><span class="p">)</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Using </span><span class="si">{</span><span class="n">max_threads</span><span class="si">}</span><span class="s2"> threads for PLINK commands&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Available memory: </span><span class="si">{</span><span class="n">available_memory_mb</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2"> MB, using </span><span class="si">{</span><span class="n">memory</span><span class="si">}</span><span class="s2"> MB for PLINK commands&quot;</span><span class="p">)</span>

        <span class="c1"># generates prune.in and prune.out files</span>
        <span class="n">plink_cmd1</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink --bfile </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">input_path</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="si">}</span><span class="s2"> --maf </span><span class="si">{</span><span class="n">maf</span><span class="si">}</span><span class="s2"> --geno </span><span class="si">{</span><span class="n">geno</span><span class="si">}</span><span class="s2"> --mind </span><span class="si">{</span><span class="n">mind</span><span class="si">}</span><span class="s2"> --hwe </span><span class="si">{</span><span class="n">hwe</span><span class="si">}</span><span class="s2"> --exclude </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">high_ld_regions</span><span class="si">}</span><span class="s2"> --range --indep-pairwise </span><span class="si">{</span><span class="n">ind_pair</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">ind_pair</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">ind_pair</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="si">}</span><span class="s2"> --threads </span><span class="si">{</span><span class="n">max_threads</span><span class="si">}</span><span class="s2"> --memory </span><span class="si">{</span><span class="n">memory</span><span class="si">}</span><span class="s2"> --out </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="c1"># prune and creates a filtered binary file</span>
        <span class="n">plink_cmd2</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink --bfile </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">input_path</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="si">}</span><span class="s2"> --keep-allele-order --extract </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="o">+</span><span class="s1">&#39;.prune.in&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2"> --make-bed --threads </span><span class="si">{</span><span class="n">max_threads</span><span class="si">}</span><span class="s2"> --memory </span><span class="si">{</span><span class="n">memory</span><span class="si">}</span><span class="s2"> --out </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="o">+</span><span class="s1">&#39;-LDpruned&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="c1"># execute plink command</span>
        <span class="n">cmds</span> <span class="o">=</span> <span class="p">[</span><span class="n">plink_cmd1</span><span class="p">,</span> <span class="n">plink_cmd2</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">cmd</span> <span class="ow">in</span> <span class="n">cmds</span><span class="p">:</span>
            <span class="n">shell_do</span><span class="p">(</span><span class="n">cmd</span><span class="p">,</span> <span class="n">log</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="k">return</span></div>


<div class="viewcode-block" id="UMAPplot.compute_pcas">
<a class="viewcode-back" href="../../PopStructure.html#ideal_genom_qc.PopStructure.UMAPplot.compute_pcas">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">compute_pcas</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">pca</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Computes Principal Component Analysis (PCA) using PLINK.</span>

<span class="sd">        This method performs PCA on the LD-pruned dataset using PLINK&#39;s --pca command.</span>
<span class="sd">        The analysis generates eigenvectors and eigenvalues that can be used for</span>
<span class="sd">        population structure analysis and visualization.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        pca : int, default=10</span>
<span class="sd">            Number of principal components to compute. Should be a positive integer.</span>
<span class="sd">            Values below 3 will trigger a warning as they may be insufficient for </span>
<span class="sd">            meaningful analysis.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None</span>
<span class="sd">            Results are written to disk in the results directory with the input_name prefix.</span>

<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        TypeError</span>
<span class="sd">            If pca parameter is not an integer.</span>
<span class="sd">        ValueError</span>
<span class="sd">            If pca parameter is not positive.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        - If recompute_pca is False, the method will skip PCA computation</span>
<span class="sd">        - Uses PLINK&#39;s --pca command on the LD-pruned dataset</span>
<span class="sd">        - Output files are saved in the results directory specified during initialization</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">recompute_pca</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;`recompuite_pca` is set to </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">recompute_pca</span><span class="si">}</span><span class="s2">. PCA will be skipped.&quot;</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;PCA already performed. Skipping this step.&quot;</span><span class="p">)</span>
            <span class="k">return</span>

        <span class="c1"># Check type of pca</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">pca</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;pca should be of type int.&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">pca</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;pca should be a positive integer.&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">pca</span> <span class="o">&lt;=</span> <span class="mi">3</span><span class="p">:</span>
            <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;The &#39;pca&#39; value </span><span class="si">{</span><span class="n">pca</span><span class="si">}</span><span class="s2"> is low. Consider increasing it for better results.&quot;</span><span class="p">,</span> <span class="ne">UserWarning</span><span class="p">)</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Executing PCA with the following parameters:&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;PCA parameters: pca=</span><span class="si">{</span><span class="n">pca</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># runs pca analysis</span>
        <span class="n">plink_cmd1</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink --bfile </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="o">+</span><span class="s1">&#39;-LDpruned&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2"> --keep-allele-order --out </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="si">}</span><span class="s2"> --pca </span><span class="si">{</span><span class="n">pca</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="n">shell_do</span><span class="p">(</span><span class="n">plink_cmd1</span><span class="p">,</span> <span class="n">log</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="k">return</span></div>

    
<div class="viewcode-block" id="UMAPplot.generate_plots">
<a class="viewcode-back" href="../../PopStructure.html#ideal_genom_qc.PopStructure.UMAPplot.generate_plots">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">generate_plots</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">color_hue_file</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Path</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">case_control_markers</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="p">:</span> <span class="nb">list</span> <span class="o">=</span> <span class="p">[</span><span class="mi">5</span><span class="p">],</span> <span class="n">min_dist</span><span class="p">:</span> <span class="nb">list</span> <span class="o">=</span> <span class="p">[</span><span class="mf">0.5</span><span class="p">],</span> <span class="n">metric</span><span class="p">:</span> <span class="nb">list</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;euclidean&#39;</span><span class="p">],</span> <span class="n">random_state</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="nb">format</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;pdf&#39;</span><span class="p">,</span> <span class="n">umap_kwargs</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">())</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Generate UMAP plots with different parameter combinations.</span>
<span class="sd">        This method generates UMAP (Uniform Manifold Approximation and Projection) plots using various </span>
<span class="sd">        combinations of parameters. It can incorporate color coding based on metadata and case-control markers.</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        color_hue_file : Path, optional</span>
<span class="sd">            Path to a tab-separated file containing color hue information. The file should have at least 3 columns,</span>
<span class="sd">            where the first two are ID1 and ID2, and the third column contains the values for color coding.</span>
<span class="sd">            Default is None.</span>
<span class="sd">        case_control_markers : bool, optional</span>
<span class="sd">            Whether to include case-control markers in the plots. If True, reads from the .fam file.</span>
<span class="sd">            Default is True. If color_hue_file is not provided, the difference between case and control will be used as hue.</span>
<span class="sd">        n_neighbors : list of int, optional</span>
<span class="sd">            List of values for the n_neighbors parameter in UMAP. Each value must be positive.</span>
<span class="sd">            Default is [5].</span>
<span class="sd">        min_dist : list of float, optional</span>
<span class="sd">            List of values for the min_dist parameter in UMAP. Each value must be non-negative.</span>
<span class="sd">            Default is [0.5].</span>
<span class="sd">        metric : list of str, optional</span>
<span class="sd">            List of distance metrics to use in UMAP.</span>
<span class="sd">            Default is [&#39;euclidean&#39;].</span>
<span class="sd">        random_state : int, optional</span>
<span class="sd">            Random seed for reproducibility. Must be non-negative.</span>
<span class="sd">            Default is None.</span>
<span class="sd">        format : str, optional</span>
<span class="sd">            Format of the output plot files. Supported formats are &#39;pdf&#39;, &#39;png&#39;, &#39;jpeg&#39;, etc.</span>
<span class="sd">        umap_kwargs : dict, optional</span>
<span class="sd">            Additional keyword arguments to pass to the UMAP constructor.</span>
<span class="sd">            Default is an empty dictionary.</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None</span>
<span class="sd">            Saves UMAP plots as pdf files (by default but can be modified) and parameters as a CSV file in the results directory.</span>
<span class="sd">        </span>
<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        TypeError</span>
<span class="sd">            If input parameters are not of the correct type.</span>
<span class="sd">        ValueError</span>
<span class="sd">            If input parameters have invalid values.</span>
<span class="sd">        FileNotFoundError</span>
<span class="sd">            If color_hue_file is specified but not found.</span>
<span class="sd">        </span>
<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        The method creates a grid of all possible parameter combinations and generates a UMAP plot for each.</span>
<span class="sd">        Parameters for each plot are saved in &#39;plots_parameters.csv&#39;.</span>
<span class="sd">        &quot;&quot;&quot;</span>


        <span class="c1"># Check type of n_neighbors</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">n_neighbors</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;n_neighbors should be of type list.&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">all</span><span class="p">(</span><span class="nb">isinstance</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="nb">int</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">n_neighbors</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;n_neighbors should be a list of integers.&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">all</span><span class="p">(</span><span class="n">i</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">n_neighbors</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;n_neighbors should be a list of positive integers.&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">n_neighbors</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;n_neighbors should not be an empty list.&quot;</span><span class="p">)</span>
        
        <span class="c1"># Check type of min_dist</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">min_dist</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;min_dist should be of type list.&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">all</span><span class="p">(</span><span class="nb">isinstance</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="nb">float</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">min_dist</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;min_dist should be a list of floats.&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">all</span><span class="p">(</span><span class="n">i</span> <span class="o">&gt;=</span> <span class="mi">0</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">min_dist</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;min_dist should be a list of non-negative floats.&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">min_dist</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;min_dist should not be an empty list.&quot;</span><span class="p">)</span>
        
        <span class="c1"># Check type of metric</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">metric</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;metric should be of type list.&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">all</span><span class="p">(</span><span class="nb">isinstance</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">metric</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;metric should be a list of strings.&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">metric</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;metric should not be an empty list.&quot;</span><span class="p">)</span>
        
        <span class="c1"># Check type of random_state</span>
        <span class="k">if</span> <span class="n">random_state</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">random_state</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
                <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;random_state should be of type int.&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">random_state</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;random_state should be a non-negative integer.&quot;</span><span class="p">)</span>
            
        <span class="c1"># Check if color_hue_file is a file</span>
        <span class="k">if</span> <span class="n">color_hue_file</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">color_hue_file</span><span class="p">,</span> <span class="n">Path</span><span class="p">):</span>
                <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;color_hue_file should be a Path object.&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">color_hue_file</span><span class="o">.</span><span class="n">is_file</span><span class="p">():</span>
                <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;color_hue_file not found at </span><span class="si">{</span><span class="n">color_hue_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        
        <span class="c1"># Check if case_control_markers is a boolean</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">case_control_markers</span><span class="p">,</span> <span class="nb">bool</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;case_control_markers should be of type bool.&quot;</span><span class="p">)</span>
        
        <span class="k">if</span> <span class="nb">format</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;pdf&#39;</span><span class="p">,</span> <span class="s1">&#39;png&#39;</span><span class="p">,</span> <span class="s1">&#39;jpeg&#39;</span><span class="p">,</span> <span class="s1">&#39;svg&#39;</span><span class="p">]:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;format should be one of &#39;pdf&#39;, &#39;png&#39;, &#39;svg&#39;, or &#39;jpeg&#39;.&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">umap_kwargs</span><span class="p">,</span> <span class="nb">dict</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;umap_kwargs should be a dictionary.&quot;</span><span class="p">)</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Generating UMAP plots with the following parameters:&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;UMAP parameters: n_neighbors=</span><span class="si">{</span><span class="n">n_neighbors</span><span class="si">}</span><span class="s2">, min_dist=</span><span class="si">{</span><span class="n">min_dist</span><span class="si">}</span><span class="s2">, metric=</span><span class="si">{</span><span class="n">metric</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Random state: </span><span class="si">{</span><span class="n">random_state</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Color hue file: </span><span class="si">{</span><span class="n">color_hue_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Case control markers: </span><span class="si">{</span><span class="n">case_control_markers</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># generate a parameter grid</span>
        <span class="n">params_dict</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s1">&#39;n_neighbors&#39;</span><span class="p">:</span> <span class="n">n_neighbors</span><span class="p">,</span>
            <span class="s1">&#39;min_dist&#39;</span>   <span class="p">:</span> <span class="n">min_dist</span><span class="p">,</span>
            <span class="s1">&#39;metric&#39;</span>     <span class="p">:</span> <span class="n">metric</span>
        <span class="p">}</span>
        <span class="n">param_grid</span> <span class="o">=</span> <span class="n">ParameterGrid</span><span class="p">(</span><span class="n">params_dict</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">color_hue_file</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>

            <span class="k">if</span> <span class="n">color_hue_file</span><span class="o">.</span><span class="n">is_file</span><span class="p">():</span>
            <span class="c1"># load color hue file</span>
                <span class="n">df_color_hue</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
                    <span class="n">color_hue_file</span><span class="p">,</span>
                    <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span>
                <span class="p">)</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Color hue file loaded from </span><span class="si">{</span><span class="n">color_hue_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Column </span><span class="si">{</span><span class="n">df_color_hue</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="si">}</span><span class="s2"> will be used for color hue&quot;</span><span class="p">)</span>
                <span class="n">df_color_hue</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;ID1&quot;</span><span class="p">,</span> <span class="s2">&quot;ID2&quot;</span><span class="p">,</span> <span class="n">df_color_hue</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">2</span><span class="p">]]</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Color hue file has </span><span class="si">{</span><span class="n">df_color_hue</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2"> rows and </span><span class="si">{</span><span class="n">df_color_hue</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s2"> columns&quot;</span><span class="p">)</span>
                <span class="n">hue_col</span> <span class="o">=</span> <span class="n">df_color_hue</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;color_hue_file not found at </span><span class="si">{</span><span class="n">color_hue_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">hue_col</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="k">if</span> <span class="n">case_control_markers</span><span class="p">:</span>
            <span class="c1"># load case control markers</span>
            <span class="n">df_fam</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">input_path</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="o">+</span><span class="s1">&#39;.fam&#39;</span><span class="p">),</span>
                <span class="n">sep</span><span class="o">=</span><span class="sa">r</span><span class="s1">&#39;\s+&#39;</span><span class="p">,</span>
                <span class="n">engine</span><span class="o">=</span><span class="s1">&#39;python&#39;</span>
            <span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Case-control labels loaded from </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">input_path</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="o">+</span><span class="s1">&#39;.fam&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            
            <span class="n">df_fam</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;ID1&quot;</span><span class="p">,</span> <span class="s2">&quot;ID2&quot;</span><span class="p">,</span> <span class="s2">&quot;F_ID&quot;</span><span class="p">,</span> <span class="s2">&quot;M_ID&quot;</span><span class="p">,</span> <span class="s2">&quot;Sex&quot;</span><span class="p">,</span> <span class="s2">&quot;Phenotype&quot;</span><span class="p">]</span>
            <span class="n">recode</span> <span class="o">=</span> <span class="p">{</span><span class="mi">1</span><span class="p">:</span><span class="s1">&#39;Control&#39;</span><span class="p">,</span> <span class="mi">2</span><span class="p">:</span><span class="s1">&#39;Patient&#39;</span><span class="p">}</span>
            <span class="n">df_fam</span><span class="p">[</span><span class="s2">&quot;Phenotype&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_fam</span><span class="p">[</span><span class="s2">&quot;Phenotype&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">recode</span><span class="p">)</span>
            <span class="n">df_fam</span> <span class="o">=</span> <span class="n">df_fam</span><span class="p">[[</span><span class="s1">&#39;ID1&#39;</span><span class="p">,</span> <span class="s1">&#39;ID2&#39;</span><span class="p">,</span> <span class="s1">&#39;Phenotype&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Case-control markers file has </span><span class="si">{</span><span class="n">df_fam</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2"> rows and </span><span class="si">{</span><span class="n">df_fam</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s2"> columns&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">color_hue_file</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">case_control_markers</span><span class="p">:</span>
            <span class="c1"># merge color hue file with case control markers</span>
            <span class="n">df_metadata</span> <span class="o">=</span> <span class="n">df_color_hue</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span>
                <span class="n">df_fam</span><span class="p">,</span>
                <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;ID1&#39;</span><span class="p">,</span> <span class="s1">&#39;ID2&#39;</span><span class="p">],</span>
                <span class="n">how</span><span class="o">=</span><span class="s1">&#39;inner&#39;</span>
            <span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Color hue file merged with case control markers file&quot;</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Merged file has </span><span class="si">{</span><span class="n">df_metadata</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2"> rows and </span><span class="si">{</span><span class="n">df_metadata</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s2"> columns&quot;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">color_hue_file</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">df_metadata</span> <span class="o">=</span> <span class="n">df_color_hue</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Color hue file used as metadata&quot;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">case_control_markers</span><span class="p">:</span>
            <span class="n">df_metadata</span> <span class="o">=</span> <span class="n">df_fam</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Case control markers file used as metadata&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">df_metadata</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No metadata file provided&quot;</span><span class="p">)</span>

        <span class="n">count</span><span class="o">=</span><span class="mi">1</span>

        <span class="c1"># create a dataframe to store parameters</span>
        <span class="n">df_params</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
            <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;n_neighbors&#39;</span><span class="p">,</span> <span class="s1">&#39;min_dist&#39;</span><span class="p">,</span> <span class="s1">&#39;metric&#39;</span><span class="p">,</span> <span class="s1">&#39;warnings&#39;</span><span class="p">]</span>
        <span class="p">)</span>

        <span class="k">for</span> <span class="n">params</span> <span class="ow">in</span> <span class="n">param_grid</span><span class="p">:</span>

            <span class="c1"># generate umap plot for data that passed QC</span>
            <span class="n">warnings</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_umap_plots</span><span class="p">(</span>
                <span class="n">plot_name</span>   <span class="o">=</span><span class="sa">f</span><span class="s2">&quot;umap_2d_</span><span class="si">{</span><span class="n">count</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="nb">format</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
                <span class="n">n_neighbors</span> <span class="o">=</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;n_neighbors&#39;</span><span class="p">],</span>
                <span class="n">min_dist</span>    <span class="o">=</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;min_dist&#39;</span><span class="p">],</span>
                <span class="n">metric</span>      <span class="o">=</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;metric&#39;</span><span class="p">],</span>
                <span class="n">random_state</span><span class="o">=</span><span class="n">random_state</span><span class="p">,</span>
                <span class="n">df_metadata</span> <span class="o">=</span><span class="n">df_metadata</span><span class="p">,</span>
                <span class="n">hue_col</span>     <span class="o">=</span><span class="n">hue_col</span><span class="p">,</span>
                <span class="n">umap_kwargs</span><span class="o">=</span><span class="n">umap_kwargs</span>
            <span class="p">)</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">files_to_keep</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;umap_2d_</span><span class="si">{</span><span class="n">count</span><span class="si">}</span><span class="s2">.jpeg&quot;</span><span class="p">)</span>

            <span class="n">df_params</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">count</span><span class="p">,</span> <span class="s1">&#39;n_neighbors&#39;</span><span class="p">]</span><span class="o">=</span> <span class="n">params</span><span class="p">[</span><span class="s1">&#39;n_neighbors&#39;</span><span class="p">]</span>
            <span class="n">df_params</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">count</span><span class="p">,</span> <span class="s1">&#39;min_dist&#39;</span><span class="p">]</span>   <span class="o">=</span> <span class="n">params</span><span class="p">[</span><span class="s1">&#39;min_dist&#39;</span><span class="p">]</span>
            <span class="n">df_params</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">count</span><span class="p">,</span> <span class="s1">&#39;metric&#39;</span><span class="p">]</span>     <span class="o">=</span> <span class="n">params</span><span class="p">[</span><span class="s1">&#39;metric&#39;</span><span class="p">]</span>
            <span class="n">df_params</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">count</span><span class="p">,</span> <span class="s1">&#39;warnings&#39;</span><span class="p">]</span>   <span class="o">=</span> <span class="n">warnings</span>

            <span class="n">count</span> <span class="o">+=</span><span class="mi">1</span>

        <span class="c1"># save parameters to a csv file</span>
        <span class="n">df_params</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span>
            <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="p">,</span> <span class="s1">&#39;plots_parameters.csv&#39;</span><span class="p">),</span>
            <span class="n">index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span>
        <span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">files_to_keep</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;plots_parameters.csv&#39;</span><span class="p">)</span>

        <span class="c1"># generate PCA plot for comparison</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_pca_plot</span><span class="p">(</span>
            <span class="n">df_metadata</span> <span class="o">=</span> <span class="n">df_metadata</span><span class="p">,</span> 
            <span class="n">plot_dir</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">plots_dir</span><span class="p">,</span>
            <span class="n">plot_name</span><span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;pca_plot.</span><span class="si">{</span><span class="nb">format</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span>
            <span class="n">hue_col</span> <span class="o">=</span> <span class="n">hue_col</span>
        <span class="p">)</span>

        <span class="k">return</span></div>

    
    <span class="k">def</span><span class="w"> </span><span class="nf">_umap_plots</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">plot_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">min_dist</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span> <span class="n">metric</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">random_state</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">df_metadata</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">hue_col</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">umap_kwargs</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">())</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Generate and save UMAP (Uniform Manifold Approximation and Projection) plots from PCA data.</span>
<span class="sd">        This method reads eigenvector data from a file, performs UMAP dimensionality reduction,</span>
<span class="sd">        and creates a 2D scatter plot with optional metadata coloring/styling.</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        plot_name : str</span>
<span class="sd">            Name of the output plot file</span>
<span class="sd">        n_neighbors : int</span>
<span class="sd">            Number of neighbors to consider for manifold approximation</span>
<span class="sd">        min_dist : float</span>
<span class="sd">            Minimum distance between points in the low dimensional representation</span>
<span class="sd">        metric : str</span>
<span class="sd">            Distance metric to use for UMAP calculation</span>
<span class="sd">        random_state : int, optional</span>
<span class="sd">            Random seed for reproducibility</span>
<span class="sd">        df_metadata : pd.DataFrame, optional</span>
<span class="sd">            DataFrame containing metadata to merge with the eigenvector data</span>
<span class="sd">        hue_col : str, optional</span>
<span class="sd">            Column name in metadata to use for point coloring</span>
<span class="sd">            Additional keyword arguments to pass to UMAP</span>
<span class="sd">        umap_kwargs : dict, optional</span>
<span class="sd">            Additional keyword arguments to pass to UMAP constructor</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        list or None</span>
<span class="sd">            List of warning messages if any were generated during execution, None otherwise</span>
<span class="sd">        </span>
<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        The method expects an eigenvector file in the results directory with the naming pattern</span>
<span class="sd">        {input_name}.eigenvec. The plot will be saved in the plots directory with the provided</span>
<span class="sd">        plot_name.</span>
<span class="sd">        If metadata is provided and contains a &#39;Phenotype&#39; column, it will be used for styling</span>
<span class="sd">        points unless hue_col is specified.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># load .eigenvec file</span>
        <span class="n">df_eigenvec</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="o">+</span><span class="s1">&#39;.eigenvec&#39;</span><span class="p">),</span>
            <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="n">sep</span><span class="o">=</span><span class="s1">&#39; &#39;</span>
        <span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Eigenvector file loaded from </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="o">+</span><span class="s1">&#39;.eigenvec&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Eigenvector file has </span><span class="si">{</span><span class="n">df_eigenvec</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2"> rows and </span><span class="si">{</span><span class="n">df_eigenvec</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s2"> columns&quot;</span><span class="p">)</span>

        <span class="c1"># rename columns</span>
        <span class="n">num_pc</span> <span class="o">=</span> <span class="n">df_eigenvec</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">-</span><span class="mi">2</span>
        <span class="n">new_cols</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;pca_</span><span class="si">{</span><span class="n">k</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="n">num_pc</span><span class="o">+</span><span class="mi">1</span><span class="p">)]</span>
        <span class="n">df_eigenvec</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ID1&#39;</span><span class="p">,</span> <span class="s1">&#39;ID2&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">new_cols</span>

        <span class="n">df_ids</span> <span class="o">=</span> <span class="n">df_eigenvec</span><span class="p">[[</span><span class="s1">&#39;ID1&#39;</span><span class="p">,</span> <span class="s1">&#39;ID2&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">df_vals</span><span class="o">=</span> <span class="n">df_eigenvec</span><span class="p">[</span><span class="n">new_cols</span><span class="p">]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">df_metadata</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="c1"># merge metadata with eigenvector data</span>
            <span class="n">df_ids</span> <span class="o">=</span> <span class="n">df_ids</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span>
                <span class="n">df_metadata</span><span class="p">,</span>
                <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;ID1&#39;</span><span class="p">,</span> <span class="s1">&#39;ID2&#39;</span><span class="p">],</span>
                <span class="n">how</span><span class="o">=</span><span class="s1">&#39;inner&#39;</span>
            <span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Metadata file merged with eigenvector file&quot;</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Merged file has </span><span class="si">{</span><span class="n">df_ids</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2"> rows and </span><span class="si">{</span><span class="n">df_ids</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s2"> columns&quot;</span><span class="p">)</span>

            <span class="k">if</span> <span class="s1">&#39;Phenotype&#39;</span> <span class="ow">in</span> <span class="n">df_ids</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
                <span class="n">style_col</span> <span class="o">=</span> <span class="s1">&#39;Phenotype&#39;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">style_col</span> <span class="o">=</span> <span class="kc">None</span>
            
            <span class="k">if</span> <span class="n">style_col</span> <span class="ow">and</span> <span class="n">hue_col</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">hue_col</span><span class="p">,</span> <span class="n">style_col</span> <span class="o">=</span> <span class="n">style_col</span><span class="p">,</span> <span class="kc">None</span>

        <span class="k">del</span> <span class="n">df_eigenvec</span>

        <span class="c1"># instantiate umap class</span>
        <span class="n">D2_redux</span> <span class="o">=</span> <span class="n">umap</span><span class="o">.</span><span class="n">UMAP</span><span class="p">(</span>
            <span class="n">n_components</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
            <span class="n">n_neighbors</span> <span class="o">=</span><span class="n">n_neighbors</span><span class="p">,</span>
            <span class="n">min_dist</span>    <span class="o">=</span><span class="n">min_dist</span><span class="p">,</span>
            <span class="n">metric</span>      <span class="o">=</span><span class="n">metric</span><span class="p">,</span>
            <span class="n">random_state</span><span class="o">=</span><span class="n">random_state</span><span class="p">,</span>
            <span class="o">**</span><span class="n">umap_kwargs</span>
        <span class="p">)</span>

        <span class="k">with</span> <span class="n">warnings</span><span class="o">.</span><span class="n">catch_warnings</span><span class="p">(</span><span class="n">record</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="k">as</span> <span class="n">w</span><span class="p">:</span>

            <span class="n">warnings</span><span class="o">.</span><span class="n">simplefilter</span><span class="p">(</span><span class="s2">&quot;always&quot;</span><span class="p">)</span>
            
            <span class="c1"># generates umap projection</span>
            <span class="n">umap_2D_proj</span> <span class="o">=</span> <span class="n">D2_redux</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">df_vals</span><span class="p">)</span>

            <span class="n">df_2D</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">df_ids</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">umap_2D_proj</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;umap1&#39;</span><span class="p">,</span> <span class="s1">&#39;umap2&#39;</span><span class="p">])],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

            <span class="k">del</span> <span class="n">df_vals</span>

            <span class="c1"># generates and saves a 2D scatter plot</span>
            <span class="c1"># size given in inches</span>
            <span class="n">sns</span><span class="o">.</span><span class="n">set_context</span><span class="p">(</span><span class="n">font_scale</span><span class="o">=</span><span class="mf">0.9</span><span class="p">)</span>
            <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>

            <span class="n">scatter_kwargs</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="k">if</span> <span class="n">style_col</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">scatter_kwargs</span><span class="p">[</span><span class="s1">&#39;style&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">style_col</span>
                
            <span class="n">scatter_plot</span><span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span>
                <span class="n">data</span><span class="o">=</span><span class="n">df_2D</span><span class="p">,</span> 
                <span class="n">x</span><span class="o">=</span><span class="s1">&#39;umap1&#39;</span><span class="p">,</span> 
                <span class="n">y</span><span class="o">=</span><span class="s1">&#39;umap2&#39;</span><span class="p">,</span> 
                <span class="n">hue</span><span class="o">=</span><span class="n">hue_col</span><span class="p">,</span>
                <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;.&#39;</span><span class="p">,</span>
                <span class="n">s</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
                <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span>
                <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span>
                <span class="o">**</span><span class="n">scatter_kwargs</span>
            <span class="p">)</span>
            <span class="k">if</span> <span class="n">df_metadata</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span>
                    <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">0.</span><span class="p">,</span> <span class="mf">1.02</span><span class="p">,</span> <span class="mf">1.</span><span class="p">,</span> <span class="mf">.102</span><span class="p">),</span> 
                    <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;lower left&#39;</span><span class="p">,</span>
                    <span class="n">ncols</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> 
                    <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;expand&quot;</span><span class="p">,</span> 
                    <span class="n">borderaxespad</span><span class="o">=</span><span class="mf">0.</span><span class="p">,</span>
                    <span class="n">fontsize</span><span class="o">=</span><span class="mi">7</span><span class="p">,</span>
                    <span class="n">markerscale</span><span class="o">=</span><span class="mi">2</span>
                <span class="p">)</span>
                
            <span class="c1"># Set tick label size</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;both&#39;</span><span class="p">,</span> <span class="n">labelsize</span><span class="o">=</span><span class="mi">7</span><span class="p">)</span>

            <span class="c1"># Set axis label and size</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;UMAP1&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">7</span><span class="p">)</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;UMAP2&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">7</span><span class="p">)</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="n">adjustable</span><span class="o">=</span><span class="s1">&#39;datalim&#39;</span><span class="p">)</span>

            <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>

            <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">plots_dir</span> <span class="o">/</span> <span class="n">plot_name</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">500</span><span class="p">)</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>


            <span class="n">warning</span> <span class="o">=</span> <span class="p">[</span><span class="n">warn</span><span class="o">.</span><span class="n">message</span><span class="o">.</span><span class="n">args</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">warn</span> <span class="ow">in</span> <span class="n">w</span><span class="p">]</span> <span class="c1"># type: ignore</span>
            <span class="k">return</span> <span class="n">warning</span>
        
    <span class="k">def</span><span class="w"> </span><span class="nf">_pca_plot</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">df_metadata</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">hue_col</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">plot_dir</span><span class="p">:</span> <span class="n">Path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(),</span> <span class="n">plot_name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;pca_plot.svg&#39;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">plot_dir</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;STEP: Generating PCA plots: `plot_dir` does not exist.&#39;</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;STEP: Generating PCA plots: pca plots will be saved in </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">output_path</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="n">plot_dir</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_path</span>

        <span class="n">df_eigenvec</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="o">+</span><span class="s1">&#39;.eigenvec&#39;</span><span class="p">),</span>
            <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="n">sep</span><span class="o">=</span><span class="s1">&#39; &#39;</span>
        <span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Eigenvector file loaded from </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="o">+</span><span class="s1">&#39;.eigenvec&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Eigenvector file has </span><span class="si">{</span><span class="n">df_eigenvec</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2"> rows and </span><span class="si">{</span><span class="n">df_eigenvec</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s2"> columns&quot;</span><span class="p">)</span>

        <span class="c1"># load .eigenval file and calculate variance explained by the first two PCs</span>
        <span class="n">df_eigenval</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="o">+</span><span class="s1">&#39;.eigenval&#39;</span><span class="p">),</span>
            <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="n">sep</span>   <span class="o">=</span><span class="sa">r</span><span class="s2">&quot;\s+&quot;</span><span class="p">,</span>
            <span class="n">engine</span><span class="o">=</span><span class="s1">&#39;python&#39;</span>
        <span class="p">)</span>

        <span class="n">total_variance</span> <span class="o">=</span> <span class="n">df_eigenval</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
        <span class="n">pc1_var</span> <span class="o">=</span> <span class="n">df_eigenval</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">pc2_var</span> <span class="o">=</span> <span class="n">df_eigenval</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>

        <span class="n">pc1_var_perc</span> <span class="o">=</span> <span class="nb">round</span><span class="p">((</span><span class="n">pc1_var</span> <span class="o">/</span> <span class="n">total_variance</span><span class="p">)</span> <span class="o">*</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
        <span class="n">pc2_var_perc</span> <span class="o">=</span> <span class="nb">round</span><span class="p">((</span><span class="n">pc2_var</span> <span class="o">/</span> <span class="n">total_variance</span><span class="p">)</span> <span class="o">*</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>

        <span class="c1"># rename columns</span>
        <span class="n">num_pc</span> <span class="o">=</span> <span class="n">df_eigenvec</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">-</span><span class="mi">2</span>
        <span class="n">new_cols</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;pc_</span><span class="si">{</span><span class="n">k</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="n">num_pc</span><span class="o">+</span><span class="mi">1</span><span class="p">)]</span>
        <span class="n">df_eigenvec</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ID1&#39;</span><span class="p">,</span> <span class="s1">&#39;ID2&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">new_cols</span>

        <span class="k">if</span> <span class="n">df_metadata</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="c1"># merge metadata with eigenvector data</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">df_eigenvec</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span>
                <span class="n">df_metadata</span><span class="p">,</span>
                <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;ID1&#39;</span><span class="p">,</span> <span class="s1">&#39;ID2&#39;</span><span class="p">],</span>
                <span class="n">how</span><span class="o">=</span><span class="s1">&#39;inner&#39;</span>
            <span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Metadata file merged with eigenvector file&quot;</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Merged file has </span><span class="si">{</span><span class="n">df</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2"> rows and </span><span class="si">{</span><span class="n">df</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s2"> columns&quot;</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">hue_col</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">hue_col</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
                    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Using column &#39;</span><span class="si">{</span><span class="n">hue_col</span><span class="si">}</span><span class="s2">&#39; for coloring the PCA plot&quot;</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Column &#39;</span><span class="si">{</span><span class="n">hue_col</span><span class="si">}</span><span class="s2">&#39; not found in metadata file&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">hue_col</span> <span class="o">=</span> <span class="n">df_metadata</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="k">if</span> <span class="n">df_metadata</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">2</span> <span class="k">else</span> <span class="kc">None</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">df_eigenvec</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No metadata file provided. Using eigenvector file only.&quot;</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Eigenvector file has </span><span class="si">{</span><span class="n">df</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2"> rows and </span><span class="si">{</span><span class="n">df</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s2"> columns&quot;</span><span class="p">)</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Generating PCA plot with </span><span class="si">{</span><span class="n">df</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2"> samples and </span><span class="si">{</span><span class="n">df</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s2"> features&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Variance explained by PC1: </span><span class="si">{</span><span class="n">pc1_var_perc</span><span class="si">}</span><span class="s2">%, PC2: </span><span class="si">{</span><span class="n">pc2_var_perc</span><span class="si">}</span><span class="s2">%&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;df columns: </span><span class="si">{</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># generates a 2D scatter plot</span>
        <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span><span class="mi">10</span><span class="p">))</span>
        <span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">df</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s1">&#39;pc_1&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;pc_2&#39;</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="n">hue_col</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">70</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="n">adjustable</span><span class="o">=</span><span class="s1">&#39;datalim&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;PC_1 (</span><span class="si">{</span><span class="n">pc1_var_perc</span><span class="si">}</span><span class="s1">%)&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;PC_2 (</span><span class="si">{</span><span class="n">pc2_var_perc</span><span class="si">}</span><span class="s1">%)&#39;</span><span class="p">)</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">plot_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;2D-aspect-equal-</span><span class="si">{</span><span class="n">plot_name</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">400</span><span class="p">)</span>

        <span class="n">fig</span><span class="o">.</span><span class="n">clf</span><span class="p">()</span>

        <span class="k">return</span></div>


<div class="viewcode-block" id="FstSummary">
<a class="viewcode-back" href="../../PopStructure.html#ideal_genom_qc.PopStructure.FstSummary">[docs]</a>
<span class="k">class</span><span class="w"> </span><span class="nc">FstSummary</span><span class="p">:</span>

<div class="viewcode-block" id="FstSummary.__init__">
<a class="viewcode-back" href="../../PopStructure.html#ideal_genom_qc.PopStructure.FstSummary.__init__">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">input_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">input_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">output_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">high_ld_file</span><span class="p">:</span> <span class="n">Path</span><span class="o">=</span><span class="n">Path</span><span class="p">(),</span> <span class="n">built</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;38&#39;</span><span class="p">,</span> <span class="n">recompute_merge</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">reference_files</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">())</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initialize FstSummary object for Fst analysis.</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        input_path : Path</span>
<span class="sd">            Path to the directory containing input files</span>
<span class="sd">        input_name : str</span>
<span class="sd">            Name of the input file</span>
<span class="sd">        output_path : Path</span>
<span class="sd">            Path to the directory where results will be saved</span>
<span class="sd">        </span>
<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        TypeError</span>
<span class="sd">            If input types are incorrect for any parameter</span>
<span class="sd">        FileNotFoundError</span>
<span class="sd">            If input_path or output_path do not exist</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">input_path</span><span class="p">,</span> <span class="n">Path</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;input_path should be a Path object&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="n">Path</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;output_path should be a Path object&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">input_name</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span> 
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;input_name should be a string&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">input_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="s2">&quot;input_path does not exist&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">output_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="s2">&quot;output_path does not exist&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">built</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;built should be a string&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">built</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;37&#39;</span><span class="p">,</span> <span class="s1">&#39;38&#39;</span><span class="p">]:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;built should be either &#39;37&#39; or &#39;38&#39;&quot;</span><span class="p">)</span> 
        <span class="k">if</span> <span class="ow">not</span> <span class="n">high_ld_file</span><span class="o">.</span><span class="n">is_file</span><span class="p">():</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;High LD file not found at </span><span class="si">{</span><span class="n">high_ld_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;High LD file will be fetched from the package&#39;</span><span class="p">)</span>
            
            <span class="n">ld_fetcher</span> <span class="o">=</span> <span class="n">FetcherLDRegions</span><span class="p">()</span>
            <span class="n">ld_fetcher</span><span class="o">.</span><span class="n">get_ld_regions</span><span class="p">()</span>

            <span class="k">if</span> <span class="n">ld_fetcher</span><span class="o">.</span><span class="n">ld_regions</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="s2">&quot;Could not fetch LD regions file.&quot;</span><span class="p">)</span>
                
            <span class="n">high_ld_file</span> <span class="o">=</span> <span class="n">ld_fetcher</span><span class="o">.</span><span class="n">ld_regions</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;High LD file fetched from the package and saved at </span><span class="si">{</span><span class="n">high_ld_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">input_path</span> <span class="o">=</span> <span class="n">input_path</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">input_name</span> <span class="o">=</span> <span class="n">input_name</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">output_path</span><span class="o">=</span> <span class="n">output_path</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">recompute_merge</span> <span class="o">=</span> <span class="n">recompute_merge</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">high_ld_regions</span> <span class="o">=</span> <span class="n">high_ld_file</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">built</span> <span class="o">=</span> <span class="n">built</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">reference_files</span><span class="p">:</span>

            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No reference files provided. Fetching 1000 Genomes reference data for built </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">built</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="n">fetcher</span> <span class="o">=</span> <span class="n">Fetcher1000Genome</span><span class="p">(</span><span class="n">built</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">built</span><span class="p">)</span>
            <span class="n">fetcher</span><span class="o">.</span><span class="n">get_1000genomes</span><span class="p">()</span>
            <span class="n">fetcher</span><span class="o">.</span><span class="n">get_1000genomes_binaries</span><span class="p">()</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">reference_files</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s1">&#39;bim&#39;</span><span class="p">:</span> <span class="n">fetcher</span><span class="o">.</span><span class="n">bim_file</span><span class="p">,</span>
                <span class="s1">&#39;bed&#39;</span><span class="p">:</span> <span class="n">fetcher</span><span class="o">.</span><span class="n">bed_file</span><span class="p">,</span>
                <span class="s1">&#39;fam&#39;</span><span class="p">:</span> <span class="n">fetcher</span><span class="o">.</span><span class="n">fam_file</span><span class="p">,</span>
                <span class="s1">&#39;psam&#39;</span><span class="p">:</span> <span class="n">fetcher</span><span class="o">.</span><span class="n">psam_file</span>
            <span class="p">}</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_path</span> <span class="o">/</span> <span class="s1">&#39;fst_results&#39;</span> 
        <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">merging_dir</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="s1">&#39;merging&#39;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">merging_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></div>


    <span class="k">pass</span>

<div class="viewcode-block" id="FstSummary.merge_reference_study">
<a class="viewcode-back" href="../../PopStructure.html#ideal_genom_qc.PopStructure.FstSummary.merge_reference_study">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">merge_reference_study</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ind_pair</span><span class="p">:</span> <span class="nb">list</span> <span class="o">=</span> <span class="p">[</span><span class="mi">50</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mf">0.2</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Merge reference and study data by applying quality control filters and merging steps.</span>
<span class="sd">        This method performs a series of quality control steps to merge study data with reference data:</span>
<span class="sd">        1. Filters problematic SNPs</span>
<span class="sd">        2. Performs LD pruning</span>
<span class="sd">        3. Fixes chromosome mismatches</span>
<span class="sd">        4. Fixes position mismatches  </span>
<span class="sd">        5. Fixes allele flips</span>
<span class="sd">        6. Removes remaining mismatches</span>
<span class="sd">        7. Merges the datasets</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ind_pair : list, default [50, 5, 0.2]</span>
<span class="sd">            Parameters for LD pruning: [window size, step size, r2 threshold]</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None</span>
<span class="sd">        </span>
<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        If recompute_merge is False, the method will skip the merging process and expect</span>
<span class="sd">        merged data to already exist in the merging directory.</span>
<span class="sd">        </span>
<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        TypeError</span>
<span class="sd">            If ind_pair is not a list</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ind_pair</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;ind_pair should be a list&quot;</span><span class="p">)</span>
        
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">recompute_merge</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;STEP: Merging study and reference data: recompute_merge is set to False. Skipping merging step&quot;</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;STEP: Merging study and reference data: merged data is expected to be in </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">merging_dir</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">return</span>

        <span class="n">rgm</span> <span class="o">=</span> <span class="n">ReferenceGenomicMerger</span><span class="p">(</span>
            <span class="n">input_path</span><span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">input_path</span><span class="p">,</span>
            <span class="n">input_name</span><span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="p">,</span>
            <span class="n">output_path</span><span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">merging_dir</span><span class="p">,</span> 
            <span class="n">output_name</span><span class="o">=</span> <span class="s1">&#39;cleaned-with-ref&#39;</span><span class="p">,</span>
            <span class="n">high_ld_regions</span> <span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">high_ld_regions</span><span class="p">,</span> 
            <span class="n">reference_files</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_files</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="n">rgm</span><span class="o">.</span><span class="n">execute_rename_snpid</span><span class="p">()</span>
        <span class="n">rgm</span><span class="o">.</span><span class="n">execute_filter_prob_snps</span><span class="p">()</span>
        <span class="n">rgm</span><span class="o">.</span><span class="n">execute_ld_pruning</span><span class="p">(</span><span class="n">ind_pair</span><span class="o">=</span><span class="n">ind_pair</span><span class="p">)</span>
        <span class="n">rgm</span><span class="o">.</span><span class="n">execute_fix_chromosome_mismatch</span><span class="p">()</span>
        <span class="n">rgm</span><span class="o">.</span><span class="n">execute_fix_possition_mismatch</span><span class="p">()</span>
        <span class="n">rgm</span><span class="o">.</span><span class="n">execute_fix_allele_flip</span><span class="p">()</span>
        <span class="n">rgm</span><span class="o">.</span><span class="n">execute_remove_mismatches</span><span class="p">()</span>
        <span class="n">rgm</span><span class="o">.</span><span class="n">execute_merge_data</span><span class="p">()</span>

        <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">merging_dir</span><span class="o">.</span><span class="n">iterdir</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">file</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="ow">and</span> <span class="s1">&#39;-merged&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">file</span><span class="o">.</span><span class="n">name</span> <span class="ow">and</span> <span class="n">file</span><span class="o">.</span><span class="n">suffix</span> <span class="o">!=</span> <span class="s1">&#39;.log&#39;</span><span class="p">:</span>
                <span class="n">file</span><span class="o">.</span><span class="n">unlink</span><span class="p">()</span>

        <span class="k">return</span></div>

    
<div class="viewcode-block" id="FstSummary.add_population_tags">
<a class="viewcode-back" href="../../PopStructure.html#ideal_genom_qc.PopStructure.FstSummary.add_population_tags">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">add_population_tags</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Add population tags to the merged dataset.</span>
<span class="sd">        This method adds population super-population tags from the reference dataset to</span>
<span class="sd">        the merged dataset. It reads population information from the reference PSAM file,</span>
<span class="sd">        merges it with the study dataset, and assigns &#39;StPop&#39; (study population) to samples</span>
<span class="sd">        not present in the reference dataset.</span>

<span class="sd">        Requirements</span>
<span class="sd">        ------------</span>
<span class="sd">            - Merged dataset files (.bed, .bim, .fam) must exist in the merging directory</span>
<span class="sd">            - Reference files dictionary must contain a valid &#39;psam&#39; Path</span>
<span class="sd">        </span>
<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">            FileNotFoundError: If any of the required merged files are not found</span>
<span class="sd">            ValueError: If the reference files dictionary doesn&#39;t contain a valid &#39;psam&#39; Path</span>
<span class="sd">        </span>
<span class="sd">        Side Effects</span>
<span class="sd">        ------------</span>
<span class="sd">            - Creates a new tab-separated file with population tags at </span>
<span class="sd">              {merging_dir}/cleaned-with-ref-merged-pop-tags.csv</span>
<span class="sd">            - Sets self.population_tags to the path of the created file</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">            None</span>
<span class="sd">        &quot;&quot;&quot;</span>
       
        <span class="n">merged_bed</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">merging_dir</span> <span class="o">/</span> <span class="s1">&#39;cleaned-with-ref-merged.bed&#39;</span>
        <span class="n">merged_bim</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">merging_dir</span> <span class="o">/</span> <span class="s1">&#39;cleaned-with-ref-merged.bim&#39;</span>
        <span class="n">merged_fam</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">merging_dir</span> <span class="o">/</span> <span class="s1">&#39;cleaned-with-ref-merged.fam&#39;</span>
        
        <span class="k">if</span> <span class="ow">not</span> <span class="n">merged_bed</span><span class="o">.</span><span class="n">is_file</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Merged data file not found at </span><span class="si">{</span><span class="n">merged_bed</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">merged_bim</span><span class="o">.</span><span class="n">is_file</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Merged BIM file not found at </span><span class="si">{</span><span class="n">merged_bim</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">merged_fam</span><span class="o">.</span><span class="n">is_file</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Merged FAM file not found at </span><span class="si">{</span><span class="n">merged_fam</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="s1">&#39;psam&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_files</span> <span class="ow">or</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_files</span><span class="p">[</span><span class="s1">&#39;psam&#39;</span><span class="p">],</span> <span class="n">Path</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Reference files dictionary must contain a valid &#39;psam&#39; Path&quot;</span><span class="p">)</span>

        <span class="n">reference_tags</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_files</span><span class="p">[</span><span class="s1">&#39;psam&#39;</span><span class="p">]</span>
        
        <span class="n">df_tags</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">reference_tags</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="sa">r</span><span class="s2">&quot;\s+&quot;</span><span class="p">,</span> <span class="n">engine</span><span class="o">=</span><span class="s1">&#39;python&#39;</span><span class="p">)</span>
        <span class="n">df_tags</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;0&#39;</span>
        <span class="n">df_tags</span> <span class="o">=</span> <span class="n">df_tags</span><span class="p">[[</span><span class="s1">&#39;ID&#39;</span><span class="p">,</span> <span class="s1">&#39;#IID&#39;</span><span class="p">,</span> <span class="s1">&#39;SuperPop&#39;</span><span class="p">]]</span>
        <span class="n">df_tags</span> <span class="o">=</span> <span class="n">df_tags</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;ID&#39;</span><span class="p">:</span> <span class="s1">&#39;ID1&#39;</span><span class="p">,</span> <span class="s1">&#39;#IID&#39;</span><span class="p">:</span> <span class="s1">&#39;ID2&#39;</span><span class="p">,</span> <span class="s1">&#39;SuperPop&#39;</span><span class="p">:</span> <span class="s1">&#39;SuperPop&#39;</span><span class="p">})</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Population tags loaded from </span><span class="si">{</span><span class="n">reference_tags</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Population tags columns: </span><span class="si">{</span><span class="n">df_tags</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>

        <span class="n">df_merged_fam</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">merged_fam</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="sa">r</span><span class="s2">&quot;\s+&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">engine</span><span class="o">=</span><span class="s1">&#39;python&#39;</span><span class="p">)</span>
        <span class="n">df_merged_fam</span> <span class="o">=</span> <span class="n">df_merged_fam</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="mi">0</span><span class="p">:</span> <span class="s1">&#39;ID1&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">:</span> <span class="s1">&#39;ID2&#39;</span><span class="p">})</span>
        <span class="c1">#df_merged_fam[&#39;SuperPop&#39;] = &#39;StPop&#39;</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Merged BIM file loaded from </span><span class="si">{</span><span class="n">merged_fam</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Merged BIM file columns: </span><span class="si">{</span><span class="n">df_merged_fam</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>

        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span>
            <span class="n">df_merged_fam</span><span class="p">[[</span><span class="s1">&#39;ID1&#39;</span><span class="p">,</span> <span class="s1">&#39;ID2&#39;</span><span class="p">]],</span>
            <span class="n">df_tags</span><span class="p">,</span>
            <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;ID1&#39;</span><span class="p">,</span> <span class="s1">&#39;ID2&#39;</span><span class="p">],</span>
            <span class="n">how</span><span class="o">=</span><span class="s1">&#39;left&#39;</span>
        <span class="p">)</span>
        <span class="n">df</span><span class="p">[</span><span class="s1">&#39;SuperPop&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;SuperPop&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="s1">&#39;StPop&#39;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Added population tags to the merged dataset&quot;</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">population_tags</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">merging_dir</span> <span class="o">/</span> <span class="s1">&#39;cleaned-with-ref-merged-pop-tags.csv&#39;</span>
        <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">population_tags</span><span class="p">,</span>
            <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span>
        <span class="p">)</span>

        <span class="k">return</span></div>

    
<div class="viewcode-block" id="FstSummary.compute_fst">
<a class="viewcode-back" href="../../PopStructure.html#ideal_genom_qc.PopStructure.FstSummary.compute_fst">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">compute_fst</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Compute FST (fixation index) statistics between populations.</span>

<span class="sd">        This method calculates FST statistics between each super-population in the dataset</span>
<span class="sd">        and a study population (&#39;StPop&#39;). The process involves:</span>
<span class="sd">        1. Reading population tags from the specified file</span>
<span class="sd">        2. For each unique super-population (except &#39;StPop&#39;):</span>
<span class="sd">        - Creating population filter files (keep and within files)</span>
<span class="sd">        - Running PLINK commands to filter the dataset and compute FST statistics</span>

<span class="sd">        The method requires the following instance variables to be set:</span>
<span class="sd">            - population_tags: Path to a file containing population information</span>
<span class="sd">            - results_dir: Directory where results will be stored</span>
<span class="sd">            - merging_dir: Directory containing the merged genotype data</span>

<span class="sd">        Returns:</span>
<span class="sd">        --------</span>
<span class="sd">        None</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">df_tags</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">population_tags</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="sa">r</span><span class="s2">&quot;\s+&quot;</span><span class="p">,</span> <span class="n">engine</span><span class="o">=</span><span class="s1">&#39;python&#39;</span><span class="p">)</span>

        <span class="n">files</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>

        <span class="k">for</span> <span class="n">pop</span> <span class="ow">in</span> <span class="n">df_tags</span><span class="p">[</span><span class="s1">&#39;SuperPop&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">pop</span> <span class="o">!=</span> <span class="s1">&#39;StPop&#39;</span><span class="p">:</span>
                <span class="n">df_temp</span> <span class="o">=</span> <span class="n">df_tags</span><span class="p">[(</span><span class="n">df_tags</span><span class="p">[</span><span class="s1">&#39;SuperPop&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">pop</span><span class="p">)</span> <span class="o">|</span> <span class="p">(</span><span class="n">df_tags</span><span class="p">[</span><span class="s1">&#39;SuperPop&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;StPop&#39;</span><span class="p">)]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
                <span class="n">df_temp</span><span class="p">[[</span><span class="s1">&#39;ID1&#39;</span><span class="p">,</span> <span class="s1">&#39;ID2&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;keep-</span><span class="si">{</span><span class="n">pop</span><span class="si">}</span><span class="s1">_StPop.txt&#39;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
                <span class="n">df_temp</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;within-</span><span class="si">{</span><span class="n">pop</span><span class="si">}</span><span class="s1">_StPop.txt&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,)</span>

                <span class="n">files</span><span class="p">[</span><span class="n">pop</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;keep-</span><span class="si">{</span><span class="n">pop</span><span class="si">}</span><span class="s1">_StPop.txt&#39;</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;within-</span><span class="si">{</span><span class="n">pop</span><span class="si">}</span><span class="s1">_StPop.txt&#39;</span><span class="p">)</span>

                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Created keep and within files for population </span><span class="si">{</span><span class="n">pop</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="n">input_file</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">merging_dir</span> <span class="o">/</span> <span class="s1">&#39;cleaned-with-ref-merged&#39;</span>

        <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">files</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>

            <span class="n">keep_file</span><span class="p">,</span> <span class="n">within_file</span> <span class="o">=</span> <span class="n">files</span><span class="p">[</span><span class="n">key</span><span class="p">]</span>
            <span class="n">output_file</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;keep-</span><span class="si">{</span><span class="n">key</span><span class="si">}</span><span class="s1">-StPop&#39;</span>

            <span class="n">plink_cmd1</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink --bfile </span><span class="si">{</span><span class="n">input_file</span><span class="si">}</span><span class="s2"> --keep </span><span class="si">{</span><span class="n">keep_file</span><span class="si">}</span><span class="s2"> --make-bed --out </span><span class="si">{</span><span class="n">output_file</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="n">plink_cmd2</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink --bfile </span><span class="si">{</span><span class="n">output_file</span><span class="si">}</span><span class="s2"> --fst --within </span><span class="si">{</span><span class="n">within_file</span><span class="si">}</span><span class="s2"> --out </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="sa">f</span><span class="s1">&#39;fst-</span><span class="si">{</span><span class="n">key</span><span class="si">}</span><span class="s1">-StPop&#39;</span><span class="si">}</span><span class="s2">&quot;</span>

            <span class="n">plink_cmds</span> <span class="o">=</span> <span class="p">[</span><span class="n">plink_cmd1</span><span class="p">,</span> <span class="n">plink_cmd2</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">cmd</span> <span class="ow">in</span> <span class="n">plink_cmds</span><span class="p">:</span>
                <span class="n">shell_do</span><span class="p">(</span><span class="n">cmd</span><span class="p">,</span> <span class="n">log</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Fst computation completed for all populations.&quot;</span><span class="p">)</span>

        <span class="k">return</span></div>

    
<div class="viewcode-block" id="FstSummary.report_fst">
<a class="viewcode-back" href="../../PopStructure.html#ideal_genom_qc.PopStructure.FstSummary.report_fst">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">report_fst</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Generate a report of Fst results.</span>
<span class="sd">        This method reads the Fst results from the results directory and generates a summary report.</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        pd.DataFrame</span>
<span class="sd">            DataFrame containing the Fst results summary</span>
<span class="sd">        </span>
<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        FileNotFoundError</span>
<span class="sd">            If no Fst result files are found in the results directory.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">df_summary</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;SuperPop&#39;</span><span class="p">,</span> <span class="s1">&#39;Fst&#39;</span><span class="p">,</span> <span class="s1">&#39;WeightedFst&#39;</span><span class="p">])</span>

        <span class="c1"># Get a list of all log files in the results directory</span>
        <span class="n">log_files</span> <span class="o">=</span> <span class="p">[</span><span class="n">f</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="o">.</span><span class="n">iterdir</span><span class="p">()</span> <span class="k">if</span> <span class="n">f</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="ow">and</span> <span class="n">f</span><span class="o">.</span><span class="n">suffix</span> <span class="o">==</span> <span class="s1">&#39;.log&#39;</span><span class="p">]</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">log_files</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No log files found in </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Found </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">log_files</span><span class="p">)</span><span class="si">}</span><span class="s2"> log files in </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># Extract the population names from log file names</span>
        <span class="c1"># Assuming log files follow the pattern &#39;fst-{population}-StPop.log&#39;</span>
        <span class="n">files</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">log_file</span> <span class="ow">in</span> <span class="n">log_files</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">log_file</span><span class="o">.</span><span class="n">stem</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;fst-&#39;</span><span class="p">)</span> <span class="ow">and</span> <span class="n">log_file</span><span class="o">.</span><span class="n">stem</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;-StPop&#39;</span><span class="p">):</span>
                <span class="n">pop</span> <span class="o">=</span> <span class="n">log_file</span><span class="o">.</span><span class="n">stem</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
                <span class="n">files</span><span class="p">[</span><span class="n">pop</span><span class="p">]</span> <span class="o">=</span> <span class="n">log_file</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Found Fst result file for population </span><span class="si">{</span><span class="n">pop</span><span class="si">}</span><span class="s2"> at </span><span class="si">{</span><span class="n">log_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">files</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No Fst result files found in </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">files</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
        
            <span class="n">log_file</span> <span class="o">=</span> <span class="n">files</span><span class="p">[</span><span class="n">key</span><span class="p">]</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">log_file</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>

                <span class="n">lines</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
                <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">lines</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;Mean Fst&#39;</span><span class="p">):</span>
                        <span class="n">fst</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;:&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
                    <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;Weighted Fst&#39;</span><span class="p">):</span>
                        <span class="n">weighted_fst</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;:&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
                <span class="n">df_summary</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">df_summary</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;SuperPop&#39;</span><span class="p">:</span> <span class="p">[</span><span class="n">key</span><span class="p">],</span> <span class="s1">&#39;Fst&#39;</span><span class="p">:</span> <span class="p">[</span><span class="n">fst</span><span class="p">],</span> <span class="s1">&#39;WeightedFst&#39;</span><span class="p">:</span> <span class="p">[</span><span class="n">weighted_fst</span><span class="p">]})],</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="n">df_summary</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="s1">&#39;fst_summary.csv&#39;</span><span class="p">,</span>
            <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span>
        <span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Fst summary report generated at </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="s1">&#39;fst_summary.csv&#39;</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="o">.</span><span class="n">iterdir</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">file</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="ow">and</span> <span class="p">(</span><span class="n">file</span><span class="o">.</span><span class="n">suffix</span> <span class="o">==</span> <span class="s1">&#39;.bed&#39;</span> <span class="ow">or</span> <span class="n">file</span><span class="o">.</span><span class="n">suffix</span> <span class="o">==</span> <span class="s1">&#39;.bim&#39;</span> <span class="ow">or</span> <span class="n">file</span><span class="o">.</span><span class="n">suffix</span> <span class="o">==</span> <span class="s1">&#39;.fam&#39;</span><span class="p">):</span>
                <span class="n">file</span><span class="o">.</span><span class="n">unlink</span><span class="p">()</span>
        
        <span class="k">return</span> <span class="n">df_summary</span></div>
</div>

</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, Luis Giraldo Gonzalez Ricardo, Amabel Tenghe, and Ashwin Ashok Kumar Sreelatha.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>