<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>ideal_genom_qc.SampleQC &mdash; IDEAL-GENOM-QC v0.1.0</title>
      <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../../_static/css/theme.css?v=19f00094" />

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../_static/jquery.js?v=5d32c60e"></script>
        <script src="../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../../_static/documentation_options.js?v=01f34227"></script>
        <script src="../../_static/doctools.js?v=9a2dae69"></script>
        <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #2980B9" >

          
          
          <a href="../../index.html" class="icon icon-home">
            IDEAL-GENOM-QC
          </a>
              <div class="version">
                0.1.0
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">User Guide</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../getting_started.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../configuration.html">Configuration Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../examples.html">Examples</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../SampleQC.html">SampleQC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../AncestryQC.html">AncestryQC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../PopStructure.html">Population Structure</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../VariantQC.html">VariantQC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../Helpers.html">Helpers</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../get_references.html">Get References</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Additional Resources</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../faq.html">Frequently Asked Questions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../troubleshooting.html">Troubleshooting Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../contributing.html">Contributing Guide</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #2980B9" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">IDEAL-GENOM-QC</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../index.html">Module code</a></li>
      <li class="breadcrumb-item active">ideal_genom_qc.SampleQC</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for ideal_genom_qc.SampleQC</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Module to perform sample quality control</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">psutil</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">warnings</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">logging</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">time</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">matplotlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">colormaps</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">seaborn</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sns</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">ideal_genom_qc.Helpers</span><span class="w"> </span><span class="kn">import</span> <span class="n">shell_do</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">ideal_genom_qc.get_references</span><span class="w"> </span><span class="kn">import</span> <span class="n">FetcherLDRegions</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">typing</span><span class="w"> </span><span class="kn">import</span> <span class="n">Optional</span><span class="p">,</span> <span class="n">Union</span>

<span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">%(asctime)s</span><span class="s2"> - </span><span class="si">%(levelname)s</span><span class="s2"> - </span><span class="si">%(message)s</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span>

<div class="viewcode-block" id="SampleQC">
<a class="viewcode-back" href="../../SampleQC.html#ideal_genom_qc.SampleQC.SampleQC">[docs]</a>
<span class="k">class</span><span class="w"> </span><span class="nc">SampleQC</span><span class="p">:</span>

<div class="viewcode-block" id="SampleQC.__init__">
<a class="viewcode-back" href="../../SampleQC.html#ideal_genom_qc.SampleQC.SampleQC.__init__">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">input_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">input_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">output_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">output_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">high_ld_file</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">built</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;38&#39;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initialize SampleQC class for quality control of genetic data.</span>
<span class="sd">        This class handles quality control procedures for genetic data files in PLINK binary format</span>
<span class="sd">        (bed, bim, fam). It sets up the directory structure and validates input files.</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        input_path : Path</span>
<span class="sd">            Directory path containing the input PLINK files</span>
<span class="sd">        input_name : str</span>
<span class="sd">            Base name of the input PLINK files (without extension)</span>
<span class="sd">        output_path : Path</span>
<span class="sd">            Directory path where output files will be saved</span>
<span class="sd">        output_name : str</span>
<span class="sd">            Base name for output files (without extension)</span>
<span class="sd">        high_ld_file : Path</span>
<span class="sd">            Path to file containing high LD regions. If not found, will be fetched from package</span>
<span class="sd">        built : str, optional</span>
<span class="sd">            Genome build version, either &#39;37&#39; or &#39;38&#39; (default=&#39;38&#39;)</span>
<span class="sd">        </span>
<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        TypeError</span>
<span class="sd">            If input types are incorrect</span>
<span class="sd">        ValueError</span>
<span class="sd">            If genome build version is not &#39;37&#39; or &#39;38&#39;</span>
<span class="sd">        FileNotFoundError</span>
<span class="sd">            If input paths or required PLINK files are not found</span>
<span class="sd">        </span>
<span class="sd">        Attributes</span>
<span class="sd">        ----------</span>
<span class="sd">        renamed_snps : bool</span>
<span class="sd">            Flag indicating if SNPs should be renamed</span>
<span class="sd">        hh_to_missing : bool</span>
<span class="sd">            Flag indicating if heterozygous haploid genotypes should be set to missing</span>
<span class="sd">        pruned_file : None</span>
<span class="sd">            Placeholder for pruned file path</span>
<span class="sd">        results_dir : Path</span>
<span class="sd">            Directory for all QC results</span>
<span class="sd">        fails_dir : Path</span>
<span class="sd">            Directory for failed samples</span>
<span class="sd">        clean_dir : Path</span>
<span class="sd">            Directory for clean files</span>
<span class="sd">        plots_dir : Path</span>
<span class="sd">            Directory for QC plots</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">input_path</span><span class="p">,</span> <span class="n">Path</span><span class="p">)</span> <span class="ow">or</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="n">Path</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;input_path and output_path should be of type Path&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">input_name</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span> <span class="ow">or</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">output_name</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;input_name and output_name should be of type str&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">high_ld_file</span><span class="p">,</span> <span class="n">Path</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;high_ld_file should be of type Path&quot;</span><span class="p">)</span>
        
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">built</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;built should be of type str&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">built</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;37&#39;</span><span class="p">,</span> <span class="s1">&#39;38&#39;</span><span class="p">]:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;built should be either 37 or 38&quot;</span><span class="p">)</span>
        
        <span class="k">if</span> <span class="ow">not</span> <span class="n">input_path</span><span class="o">.</span><span class="n">exists</span><span class="p">()</span> <span class="ow">or</span> <span class="ow">not</span> <span class="n">output_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="s2">&quot;input_path or output_path is not a valid path&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="n">input_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">input_name</span><span class="si">}</span><span class="s2">.bed&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="s2">&quot;.bed file not found&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="n">input_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">input_name</span><span class="si">}</span><span class="s2">.fam&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="s2">&quot;.fam file not found&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="n">input_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">input_name</span><span class="si">}</span><span class="s2">.bim&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="s2">&quot;.bim file not found&quot;</span><span class="p">)</span>
        
        <span class="k">if</span> <span class="ow">not</span> <span class="n">high_ld_file</span><span class="o">.</span><span class="n">is_file</span><span class="p">():</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;High LD file not found at </span><span class="si">{</span><span class="n">high_ld_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;High LD file will be fetched from the package&#39;</span><span class="p">)</span>
            
            <span class="n">ld_fetcher</span> <span class="o">=</span> <span class="n">FetcherLDRegions</span><span class="p">(</span><span class="n">built</span><span class="o">=</span><span class="n">built</span><span class="p">)</span>
            <span class="n">ld_fetcher</span><span class="o">.</span><span class="n">get_ld_regions</span><span class="p">()</span>

            <span class="n">ld_regions</span> <span class="o">=</span> <span class="n">ld_fetcher</span><span class="o">.</span><span class="n">ld_regions</span>
            <span class="k">if</span> <span class="n">ld_regions</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Failed to fetch high LD regions file&quot;</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;High LD file fetched from the package and saved at </span><span class="si">{</span><span class="n">ld_regions</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;High LD file found at </span><span class="si">{</span><span class="n">high_ld_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">ld_regions</span> <span class="o">=</span> <span class="n">high_ld_file</span>
        
        <span class="bp">self</span><span class="o">.</span><span class="n">input_path</span>  <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">input_path</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">output_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">output_path</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">input_name</span>  <span class="o">=</span> <span class="n">input_name</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">output_name</span> <span class="o">=</span> <span class="n">output_name</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">high_ld_file</span> <span class="o">=</span> <span class="n">ld_regions</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">renamed_snps</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">hh_to_missing</span><span class="o">=</span> <span class="kc">False</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pruned_file</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="c1"># create results folder</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_path</span> <span class="o">/</span> <span class="s1">&#39;sample_qc_results&#39;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="c1"># create fails folder</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">fails_dir</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="s1">&#39;fail_samples&#39;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">fails_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="c1"># create clean files folder</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">clean_dir</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="s1">&#39;clean_files&#39;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">clean_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        
        <span class="c1"># create figures folder</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">plots_dir</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="s1">&#39;plots&#39;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">plots_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></div>


<div class="viewcode-block" id="SampleQC.execute_rename_snpid">
<a class="viewcode-back" href="../../SampleQC.html#ideal_genom_qc.SampleQC.SampleQC.execute_rename_snpid">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">execute_rename_snpid</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">rename</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Executes the SNP ID renaming process using PLINK2.</span>
<span class="sd">        This method renames SNP IDs in the PLINK binary files to a standardized format of &#39;chr:pos:a1:a2&#39;.</span>
<span class="sd">        The renaming is performed using PLINK2&#39;s --set-all-var-ids parameter.</span>
<span class="sd">        </span>
<span class="sd">        Parameter:</span>
<span class="sd">        ----------</span>
<span class="sd">        rename (bool, optional): Flag to control whether SNP renaming should be performed. </span>
<span class="sd">            Defaults to True.</span>

<span class="sd">        Returns:</span>
<span class="sd">        --------</span>
<span class="sd">            None</span>

<span class="sd">        Raises:</span>
<span class="sd">        -------</span>
<span class="sd">            TypeError: If rename parameter is not a boolean.</span>

<span class="sd">        Notes:</span>
<span class="sd">        ------</span>
<span class="sd">            - The renamed files will be saved with &#39;-renamed&#39; suffix</span>
<span class="sd">            - Thread count is optimized based on available CPU cores</span>
<span class="sd">            - The new SNP ID format will be: chromosome:position:allele1:allele2</span>
<span class="sd">            - Sets self.renamed_snps to True if renaming is performed</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">rename</span><span class="p">,</span> <span class="nb">bool</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;rename must be a boolean&quot;</span><span class="p">)</span>
        
        <span class="k">if</span> <span class="ow">not</span> <span class="n">rename</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;STEP: Rename SNPs. `rename` set to </span><span class="si">{</span><span class="n">rename</span><span class="si">}</span><span class="s2">. Skipping renaming of SNPs in the study data&quot;</span><span class="p">)</span>
            <span class="k">return</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;STEP: Rename SNPs. `rename` set to </span><span class="si">{</span><span class="n">rename</span><span class="si">}</span><span class="s2">. Renaming SNPs in the study data to the format chr_pos_a1_a2&quot;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">renamed_snps</span> <span class="o">=</span> <span class="kc">True</span>

        <span class="n">cpu_count</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">cpu_count</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">max_threads</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">cpu_count</span> <span class="o">-</span> <span class="mi">2</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Dynamically calculate fallback as half of available cores or default to 2</span>
            <span class="n">max_threads</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="p">(</span><span class="n">psutil</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">(</span><span class="n">logical</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="ow">or</span> <span class="mi">2</span><span class="p">)</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span>

        <span class="n">plink2_cmd</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink2 --bfile </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">input_path</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="si">}</span><span class="s2"> --set-all-var-ids @:#:$r:$a --threads </span><span class="si">{</span><span class="n">max_threads</span><span class="si">}</span><span class="s2"> --make-bed --out </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">input_path</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="o">+</span><span class="w"> </span><span class="s1">&#39;-renamed&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="c1"># Execute PLINK2 command</span>
        <span class="n">shell_do</span><span class="p">(</span><span class="n">plink2_cmd</span><span class="p">,</span> <span class="n">log</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="k">return</span></div>

    
<div class="viewcode-block" id="SampleQC.execute_haploid_to_missing">
<a class="viewcode-back" href="../../SampleQC.html#ideal_genom_qc.SampleQC.SampleQC.execute_haploid_to_missing">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">execute_haploid_to_missing</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">hh_to_missing</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Convert haploid genotypes to missing values in PLINK binary files.</span>
<span class="sd">        This method uses PLINK&#39;s --set-hh-missing flag to convert haploid genotypes to </span>
<span class="sd">        missing values in the genotype data. This is often useful for quality control </span>
<span class="sd">        of genetic data, particularly for variants on sex chromosomes.</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        hh_to_missing : bool, default=True</span>
<span class="sd">            If True, converts haploid genotypes to missing values.</span>
<span class="sd">            If False, skips the conversion step.</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None</span>
<span class="sd">        </span>
<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        TypeError</span>
<span class="sd">            If hh_to_missing is not a boolean value.</span>
<span class="sd">        </span>
<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        The method uses PLINK to process the binary files (.bed, .bim, .fam) and creates</span>
<span class="sd">        new files with suffix &#39;-hh-missing&#39;. The input files are determined based on whether</span>
<span class="sd">        SNPs have been previously renamed (checks self.renamed_snps).</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">hh_to_missing</span><span class="p">,</span> <span class="nb">bool</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;hh_to_missing must be a boolean&quot;</span><span class="p">)</span>
        
        <span class="k">if</span> <span class="ow">not</span> <span class="n">hh_to_missing</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;STEP: Convert haploid genotypes to missing values. `hh_to_missing` set to </span><span class="si">{</span><span class="n">hh_to_missing</span><span class="si">}</span><span class="s2">. Skipping conversion of haploid genotypes to missing values&quot;</span><span class="p">)</span>
            <span class="k">return</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;STEP: Convert haploid genotypes to missing values. `hh_to_missing` set to </span><span class="si">{</span><span class="n">hh_to_missing</span><span class="si">}</span><span class="s2">. Converting haploid genotypes to missing values in the study data&quot;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">hh_to_missing</span> <span class="o">=</span> <span class="kc">True</span>
        
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;STEP: Convert haploid genotypes to missing values&quot;</span><span class="p">)</span>

        <span class="c1"># Dynamically set the input file name based on whether SNPs are renamed</span>
        <span class="n">input_file</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">input_name</span> <span class="o">+</span> <span class="s1">&#39;-renamed&#39;</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">renamed_snps</span> <span class="k">else</span> <span class="bp">self</span><span class="o">.</span><span class="n">input_name</span>

        <span class="c1"># PLINK command: convert haploid genotypes to missing</span>
        <span class="n">plink_cmd</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink2 --bfile </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">input_path</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="n">input_file</span><span class="si">}</span><span class="s2"> --set-invalid-haploid-missing --keep-allele-order --make-bed --out </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">input_path</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="o">+</span><span class="s1">&#39;-hh-missing&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="c1"># execute PLINK command</span>
        <span class="n">shell_do</span><span class="p">(</span><span class="n">plink_cmd</span><span class="p">,</span> <span class="n">log</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="k">return</span></div>

    
<div class="viewcode-block" id="SampleQC.execute_ld_pruning">
<a class="viewcode-back" href="../../SampleQC.html#ideal_genom_qc.SampleQC.SampleQC.execute_ld_pruning">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">execute_ld_pruning</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ind_pair</span><span class="p">:</span> <span class="nb">list</span> <span class="o">=</span> <span class="p">[</span><span class="mi">50</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mf">0.2</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Execute LD (Linkage Disequilibrium) pruning on genetic data using PLINK.</span>
<span class="sd">        This method performs LD pruning in three steps:</span>
<span class="sd">        1. Excludes complex/high LD regions</span>
<span class="sd">        2. Identifies SNPs for pruning using indep-pairwise test</span>
<span class="sd">        3. Creates final pruned dataset</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ind_pair : list, optional</span>
<span class="sd">            List of three elements for LD pruning parameters:</span>
<span class="sd">            - Window size (int): Number of SNPs to analyze in each window</span>
<span class="sd">            - Step size (int): Number of SNPs to shift window at each step</span>
<span class="sd">            - r² threshold (float): Correlation coefficient threshold for pruning</span>
<span class="sd">            Default is [50, 5, 0.2]</span>
<span class="sd">        </span>
<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        TypeError</span>
<span class="sd">            If ind_pair is not a list</span>
<span class="sd">            If first two elements of ind_pair are not integers</span>
<span class="sd">            If third element of ind_pair is not float</span>
<span class="sd">        ValueError</span>
<span class="sd">            If ind_pair does not contain exactly three elements</span>
<span class="sd">            If window size or step size is not positive</span>
<span class="sd">            If r² threshold is not between 0 and 1</span>
<span class="sd">        FileNotFoundError</span>
<span class="sd">            If required pruning input file is not found</span>
<span class="sd">        </span>
<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        - Uses available CPU cores (leaving 2 cores free) and 2/3 of available memory</span>
<span class="sd">        - Creates intermediate and final files with suffixes:</span>
<span class="sd">          * &#39;-LDregionExcluded&#39;</span>
<span class="sd">          * &#39;-LDregionExcluded-prunning&#39;</span>
<span class="sd">          * &#39;-LDpruned&#39;</span>
<span class="sd">        - Updates self.pruned_file with path to final pruned dataset</span>
<span class="sd">        &quot;&quot;&quot;</span>
        
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ind_pair</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;ind_pair should be a list&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">ind_pair</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">3</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;ind_pair must have exactly three elements&quot;</span><span class="p">)</span>
        
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ind_pair</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="nb">int</span><span class="p">)</span> <span class="ow">or</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ind_pair</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="nb">int</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;The first two elements in ind_pair values should be integers (window size and step size)&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">ind_pair</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">ind_pair</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Window size and step size must be positive integers&quot;</span><span class="p">)</span>
        
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ind_pair</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="nb">float</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;The third element in ind_pair should be a float (r^2 threshold)&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="mi">0</span> <span class="o">&lt;</span> <span class="n">ind_pair</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="mi">1</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;The r^2 threshold must be a float between 0 and 1&quot;</span><span class="p">)</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;STEP: LD pruning&quot;</span><span class="p">)</span>

        <span class="n">cpu_count</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">cpu_count</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">max_threads</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">cpu_count</span> <span class="o">-</span> <span class="mi">2</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Dynamically calculate fallback as half of available cores or default to 2</span>
            <span class="n">max_threads</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="p">(</span><span class="n">psutil</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">(</span><span class="n">logical</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="ow">or</span> <span class="mi">2</span><span class="p">)</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span>

        <span class="c1"># Get the virtual memory details</span>
        <span class="n">memory_info</span> <span class="o">=</span> <span class="n">psutil</span><span class="o">.</span><span class="n">virtual_memory</span><span class="p">()</span>
        <span class="n">available_memory_mb</span> <span class="o">=</span> <span class="n">memory_info</span><span class="o">.</span><span class="n">available</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1024</span> <span class="o">*</span> <span class="mi">1024</span><span class="p">)</span>
        <span class="n">memory</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="mi">2</span><span class="o">*</span><span class="n">available_memory_mb</span><span class="o">/</span><span class="mi">3</span><span class="p">,</span><span class="mi">0</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">hh_to_missing</span><span class="p">:</span>
            <span class="n">ld_input</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="o">+</span><span class="s1">&#39;-hh-missing&#39;</span>
        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">renamed_snps</span><span class="p">:</span>
            <span class="n">ld_input</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="o">+</span><span class="s1">&#39;-renamed&#39;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">ld_input</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">input_name</span>

        <span class="c1"># exclude complex regions</span>
        <span class="n">plink_cmd1</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink2 --bfile </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">input_path</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="n">ld_input</span><span class="si">}</span><span class="s2"> --exclude </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">high_ld_file</span><span class="si">}</span><span class="s2"> --memory </span><span class="si">{</span><span class="n">memory</span><span class="si">}</span><span class="s2"> --threads </span><span class="si">{</span><span class="n">max_threads</span><span class="si">}</span><span class="s2"> --make-bed --out </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="o">+</span><span class="s1">&#39;-LDregionExcluded&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>
        
        <span class="n">prune_in_file</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="o">+</span><span class="s1">&#39;-LDregionExcluded-prunning&#39;</span><span class="p">))</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s1">&#39;.prune.in&#39;</span><span class="p">)</span>


        <span class="c1"># LD prune indep-pairwise test</span>
        <span class="n">plink_cmd2</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink2 --bfile </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="o">+</span><span class="s1">&#39;-LDregionExcluded&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2"> --indep-pairwise </span><span class="si">{</span><span class="n">ind_pair</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">ind_pair</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">ind_pair</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="si">}</span><span class="s2"> --keep-allele-order --memory </span><span class="si">{</span><span class="n">memory</span><span class="si">}</span><span class="s2"> --threads </span><span class="si">{</span><span class="n">max_threads</span><span class="si">}</span><span class="s2"> --out </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="o">+</span><span class="s1">&#39;-LDregionExcluded-prunning&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>


        <span class="n">plink_cmd3</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink2 --bfile </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="o">+</span><span class="s1">&#39;-LDregionExcluded&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2"> --extract </span><span class="si">{</span><span class="n">prune_in_file</span><span class="si">}</span><span class="s2"> --keep-allele-order --make-bed --out </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="s1">&#39;-LDpruned&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2"> --memory </span><span class="si">{</span><span class="n">memory</span><span class="si">}</span><span class="s2"> --threads </span><span class="si">{</span><span class="n">max_threads</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="c1">#plink_cmd3 = f&quot;plink --bfile {self.results_dir / (self.input_name+&#39;-LDregionExcluded&#39;)} --extract {(self.results_dir / (self.input_name+&#39;-LDregionExcluded-prunning&#39;)).with_suffix(&#39;.prune.in&#39;)} --keep-allele-order --make-bed --out {self.results_dir / (self.input_name + &#39;-LDpruned&#39;)} --memory {memory} --threads {max_threads}&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">pruned_file</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">input_name</span> <span class="o">+</span> <span class="s1">&#39;-LDpruned&#39;</span><span class="p">)</span>

        <span class="c1"># execute PLINK commands</span>
        <span class="n">cmds</span> <span class="o">=</span> <span class="p">[</span><span class="n">plink_cmd1</span><span class="p">,</span> <span class="n">plink_cmd2</span><span class="p">,</span> <span class="n">plink_cmd3</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">cmd</span> <span class="ow">in</span> <span class="n">cmds</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Executing PLINK command: </span><span class="si">{</span><span class="n">cmd</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">shell_do</span><span class="p">(</span><span class="n">cmd</span><span class="p">,</span> <span class="n">log</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">time</span><span class="o">.</span><span class="n">sleep</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>  <span class="c1"># Adding a small delay to ensure commands are executed sequentially</span>

        <span class="k">return</span></div>

    
<div class="viewcode-block" id="SampleQC.execute_miss_genotype">
<a class="viewcode-back" href="../../SampleQC.html#ideal_genom_qc.SampleQC.SampleQC.execute_miss_genotype">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">execute_miss_genotype</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Execute missing genotype analysis using PLINK to identify and filter samples with high missingness rates.</span>
<span class="sd">        </span>
<span class="sd">        This method performs two main operations:</span>
<span class="sd">        1. Generates missingness statistics for all samples</span>
<span class="sd">        2. Filters samples based on the specified missingness threshold (mind)</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        mind : float, optional</span>
<span class="sd">            The missingness threshold for sample filtering (default is 0.2).</span>
<span class="sd">            Samples with missingness rates above this threshold will be removed.</span>
<span class="sd">            Recommended range is between 0.02 and 0.1.</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        dict</span>
<span class="sd">            Dictionary containing missingness analysis results</span>
<span class="sd">        </span>
<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        TypeError</span>
<span class="sd">            If mind parameter is not a float</span>
<span class="sd">        ValueError</span>
<span class="sd">            If mind parameter is not between 0 and 1</span>
<span class="sd">        FileNotFoundError</span>
<span class="sd">            If the output .imiss file is not generated</span>
<span class="sd">            If mind value is outside recommended range (0.02-0.1)</span>
<span class="sd">        </span>
<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        This function creates two files:</span>
<span class="sd">        - {input_name}-missing.imiss: Contains missingness statistics for all samples</span>
<span class="sd">        - {output_name}-mind.bed: New binary file with filtered samples</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;STEP: Missing genotype check.&quot;</span><span class="p">)</span>

        <span class="n">cpu_count</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">cpu_count</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">max_threads</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">cpu_count</span> <span class="o">-</span> <span class="mi">2</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Dynamically calculate fallback as half of available cores or default to 2</span>
            <span class="n">max_threads</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="p">(</span><span class="n">psutil</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">(</span><span class="n">logical</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="ow">or</span> <span class="mi">2</span><span class="p">)</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span>

        <span class="c1"># Get the virtual memory details</span>
        <span class="n">memory_info</span> <span class="o">=</span> <span class="n">psutil</span><span class="o">.</span><span class="n">virtual_memory</span><span class="p">()</span>
        <span class="n">available_memory_mb</span> <span class="o">=</span> <span class="n">memory_info</span><span class="o">.</span><span class="n">available</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1024</span> <span class="o">*</span> <span class="mi">1024</span><span class="p">)</span>
        <span class="n">memory</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="mi">2</span><span class="o">*</span><span class="n">available_memory_mb</span><span class="o">/</span><span class="mi">3</span><span class="p">,</span><span class="mi">0</span><span class="p">)</span>

        <span class="c1"># PLINK command: run mssingness across file genome-wide </span>
        <span class="n">plink_cmd1</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink2 --bfile </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">pruned_file</span><span class="si">}</span><span class="s2"> --missing --memory </span><span class="si">{</span><span class="n">memory</span><span class="si">}</span><span class="s2"> --threads </span><span class="si">{</span><span class="n">max_threads</span><span class="si">}</span><span class="s2"> --out </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="o">+</span><span class="s1">&#39;-missing&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="n">shell_do</span><span class="p">(</span><span class="n">plink_cmd1</span><span class="p">,</span> <span class="n">log</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">call_rate_miss</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="o">+</span><span class="s1">&#39;-missing&#39;</span><span class="p">))</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s1">&#39;.smiss&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">call_rate_miss</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Missing file: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">call_rate_miss</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">return</span></div>

    
<div class="viewcode-block" id="SampleQC.execute_sex_check">
<a class="viewcode-back" href="../../SampleQC.html#ideal_genom_qc.SampleQC.SampleQC.execute_sex_check">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">execute_sex_check</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sex_check</span><span class="p">:</span> <span class="nb">list</span> <span class="o">=</span> <span class="p">[</span><span class="mf">0.2</span><span class="p">,</span> <span class="mf">0.8</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Execute sex check using PLINK to identify potential sex discrepancies in genetic data.</span>
<span class="sd">        </span>
<span class="sd">        This method performs sex check analysis by:</span>
<span class="sd">        1. Running PLINK&#39;s --check-sex command on pruned data</span>
<span class="sd">        2. Extracting X chromosome SNPs</span>
<span class="sd">        3. Calculating missingness rates for X chromosome SNPs</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        sex_check : list of float, default=[0.2, 0.8]</span>
<span class="sd">            List containing two float values that define the F-statistic boundaries for sex determination.</span>
<span class="sd">            The values must sum to 1.0. First value is the lower bound, second is the upper bound.</span>
<span class="sd">            Samples with F-statistics below the first value are called female, above the second value are called male.</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None</span>
<span class="sd">        </span>
<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        TypeError</span>
<span class="sd">            If sex_check is not a list or if its elements are not floats</span>
<span class="sd">        ValueError</span>
<span class="sd">            If sex_check doesn&#39;t contain exactly 2 elements or if they don&#39;t sum to 1</span>
<span class="sd">        </span>
<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        The method creates the following output files:</span>
<span class="sd">        - {output_name}-sexcheck.sexcheck : Contains sex check results</span>
<span class="sd">        - {output_name}-xchr.bed/bim/fam : X chromosome SNP data</span>
<span class="sd">        - {output_name}-xchr-missing.imiss : X chromosome missingness data</span>
<span class="sd">        The number of threads used is automatically determined based on available CPU cores,</span>
<span class="sd">        using max(available cores - 2, 1) or falling back to half of logical cores if CPU count</span>
<span class="sd">        cannot be determined.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">sex_check</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;sex_check should be a list&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">sex_check</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">2</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;sex_check must have two elements&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">all</span><span class="p">(</span><span class="nb">isinstance</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="nb">float</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">sex_check</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;All elements in sex_check must be floats&quot;</span><span class="p">)</span>
        
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;STEP: Check discordant sex information.&quot;</span><span class="p">)</span>

        <span class="n">cpu_count</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">cpu_count</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">max_threads</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">cpu_count</span> <span class="o">-</span> <span class="mi">2</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Dynamically calculate fallback as half of available cores or default to 2</span>
            <span class="n">max_threads</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="p">(</span><span class="n">psutil</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">(</span><span class="n">logical</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="ow">or</span> <span class="mi">2</span><span class="p">)</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span>

        <span class="n">plink_cmd1</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink2 --bfile </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">pruned_file</span><span class="si">}</span><span class="s2"> --check-sex max-female-xf=</span><span class="si">{</span><span class="n">sex_check</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2"> min-male-xf=</span><span class="si">{</span><span class="n">sex_check</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s2"> --threads </span><span class="si">{</span><span class="n">max_threads</span><span class="si">}</span><span class="s2"> --out </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-sexcheck&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="nb">print</span><span class="p">(</span><span class="n">plink_cmd1</span><span class="p">)</span>

        <span class="c1"># extract xchr SNPs</span>
        <span class="n">plink_cmd2</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink2 --bfile </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">pruned_file</span><span class="si">}</span><span class="s2"> --chr 23 --keep-allele-order --threads </span><span class="si">{</span><span class="n">max_threads</span><span class="si">}</span><span class="s2">  --make-bed --out </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-xchr&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="c1"># run missingness on xchr SNPs</span>
        <span class="n">plink_cmd3</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink2 --bfile </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-xchr&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2"> --threads </span><span class="si">{</span><span class="n">max_threads</span><span class="si">}</span><span class="s2">  --missing --out </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-xchr-missing&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="c1"># execute PLINK commands</span>
        <span class="n">cmds</span> <span class="o">=</span> <span class="p">[</span><span class="n">plink_cmd1</span><span class="p">,</span> <span class="n">plink_cmd2</span><span class="p">,</span> <span class="n">plink_cmd3</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">cmd</span> <span class="ow">in</span> <span class="n">cmds</span><span class="p">:</span>
            <span class="n">shell_do</span><span class="p">(</span><span class="n">cmd</span><span class="p">,</span> <span class="n">log</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">time</span><span class="o">.</span><span class="n">sleep</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>  <span class="c1"># Adding a small delay to ensure commands are executed sequentially</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">sexcheck_miss</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span> <span class="o">+</span> <span class="s1">&#39;-sexcheck.sexcheck&#39;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">xchr_miss</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span> <span class="o">+</span> <span class="s1">&#39;-xchr-missing.smiss&#39;</span><span class="p">)</span>

        <span class="k">return</span></div>


<div class="viewcode-block" id="SampleQC.execute_heterozygosity_rate">
<a class="viewcode-back" href="../../SampleQC.html#ideal_genom_qc.SampleQC.SampleQC.execute_heterozygosity_rate">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">execute_heterozygosity_rate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">maf</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.01</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Executes heterozygosity rate analysis on genetic data using PLINK.</span>

<span class="sd">        This method performs a series of PLINK commands to analyze heterozygosity rates in genetic data,</span>
<span class="sd">        separating SNPs based on minor allele frequency (MAF) threshold and computing heterozygosity</span>
<span class="sd">        for both groups.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        maf : float, optional</span>
<span class="sd">            Minor allele frequency threshold used to split SNPs into two groups.</span>
<span class="sd">            Must be between 0 and 0.5. Default is 0.01.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None</span>

<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        TypeError</span>
<span class="sd">            If maf is not a float</span>
<span class="sd">        ValueError</span>
<span class="sd">            If maf is not between 0 and 0.5</span>
<span class="sd">        FileNotFoundError</span>
<span class="sd">            If any of the expected output files are not created</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        The method:</span>
<span class="sd">        1. Extracts autosomal SNPs</span>
<span class="sd">        2. Splits SNPs based on MAF threshold</span>
<span class="sd">        3. Computes missingness</span>
<span class="sd">        4. Converts to PED/MAP format</span>
<span class="sd">        5. Computes heterozygosity for both MAF groups</span>

<span class="sd">        The computation uses optimized threading based on available CPU cores and memory.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">maf</span><span class="p">,</span> <span class="nb">float</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;maf should be a float&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">maf</span> <span class="o">&lt;=</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">maf</span> <span class="o">&gt;=</span> <span class="mf">0.5</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;maf should be between 0 and 0.5&quot;</span><span class="p">)</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;STEP: Heterozygosity rate check. `maf` set to </span><span class="si">{</span><span class="n">maf</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="n">cpu_count</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">cpu_count</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">max_threads</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">cpu_count</span> <span class="o">-</span> <span class="mi">2</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Dynamically calculate fallback as half of available cores or default to 2</span>
            <span class="n">max_threads</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="p">(</span><span class="n">psutil</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">(</span><span class="n">logical</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="ow">or</span> <span class="mi">2</span><span class="p">)</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span>

        <span class="c1"># Get the virtual memory details</span>
        <span class="n">memory_info</span> <span class="o">=</span> <span class="n">psutil</span><span class="o">.</span><span class="n">virtual_memory</span><span class="p">()</span>
        <span class="n">available_memory_mb</span> <span class="o">=</span> <span class="n">memory_info</span><span class="o">.</span><span class="n">available</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1024</span> <span class="o">*</span> <span class="mi">1024</span><span class="p">)</span>
        <span class="n">memory</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="mi">2</span><span class="o">*</span><span class="n">available_memory_mb</span><span class="o">/</span><span class="mi">3</span><span class="p">,</span><span class="mi">0</span><span class="p">)</span>

        <span class="c1"># extract autosomal SNPS</span>
        <span class="n">plink_cmd1</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink2 --bfile </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">pruned_file</span><span class="si">}</span><span class="s2"> --autosome --keep-allele-order --memory </span><span class="si">{</span><span class="n">memory</span><span class="si">}</span><span class="s2"> --make-bed --out </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-chr1-22&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="c1"># extract SNPs with minor allele frequency greater than threshold</span>
        <span class="n">plink_cmd2</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink2 --bfile </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-chr1-22&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2"> --maf </span><span class="si">{</span><span class="n">maf</span><span class="si">}</span><span class="s2"> --keep-allele-order --make-bed --out </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-chr1-22-mafgreater&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="c1"># extract SNPs with minor allele frequency less than threshold</span>
        <span class="n">plink_cmd3</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink2 --bfile </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-chr1-22&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2"> --exclude </span><span class="si">{</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-chr1-22-mafgreater&#39;</span><span class="p">))</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s1">&#39;.bim&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2"> --keep-allele-order --make-bed --out </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-chr1-22-mafless&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="c1"># get missingness to plot against het</span>
        <span class="n">plink_cmd4</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink2 --bfile </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-chr1-22-mafgreater&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2"> --missing --out </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-chr1-22-mafgreater-missing&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="n">plink_cmd5</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink2 --bfile </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-chr1-22-mafless&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2"> --missing --out </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-chr1-22-mafless-missing&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="c1"># compute heterozigosity for both MAF groups</span>
        <span class="n">plink_cmd6</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink2 --bfile </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-chr1-22-mafgreater&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2"> --het --out </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-chr1-22-mafgreater&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2"> --memory </span><span class="si">{</span><span class="n">memory</span><span class="si">}</span><span class="s2"> --threads </span><span class="si">{</span><span class="n">max_threads</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="n">plink_cmd7</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink2 --bfile </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-chr1-22-mafless&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2"> --het --out </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-chr1-22-mafless&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2"> --memory </span><span class="si">{</span><span class="n">memory</span><span class="si">}</span><span class="s2"> --threads </span><span class="si">{</span><span class="n">max_threads</span><span class="si">}</span><span class="s2">&quot;</span>

<span class="c1">#        # convert both to ped/map files for heterozigosity computation</span>
<span class="c1">#        plink_cmd6 = f&quot;plink --bfile {self.results_dir / (self.output_name+&#39;-chr1-22-mafgreater&#39;)} --recode --out {self.results_dir / (self.output_name+&#39;-chr1-22-mafgreater-recode&#39;)} --memory {memory} --threads {max_threads}&quot;</span>
<span class="c1">#        plink_cmd7 = f&quot;plink --bfile {self.results_dir / (self.output_name+&#39;-chr1-22-mafless&#39;)} --recode --out {self.results_dir / (self.output_name+&#39;-chr1-22-mafless-recode&#39;)} --memory {memory} --threads {max_threads}&quot;</span>
<span class="c1">#</span>
        <span class="c1"># execute PLINK commands</span>
        <span class="n">cmds</span> <span class="o">=</span> <span class="p">[</span><span class="n">plink_cmd1</span><span class="p">,</span> <span class="n">plink_cmd2</span><span class="p">,</span> <span class="n">plink_cmd3</span><span class="p">,</span> <span class="n">plink_cmd4</span><span class="p">,</span> <span class="n">plink_cmd5</span><span class="p">,</span> <span class="n">plink_cmd6</span><span class="p">,</span> <span class="n">plink_cmd7</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">cmd</span> <span class="ow">in</span> <span class="n">cmds</span><span class="p">:</span>
            <span class="n">shell_do</span><span class="p">(</span><span class="n">cmd</span><span class="p">,</span> <span class="n">log</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">time</span><span class="o">.</span><span class="n">sleep</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>  <span class="c1"># Adding a small delay to ensure commands are executed sequentially</span>

<span class="c1">#        self._compute_heterozigozity(</span>
<span class="c1">#            ped_file=(self.results_dir / (self.output_name+&#39;-chr1-22-mafgreater-recode&#39;)).with_suffix(&#39;.ped&#39;)</span>
<span class="c1">#        )</span>
<span class="c1">#        self._compute_heterozigozity(</span>
<span class="c1">#            ped_file=(self.results_dir / (self.output_name+&#39;-chr1-22-mafless-recode&#39;)).with_suffix(&#39;.ped&#39;)</span>
<span class="c1">#        )</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">maf_greater_het</span><span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-chr1-22-mafgreater.het&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">maf_greater_het</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Missing file: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">maf_greater_het</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">maf_less_het</span>   <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-chr1-22-mafless.het&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">maf_less_het</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Missing file: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">maf_less_het</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="c1">#        self.summary_greater = self.results_dir / (&#39;Summary-&#39;+self.output_name+&#39;-chr1-22-mafgreater-recode.ped&#39;)</span>
<span class="c1">#        if not self.summary_greater.exists():</span>
<span class="c1">#            raise FileNotFoundError(f&quot;Missing file: {self.summary_greater}&quot;)</span>
<span class="c1">#        self.summary_less    = self.results_dir / (&#39;Summary-&#39;+self.output_name+&#39;-chr1-22-mafless-recode.ped&#39;)</span>
<span class="c1">#        if not self.summary_less.exists():</span>
<span class="c1">#            raise FileNotFoundError(f&quot;Missing file: {self.summary_less}&quot;)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">maf_greater_smiss</span><span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-chr1-22-mafgreater-missing.smiss&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">maf_greater_smiss</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Missing file: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">maf_greater_smiss</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">maf_less_smiss</span>   <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-chr1-22-mafless-missing.smiss&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">maf_less_smiss</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Missing file: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">maf_less_smiss</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">return</span></div>


<div class="viewcode-block" id="SampleQC.execute_ibd">
<a class="viewcode-back" href="../../SampleQC.html#ideal_genom_qc.SampleQC.SampleQC.execute_ibd">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">execute_ibd</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Execute Identity by Descent (IBD) analysis using PLINK.</span>

<span class="sd">        This method performs duplicate and relatedness checks using IBD analysis. It runs two PLINK commands:</span>
<span class="sd">        1. Generates genome-wide IBD estimates</span>
<span class="sd">        2. Calculates missing genotype rates</span>

<span class="sd">        The method uses optimal thread count based on available CPU cores and validates input/output files.</span>

<span class="sd">        Returns:</span>
<span class="sd">        --------</span>
<span class="sd">            None</span>

<span class="sd">        Raises:</span>
<span class="sd">        -------</span>
<span class="sd">            FileNotFoundError: If required input pruned file is missing or if expected output files are not generated</span>

<span class="sd">        Required instance attributes:</span>
<span class="sd">            pruned_file: Path to pruned PLINK binary file</span>
<span class="sd">            results_dir: Directory path for output files</span>
<span class="sd">            output_name: Base name for output files</span>
<span class="sd">            ibd_miss: Path to missing genotype rate file (set by method)</span>
<span class="sd">            genome: Path to IBD estimates file (set by method)</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;STEP: Duplicates and relatedness check with IBD&quot;</span><span class="p">)</span>

        <span class="n">cpu_count</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">cpu_count</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">max_threads</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">cpu_count</span> <span class="o">-</span> <span class="mi">2</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Dynamically calculate fallback as half of available cores or default to 2</span>
            <span class="n">max_threads</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="p">(</span><span class="n">psutil</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">(</span><span class="n">logical</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="ow">or</span> <span class="mi">2</span><span class="p">)</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">pruned_file</span> <span class="ow">or</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">pruned_file</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Missing file: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">pruned_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># PLINK command</span>
        <span class="n">plink_cmd1</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink2 --bfile </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">pruned_file</span><span class="si">}</span><span class="s2"> --genome --out </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-ibd&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2"> --threads </span><span class="si">{</span><span class="n">max_threads</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="c1"># PLINK command</span>
        <span class="n">plink_cmd2</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink2 --bfile </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">pruned_file</span><span class="si">}</span><span class="s2"> --allow-no-sex --missing --out </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-ibd-missing&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="c1"># execute PLINK commands</span>
        <span class="n">cmds</span> <span class="o">=</span> <span class="p">[</span><span class="n">plink_cmd1</span><span class="p">,</span> <span class="n">plink_cmd2</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">cmd</span> <span class="ow">in</span> <span class="n">cmds</span><span class="p">:</span>
            <span class="n">shell_do</span><span class="p">(</span><span class="n">cmd</span><span class="p">,</span> <span class="n">log</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">ibd_miss</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-ibd-missing.imiss&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">ibd_miss</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Missing file: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">ibd_miss</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genome</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-ibd.genome&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Missing file: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">return</span></div>


<div class="viewcode-block" id="SampleQC.execute_kinship">
<a class="viewcode-back" href="../../SampleQC.html#ideal_genom_qc.SampleQC.SampleQC.execute_kinship">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">execute_kinship</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">kinship</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.354</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Execute kinship analysis to identify and handle sample relatedness.</span>

<span class="sd">        This method performs kinship analysis using PLINK2 to identify duplicate samples and related individuals.</span>
<span class="sd">        It first computes a kinship coefficient matrix for all samples and then prunes samples based on the</span>
<span class="sd">        specified kinship threshold.</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        kinship : float, optional</span>
<span class="sd">            The kinship coefficient threshold used to identify related samples. Must be between 0 and 1.</span>
<span class="sd">            Samples with kinship coefficients above this threshold will be marked for removal.</span>
<span class="sd">            Default is 0.354 (equivalent to first-degree relatives).</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None</span>
<span class="sd">        </span>
<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        TypeError</span>
<span class="sd">            If kinship parameter is not a float.</span>
<span class="sd">        ValueError</span>
<span class="sd">            If kinship parameter is not between 0 and 1.</span>
<span class="sd">        FileNotFoundError</span>
<span class="sd">            If the expected output file from PLINK2 is not created.</span>
<span class="sd">        </span>
<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        - Uses PLINK2 to compute kinship coefficients and perform sample pruning</span>
<span class="sd">        - Automatically determines optimal thread count and memory usage based on system resources</span>
<span class="sd">        - Creates output files with kinship coefficient matrix and list of samples to be removed</span>
<span class="sd">        - Updates self.kinship_miss with path to file containing samples to be removed</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">kinship</span><span class="p">,</span> <span class="nb">float</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;kinship should be a float&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">kinship</span> <span class="o">&lt;</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">kinship</span> <span class="o">&gt;</span><span class="mi">1</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;kinship should be between 0 and 1&quot;</span><span class="p">)</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;STEP: Duplicates and relatedness check with Kinship. `kinship` set to </span><span class="si">{</span><span class="n">kinship</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">hh_to_missing</span><span class="p">:</span>
            <span class="n">kinship_input</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="o">+</span><span class="s1">&#39;-hh-missing&#39;</span>
        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">renamed_snps</span><span class="p">:</span>
            <span class="n">kinship_input</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="o">+</span><span class="s1">&#39;-renamed&#39;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">kinship_input</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">input_name</span>

        <span class="n">cpu_count</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">cpu_count</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">max_threads</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">cpu_count</span> <span class="o">-</span> <span class="mi">2</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Dynamically calculate fallback as half of available cores or default to 2</span>
            <span class="n">max_threads</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="p">(</span><span class="n">psutil</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">(</span><span class="n">logical</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="ow">or</span> <span class="mi">2</span><span class="p">)</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span>

        <span class="c1"># Get the virtual memory details</span>
        <span class="n">memory_info</span> <span class="o">=</span> <span class="n">psutil</span><span class="o">.</span><span class="n">virtual_memory</span><span class="p">()</span>
        <span class="n">available_memory_mb</span> <span class="o">=</span> <span class="n">memory_info</span><span class="o">.</span><span class="n">available</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1024</span> <span class="o">*</span> <span class="mi">1024</span><span class="p">)</span>
        <span class="n">memory</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="mi">2</span><span class="o">*</span><span class="n">available_memory_mb</span><span class="o">/</span><span class="mi">3</span><span class="p">,</span><span class="mi">0</span><span class="p">)</span>
        
        <span class="c1"># Compute kinship-coefficient matrix for all samples</span>
        <span class="n">plink2_cmd1</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink2 --bfile </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">input_path</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="n">kinship_input</span><span class="si">}</span><span class="s2"> --make-king triangle bin --out </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-kinship-coefficient-matrix&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2"> --memory </span><span class="si">{</span><span class="n">memory</span><span class="si">}</span><span class="s2"> --threads </span><span class="si">{</span><span class="n">max_threads</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="c1"># Prune for Monozygotic Twins OR Duplicates</span>
        <span class="n">plink2_cmd2</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink2 --bfile </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">input_path</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="n">kinship_input</span><span class="si">}</span><span class="s2"> --king-cutoff </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-kinship-coefficient-matrix&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">kinship</span><span class="si">}</span><span class="s2"> --out </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-kinship-pruned-duplicates&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2"> --memory </span><span class="si">{</span><span class="n">memory</span><span class="si">}</span><span class="s2"> --threads </span><span class="si">{</span><span class="n">max_threads</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="c1"># execute PLINK commands</span>
        <span class="n">cmds</span> <span class="o">=</span> <span class="p">[</span><span class="n">plink2_cmd1</span><span class="p">,</span> <span class="n">plink2_cmd2</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">cmd</span> <span class="ow">in</span> <span class="n">cmds</span><span class="p">:</span>
            <span class="n">shell_do</span><span class="p">(</span><span class="n">cmd</span><span class="p">,</span> <span class="n">log</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">kinship_miss</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-kinship-pruned-duplicates&#39;</span><span class="p">))</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s1">&#39;.king.cutoff.out.id&#39;</span><span class="p">)</span>

        <span class="c1"># Check if the file exists</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">kinship_miss</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Expected file </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">kinship_miss</span><span class="si">}</span><span class="s2"> was not created. Ensure the PLINK2 command executed successfully.&quot;</span><span class="p">)</span>

        <span class="k">return</span></div>

    
<div class="viewcode-block" id="SampleQC.execute_duplicate_relatedness">
<a class="viewcode-back" href="../../SampleQC.html#ideal_genom_qc.SampleQC.SampleQC.execute_duplicate_relatedness">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">execute_duplicate_relatedness</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">kinship</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.354</span><span class="p">,</span> <span class="n">use_kinship</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Execute duplicate and relatedness analysis on the genotype data.</span>
<span class="sd">        This method performs either IBD (Identity by Descent) or KING kinship coefficient</span>
<span class="sd">        analysis to identify duplicate samples and related individuals in the dataset.</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        kinship : float, optional</span>
<span class="sd">            The KING kinship coefficient threshold for identifying related samples.</span>
<span class="sd">            Default is 0.354, which corresponds to duplicates/MZ twins.</span>
<span class="sd">        use_kinship : bool, optional</span>
<span class="sd">            If True, uses KING algorithm for relatedness analysis.</span>
<span class="sd">            If False, uses traditional IBD analysis.</span>
<span class="sd">            Default is True.</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None</span>

<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        TypeError</span>
<span class="sd">            If kinship is not a float or use_kinship is not a boolean.</span>
<span class="sd">        </span>
<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        The method will store the analysis type (KING or IBD) in the use_kinship attribute.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">use_kinship</span><span class="p">,</span> <span class="nb">bool</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;use_kinship must be a boolean&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">kinship</span><span class="p">,</span> <span class="nb">float</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;kinship must be a float&quot;</span><span class="p">)</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;STEP: Duplicates and relatedness check&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">use_kinship</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">execute_kinship</span><span class="p">(</span><span class="n">kinship</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">execute_ibd</span><span class="p">()</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">use_kinship</span> <span class="o">=</span> <span class="n">use_kinship</span>

        <span class="k">return</span></div>


    <span class="k">def</span><span class="w"> </span><span class="nf">_compute_heterozigozity</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ped_file</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">map_file</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Path</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Computes heterozygosity statistics from a PED file and writes results to a summary file.</span>
<span class="sd">        This method analyzes a PED file to calculate homozygosity and heterozygosity rates</span>
<span class="sd">        for each individual. The results are written to a summary file in the same directory</span>
<span class="sd">        as the input PED file.</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ped_file : Path</span>
<span class="sd">            Path to the input PED file containing genotype data</span>
<span class="sd">        map_file : Path, optional</span>
<span class="sd">            Path to the MAP file (not used in current implementation)</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None</span>
<span class="sd">            Results are written to a summary file named &quot;Summary-{ped_filename}&quot;</span>
<span class="sd">        The output summary file contains the following columns:</span>
<span class="sd">        - ID: Individual identifier</span>
<span class="sd">        - total: Total number of valid genotypes</span>
<span class="sd">        - num_hom: Number of homozygous genotypes</span>
<span class="sd">        - num_het: Number of heterozygous genotypes  </span>
<span class="sd">        - Percent_hom: Percentage of homozygous genotypes</span>
<span class="sd">        - Percent_het: Percentage of heterozygous genotypes</span>
<span class="sd">        </span>
<span class="sd">        sNotes</span>
<span class="sd">        -----</span>
<span class="sd">        - Missing alleles (coded as &#39;0&#39; or &#39;N&#39;) are excluded from calculations</span>
<span class="sd">        - The method assumes PED file format with genotype data starting from column 7</span>
<span class="sd">        - Handles FileNotFound and IOError exceptions with appropriate error messages</span>
<span class="sd">        &quot;&quot;&quot;</span>
        
        <span class="c1"># Define output file name</span>
        <span class="n">summary_file</span><span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;Summary-</span><span class="si">{</span><span class="n">ped_file</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="n">output_path</span> <span class="o">=</span> <span class="n">ped_file</span><span class="o">.</span><span class="n">parent</span> <span class="o">/</span> <span class="n">summary_file</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">ped_file</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">ped</span><span class="p">,</span> <span class="nb">open</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">output</span><span class="p">:</span>
                <span class="c1"># Write the header to the summary file</span>
                <span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;ID</span><span class="se">\t</span><span class="s2">total</span><span class="se">\t</span><span class="s2">num_hom</span><span class="se">\t</span><span class="s2">num_het</span><span class="se">\t</span><span class="s2">Percent_hom</span><span class="se">\t</span><span class="s2">Percent_het</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

                <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">ped</span><span class="p">:</span>
                    <span class="n">line</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="n">line</span><span class="p">:</span>
                        <span class="k">continue</span>
                    
                    <span class="c1"># Split the line into columns</span>
                    <span class="n">columns</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
                    <span class="n">individual_id</span> <span class="o">=</span> <span class="n">columns</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>  <span class="c1"># Individual ID (second column)</span>
                    <span class="n">genotype_data</span> <span class="o">=</span> <span class="n">columns</span><span class="p">[</span><span class="mi">6</span><span class="p">:]</span>  <span class="c1"># Genotype data starts at the 7th column</span>

                    <span class="c1"># Initialize counters</span>
                    <span class="n">total</span><span class="o">=</span> <span class="mi">0</span>
                    <span class="n">hom</span>  <span class="o">=</span> <span class="mi">0</span>
                    <span class="n">het</span>  <span class="o">=</span> <span class="mi">0</span>

                    <span class="c1"># Iterate through genotype pairs</span>
                    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">genotype_data</span><span class="p">),</span> <span class="mi">2</span><span class="p">):</span>
                        <span class="n">allele1</span> <span class="o">=</span> <span class="n">genotype_data</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
                        <span class="n">allele2</span> <span class="o">=</span> <span class="n">genotype_data</span><span class="p">[</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">]</span>

                        <span class="k">if</span> <span class="n">allele1</span> <span class="o">==</span> <span class="n">allele2</span><span class="p">:</span>
                            <span class="k">if</span> <span class="n">allele1</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;0&#39;</span><span class="p">,</span> <span class="s1">&#39;N&#39;</span><span class="p">]:</span>  <span class="c1"># Exclude missing alleles</span>
                                <span class="n">hom</span> <span class="o">+=</span> <span class="mi">1</span>
                                <span class="n">total</span> <span class="o">+=</span> <span class="mi">1</span>
                        <span class="k">elif</span> <span class="n">allele1</span> <span class="o">!=</span> <span class="n">allele2</span><span class="p">:</span>
                            <span class="n">het</span> <span class="o">+=</span> <span class="mi">1</span>
                            <span class="n">total</span> <span class="o">+=</span> <span class="mi">1</span>

                    <span class="c1"># Calculate percentages</span>
                    <span class="n">hom_percent</span> <span class="o">=</span> <span class="p">(</span><span class="n">hom</span> <span class="o">/</span> <span class="n">total</span><span class="p">)</span> <span class="o">*</span> <span class="mi">100</span> <span class="k">if</span> <span class="n">total</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="mf">0.0</span>
                    <span class="n">het_percent</span> <span class="o">=</span> <span class="p">(</span><span class="n">het</span> <span class="o">/</span> <span class="n">total</span><span class="p">)</span> <span class="o">*</span> <span class="mi">100</span> <span class="k">if</span> <span class="n">total</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="mf">0.0</span>

                    <span class="c1"># Write the statistics to the output file</span>
                    <span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">individual_id</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">total</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">hom</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">het</span><span class="si">}</span><span class="se">\t</span><span class="s2">&quot;</span>
                                 <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">hom_percent</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">het_percent</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Summary written to </span><span class="si">{</span><span class="n">summary_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">FileNotFoundError</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error: File </span><span class="si">{</span><span class="n">ped_file</span><span class="si">}</span><span class="s2"> not found.&quot;</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">IOError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<div class="viewcode-block" id="SampleQC.get_fail_samples">
<a class="viewcode-back" href="../../SampleQC.html#ideal_genom_qc.SampleQC.SampleQC.get_fail_samples">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">get_fail_samples</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">call_rate_thres</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span> <span class="n">std_deviation_het</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span> <span class="n">maf_het</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span> <span class="n">ibd_threshold</span><span class="p">:</span> <span class="nb">float</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get samples that failed quality control checks and generate a summary DataFrame.</span>
<span class="sd">        This method performs multiple QC checks on samples:</span>
<span class="sd">        1. Call rate check</span>
<span class="sd">        2. Sex check</span>
<span class="sd">        3. Heterozygosity rate check (for MAF &gt; and &lt; threshold)</span>
<span class="sd">        4. Duplicates/Relatedness check (using either KING or IBD)</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        call_rate_thres : float</span>
<span class="sd">            Threshold for call rate filtering</span>
<span class="sd">        std_deviation_het : float </span>
<span class="sd">            Number of standard deviations to use for heterozygosity filtering</span>
<span class="sd">        maf_het : float</span>
<span class="sd">            Minor allele frequency threshold for heterozygosity check</span>
<span class="sd">        ibd_threshold : float</span>
<span class="sd">            Threshold for IBD filtering (only used if use_king=False)</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        pd.DataFrame</span>
<span class="sd">            Summary DataFrame containing:</span>
<span class="sd">            - Counts of samples failing each QC check</span>
<span class="sd">            - Number of duplicated sample IDs</span>
<span class="sd">            - Total failures</span>
<span class="sd">        </span>
<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        FileNotFoundError</span>
<span class="sd">            If any required input files are missing</span>
<span class="sd">        TypeError</span>
<span class="sd">            If unexpected column types are found in summary DataFrame</span>
<span class="sd">        </span>
<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        The method saves a detailed fail_samples.txt file with all failed samples and their failure reasons.</span>
<span class="sd">        Samples failing multiple checks are only counted once in the final summary.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Check if required files exist</span>
        <span class="n">required_files</span> <span class="o">=</span> <span class="p">[</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">call_rate_miss</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span> <span class="o">+</span> <span class="s1">&#39;-sexcheck.sexcheck&#39;</span><span class="p">),</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span> <span class="o">+</span> <span class="s1">&#39;-xchr-missing.imiss&#39;</span><span class="p">),</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="p">(</span><span class="s1">&#39;Summary-&#39;</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_name</span> <span class="o">+</span> <span class="s1">&#39;-chr1-22-mafgreater-recode.ped&#39;</span><span class="p">),</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span> <span class="o">+</span> <span class="s1">&#39;-chr1-22-mafgreater-missing.imiss&#39;</span><span class="p">),</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="p">(</span><span class="s1">&#39;Summary-&#39;</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_name</span> <span class="o">+</span> <span class="s1">&#39;-chr1-22-mafless-recode.ped&#39;</span><span class="p">),</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span> <span class="o">+</span> <span class="s1">&#39;-chr1-22-mafless-missing.imiss&#39;</span><span class="p">)</span>
        <span class="p">]</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">use_kinship</span><span class="p">:</span>
            <span class="n">required_files</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span> <span class="o">+</span> <span class="s1">&#39;-ibd-missing.imiss&#39;</span><span class="p">))</span>
            <span class="n">required_files</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span> <span class="o">+</span> <span class="s1">&#39;-ibd.genome&#39;</span><span class="p">))</span>

        <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">required_files</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">file</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
                <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Required file not found: </span><span class="si">{</span><span class="n">file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># ==========================================================================================================</span>
        <span class="c1">#                                             CALL RATE CHECK</span>
        <span class="c1"># ==========================================================================================================</span>

        <span class="c1"># load samples who failed call rate check</span>
        <span class="n">fail_call_rate</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">report_call_rate</span><span class="p">(</span>
            <span class="n">file_path</span>     <span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">call_rate_miss</span><span class="p">,</span>
            <span class="n">threshold</span>    <span class="o">=</span><span class="n">call_rate_thres</span><span class="p">,</span> 
            <span class="n">plots_dir</span>    <span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">plots_dir</span><span class="p">,</span>
            <span class="n">y_axis_cap</span>   <span class="o">=</span><span class="mi">10</span>
        <span class="p">)</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Call rate check done&#39;</span><span class="p">)</span>

        <span class="c1"># ==========================================================================================================</span>
        <span class="c1">#                                             SEX CHECK</span>
        <span class="c1"># ==========================================================================================================</span>

        <span class="n">fail_sexcheck</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">report_sex_check</span><span class="p">(</span>
            <span class="n">directory</span>          <span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="p">,</span> 
            <span class="n">sex_check_filename</span> <span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-sexcheck.sexcheck&#39;</span><span class="p">,</span> 
            <span class="n">xchr_imiss_filename</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-xchr-missing.imiss&#39;</span><span class="p">,</span>
            <span class="n">plots_dir</span>          <span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">plots_dir</span>
        <span class="p">)</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Sex check done&#39;</span><span class="p">)</span>

        <span class="c1"># ==========================================================================================================</span>
        <span class="c1">#                                       HETETROZYGOSITY RATE CHECK</span>
        <span class="c1"># ==========================================================================================================</span>

        <span class="n">fail_het_greater</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">report_heterozygosity_rate</span><span class="p">(</span>
            <span class="n">het_filename</span>        <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">maf_greater_het</span><span class="p">,</span> 
            <span class="n">autosomal_filename</span>  <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">maf_greater_smiss</span><span class="p">,</span> 
            <span class="n">std_deviation_het</span>   <span class="o">=</span> <span class="n">std_deviation_het</span><span class="p">,</span>
            <span class="n">maf</span>                 <span class="o">=</span> <span class="n">maf_het</span><span class="p">,</span>
            <span class="n">split</span>               <span class="o">=</span> <span class="s1">&#39;&gt;&#39;</span><span class="p">,</span>
            <span class="n">plots_dir</span>           <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">plots_dir</span>
        <span class="p">)</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Heterozygosity rate check done for MAF &gt; </span><span class="si">{</span><span class="n">maf_het</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>

        <span class="n">fail_het_less</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">report_heterozygosity_rate</span><span class="p">(</span>
            <span class="n">het_filename</span>        <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">maf_less_het</span><span class="p">,</span> 
            <span class="n">autosomal_filename</span>  <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">maf_less_smiss</span><span class="p">,</span> 
            <span class="n">std_deviation_het</span>   <span class="o">=</span> <span class="n">std_deviation_het</span><span class="p">,</span>
            <span class="n">maf</span>                 <span class="o">=</span> <span class="n">maf_het</span><span class="p">,</span>
            <span class="n">split</span>               <span class="o">=</span> <span class="s1">&#39;&lt;&#39;</span><span class="p">,</span>
            <span class="n">plots_dir</span>           <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">plots_dir</span>
        <span class="p">)</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Heterozygosity rate check done for MAF &lt; </span><span class="si">{</span><span class="n">maf_het</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>

        <span class="c1"># ==========================================================================================================</span>
        <span class="c1">#                                       DUPLICATES-RELATEDNESS CHECK</span>
        <span class="c1"># ==========================================================================================================</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">use_kinship</span><span class="p">:</span>

            <span class="c1"># load samples that failed duplicates and relatedness check</span>
            <span class="n">duplicates_file</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-kinship-pruned-duplicates.king.cutoff.out.id&#39;</span><span class="p">)</span>
            <span class="n">df_duplicates</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
                <span class="n">duplicates_file</span><span class="p">,</span>
                <span class="n">sep</span>   <span class="o">=</span><span class="sa">r</span><span class="s1">&#39;\s+&#39;</span><span class="p">,</span>
                <span class="n">engine</span><span class="o">=</span><span class="s1">&#39;python&#39;</span>
            <span class="p">)</span>

            <span class="c1"># filter samples that failed duplicates and relatedness check</span>
            <span class="n">df_duplicates</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;FID&#39;</span><span class="p">,</span> <span class="s1">&#39;IID&#39;</span><span class="p">]</span>
            <span class="n">fail_duplicates</span> <span class="o">=</span> <span class="n">df_duplicates</span><span class="p">[[</span><span class="s1">&#39;FID&#39;</span><span class="p">,</span> <span class="s1">&#39;IID&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">fail_duplicates</span><span class="p">[</span><span class="s1">&#39;Failure&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Duplicates and relatedness (Kinship)&#39;</span>

            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Duplicates and relatedness check done with kinship&#39;</span><span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>

            <span class="n">fail_duplicates</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">report_ibd_analysis</span><span class="p">(</span><span class="n">ibd_threshold</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Duplicates and relatedness check done with IBD&#39;</span><span class="p">)</span>

        <span class="c1"># ==========================================================================================================</span>
        <span class="c1">#                                       MERGE ALL FAILURES</span>
        <span class="c1"># ==========================================================================================================</span>

        <span class="n">fails</span> <span class="o">=</span> <span class="p">[</span><span class="n">fail_call_rate</span><span class="p">,</span> <span class="n">fail_sexcheck</span><span class="p">,</span> <span class="n">fail_het_greater</span><span class="p">,</span> <span class="n">fail_het_less</span><span class="p">,</span> <span class="n">fail_duplicates</span><span class="p">]</span> 

        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">fails</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="n">summary</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Failure&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
        <span class="n">num_dup</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">duplicated</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;FID&#39;</span><span class="p">,</span> <span class="s1">&#39;IID&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>

        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;FID&#39;</span><span class="p">,</span> <span class="s1">&#39;IID&#39;</span><span class="p">])</span>

        <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fails_dir</span> <span class="o">/</span> <span class="s1">&#39;fail_samples.txt&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>

        <span class="n">totals</span> <span class="o">=</span> <span class="n">summary</span><span class="o">.</span><span class="n">select_dtypes</span><span class="p">(</span><span class="n">include</span><span class="o">=</span><span class="s2">&quot;number&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="o">-</span> <span class="n">num_dup</span>
        <span class="n">dups_row</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="n">summary</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span> <span class="p">[</span><span class="s1">&#39;Duplicated Sample IDs&#39;</span><span class="p">],</span> <span class="n">summary</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span> <span class="p">[</span><span class="o">-</span><span class="n">num_dup</span><span class="p">]})</span>
        <span class="c1"># Create the total row</span>
        <span class="n">dups_row</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;Failure&#39;</span><span class="p">:[</span><span class="s1">&#39;Duplicated Sample IDs&#39;</span><span class="p">],</span> <span class="s1">&#39;count&#39;</span><span class="p">:[</span><span class="o">-</span><span class="n">num_dup</span><span class="p">]})</span>
        <span class="c1"># Validate column types in the summary DataFrame</span>
        <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">summary</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">api</span><span class="o">.</span><span class="n">types</span><span class="o">.</span><span class="n">is_numeric_dtype</span><span class="p">(</span><span class="n">summary</span><span class="p">[</span><span class="n">col</span><span class="p">])</span> <span class="ow">or</span> <span class="n">summary</span><span class="p">[</span><span class="n">col</span><span class="p">]</span><span class="o">.</span><span class="n">dtype</span> <span class="o">==</span> <span class="s1">&#39;object&#39;</span><span class="p">):</span>
                <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unexpected column type in summary DataFrame: </span><span class="si">{</span><span class="n">col</span><span class="si">}</span><span class="s2"> has type </span><span class="si">{</span><span class="n">summary</span><span class="p">[</span><span class="n">col</span><span class="p">]</span><span class="o">.</span><span class="n">dtype</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># Construct the total_row DataFrame</span>
        <span class="n">total_row</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="n">col</span><span class="p">:</span> <span class="p">[</span><span class="n">totals</span><span class="p">[</span><span class="n">col</span><span class="p">]</span> <span class="k">if</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">totals</span><span class="o">.</span><span class="n">index</span> <span class="k">else</span> <span class="s2">&quot;Total&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">summary</span><span class="o">.</span><span class="n">columns</span><span class="p">})</span>

        <span class="c1"># Append the total row to the DataFrame</span>
        <span class="n">summary</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">summary</span><span class="p">,</span> <span class="n">dups_row</span><span class="p">,</span> <span class="n">total_row</span><span class="p">],</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        
        <span class="k">return</span> <span class="n">summary</span></div>

    
<div class="viewcode-block" id="SampleQC.execute_drop_samples">
<a class="viewcode-back" href="../../SampleQC.html#ideal_genom_qc.SampleQC.SampleQC.execute_drop_samples">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">execute_drop_samples</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Execute the removal of samples that failed quality control checks using PLINK.</span>
<span class="sd">        This method performs the following steps:</span>
<span class="sd">        1. Determines the appropriate binary file name based on previous processing steps</span>
<span class="sd">        2. Reads the fail_samples.txt file containing samples to be removed</span>
<span class="sd">        3. Executes PLINK command to create new binary files excluding failed samples</span>
<span class="sd">        </span>
<span class="sd">        Raises:</span>
<span class="sd">        -------</span>
<span class="sd">            FileNotFoundError: If the fail_samples.txt file is not found in the fails directory</span>

<span class="sd">        Returns:</span>
<span class="sd">        --------</span>
<span class="sd">            None</span>

<span class="sd">        Notes:</span>
<span class="sd">        ------</span>
<span class="sd">            - The output files will be created with suffix &#39;-clean-samples&#39;</span>
<span class="sd">            - The method preserves allele order during the operation</span>
<span class="sd">            - Input files must be in PLINK binary format (.bed, .bim, .fam)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;STEP: Drop samples that failed quality control checks&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">hh_to_missing</span><span class="p">:</span>
            <span class="n">binary_name</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="o">+</span><span class="s1">&#39;-hh-missing&#39;</span>
        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">renamed_snps</span><span class="p">:</span>
            <span class="n">binary_name</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="o">+</span><span class="s1">&#39;-renamed&#39;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">binary_name</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">input_name</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Binary file name: </span><span class="si">{</span><span class="n">binary_name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># drop samples</span>
        <span class="n">fail_samples_file</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">fails_dir</span> <span class="o">/</span> <span class="s1">&#39;fail_samples.txt&#39;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">fail_samples_file</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Required file </span><span class="si">{</span><span class="n">fail_samples_file</span><span class="si">}</span><span class="s2"> not found. Ensure the fail_samples.txt file is generated before executing this step.&quot;</span><span class="p">)</span>

        <span class="n">plink_cmd</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink --bfile </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">input_path</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="n">binary_name</span><span class="si">}</span><span class="s2"> --remove </span><span class="si">{</span><span class="n">fail_samples_file</span><span class="si">}</span><span class="s2"> --keep-allele-order --make-bed --out </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">clean_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-clean-samples&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="c1"># execute PLINK command</span>
        <span class="n">shell_do</span><span class="p">(</span><span class="n">plink_cmd</span><span class="p">,</span> <span class="n">log</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="k">return</span></div>

  
<div class="viewcode-block" id="SampleQC.report_call_rate">
<a class="viewcode-back" href="../../SampleQC.html#ideal_genom_qc.SampleQC.SampleQC.report_call_rate">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">report_call_rate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">file_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">threshold</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span> <span class="n">plots_dir</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Path</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">y_axis_cap</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span> <span class="n">color</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;#1B9E77&#39;</span><span class="p">,</span> <span class="n">line_color</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;#D95F02&#39;</span><span class="p">,</span> <span class="nb">format</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;png&#39;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Generate sample call rate analysis plots and identify samples failing the call rate threshold.</span>
<span class="sd">        This method reads a PLINK-format missing rate file, creates visualization plots, and identifies</span>
<span class="sd">        samples that fail the specified call rate threshold. It generates two sets of plots:</span>
<span class="sd">        1. Histograms showing the distribution of missing SNPs (F_MISS)</span>
<span class="sd">        2. Scatterplots showing different views of the call rate data</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        directory : Path</span>
<span class="sd">            Directory containing the input file</span>
<span class="sd">        filename : str</span>
<span class="sd">            Name of the PLINK format missing rate file</span>
<span class="sd">        threshold : float</span>
<span class="sd">            Call rate threshold for sample filtering (in terms of F_MISS)</span>
<span class="sd">        plots_dir : Path, optional</span>
<span class="sd">            Directory where plots will be saved. If None, uses default plots directory</span>
<span class="sd">        y_axis_cap : int, optional</span>
<span class="sd">            Maximum value for y-axis in capped histogram plots. Default is 10</span>
<span class="sd">        color : str, optional</span>
<span class="sd">            Color for the main plot elements. Default is &#39;#1B9E77&#39;</span>
<span class="sd">        line_color : str, optional</span>
<span class="sd">            Color for threshold lines in plots. Default is &#39;#D95F02&#39;</span>
<span class="sd">        format : str, optional</span>
<span class="sd">            Format for saving plots. Default is &#39;png&#39;</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        pd.DataFrame</span>
<span class="sd">            DataFrame containing samples that failed the call rate threshold with columns:</span>
<span class="sd">            - FID: Family ID</span>
<span class="sd">            - IID: Individual ID</span>
<span class="sd">            - Failure: Always set to &#39;Call rate&#39;</span>
<span class="sd">        </span>
<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        The method generates two image files:</span>
<span class="sd">        - call_rate_{threshold}_histogram.&lt;format&gt;: Contains histogram plots</span>
<span class="sd">        - call_rate_{threshold}_scatterplot.&lt;format&gt;: Contains scatter plots</span>
<span class="sd">        &quot;&quot;&quot;</span>
        
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">Path</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;directory should be a Path object&quot;</span><span class="p">)</span>
        <span class="k">if</span>  <span class="ow">not</span> <span class="n">file_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;File </span><span class="si">{</span><span class="n">file_path</span><span class="si">}</span><span class="s2"> does not exist&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">file_path</span><span class="o">.</span><span class="n">is_file</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Path </span><span class="si">{</span><span class="n">file_path</span><span class="si">}</span><span class="s2"> is not a file&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">threshold</span><span class="p">,</span> <span class="nb">float</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;threshold should be a float object&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="mi">0</span> <span class="o">&lt;=</span> <span class="n">threshold</span> <span class="o">&lt;=</span> <span class="mi">1</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;threshold should be between 0 and 1&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">plots_dir</span><span class="p">,</span> <span class="p">(</span><span class="n">Path</span><span class="p">,</span> <span class="nb">type</span><span class="p">(</span><span class="kc">None</span><span class="p">))):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;plots_dir should be a Path object or None&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">y_axis_cap</span><span class="p">,</span> <span class="p">(</span><span class="nb">int</span><span class="p">,</span> <span class="nb">float</span><span class="p">)):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;y_axis_cap should be an int or float&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">color</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;color should be a string representing a color&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">line_color</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;line_color should be a string representing a color&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="nb">format</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;format should be a string representing the file format (e.g., &#39;png&#39;, &#39;pdf&#39;)&quot;</span><span class="p">)</span>
        
        <span class="k">if</span> <span class="ow">not</span> <span class="n">plots_dir</span><span class="p">:</span>
            <span class="n">plots_dir</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">plots_dir</span>

        <span class="c1"># load samples that failed sex check</span>
        <span class="n">df_call_rate</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
            <span class="n">file_path</span><span class="p">,</span>
            <span class="n">sep</span><span class="o">=</span><span class="sa">r</span><span class="s1">&#39;\s+&#39;</span><span class="p">,</span>
            <span class="n">engine</span><span class="o">=</span><span class="s1">&#39;python&#39;</span>
        <span class="p">)</span>
        <span class="n">df_call_rate</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span><span class="o">.</span><span class="n">lstrip</span><span class="p">(</span><span class="s1">&#39;#&#39;</span><span class="p">)</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">df_call_rate</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>

        <span class="c1"># filter samples that fail call rate</span>
        <span class="n">fail_call_rate</span> <span class="o">=</span> <span class="n">df_call_rate</span><span class="p">[</span><span class="n">df_call_rate</span><span class="p">[</span><span class="s1">&#39;F_MISS&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">threshold</span><span class="p">][[</span><span class="s1">&#39;FID&#39;</span><span class="p">,</span> <span class="s1">&#39;IID&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">fail_call_rate</span><span class="p">[</span><span class="s1">&#39;Failure&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Call rate&#39;</span>

        <span class="c1"># Create the figure and subplots</span>
        <span class="n">fig1</span><span class="p">,</span> <span class="n">axes1</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">5</span><span class="p">),</span> <span class="n">sharey</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="c1"># First subplot: Full histogram</span>
        <span class="n">axes1</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span><span class="n">df_call_rate</span><span class="p">[</span><span class="s1">&#39;F_MISS&#39;</span><span class="p">],</span> <span class="n">bins</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes1</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="c1"># type: ignore</span>
        <span class="n">axes1</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Sample Call Rate Distribution&quot;</span><span class="p">)</span>
        <span class="n">axes1</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Proportion of missing SNPs (F_MISS)&quot;</span><span class="p">)</span>
        <span class="n">axes1</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Frequency&quot;</span><span class="p">)</span>

        <span class="c1"># Second subplot: Histogram with capped y-axis</span>
        <span class="n">axes1</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span><span class="n">df_call_rate</span><span class="p">[</span><span class="s1">&#39;F_MISS&#39;</span><span class="p">],</span> <span class="n">bins</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes1</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="c1"># type: ignore</span>
        <span class="n">axes1</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">y_axis_cap</span><span class="p">)</span>  <span class="c1"># Cap y-axis</span>
        <span class="n">axes1</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Sample Call Rate Distribution (Capped)&quot;</span><span class="p">)</span>
        <span class="n">axes1</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Proportion of missing SNPs (F_MISS)&quot;</span><span class="p">)</span>

        <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">plots_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;call_rate_</span><span class="si">{</span><span class="n">threshold</span><span class="si">}</span><span class="s2">_histogram.</span><span class="si">{</span><span class="nb">format</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">400</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">block</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="n">fig2</span><span class="p">,</span> <span class="n">axes2</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">15</span><span class="p">,</span> <span class="mi">5</span><span class="p">),</span> <span class="n">sharey</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="c1"># First subplot: capped y-axis</span>
        <span class="n">axes2</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span><span class="n">df_call_rate</span><span class="p">[</span><span class="s1">&#39;F_MISS&#39;</span><span class="p">],</span> <span class="n">bins</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes2</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="c1"># type: ignore</span>
        <span class="n">axes2</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">y_axis_cap</span><span class="p">)</span>  <span class="c1"># Cap y-axis</span>
        <span class="n">axes2</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Sample Call Rate Distribution (Capped)&quot;</span><span class="p">)</span>
        <span class="n">axes2</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Proportion of missing SNPs (F_MISS)&quot;</span><span class="p">)</span>

        <span class="c1"># Add a vertical line at the threshold</span>
        <span class="n">axes2</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">threshold</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">line_color</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;dashed&#39;</span><span class="p">)</span>

        <span class="c1"># Second subplot: Number of samples vs F_MISS</span>
        <span class="n">df_call_rate_sorted</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
            <span class="s1">&#39;Index&#39;</span><span class="p">:</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">df_call_rate</span><span class="p">[</span><span class="s1">&#39;F_MISS&#39;</span><span class="p">])),</span>
            <span class="s1">&#39;F_MISS&#39;</span><span class="p">:</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">df_call_rate</span><span class="p">[</span><span class="s1">&#39;F_MISS&#39;</span><span class="p">])</span>
        <span class="p">})</span>

        <span class="n">axes2</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span>
            <span class="n">data</span>  <span class="o">=</span><span class="n">df_call_rate_sorted</span><span class="p">,</span>
            <span class="n">x</span>     <span class="o">=</span><span class="s1">&#39;Index&#39;</span><span class="p">,</span>
            <span class="n">y</span>     <span class="o">=</span><span class="s1">&#39;F_MISS&#39;</span><span class="p">,</span>
            <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span>
            <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;none&#39;</span><span class="p">,</span>
            <span class="n">color</span> <span class="o">=</span><span class="n">color</span><span class="p">,</span>
            <span class="n">ax</span>    <span class="o">=</span><span class="n">axes2</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
        <span class="p">)</span> 
        <span class="n">axes2</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Sample Call Rate&quot;</span><span class="p">)</span>
        <span class="n">axes2</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Number of samples&quot;</span><span class="p">)</span>
        <span class="n">axes2</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;F_MISS&quot;</span><span class="p">)</span>

        <span class="c1"># Add a vertical line at the threshold</span>
        <span class="n">axes2</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span><span class="n">threshold</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">line_color</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;dashed&#39;</span><span class="p">)</span>

        <span class="c1"># third subplot: Number of samples vs F_MISS</span>
        <span class="n">axes2</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span>
            <span class="n">x</span>      <span class="o">=</span><span class="n">df_call_rate</span><span class="p">[</span><span class="s1">&#39;F_MISS&#39;</span><span class="p">],</span>
            <span class="n">y</span>      <span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">df_call_rate</span><span class="p">[</span><span class="s1">&#39;F_MISS&#39;</span><span class="p">])),</span>
            <span class="n">markers</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span>
            <span class="n">s</span>      <span class="o">=</span><span class="mi">20</span><span class="p">,</span>
            <span class="n">color</span> <span class="o">=</span><span class="n">color</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">axes2</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Sample Call Rate&quot;</span><span class="p">)</span>
        <span class="n">axes2</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Proportion of missing SNPs (F_MISS)&quot;</span><span class="p">)</span>
        <span class="n">axes2</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Samples&quot;</span><span class="p">)</span>
        <span class="n">axes2</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
    

        <span class="c1"># Add a vertical line at the threshold</span>
        <span class="n">axes2</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">threshold</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">line_color</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;dashed&#39;</span><span class="p">)</span>

        <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">plots_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;call_rate_</span><span class="si">{</span><span class="n">threshold</span><span class="si">}</span><span class="s2">_scatterplot.</span><span class="si">{</span><span class="nb">format</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">400</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">block</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">fail_call_rate</span></div>

    
<div class="viewcode-block" id="SampleQC.report_sex_check">
<a class="viewcode-back" href="../../SampleQC.html#ideal_genom_qc.SampleQC.SampleQC.report_sex_check">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">report_sex_check</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">directory</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">sex_check_filename</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">xchr_imiss_filename</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">f_coeff_thresholds</span><span class="p">:</span> <span class="nb">list</span> <span class="o">=</span> <span class="p">[</span><span class="mf">0.2</span><span class="p">,</span> <span class="mf">0.8</span><span class="p">],</span>  <span class="n">plots_dir</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Path</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="nb">format</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;png&#39;</span><span class="p">,</span> <span class="n">fig_size</span><span class="p">:</span> <span class="nb">tuple</span> <span class="o">=</span> <span class="p">(</span><span class="mi">8</span><span class="p">,</span><span class="mi">6</span><span class="p">))</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Creates a sex check report and visualization based on PLINK&#39;s sex check results.</span>
<span class="sd">        This function reads sex check data and X chromosome missingness data, merges them,</span>
<span class="sd">        and generates a scatter plot to visualize potential sex discrepancies. It also identifies</span>
<span class="sd">        samples that fail sex check quality control.</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        directory : Path</span>
<span class="sd">            Path to the directory containing input files</span>
<span class="sd">        sex_check_filename : str</span>
<span class="sd">            Filename of PLINK&#39;s sex check results (typically .sexcheck file)</span>
<span class="sd">        xchr_imiss_filename : str</span>
<span class="sd">            Filename of X chromosome missingness data</span>
<span class="sd">        plots_dir : Path, optional</span>
<span class="sd">            Directory where the plot should be saved. If None, uses default plots directory</span>
<span class="sd">        format : str, optional</span>
<span class="sd">            Format for saving the plot. Default is &#39;png&#39;</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        pd.DataFrame</span>
<span class="sd">            DataFrame containing samples that failed sex check QC with columns:</span>
<span class="sd">            - FID: Family ID</span>
<span class="sd">            - IID: Individual ID</span>
<span class="sd">            - Failure: Type of failure (always &#39;Sex check&#39;)</span>
<span class="sd">        </span>
<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        The function creates a scatter plot with:</span>
<span class="sd">        - Blue hollow circles for samples with Male PEDSEX</span>
<span class="sd">        - Green hollow circles for samples with Female PEDSEX</span>
<span class="sd">        - Red filled circles for problematic samples</span>
<span class="sd">        - Dotted red vertical lines at F=0.2 and F=0.8</span>
<span class="sd">        The plot is saved as &#39;sex_check.jpeg&#39; in the specified plots directory.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">directory</span><span class="p">,</span> <span class="n">Path</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;directory should be a Path object&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">sex_check_filename</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;sex_check_filename should be a string&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">xchr_imiss_filename</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;xchr_imiss_filename should be a string&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="nb">format</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;format should be a string&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">plots_dir</span><span class="p">,</span> <span class="p">(</span><span class="n">Path</span><span class="p">,</span> <span class="nb">type</span><span class="p">(</span><span class="kc">None</span><span class="p">))):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;plots_dir should be a Path object or None&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">directory</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Directory </span><span class="si">{</span><span class="n">directory</span><span class="si">}</span><span class="s2"> does not exist&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">format</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;png&#39;</span><span class="p">,</span> <span class="s1">&#39;jpeg&#39;</span><span class="p">,</span> <span class="s1">&#39;jpg&#39;</span><span class="p">,</span> <span class="s1">&#39;svg&#39;</span><span class="p">,</span> <span class="s1">&#39;pdf&#39;</span><span class="p">,</span> <span class="s1">&#39;ps&#39;</span><span class="p">]:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;format should be one of [&#39;png&#39;, &#39;jpeg&#39;, &#39;jpg&#39;, &#39;svg&#39;, &#39;pdf&#39;, &#39;ps]&quot;</span><span class="p">)</span>
        
        <span class="k">if</span> <span class="ow">not</span> <span class="n">plots_dir</span><span class="p">:</span>
            <span class="n">plots_dir</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">plots_dir</span>

        <span class="n">df_sexcheck</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
            <span class="n">directory</span> <span class="o">/</span> <span class="n">sex_check_filename</span><span class="p">,</span>
            <span class="n">sep</span>   <span class="o">=</span><span class="sa">r</span><span class="s1">&#39;\s+&#39;</span><span class="p">,</span>
            <span class="n">engine</span><span class="o">=</span><span class="s1">&#39;python&#39;</span>
        <span class="p">)</span>
        <span class="n">df_sexcheck</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span><span class="o">.</span><span class="n">lstrip</span><span class="p">(</span><span class="s1">&#39;#&#39;</span><span class="p">)</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">df_sexcheck</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>

        <span class="n">df_xchr_smiss</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
            <span class="n">directory</span> <span class="o">/</span> <span class="n">xchr_imiss_filename</span><span class="p">,</span>
            <span class="n">sep</span>   <span class="o">=</span><span class="sa">r</span><span class="s1">&#39;\s+&#39;</span><span class="p">,</span>
            <span class="n">engine</span><span class="o">=</span><span class="s1">&#39;python&#39;</span>
        <span class="p">)</span>
        <span class="n">df_xchr_smiss</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span><span class="o">.</span><span class="n">lstrip</span><span class="p">(</span><span class="s1">&#39;#&#39;</span><span class="p">)</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">df_xchr_smiss</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>

        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">df_sexcheck</span><span class="p">,</span> <span class="n">df_xchr_smiss</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;FID&#39;</span><span class="p">,</span> <span class="s1">&#39;IID&#39;</span><span class="p">],</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;inner&#39;</span><span class="p">)</span>

        <span class="n">fail_sexcheck</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;STATUS&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;PROBLEM&#39;</span><span class="p">][[</span><span class="s1">&#39;FID&#39;</span><span class="p">,</span> <span class="s1">&#39;IID&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">fail_sexcheck</span><span class="p">[</span><span class="s1">&#39;Failure&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Sex check&#39;</span>

        <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Category&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;General&#39;</span>
        <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;PEDSEX&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">,</span> <span class="s1">&#39;Category&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Male PEDSEX&#39;</span>
        <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;PEDSEX&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">2</span><span class="p">,</span> <span class="s1">&#39;Category&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Female PEDSEX&#39;</span>

        <span class="n">df_problem</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;STATUS&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;PROBLEM&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;STATUS&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;PROBLEM&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="c1"># Define the palette (color mapping)</span>
        <span class="n">palette</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;Male PEDSEX&quot;</span>  <span class="p">:</span> <span class="s2">&quot;blue&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Female PEDSEX&quot;</span><span class="p">:</span> <span class="s2">&quot;green&quot;</span>
        <span class="p">}</span>

        <span class="c1"># Define the size mapping</span>
        <span class="n">size_mapping</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;Male PEDSEX&quot;</span>  <span class="p">:</span> <span class="mi">40</span><span class="p">,</span>
            <span class="s2">&quot;Female PEDSEX&quot;</span><span class="p">:</span> <span class="mi">40</span>
        <span class="p">}</span>

        <span class="c1"># Create the Matplotlib scatter plot</span>
        <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="n">fig_size</span><span class="p">)</span>

        <span class="c1"># Iterate through categories to plot each group separately</span>
        <span class="k">for</span> <span class="n">category</span><span class="p">,</span> <span class="n">group</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;Category&quot;</span><span class="p">):</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span>
                <span class="n">group</span><span class="p">[</span><span class="s2">&quot;F&quot;</span><span class="p">],</span> 
                <span class="n">group</span><span class="p">[</span><span class="s2">&quot;F_MISS&quot;</span><span class="p">],</span> 
                <span class="n">edgecolors</span><span class="o">=</span><span class="n">palette</span><span class="p">[</span><span class="n">category</span><span class="p">],</span>     <span class="c1"># Map color # type: ignore</span>
                <span class="n">facecolors</span><span class="o">=</span><span class="s1">&#39;none&#39;</span><span class="p">,</span>                <span class="c1"># Hollow circles</span>
                <span class="n">s</span>         <span class="o">=</span><span class="n">size_mapping</span><span class="p">[</span><span class="n">category</span><span class="p">],</span><span class="c1"># Map size # type: ignore</span>
                <span class="n">label</span>     <span class="o">=</span><span class="n">category</span>               <span class="c1"># Add label for legend</span>
            <span class="p">)</span>

        <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span>
            <span class="n">df_problem</span><span class="p">[</span><span class="s2">&quot;F&quot;</span><span class="p">],</span> 
            <span class="n">df_problem</span><span class="p">[</span><span class="s2">&quot;F_MISS&quot;</span><span class="p">],</span> 
            <span class="n">color</span>     <span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span>
            <span class="n">s</span>         <span class="o">=</span><span class="mi">25</span><span class="p">,</span>
            <span class="n">marker</span>    <span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span>
            <span class="n">label</span>     <span class="o">=</span><span class="s1">&#39;Problem Status&#39;</span><span class="p">,</span>
            <span class="n">edgecolors</span><span class="o">=</span><span class="n">palette</span><span class="p">[</span><span class="s1">&#39;Female PEDSEX&#39;</span><span class="p">],</span>
        <span class="p">)</span>

        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="n">bottom</span><span class="o">=-</span><span class="mf">0.01</span><span class="p">)</span>  

        <span class="c1"># Add vertical lines</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">f_coeff_thresholds</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;dotted&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">f_coeff_thresholds</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;dotted&#39;</span><span class="p">)</span>

        <span class="c1"># Customize labels and legend</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Sex Check&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;X chr inbreeding (homozygosity) estimate F&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Proportion of missing SNPs for the X chr&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;best&#39;</span><span class="p">)</span>
        
        <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">plots_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;sex_check.</span><span class="si">{</span><span class="nb">format</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">400</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">fail_sexcheck</span></div>

    
<div class="viewcode-block" id="SampleQC.report_heterozygosity_rate">
<a class="viewcode-back" href="../../SampleQC.html#ideal_genom_qc.SampleQC.SampleQC.report_heterozygosity_rate">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">report_heterozygosity_rate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">het_filename</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">autosomal_filename</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">std_deviation_het</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">float</span><span class="p">,</span> <span class="nb">int</span><span class="p">],</span> <span class="n">maf</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span> <span class="n">split</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">plots_dir</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">y_axis_cap</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">float</span><span class="p">,</span> <span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="mi">80</span><span class="p">,</span> <span class="nb">format</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;png&#39;</span><span class="p">,</span> <span class="n">scatter_fig_size</span><span class="p">:</span> <span class="nb">tuple</span> <span class="o">=</span> <span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Analyze and report heterozygosity rates for samples, creating visualization plots and identifying samples that fail heterozygosity rate checks.</span>
<span class="sd">        This function loads heterozygosity and autosomal call rate data, merges them, identifies samples with deviant heterozygosity rates,</span>
<span class="sd">        and generates visualization plots to aid in quality control analysis.</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        directory : str</span>
<span class="sd">            Path to the directory containing input files</span>
<span class="sd">        summary_ped_filename : str</span>
<span class="sd">            Filename of the summary PED file containing heterozygosity information</span>
<span class="sd">        autosomal_filename : str</span>
<span class="sd">            Filename of the autosomal file containing call rate information</span>
<span class="sd">        std_deviation_het : float</span>
<span class="sd">            Number of standard deviations to use as threshold for identifying deviant samples</span>
<span class="sd">        maf : float</span>
<span class="sd">            Minor allele frequency threshold used in the analysis</span>
<span class="sd">        split : str</span>
<span class="sd">            Direction of MAF comparison (&#39;&gt;&#39; or &#39;&lt;&#39;)</span>
<span class="sd">        plots_dir : str</span>
<span class="sd">            Directory where plot files will be saved</span>
<span class="sd">        y_axis_cap : float, optional</span>
<span class="sd">            Maximum value for y-axis in capped histogram plot (default: 80)</span>
<span class="sd">        format : str, optional</span>
<span class="sd">            Format for saving plots (default: &#39;png&#39;)</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        pd.DataFrame</span>
<span class="sd">            DataFrame containing samples that failed heterozygosity rate check with columns:</span>
<span class="sd">            - FID: Family ID</span>
<span class="sd">            - IID: Individual ID</span>
<span class="sd">            - Failure: Description of failure type</span>
<span class="sd">        </span>
<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        The function generates two types of plots:</span>
<span class="sd">        1. Histograms of heterozygosity rates (both uncapped and capped)</span>
<span class="sd">        2. Scatter plot of heterozygosity rate vs missing SNP proportion</span>
<span class="sd">        Files are saved as JPEG images in the specified plots directory.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">het_filename</span><span class="p">,</span> <span class="n">Path</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;het_filename should be a Path object&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">autosomal_filename</span><span class="p">,</span> <span class="n">Path</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;autosomal_filename should be a Path object&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">std_deviation_het</span><span class="p">,</span> <span class="p">(</span><span class="nb">float</span><span class="p">,</span> <span class="nb">int</span><span class="p">)):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;std_deviation_het should be a float or int&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">maf</span><span class="p">,</span> <span class="nb">float</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;maf should be a float&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">maf</span><span class="o">&lt;</span><span class="mi">0</span> <span class="ow">or</span> <span class="n">maf</span><span class="o">&gt;</span><span class="mf">0.5</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;maf should be between 0 and 0.5&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">split</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;split should be a string&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">plots_dir</span><span class="p">,</span> <span class="n">Path</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;plots_dir should be a Path object&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">y_axis_cap</span><span class="p">,</span> <span class="p">(</span><span class="nb">float</span><span class="p">,</span> <span class="nb">int</span><span class="p">)):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;y_axis_cap should be a float or int&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="nb">format</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;format should be a string&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">format</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;png&#39;</span><span class="p">,</span> <span class="s1">&#39;jpeg&#39;</span><span class="p">,</span> <span class="s1">&#39;jpg&#39;</span><span class="p">,</span> <span class="s1">&#39;svg&#39;</span><span class="p">,</span> <span class="s1">&#39;pdf&#39;</span><span class="p">,</span> <span class="s1">&#39;ps&#39;</span><span class="p">]:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;format should be one of [&#39;png&#39;, &#39;jpeg&#39;, &#39;jpg&#39;, &#39;svg&#39;, &#39;pdf&#39;, &#39;ps]&quot;</span><span class="p">)</span>
        
        <span class="c1"># load samples that failed heterozygosity rate check with MAF &gt; threshold</span>
        <span class="n">df_het</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
            <span class="n">het_filename</span><span class="p">,</span>
            <span class="n">sep</span>   <span class="o">=</span><span class="sa">r</span><span class="s1">&#39;\s+&#39;</span><span class="p">,</span>
            <span class="n">engine</span><span class="o">=</span><span class="s1">&#39;python&#39;</span>
        <span class="p">)</span>
        <span class="n">df_het</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span><span class="o">.</span><span class="n">lstrip</span><span class="p">(</span><span class="s1">&#39;#&#39;</span><span class="p">)</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">df_het</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>
        <span class="n">df_het</span><span class="p">[</span><span class="s2">&quot;HET_RATE&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">-</span> <span class="p">(</span><span class="n">df_het</span><span class="p">[</span><span class="s2">&quot;O(HOM)&quot;</span><span class="p">]</span> <span class="o">/</span> <span class="n">df_het</span><span class="p">[</span><span class="s2">&quot;OBS_CT&quot;</span><span class="p">])</span>


        <span class="c1"># autosomal call rate per individual</span>
        <span class="n">df_autosomal</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
            <span class="n">autosomal_filename</span><span class="p">,</span>
            <span class="n">sep</span>   <span class="o">=</span><span class="sa">r</span><span class="s1">&#39;\s+&#39;</span><span class="p">,</span>
            <span class="n">engine</span><span class="o">=</span><span class="s1">&#39;python&#39;</span>
        <span class="p">)</span>
        <span class="n">df_autosomal</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span><span class="o">.</span><span class="n">lstrip</span><span class="p">(</span><span class="s1">&#39;#&#39;</span><span class="p">)</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">df_autosomal</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>

        <span class="c1"># merge both dataframes</span>
        <span class="n">df_het</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span>
            <span class="n">df_het</span><span class="p">[[</span><span class="s1">&#39;FID&#39;</span><span class="p">,</span> <span class="s1">&#39;IID&#39;</span><span class="p">,</span> <span class="s1">&#39;HET_RATE&#39;</span><span class="p">]],</span>
            <span class="n">df_autosomal</span><span class="p">[[</span><span class="s1">&#39;FID&#39;</span><span class="p">,</span> <span class="s1">&#39;IID&#39;</span><span class="p">,</span> <span class="s1">&#39;F_MISS&#39;</span><span class="p">]],</span>
            <span class="n">on</span> <span class="o">=</span><span class="p">[</span><span class="s1">&#39;FID&#39;</span><span class="p">,</span> <span class="s1">&#39;IID&#39;</span><span class="p">],</span>
            <span class="n">how</span><span class="o">=</span><span class="s1">&#39;inner&#39;</span>
        <span class="p">)</span>

        <span class="n">mean_percent</span><span class="o">=</span> <span class="n">df_het</span><span class="p">[</span><span class="s1">&#39;HET_RATE&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
        <span class="n">sd_percent</span>  <span class="o">=</span> <span class="n">df_het</span><span class="p">[</span><span class="s1">&#39;HET_RATE&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">std</span><span class="p">()</span>

        <span class="n">mask_plus</span> <span class="o">=</span> <span class="n">df_het</span><span class="p">[</span><span class="s1">&#39;HET_RATE&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">mean_percent</span> <span class="o">+</span> <span class="n">std_deviation_het</span><span class="o">*</span><span class="n">sd_percent</span>
        <span class="n">mask_minus</span><span class="o">=</span> <span class="n">df_het</span><span class="p">[</span><span class="s1">&#39;HET_RATE&#39;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">mean_percent</span> <span class="o">-</span> <span class="n">std_deviation_het</span><span class="o">*</span><span class="n">sd_percent</span>

        <span class="n">fail_het</span> <span class="o">=</span> <span class="n">df_het</span><span class="p">[</span><span class="n">mask_plus</span> <span class="o">|</span> <span class="n">mask_minus</span><span class="p">][[</span><span class="s1">&#39;FID&#39;</span><span class="p">,</span> <span class="s1">&#39;IID&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">split</span> <span class="o">==</span> <span class="s1">&#39;&gt;&#39;</span><span class="p">:</span>
            <span class="n">fail_het</span><span class="p">[</span><span class="s1">&#39;Failure&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Heterozygosity rate greater&#39;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">fail_het</span><span class="p">[</span><span class="s1">&#39;Failure&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Heterozygosity rate less&#39;</span>

        <span class="c1"># plots</span>

        <span class="n">fig1</span><span class="p">,</span> <span class="n">axes1</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">5</span><span class="p">),</span> <span class="n">sharey</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="n">axes1</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span><span class="n">df_het</span><span class="p">[</span><span class="s1">&#39;HET_RATE&#39;</span><span class="p">],</span> <span class="n">bins</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;green&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes1</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="c1"># type: ignore</span>
        <span class="n">axes1</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Autosomal heterozygosity&quot;</span><span class="p">)</span>
        <span class="n">axes1</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;% Heterozygosity MAF </span><span class="si">{</span><span class="n">split</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">maf</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">axes1</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Frequency&quot;</span><span class="p">)</span>

        <span class="n">axes1</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span><span class="n">df_het</span><span class="p">[</span><span class="s1">&#39;HET_RATE&#39;</span><span class="p">],</span> <span class="n">bins</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;green&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes1</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="c1"># type: ignore</span>
        <span class="n">axes1</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Autosomal heterozygosity (capped)&quot;</span><span class="p">)</span>
        <span class="n">axes1</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;% Heterozygosity MAF </span><span class="si">{</span><span class="n">split</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">maf</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">axes1</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">y_axis_cap</span><span class="p">)</span>  <span class="c1"># Cap y-axis</span>
        <span class="n">axes1</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Frequency&quot;</span><span class="p">)</span>

        <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        
        <span class="k">if</span> <span class="n">split</span> <span class="o">==</span> <span class="s1">&#39;&gt;&#39;</span><span class="p">:</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">plots_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;heterozygosity_rate_greater_</span><span class="si">{</span><span class="n">maf</span><span class="si">}</span><span class="s2">_histogram.</span><span class="si">{</span><span class="nb">format</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">400</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">plots_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;heterozygosity_rate_less_</span><span class="si">{</span><span class="n">maf</span><span class="si">}</span><span class="s2">_histogram.</span><span class="si">{</span><span class="nb">format</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">400</span><span class="p">)</span>
        
        <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">block</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="n">df_het</span><span class="p">[</span><span class="s1">&#39;Deviated&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Not Excluded&#39;</span>
        <span class="n">df_het</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">mask_plus</span><span class="p">,</span> <span class="s1">&#39;Deviated&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">std_deviation_het</span><span class="si">}</span><span class="s1">xSD Excluded&#39;</span>
        <span class="n">df_het</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">mask_minus</span><span class="p">,</span> <span class="s1">&#39;Deviated&#39;</span><span class="p">]</span><span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">std_deviation_het</span><span class="si">}</span><span class="s1">xSD Excluded&#39;</span>

        <span class="c1"># Create the scatter plot</span>
        <span class="n">fig</span><span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="n">scatter_fig_size</span><span class="p">)</span>
        <span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">111</span><span class="p">)</span>

        <span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span>
            <span class="n">data</span>   <span class="o">=</span><span class="n">df_het</span><span class="p">,</span>
            <span class="n">x</span>      <span class="o">=</span><span class="s1">&#39;HET_RATE&#39;</span><span class="p">,</span>
            <span class="n">y</span>      <span class="o">=</span><span class="s1">&#39;F_MISS&#39;</span><span class="p">,</span>
            <span class="n">hue</span>    <span class="o">=</span><span class="s1">&#39;Deviated&#39;</span><span class="p">,</span>
            <span class="n">palette</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;Not Excluded&#39;</span><span class="p">:</span> <span class="s1">&#39;blue&#39;</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">std_deviation_het</span><span class="si">}</span><span class="s1">xSD Excluded&#39;</span><span class="p">:</span> <span class="s1">&#39;red&#39;</span><span class="p">},</span>
            <span class="n">markers</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;Not Excluded&#39;</span><span class="p">:</span> <span class="s1">&#39;o&#39;</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">std_deviation_het</span><span class="si">}</span><span class="s1">xSD Excluded&#39;</span><span class="p">:</span> <span class="s1">&#39;o&#39;</span><span class="p">},</span>
            <span class="n">size</span>   <span class="o">=</span><span class="s1">&#39;Deviated&#39;</span><span class="p">,</span>
            <span class="n">sizes</span>  <span class="o">=</span><span class="p">{</span><span class="s1">&#39;Not Excluded&#39;</span><span class="p">:</span> <span class="mi">20</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">std_deviation_het</span><span class="si">}</span><span class="s1">xSD Excluded&#39;</span><span class="p">:</span> <span class="mi">30</span><span class="p">}</span>
        <span class="p">)</span>

        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="n">bottom</span><span class="o">=-</span><span class="mf">0.01</span><span class="p">)</span> 

        <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Autosomal heterozygosity and call rate&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;% Heterozygosity MAF </span><span class="si">{</span><span class="n">split</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">maf</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Proportion of missing SNPs&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="s1">&#39;Exclusion&#39;</span><span class="p">,</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;best&#39;</span><span class="p">)</span>

        <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">split</span> <span class="o">==</span> <span class="s1">&#39;&gt;&#39;</span><span class="p">:</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">plots_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;heterozygosity_rate_greater_</span><span class="si">{</span><span class="n">maf</span><span class="si">}</span><span class="s2">_scatterplot.</span><span class="si">{</span><span class="nb">format</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">400</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">plots_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;heterozygosity_rate_less_</span><span class="si">{</span><span class="n">maf</span><span class="si">}</span><span class="s2">_scatterplot.</span><span class="si">{</span><span class="nb">format</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">400</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">block</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">fail_het</span></div>


<div class="viewcode-block" id="SampleQC.report_ibd_analysis">
<a class="viewcode-back" href="../../SampleQC.html#ideal_genom_qc.SampleQC.SampleQC.report_ibd_analysis">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">report_ibd_analysis</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ibd_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.185</span><span class="p">,</span> <span class="n">chunk_size</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">100000</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Analyze IBD (Identity By Descent) to identify duplicated or related samples.</span>
<span class="sd">        </span>
<span class="sd">        This method processes IBD analysis results to identify sample pairs with IBD scores</span>
<span class="sd">        above a specified threshold, indicating potential duplicates or related individuals.</span>
<span class="sd">        For identified pairs, it uses missingness data to determine which sample should be</span>
<span class="sd">        removed (keeping the sample with lower missingness rate).</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ibd_threshold : float, default=0.185</span>
<span class="sd">            The PI_HAT threshold above which samples are considered related.</span>
<span class="sd">            Typical values: &gt;0.98 for duplicates, &gt;0.5 for first-degree relatives.</span>
<span class="sd">        chunk_size : int, default=100000</span>
<span class="sd">            Number of rows to process at a time when reading the genome file.</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        pd.DataFrame</span>
<span class="sd">            A DataFrame containing samples to be removed, with columns:</span>
<span class="sd">            - FID: Family ID</span>
<span class="sd">            - IID: Individual ID</span>
<span class="sd">            - Failure: Reason for removal (&#39;Duplicates and relatedness (IBD)&#39;)</span>
<span class="sd">            Returns empty DataFrame if no related pairs are found or if KING is used.</span>
<span class="sd">        </span>
<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        TypeError</span>
<span class="sd">            If ibd_threshold is not a float.</span>
<span class="sd">        FileNotFoundError</span>
<span class="sd">            If required input files (``*.smiss`` or ``*.genome``) are not found.</span>
<span class="sd">        </span>
<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        The method requires two input files:</span>
<span class="sd">        - {output_name}-ibd-missing.imiss: Contains sample missingness information</span>
<span class="sd">        - {output_name}-ibd.genome: Contains pairwise IBD estimates</span>
<span class="sd">        For each related pair, the sample with higher missingness rate is marked for removal.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ibd_threshold</span><span class="p">,</span> <span class="nb">float</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;ibd_threshold should be a float&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">chunk_size</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;chunk_size should be an integer&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">use_kinship</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>

        <span class="c1"># File paths</span>

        <span class="n">imiss_path</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span> <span class="o">+</span> <span class="s1">&#39;-ibd-missing.imiss&#39;</span><span class="p">)</span>
        <span class="n">genome_path</span><span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span> <span class="o">+</span> <span class="s1">&#39;-ibd.genome&#39;</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">imiss_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Missing file: </span><span class="si">{</span><span class="n">imiss_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">genome_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Missing file: </span><span class="si">{</span><span class="n">genome_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># Load .imiss file</span>
        <span class="n">df_imiss</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">imiss_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="sa">r</span><span class="s1">&#39;\s+&#39;</span><span class="p">,</span> <span class="n">engine</span><span class="o">=</span><span class="s1">&#39;python&#39;</span><span class="p">)</span>

        <span class="c1"># Initialize dataframe for duplicates</span>
        <span class="n">duplicates</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="c1"># Process the .genome file in chunks</span>
        <span class="k">for</span> <span class="n">chunk</span> <span class="ow">in</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
            <span class="n">genome_path</span><span class="p">,</span>
            <span class="n">usecols</span>  <span class="o">=</span><span class="p">[</span><span class="s1">&#39;FID1&#39;</span><span class="p">,</span> <span class="s1">&#39;IID1&#39;</span><span class="p">,</span> <span class="s1">&#39;FID2&#39;</span><span class="p">,</span> <span class="s1">&#39;IID2&#39;</span><span class="p">,</span> <span class="s1">&#39;PI_HAT&#39;</span><span class="p">],</span>
            <span class="n">sep</span>      <span class="o">=</span><span class="sa">r</span><span class="s1">&#39;\s+&#39;</span><span class="p">,</span>
            <span class="n">engine</span>   <span class="o">=</span><span class="s1">&#39;python&#39;</span><span class="p">,</span>
            <span class="n">chunksize</span><span class="o">=</span><span class="n">chunk_size</span><span class="p">,</span>
        <span class="p">):</span>
            <span class="c1"># Filter rows with PI_HAT &gt; ibd_threshold</span>
            <span class="n">filtered_chunk</span> <span class="o">=</span> <span class="n">chunk</span><span class="p">[</span><span class="n">chunk</span><span class="p">[</span><span class="s1">&#39;PI_HAT&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">ibd_threshold</span><span class="p">]</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">filtered_chunk</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
                <span class="n">duplicates</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">filtered_chunk</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">duplicates</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;FID&#39;</span><span class="p">,</span> <span class="s1">&#39;IID&#39;</span><span class="p">,</span> <span class="s1">&#39;Failure&#39;</span><span class="p">])</span>

        <span class="c1"># Concatenate all filtered chunks</span>
        <span class="n">df_dup</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">duplicates</span><span class="p">,</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="c1"># Merge with missingness data</span>
        <span class="n">imiss_related1</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span>
            <span class="n">df_dup</span><span class="p">[[</span><span class="s1">&#39;FID1&#39;</span><span class="p">,</span> <span class="s1">&#39;IID1&#39;</span><span class="p">]],</span>
            <span class="n">df_imiss</span><span class="p">[[</span><span class="s1">&#39;FID&#39;</span><span class="p">,</span> <span class="s1">&#39;IID&#39;</span><span class="p">,</span> <span class="s1">&#39;F_MISS&#39;</span><span class="p">]],</span>
            <span class="n">left_on</span> <span class="o">=</span><span class="p">[</span><span class="s1">&#39;FID1&#39;</span><span class="p">,</span> <span class="s1">&#39;IID1&#39;</span><span class="p">],</span>
            <span class="n">right_on</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;FID&#39;</span><span class="p">,</span> <span class="s1">&#39;IID&#39;</span><span class="p">],</span>
        <span class="p">)</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;F_MISS&#39;</span><span class="p">:</span> <span class="s1">&#39;F_MISS_1&#39;</span><span class="p">})</span>

        <span class="n">imiss_related2</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span>
            <span class="n">df_dup</span><span class="p">[[</span><span class="s1">&#39;FID2&#39;</span><span class="p">,</span> <span class="s1">&#39;IID2&#39;</span><span class="p">]],</span>
            <span class="n">df_imiss</span><span class="p">[[</span><span class="s1">&#39;FID&#39;</span><span class="p">,</span> <span class="s1">&#39;IID&#39;</span><span class="p">,</span> <span class="s1">&#39;F_MISS&#39;</span><span class="p">]],</span>
            <span class="n">left_on</span> <span class="o">=</span><span class="p">[</span><span class="s1">&#39;FID2&#39;</span><span class="p">,</span> <span class="s1">&#39;IID2&#39;</span><span class="p">],</span>
            <span class="n">right_on</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;FID&#39;</span><span class="p">,</span> <span class="s1">&#39;IID&#39;</span><span class="p">],</span>
        <span class="p">)</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;F_MISS&#39;</span><span class="p">:</span> <span class="s1">&#39;F_MISS_2&#39;</span><span class="p">})</span>

        <span class="c1"># Decide which samples to remove</span>
        <span class="n">to_remove</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span>
            <span class="p">[</span>
                <span class="n">imiss_related1</span><span class="p">[[</span><span class="s1">&#39;FID1&#39;</span><span class="p">,</span> <span class="s1">&#39;IID1&#39;</span><span class="p">,</span> <span class="s1">&#39;F_MISS_1&#39;</span><span class="p">]],</span>
                <span class="n">imiss_related2</span><span class="p">[[</span><span class="s1">&#39;FID2&#39;</span><span class="p">,</span> <span class="s1">&#39;IID2&#39;</span><span class="p">,</span> <span class="s1">&#39;F_MISS_2&#39;</span><span class="p">]],</span>
            <span class="p">],</span>
            <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="n">to_remove</span><span class="p">[</span><span class="s1">&#39;FID&#39;</span><span class="p">],</span> <span class="n">to_remove</span><span class="p">[</span><span class="s1">&#39;IID&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span>
            <span class="n">to_remove</span><span class="p">[</span><span class="s1">&#39;F_MISS_1&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">to_remove</span><span class="p">[</span><span class="s1">&#39;F_MISS_2&#39;</span><span class="p">],</span>
            <span class="p">(</span><span class="n">to_remove</span><span class="p">[</span><span class="s1">&#39;FID1&#39;</span><span class="p">],</span> <span class="n">to_remove</span><span class="p">[</span><span class="s1">&#39;IID1&#39;</span><span class="p">]),</span>
            <span class="p">(</span><span class="n">to_remove</span><span class="p">[</span><span class="s1">&#39;FID2&#39;</span><span class="p">],</span> <span class="n">to_remove</span><span class="p">[</span><span class="s1">&#39;IID2&#39;</span><span class="p">]),</span>
        <span class="p">)</span>

        <span class="n">to_remove</span> <span class="o">=</span> <span class="n">to_remove</span><span class="p">[[</span><span class="s1">&#39;FID&#39;</span><span class="p">,</span> <span class="s1">&#39;IID&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">to_remove</span><span class="p">[</span><span class="s1">&#39;Failure&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Duplicates and relatedness (IBD)&#39;</span>

        <span class="k">return</span> <span class="n">to_remove</span></div>

    
<div class="viewcode-block" id="SampleQC.clean_input_folder">
<a class="viewcode-back" href="../../SampleQC.html#ideal_genom_qc.SampleQC.SampleQC.clean_input_folder">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">clean_input_folder</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Removes specific files from the input folder.</span>

<span class="sd">        This method cleans the input folder by removing files that contain &#39;missing&#39; or</span>
<span class="sd">        &#39;renamed&#39; in their names, excluding log files. The cleaning process helps maintain</span>
<span class="sd">        organization by removing temporary or processed files.</span>

<span class="sd">        Returns:</span>
<span class="sd">            None</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;STEP: Clean input folder&quot;</span><span class="p">)</span>

        <span class="c1"># remove all files from input folder</span>
        <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">input_path</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="s1">&#39;*&#39;</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">file</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="ow">and</span> <span class="n">file</span><span class="o">.</span><span class="n">suffix</span> <span class="o">!=</span> <span class="s1">&#39;.log&#39;</span> <span class="ow">and</span> <span class="s1">&#39;missing&#39;</span> <span class="ow">in</span> <span class="n">file</span><span class="o">.</span><span class="n">name</span><span class="p">:</span>
                <span class="n">file</span><span class="o">.</span><span class="n">unlink</span><span class="p">()</span>
            <span class="k">elif</span> <span class="n">file</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="ow">and</span> <span class="n">file</span><span class="o">.</span><span class="n">suffix</span> <span class="o">!=</span> <span class="s1">&#39;.log&#39;</span> <span class="ow">and</span> <span class="s1">&#39;renamed&#39;</span> <span class="ow">in</span> <span class="n">file</span><span class="o">.</span><span class="n">name</span><span class="p">:</span>
                <span class="n">file</span><span class="o">.</span><span class="n">unlink</span><span class="p">()</span>

        <span class="k">return</span></div>

    
<div class="viewcode-block" id="SampleQC.clean_result_folder">
<a class="viewcode-back" href="../../SampleQC.html#ideal_genom_qc.SampleQC.SampleQC.clean_result_folder">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">clean_result_folder</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Clean the results folder by removing specific file types.</span>

<span class="sd">        This method removes all .bed, .bim, and .fam files from the results directory,</span>
<span class="sd">        while preserving log files. The cleaning process helps maintain a tidy workspace</span>
<span class="sd">        by removing intermediate or temporary files from previous runs.</span>

<span class="sd">        Returns:</span>
<span class="sd">            None</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;STEP: Clean results folder&quot;</span><span class="p">)</span>

        <span class="c1"># remove all files from output folder</span>
        <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="s1">&#39;*&#39;</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">file</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="ow">and</span> <span class="n">file</span><span class="o">.</span><span class="n">suffix</span> <span class="o">!=</span> <span class="s1">&#39;.log&#39;</span> <span class="ow">and</span> <span class="s1">&#39;.bed&#39;</span> <span class="ow">in</span> <span class="n">file</span><span class="o">.</span><span class="n">name</span><span class="p">:</span>
                <span class="n">file</span><span class="o">.</span><span class="n">unlink</span><span class="p">()</span>
            <span class="k">elif</span> <span class="n">file</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="ow">and</span> <span class="n">file</span><span class="o">.</span><span class="n">suffix</span> <span class="o">!=</span> <span class="s1">&#39;.log&#39;</span> <span class="ow">and</span> <span class="s1">&#39;.bim&#39;</span> <span class="ow">in</span> <span class="n">file</span><span class="o">.</span><span class="n">name</span><span class="p">:</span>
                <span class="n">file</span><span class="o">.</span><span class="n">unlink</span><span class="p">()</span>
            <span class="k">elif</span> <span class="n">file</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="ow">and</span> <span class="n">file</span><span class="o">.</span><span class="n">suffix</span> <span class="o">!=</span> <span class="s1">&#39;.log&#39;</span> <span class="ow">and</span> <span class="s1">&#39;.fam&#39;</span> <span class="ow">in</span> <span class="n">file</span><span class="o">.</span><span class="n">name</span><span class="p">:</span>
                <span class="n">file</span><span class="o">.</span><span class="n">unlink</span><span class="p">()</span>

        <span class="k">return</span></div>


<div class="viewcode-block" id="SampleQC.execute_sample_qc_pipeline">
<a class="viewcode-back" href="../../SampleQC.html#ideal_genom_qc.SampleQC.SampleQC.execute_sample_qc_pipeline">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">execute_sample_qc_pipeline</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sample_params</span><span class="p">:</span> <span class="nb">dict</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Execute the complete sample quality control pipeline.</span>
<span class="sd">        </span>
<span class="sd">        This method runs all sample QC steps in the correct order with proper</span>
<span class="sd">        memory management and logging. It encapsulates the entire workflow</span>
<span class="sd">        that was previously handled in the main script.</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        sample_params : dict</span>
<span class="sd">            Dictionary containing all sample QC parameters with keys:</span>
<span class="sd">            - &#39;rename_snp&#39;: bool, whether to rename SNPs</span>
<span class="sd">            - &#39;hh_to_missing&#39;: bool, whether to convert haploid to missing</span>
<span class="sd">            - &#39;ind_pair&#39;: list, LD pruning parameters [window, step, r2]</span>
<span class="sd">            - &#39;mind&#39;: float, missing genotype rate threshold</span>
<span class="sd">            - &#39;sex_check&#39;: list, sex check F-statistic thresholds</span>
<span class="sd">            - &#39;maf&#39;: float, minor allele frequency for heterozygosity</span>
<span class="sd">            - &#39;kinship&#39;: float, kinship coefficient threshold</span>
<span class="sd">            - &#39;use_kinship&#39;: bool, whether to use KING vs IBD</span>
<span class="sd">            - &#39;het_deviation&#39;: float, standard deviations for het filtering</span>
<span class="sd">            - &#39;ibd_threshold&#39;: float, IBD threshold for relatedness</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None</span>
<span class="sd">        </span>
<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        The pipeline executes steps in this order:</span>
<span class="sd">        1. Rename SNPs (optional)</span>
<span class="sd">        2. Convert haploid to missing (optional)</span>
<span class="sd">        3. LD pruning</span>
<span class="sd">        4. Missing genotype analysis</span>
<span class="sd">        5. Sex check</span>
<span class="sd">        6. Heterozygosity rate analysis</span>
<span class="sd">        7. Duplicate/relatedness analysis</span>
<span class="sd">        8. Identify failed samples</span>
<span class="sd">        9. Remove failed samples</span>
<span class="sd">        10. Clean up temporary files</span>
<span class="sd">        </span>
<span class="sd">        Memory usage is monitored after each step and garbage collection</span>
<span class="sd">        is performed to prevent memory issues with large datasets.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        
        <span class="kn">import</span><span class="w"> </span><span class="nn">time</span>
        <span class="kn">import</span><span class="w"> </span><span class="nn">gc</span>
        <span class="kn">import</span><span class="w"> </span><span class="nn">psutil</span>
        
        <span class="n">sample_qc_steps</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s1">&#39;rename SNPs&#39;</span>           <span class="p">:</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">execute_rename_snpid</span><span class="p">,</span> <span class="p">{</span><span class="s2">&quot;rename&quot;</span><span class="p">:</span> <span class="n">sample_params</span><span class="p">[</span><span class="s1">&#39;rename_snp&#39;</span><span class="p">]}),</span>
            <span class="s1">&#39;hh_to_missing&#39;</span>         <span class="p">:</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">execute_haploid_to_missing</span><span class="p">,</span> <span class="p">{</span><span class="s2">&quot;hh_to_missing&quot;</span><span class="p">:</span> <span class="n">sample_params</span><span class="p">[</span><span class="s1">&#39;hh_to_missing&#39;</span><span class="p">]}),</span>
            <span class="s1">&#39;ld_pruning&#39;</span>            <span class="p">:</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">execute_ld_pruning</span><span class="p">,</span> <span class="p">{</span><span class="s2">&quot;ind_pair&quot;</span><span class="p">:</span> <span class="n">sample_params</span><span class="p">[</span><span class="s1">&#39;ind_pair&#39;</span><span class="p">]}),</span>
            <span class="s1">&#39;miss_genotype&#39;</span>         <span class="p">:</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">execute_miss_genotype</span><span class="p">,</span> <span class="p">{}),</span>
            <span class="s1">&#39;sex_check&#39;</span>             <span class="p">:</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">execute_sex_check</span><span class="p">,</span> <span class="p">{</span><span class="s2">&quot;sex_check&quot;</span><span class="p">:</span> <span class="n">sample_params</span><span class="p">[</span><span class="s1">&#39;sex_check&#39;</span><span class="p">]}),</span>
            <span class="s1">&#39;heterozygosity&#39;</span>        <span class="p">:</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">execute_heterozygosity_rate</span><span class="p">,</span> <span class="p">{</span><span class="s2">&quot;maf&quot;</span><span class="p">:</span> <span class="n">sample_params</span><span class="p">[</span><span class="s1">&#39;maf&#39;</span><span class="p">]}),</span>
            <span class="s1">&#39;duplicates_relatedness&#39;</span><span class="p">:</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">execute_duplicate_relatedness</span><span class="p">,</span> <span class="p">{</span><span class="s2">&quot;kinship&quot;</span><span class="p">:</span> <span class="n">sample_params</span><span class="p">[</span><span class="s1">&#39;kinship&#39;</span><span class="p">],</span> <span class="s2">&quot;use_kinship&quot;</span><span class="p">:</span> <span class="n">sample_params</span><span class="p">[</span><span class="s1">&#39;use_kinship&#39;</span><span class="p">]}),</span>
            <span class="s1">&#39;get_fail_samples&#39;</span>      <span class="p">:</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">get_fail_samples</span><span class="p">,</span> <span class="p">{</span><span class="s2">&quot;call_rate_thres&quot;</span><span class="p">:</span> <span class="n">sample_params</span><span class="p">[</span><span class="s1">&#39;mind&#39;</span><span class="p">],</span> <span class="s2">&quot;std_deviation_het&quot;</span><span class="p">:</span> <span class="n">sample_params</span><span class="p">[</span><span class="s1">&#39;het_deviation&#39;</span><span class="p">],</span> <span class="s2">&quot;maf_het&quot;</span><span class="p">:</span> <span class="n">sample_params</span><span class="p">[</span><span class="s1">&#39;maf&#39;</span><span class="p">],</span> <span class="s2">&quot;ibd_threshold&quot;</span><span class="p">:</span> <span class="n">sample_params</span><span class="p">[</span><span class="s1">&#39;ibd_threshold&#39;</span><span class="p">]}),</span>
            <span class="s1">&#39;drop_fail_samples&#39;</span>     <span class="p">:</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">execute_drop_samples</span><span class="p">,</span> <span class="p">{}),</span>
            <span class="s1">&#39;clean_input_files&#39;</span>     <span class="p">:</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">clean_input_folder</span><span class="p">,</span> <span class="p">{}),</span>
            <span class="s1">&#39;clean_results_files&#39;</span>   <span class="p">:</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">clean_result_folder</span><span class="p">,</span> <span class="p">{}),</span>
        <span class="p">}</span>

        <span class="n">step_description</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s1">&#39;rename SNPs&#39;</span>           <span class="p">:</span> <span class="s1">&#39;Rename SNPs to chr:pos:ref:alt&#39;</span><span class="p">,</span>
            <span class="s1">&#39;hh_to_missing&#39;</span>         <span class="p">:</span> <span class="s1">&#39;Solve hh warnings by setting to missing&#39;</span><span class="p">,</span>
            <span class="s1">&#39;ld_pruning&#39;</span>            <span class="p">:</span> <span class="s1">&#39;Perform LD pruning&#39;</span><span class="p">,</span>
            <span class="s1">&#39;miss_genotype&#39;</span>         <span class="p">:</span> <span class="s1">&#39;Get samples with high missing rate&#39;</span><span class="p">,</span>
            <span class="s1">&#39;sex_check&#39;</span>             <span class="p">:</span> <span class="s1">&#39;Get samples with discordant sex information&#39;</span><span class="p">,</span>
            <span class="s1">&#39;heterozygosity&#39;</span>        <span class="p">:</span> <span class="s1">&#39;Get samples with high heterozygosity rate&#39;</span><span class="p">,</span>
            <span class="s1">&#39;duplicates_relatedness&#39;</span><span class="p">:</span> <span class="s1">&#39;Get samples with high relatedness rate or duplicates&#39;</span><span class="p">,</span>
            <span class="s1">&#39;get_fail_samples&#39;</span>      <span class="p">:</span> <span class="s1">&#39;Get samples that failed quality control&#39;</span><span class="p">,</span>
            <span class="s1">&#39;drop_fail_samples&#39;</span>     <span class="p">:</span> <span class="s1">&#39;Drop samples that failed quality control&#39;</span><span class="p">,</span>
            <span class="s1">&#39;clean_input_files&#39;</span>     <span class="p">:</span> <span class="s1">&#39;Clean input folder&#39;</span><span class="p">,</span>
            <span class="s1">&#39;clean_results_files&#39;</span>   <span class="p">:</span> <span class="s1">&#39;Clean results folder&#39;</span><span class="p">,</span>
        <span class="p">}</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Starting Sample Quality Control Pipeline&quot;</span><span class="p">)</span>
        
        <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="p">(</span><span class="n">func</span><span class="p">,</span> <span class="n">params</span><span class="p">)</span> <span class="ow">in</span> <span class="n">sample_qc_steps</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\033</span><span class="s2">[1m</span><span class="si">{</span><span class="n">step_description</span><span class="p">[</span><span class="n">name</span><span class="p">]</span><span class="si">}</span><span class="s2">.</span><span class="se">\033</span><span class="s2">[0m&quot;</span><span class="p">)</span>
            
            <span class="k">try</span><span class="p">:</span>
                <span class="n">func</span><span class="p">(</span><span class="o">**</span><span class="n">params</span><span class="p">)</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Successfully completed step: </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error in step &#39;</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&#39;: </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="k">raise</span>

            <span class="c1"># Memory management</span>
            <span class="n">time</span><span class="o">.</span><span class="n">sleep</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>  <span class="c1"># to avoid overwhelming the system with too many operations at once</span>
            <span class="n">gc</span><span class="o">.</span><span class="n">collect</span><span class="p">()</span>  <span class="c1"># clear memory after each step</span>

            <span class="n">mem</span> <span class="o">=</span> <span class="n">psutil</span><span class="o">.</span><span class="n">virtual_memory</span><span class="p">()</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Memory usage after </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">mem</span><span class="o">.</span><span class="n">percent</span><span class="si">}</span><span class="s2">%&quot;</span><span class="p">)</span>
            
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\033</span><span class="s2">[92mSample quality control pipeline completed successfully.</span><span class="se">\033</span><span class="s2">[0m&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Sample Quality Control Pipeline completed successfully&quot;</span><span class="p">)</span>

        <span class="k">return</span></div>
</div>

</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, Luis Giraldo Gonzalez Ricardo, Amabel Tenghe, and Ashwin Ashok Kumar Sreelatha.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>