<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>ideal_genom_qc.VariantQC &mdash; IDEAL-GENOM-QC v0.1.0</title>
      <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../../_static/css/theme.css?v=19f00094" />

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../_static/jquery.js?v=5d32c60e"></script>
        <script src="../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../../_static/documentation_options.js?v=01f34227"></script>
        <script src="../../_static/doctools.js?v=9a2dae69"></script>
        <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #2980B9" >

          
          
          <a href="../../index.html" class="icon icon-home">
            IDEAL-GENOM-QC
          </a>
              <div class="version">
                0.1.0
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">User Guide</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../getting_started.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../configuration.html">Configuration Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../examples.html">Examples</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../SampleQC.html">SampleQC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../AncestryQC.html">AncestryQC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../PopStructure.html">Population Structure</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../VariantQC.html">VariantQC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../Helpers.html">Helpers</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../get_references.html">Get References</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Additional Resources</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../faq.html">Frequently Asked Questions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../troubleshooting.html">Troubleshooting Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../contributing.html">Contributing Guide</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #2980B9" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">IDEAL-GENOM-QC</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../index.html">Module code</a></li>
      <li class="breadcrumb-item active">ideal_genom_qc.VariantQC</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for ideal_genom_qc.VariantQC</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Python module to perform variant quality control</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">psutil</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">logging</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">typing</span><span class="w"> </span><span class="kn">import</span> <span class="n">Optional</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">ideal_genom_qc.Helpers</span><span class="w"> </span><span class="kn">import</span> <span class="n">shell_do</span><span class="p">,</span> <span class="n">delete_temp_files</span>

<span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">%(asctime)s</span><span class="s2"> - </span><span class="si">%(levelname)s</span><span class="s2"> - </span><span class="si">%(message)s</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span>

<div class="viewcode-block" id="VariantQC">
<a class="viewcode-back" href="../../VariantQC.html#ideal_genom_qc.VariantQC.VariantQC">[docs]</a>
<span class="k">class</span><span class="w"> </span><span class="nc">VariantQC</span><span class="p">:</span>

<div class="viewcode-block" id="VariantQC.__init__">
<a class="viewcode-back" href="../../VariantQC.html#ideal_genom_qc.VariantQC.VariantQC.__init__">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">input_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">input_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">output_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">output_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initialize the VariantQC class.</span>
<span class="sd">        This class handles quality control for genetic variants data stored in PLINK binary format</span>
<span class="sd">        (.bed, .bim, .fam files).</span>
<span class="sd">        </span>
<span class="sd">        Parameters:</span>
<span class="sd">        -----------</span>
<span class="sd">        input_path: Path </span>
<span class="sd">            Directory path containing input PLINK files</span>
<span class="sd">        input_name: str </span>
<span class="sd">            Base name of input PLINK files (without extension)</span>
<span class="sd">        output_path: Path </span>
<span class="sd">            Directory path where output files will be saved</span>
<span class="sd">        output_name: str </span>
<span class="sd">            Base name for output files</span>
<span class="sd">        </span>
<span class="sd">        Raises:</span>
<span class="sd">        -------</span>
<span class="sd">        TypeError: </span>
<span class="sd">            If input_path/output_path are not Path objects or if input_name/output_name are not strings</span>
<span class="sd">        FileNotFoundError: </span>
<span class="sd">            If input_path/output_path don&#39;t exist or required PLINK files are not found</span>
<span class="sd">        </span>
<span class="sd">        Attributes:</span>
<span class="sd">        -----------</span>
<span class="sd">        input_path: Path </span>
<span class="sd">            Path to input directory</span>
<span class="sd">        output_path: Path </span>
<span class="sd">            Path to output directory  </span>
<span class="sd">        input_name: str </span>
<span class="sd">            Base name of input files</span>
<span class="sd">        output_name: str </span>
<span class="sd">            Base name for output files</span>
<span class="sd">        hwe_results: </span>
<span class="sd">            Storage for Hardy-Weinberg equilibrium test results</span>
<span class="sd">        results_dir: Path </span>
<span class="sd">            Directory for all QC results</span>
<span class="sd">        fails_dir: Path </span>
<span class="sd">            Directory for failed samples</span>
<span class="sd">        clean_dir: Path </span>
<span class="sd">            Directory for cleaned files</span>
<span class="sd">        plots_dir: Path </span>
<span class="sd">            Directory for QC plots</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">input_path</span><span class="p">,</span> <span class="n">Path</span><span class="p">)</span> <span class="ow">or</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="n">Path</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;input_path and output_path should be of type Path&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">input_name</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span> <span class="ow">or</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">output_name</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;input_name and output_name should be of type str&quot;</span><span class="p">)</span>
        
        <span class="k">if</span> <span class="ow">not</span> <span class="n">input_path</span><span class="o">.</span><span class="n">exists</span><span class="p">()</span> <span class="ow">or</span> <span class="ow">not</span> <span class="n">output_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="s2">&quot;input_path or output_path is not a valid path&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="n">input_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">input_name</span><span class="si">}</span><span class="s2">.bed&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="s2">&quot;.bed file not found&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="n">input_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">input_name</span><span class="si">}</span><span class="s2">.fam&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="s2">&quot;.fam file not found&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="n">input_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">input_name</span><span class="si">}</span><span class="s2">.bim&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="s2">&quot;.bim file not found&quot;</span><span class="p">)</span>
        
        <span class="bp">self</span><span class="o">.</span><span class="n">input_path</span> <span class="o">=</span> <span class="n">input_path</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">output_path</span><span class="o">=</span> <span class="n">output_path</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">input_name</span> <span class="o">=</span> <span class="n">input_name</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">=</span> <span class="n">output_name</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">hwe_results</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="c1"># create results folder</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_path</span> <span class="o">/</span> <span class="s1">&#39;variant_qc_results&#39;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="c1"># create fails folder</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">fails_dir</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="s1">&#39;fail_samples&#39;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">fails_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="c1"># create clean files folder</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">clean_dir</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="s1">&#39;clean_files&#39;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">clean_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        
        <span class="c1"># create figures folder</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">plots_dir</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="s1">&#39;plots&#39;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">plots_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></div>


<div class="viewcode-block" id="VariantQC.execute_missing_data_rate">
<a class="viewcode-back" href="../../VariantQC.html#ideal_genom_qc.VariantQC.VariantQC.execute_missing_data_rate">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">execute_missing_data_rate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chr_y</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">24</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Executes missing data rate analysis using PLINK for male and female subjects separately.</span>
<span class="sd">        This method performs two PLINK operations:</span>
<span class="sd">        1. Generates .lmiss and .imiss files for male subjects on chromosome Y</span>
<span class="sd">        2. Generates .lmiss and .imiss files for all subjects excluding chromosome Y</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        chr_y : int, default=24</span>
<span class="sd">            Chromosome Y number in the dataset. Must be between 0 and 26.</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None</span>
<span class="sd">        </span>
<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        TypeError</span>
<span class="sd">            If chr_y is not an integer</span>
<span class="sd">        ValueError</span>
<span class="sd">            If chr_y is not between 0 and 26</span>
<span class="sd">        </span>
<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        The method uses 2/3 of available system memory for PLINK operations.</span>
<span class="sd">        Output files are generated in the results directory with the following naming pattern:</span>
<span class="sd">        - {output_name}-missing-males-only.lmiss/.imiss : For male subjects</span>
<span class="sd">        - {output_name}-missing-not-y.lmiss/.imiss : For non-Y chromosome data</span>
<span class="sd">        The results are stored in self.males_missing_data and self.females_missing_data as Path objects.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># check type for chr_y</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">chr_y</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;chr_y should be of type integer.&quot;</span><span class="p">)</span>
        
        <span class="k">if</span> <span class="n">chr_y</span> <span class="o">&lt;</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">chr_y</span> <span class="o">&gt;</span> <span class="mi">26</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;chr_y should be between 1 and 26&quot;</span><span class="p">)</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Identifying markers with excessive missing rate...&quot;</span><span class="p">)</span>

        <span class="c1"># Get the virtual memory details</span>
        <span class="n">memory_info</span> <span class="o">=</span> <span class="n">psutil</span><span class="o">.</span><span class="n">virtual_memory</span><span class="p">()</span>
        <span class="n">available_memory_mb</span> <span class="o">=</span> <span class="n">memory_info</span><span class="o">.</span><span class="n">available</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1024</span> <span class="o">*</span> <span class="mi">1024</span><span class="p">)</span>
        <span class="n">memory</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="mi">2</span><span class="o">*</span><span class="n">available_memory_mb</span><span class="o">/</span><span class="mi">3</span><span class="p">,</span><span class="mi">0</span><span class="p">)</span>

        <span class="c1"># generates  .lmiss and .imiss files for male subjects</span>
        <span class="n">plink_cmd1</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink --bfile </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">input_path</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="si">}</span><span class="s2"> --missing --filter-males --chr </span><span class="si">{</span><span class="n">chr_y</span><span class="si">}</span><span class="s2"> --out </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-missing-males-only&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2"> --memory </span><span class="si">{</span><span class="n">memory</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="c1"># generates .lmiss and. imiss files for female subjects</span>
        <span class="n">plink_cmd2</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink --bfile </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">input_path</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="si">}</span><span class="s2"> --missing --not-chr </span><span class="si">{</span><span class="n">chr_y</span><span class="si">}</span><span class="s2"> --out </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-missing-not-y&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2"> --memory </span><span class="si">{</span><span class="n">memory</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">males_missing_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-missing-males-only.lmiss&#39;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">females_missing_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-missing-not-y.lmiss&#39;</span><span class="p">)</span>

        <span class="c1"># execute PLINK commands</span>
        <span class="n">cmds</span> <span class="o">=</span> <span class="p">[</span><span class="n">plink_cmd1</span><span class="p">,</span> <span class="n">plink_cmd2</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">cmd</span> <span class="ow">in</span> <span class="n">cmds</span><span class="p">:</span>
            <span class="n">shell_do</span><span class="p">(</span><span class="n">cmd</span><span class="p">,</span> <span class="n">log</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="k">return</span></div>


<div class="viewcode-block" id="VariantQC.execute_different_genotype_call_rate">
<a class="viewcode-back" href="../../VariantQC.html#ideal_genom_qc.VariantQC.VariantQC.execute_different_genotype_call_rate">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">execute_different_genotype_call_rate</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Execute test for different genotype call rates between cases and controls using PLINK.</span>

<span class="sd">        This method performs the following operations:</span>
<span class="sd">        1. Calculates available memory for PLINK execution</span>
<span class="sd">        2. Runs PLINK&#39;s --test-missing command to identify markers with significantly different</span>
<span class="sd">        missing rates between cases and controls</span>
<span class="sd">        3. Generates a .missing file with the results</span>

<span class="sd">        The method uses approximately 2/3 of available system memory for PLINK execution.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">            None</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        - Creates a .missing file in the results directory</span>
<span class="sd">        - Sets self.case_control_missing path attribute</span>
<span class="sd">        &quot;&quot;&quot;</span>


        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Identifying markers with different genotype call rates between cases and controls...&quot;</span><span class="p">)</span>

        <span class="c1"># Get the virtual memory details</span>
        <span class="n">memory_info</span> <span class="o">=</span> <span class="n">psutil</span><span class="o">.</span><span class="n">virtual_memory</span><span class="p">()</span>
        <span class="n">available_memory_mb</span> <span class="o">=</span> <span class="n">memory_info</span><span class="o">.</span><span class="n">available</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1024</span> <span class="o">*</span> <span class="mi">1024</span><span class="p">)</span>
        <span class="n">memory</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="mi">2</span><span class="o">*</span><span class="n">available_memory_mb</span><span class="o">/</span><span class="mi">3</span><span class="p">,</span><span class="mi">0</span><span class="p">)</span>

        <span class="c1"># generates .missing file</span>
        <span class="n">plink_cmd</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink --bfile </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">input_path</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="si">}</span><span class="s2"> --test-missing --out </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-case-control-missing&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2"> --memory </span><span class="si">{</span><span class="n">memory</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="c1"># execute PLINK command</span>
        <span class="n">shell_do</span><span class="p">(</span><span class="n">plink_cmd</span><span class="p">,</span> <span class="n">log</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">case_control_missing</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-case-control-missing.missing&#39;</span><span class="p">)</span>

        <span class="k">return</span></div>

    
<div class="viewcode-block" id="VariantQC.execute_hwe_test">
<a class="viewcode-back" href="../../VariantQC.html#ideal_genom_qc.VariantQC.VariantQC.execute_hwe_test">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">execute_hwe_test</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Execute Hardy-Weinberg Equilibrium (HWE) test using PLINK.</span>

<span class="sd">        This method performs the following steps:</span>
<span class="sd">        1. Calculates available memory and allocates 2/3 for the test</span>
<span class="sd">        2. Runs PLINK command to compute HWE test on the input binary PLINK files</span>
<span class="sd">        3. Saves results to a .hwe output file</span>

<span class="sd">        The HWE test is used to assess whether genotype frequencies in a population remain constant </span>
<span class="sd">        across generations under specific conditions.</span>

<span class="sd">        Returns:</span>
<span class="sd">        --------</span>
<span class="sd">            None</span>

<span class="sd">        Side effects:</span>
<span class="sd">        -------------</span>
<span class="sd">            - Creates a .hwe output file in the results directory</span>
<span class="sd">            - Sets self.hwe_results to the name of the output file</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Computing Hardy-Weinberg Equilibrium test...&#39;</span><span class="p">)</span>

        <span class="c1"># Get the virtual memory details</span>
        <span class="n">memory_info</span> <span class="o">=</span> <span class="n">psutil</span><span class="o">.</span><span class="n">virtual_memory</span><span class="p">()</span>
        <span class="n">available_memory_mb</span> <span class="o">=</span> <span class="n">memory_info</span><span class="o">.</span><span class="n">available</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1024</span> <span class="o">*</span> <span class="mi">1024</span><span class="p">)</span>
        <span class="n">memory</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="mi">2</span><span class="o">*</span><span class="n">available_memory_mb</span><span class="o">/</span><span class="mi">3</span><span class="p">,</span><span class="mi">0</span><span class="p">)</span>

        <span class="c1"># PLINK command to compute HWE test</span>
        <span class="n">plink_cmd</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink --bfile </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">input_path</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="si">}</span><span class="s2"> --hardy --out </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-hwe&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2"> --memory </span><span class="si">{</span><span class="n">memory</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="c1"># execute PLINK command</span>
        <span class="n">shell_do</span><span class="p">(</span><span class="n">plink_cmd</span><span class="p">,</span> <span class="n">log</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">hwe_results</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-hwe.hwe&#39;</span>

        <span class="k">return</span></div>

    
<div class="viewcode-block" id="VariantQC.get_fail_variants">
<a class="viewcode-back" href="../../VariantQC.html#ideal_genom_qc.VariantQC.VariantQC.get_fail_variants">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">get_fail_variants</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">marker_call_rate_thres</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.2</span><span class="p">,</span> <span class="n">case_controls_thres</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">1e-5</span><span class="p">,</span> <span class="n">hwe_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">5e-8</span><span class="p">,</span> <span class="n">male_female_y_cap</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">hwe_y_cap</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Identify and consolidate failing variants based on multiple quality control criteria.</span>
<span class="sd">        This method combines the results of three QC checks:</span>
<span class="sd">        1. Variants with high missing data rates</span>
<span class="sd">        2. Variants with significantly different genotype call rates between cases and controls</span>
<span class="sd">        3. Variants failing Hardy-Weinberg equilibrium test</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        marker_call_rate_thres : float, optional</span>
<span class="sd">            Threshold for failing variants based on missing data rate (default: 0.2)</span>
<span class="sd">        case_controls_thres : float, optional</span>
<span class="sd">            P-value threshold for differential missingness between cases and controls (default: 1e-5)</span>
<span class="sd">        hwe_threshold : float, optional</span>
<span class="sd">            P-value threshold for Hardy-Weinberg equilibrium test (default: 5e-8)</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        pd.DataFrame</span>
<span class="sd">            A summary DataFrame containing:</span>
<span class="sd">            - Counts of variants failing each QC criterion</span>
<span class="sd">            - Number of variants failing multiple criteria (duplicates)</span>
<span class="sd">            - Total number of unique failing variants</span>
<span class="sd">        </span>
<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        - Results are also written to a tab-separated file &#39;fail_markers.txt&#39;</span>
<span class="sd">        - Variants failing multiple criteria are only counted once in the final output file</span>
<span class="sd">        &quot;&quot;&quot;</span>
        

        <span class="c1"># ==========================================================================================================</span>
        <span class="c1">#                                             MARKERS WITH MISSING DATA </span>
        <span class="c1"># ==========================================================================================================</span>

        <span class="n">fail_missing_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">report_missing_data</span><span class="p">(</span>
            <span class="n">filename_male</span>  <span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">males_missing_data</span><span class="p">,</span> 
            <span class="n">filename_female</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">females_missing_data</span><span class="p">,</span>
            <span class="n">threshold</span>      <span class="o">=</span><span class="n">marker_call_rate_thres</span><span class="p">,</span>
            <span class="n">y_axis_cap</span>     <span class="o">=</span><span class="n">male_female_y_cap</span>
        <span class="p">)</span>

        <span class="c1"># ==========================================================================================================</span>
        <span class="c1">#                                             MARKERS WITH DIFFERENT GENOTYPE CALL RATE</span>
        <span class="c1"># ==========================================================================================================</span>

        <span class="n">fail_genotype</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">report_different_genotype_call_rate</span><span class="p">(</span>
            <span class="n">filename</span> <span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">case_control_missing</span><span class="p">,</span> 
            <span class="n">threshold</span><span class="o">=</span><span class="n">case_controls_thres</span><span class="p">,</span> 
        <span class="p">)</span>

        <span class="c1"># ==========================================================================================================</span>
        <span class="c1">#                                             MARKERS FAILING HWE TEST</span>
        <span class="c1"># ==========================================================================================================</span>

        <span class="c1"># Make sure self.hwe_results is not None before passing it to report_hwe</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">hwe_results</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;HWE results not available. Run execute_hwe_test first.&quot;</span><span class="p">)</span>
            
        <span class="n">fail_hwe</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">report_hwe</span><span class="p">(</span>
            <span class="n">directory</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">results_dir</span><span class="p">,</span>
            <span class="n">filename</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">hwe_results</span><span class="p">,</span>
            <span class="n">hwe_threshold</span><span class="o">=</span><span class="n">hwe_threshold</span><span class="p">,</span>
            <span class="n">y_lim_cap</span><span class="o">=</span><span class="n">hwe_y_cap</span>
        <span class="p">)</span>

        <span class="n">fails</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">fail_missing_data</span><span class="p">,</span> <span class="n">fail_genotype</span><span class="p">,</span> <span class="n">fail_hwe</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="n">summary</span> <span class="o">=</span> <span class="n">fails</span><span class="p">[</span><span class="s1">&#39;Failure&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
        <span class="n">num_dup</span> <span class="o">=</span> <span class="n">fails</span><span class="o">.</span><span class="n">duplicated</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;SNP&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>

        <span class="n">totals</span> <span class="o">=</span> <span class="n">summary</span><span class="o">.</span><span class="n">select_dtypes</span><span class="p">(</span><span class="n">include</span><span class="o">=</span><span class="s2">&quot;number&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="o">-</span> <span class="n">num_dup</span>
        <span class="n">dups_row</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;Failure&#39;</span><span class="p">:[</span><span class="s1">&#39;Duplicated SNPs&#39;</span><span class="p">],</span> <span class="s1">&#39;count&#39;</span><span class="p">:[</span><span class="o">-</span><span class="n">num_dup</span><span class="p">]})</span>
        <span class="n">total_row</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="n">col</span><span class="p">:</span> <span class="p">[</span><span class="n">totals</span><span class="p">[</span><span class="n">col</span><span class="p">]</span> <span class="k">if</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">totals</span><span class="o">.</span><span class="n">index</span> <span class="k">else</span> <span class="s2">&quot;Total&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">summary</span><span class="o">.</span><span class="n">columns</span><span class="p">})</span>

        <span class="n">fails</span> <span class="o">=</span> <span class="n">fails</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="s1">&#39;SNP&#39;</span><span class="p">,</span> <span class="n">keep</span><span class="o">=</span><span class="s1">&#39;first&#39;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="n">fails</span> <span class="o">=</span> <span class="n">fails</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Failure&#39;</span><span class="p">],</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="n">fails</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fails_dir</span> <span class="o">/</span> <span class="s1">&#39;fail_markers.txt&#39;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">summary</span><span class="p">,</span> <span class="n">dups_row</span><span class="p">,</span> <span class="n">total_row</span><span class="p">],</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></div>


<div class="viewcode-block" id="VariantQC.execute_drop_variants">
<a class="viewcode-back" href="../../VariantQC.html#ideal_genom_qc.VariantQC.VariantQC.execute_drop_variants">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">execute_drop_variants</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">maf</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">5e-8</span><span class="p">,</span> <span class="n">geno</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.1</span><span class="p">,</span> <span class="n">hwe</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">5e-8</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Execute variant filtering based on quality control parameters using PLINK.</span>

<span class="sd">        This method removes variants that fail quality control criteria including minor allele frequency (MAF),</span>
<span class="sd">        genotype missingness rate, and Hardy-Weinberg equilibrium (HWE) test.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        maf : float, optional</span>
<span class="sd">            Minor allele frequency threshold. Variants with MAF below this value are removed.</span>
<span class="sd">            Default is 5e-8.</span>
<span class="sd">        geno : float, optional</span>
<span class="sd">            Maximum per-variant missing genotype rate. Variants with missing rate above this </span>
<span class="sd">            value are removed. Default is 0.1 (10%).</span>
<span class="sd">        hwe : float, optional </span>
<span class="sd">            Hardy-Weinberg equilibrium test p-value threshold. Variants with HWE p-value below</span>
<span class="sd">            this are removed. Default is 5e-8.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None</span>
<span class="sd">            Creates quality controlled PLINK binary files (.bed, .bim, .fam) in the clean directory</span>
<span class="sd">            with suffix &#39;-variantQCed&#39;.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Removing markers failing quality control...&quot;</span><span class="p">)</span>

        <span class="c1"># create cleaned binary files</span>
        <span class="n">plink_cmd</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;plink --bfile </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">input_path</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="bp">self</span><span class="o">.</span><span class="n">input_name</span><span class="si">}</span><span class="s2"> --exclude </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">fails_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="s1">&#39;fail_markers.txt&#39;</span><span class="si">}</span><span class="s2"> --autosome --maf </span><span class="si">{</span><span class="n">maf</span><span class="si">}</span><span class="s2"> --hwe </span><span class="si">{</span><span class="n">hwe</span><span class="si">}</span><span class="s2"> --geno </span><span class="si">{</span><span class="n">geno</span><span class="si">}</span><span class="s2"> --make-bed --out </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">clean_dir</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_name</span><span class="o">+</span><span class="s1">&#39;-variantQCed&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="c1"># execute PLINK command</span>
        <span class="n">shell_do</span><span class="p">(</span><span class="n">plink_cmd</span><span class="p">,</span> <span class="n">log</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="k">return</span></div>


<div class="viewcode-block" id="VariantQC.report_missing_data">
<a class="viewcode-back" href="../../VariantQC.html#ideal_genom_qc.VariantQC.VariantQC.report_missing_data">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">report_missing_data</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename_male</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">filename_female</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">threshold</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span> <span class="n">y_axis_cap</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Analyze and report missing data rates for male and female subjects.</span>
<span class="sd">        This method processes missing data information from separate files for male and female subjects,</span>
<span class="sd">        creates visualizations of missing data distributions, and identifies SNPs that fail the missing</span>
<span class="sd">        data threshold for each sex group.</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        directory : str</span>
<span class="sd">            Path to the directory containing the input files</span>
<span class="sd">        filename_male : str</span>
<span class="sd">            Name of the file containing missing data information for male subjects (.lmiss format)</span>
<span class="sd">        filename_female : str</span>
<span class="sd">            Name of the file containing missing data information for female subjects (.lmiss format)</span>
<span class="sd">        threshold : float</span>
<span class="sd">            Maximum allowed missing data rate (between 0 and 1)</span>
<span class="sd">        y_axis_cap : int, optional</span>
<span class="sd">            Upper limit for y-axis in histogram plots (default is 10)</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        pd.DataFrame</span>
<span class="sd">            A DataFrame containing SNPs that fail the missing data threshold for either sex,</span>
<span class="sd">            with columns [&#39;SNP&#39;, &#39;Failure&#39;] where &#39;Failure&#39; indicates the failing category</span>
<span class="sd">        </span>
<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        The method generates two histogram plots saved as &#39;missing_data_male&#39; and </span>
<span class="sd">        &#39;missing_data_female&#39; showing the distribution of missing data rates for each sex.</span>
<span class="sd">        &quot;&quot;&quot;</span>
   

        <span class="c1"># load .lmiss file for male subjects</span>
        <span class="n">df_males</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
            <span class="n">filename_male</span><span class="p">,</span>
            <span class="n">sep</span><span class="o">=</span><span class="sa">r</span><span class="s2">&quot;\s+&quot;</span><span class="p">,</span>
            <span class="n">engine</span><span class="o">=</span><span class="s1">&#39;python&#39;</span>
        <span class="p">)</span>
        
        <span class="c1">## filter male subjects</span>
        <span class="n">fail_males</span> <span class="o">=</span> <span class="n">df_males</span><span class="p">[</span><span class="n">df_males</span><span class="p">[</span><span class="s1">&#39;F_MISS&#39;</span><span class="p">]</span><span class="o">&gt;=</span><span class="n">threshold</span><span class="p">]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">fail_males</span> <span class="o">=</span> <span class="n">fail_males</span><span class="p">[[</span><span class="s1">&#39;SNP&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">fail_males</span><span class="p">[</span><span class="s1">&#39;Failure&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Missing data rate on males&#39;</span>

        <span class="c1"># load .lmiss file for female subjects</span>
        <span class="n">df_females</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
            <span class="n">filename_female</span><span class="p">,</span>
            <span class="n">sep</span><span class="o">=</span><span class="sa">r</span><span class="s2">&quot;\s+&quot;</span><span class="p">,</span>
            <span class="n">engine</span><span class="o">=</span><span class="s1">&#39;python&#39;</span>
        <span class="p">)</span>
        
        <span class="c1">## filter female subjects</span>
        <span class="n">fail_females</span> <span class="o">=</span> <span class="n">df_females</span><span class="p">[</span><span class="n">df_females</span><span class="p">[</span><span class="s1">&#39;F_MISS&#39;</span><span class="p">]</span><span class="o">&gt;=</span><span class="n">threshold</span><span class="p">]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">fail_females</span> <span class="o">=</span> <span class="n">fail_females</span><span class="p">[[</span><span class="s1">&#39;SNP&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">fail_females</span><span class="p">[</span><span class="s1">&#39;Failure&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Missing data rate on females&#39;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_make_histogram</span><span class="p">(</span><span class="n">df_males</span><span class="p">[</span><span class="s1">&#39;F_MISS&#39;</span><span class="p">],</span> <span class="s1">&#39;missing_data_male&#39;</span><span class="p">,</span> <span class="n">threshold</span><span class="p">,</span> <span class="s1">&#39;Ratio of missing data&#39;</span><span class="p">,</span> <span class="s1">&#39;Missing data for males&#39;</span><span class="p">,</span> <span class="n">y_lim_cap</span><span class="o">=</span><span class="n">y_axis_cap</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_make_histogram</span><span class="p">(</span><span class="n">df_females</span><span class="p">[</span><span class="s1">&#39;F_MISS&#39;</span><span class="p">],</span> <span class="s1">&#39;missing_data_female&#39;</span><span class="p">,</span> <span class="n">threshold</span><span class="p">,</span> <span class="s1">&#39;Ratio of missing data&#39;</span><span class="p">,</span> <span class="s1">&#39;Missing data for females&#39;</span><span class="p">,</span> <span class="n">y_lim_cap</span><span class="o">=</span><span class="n">y_axis_cap</span><span class="p">)</span>

        <span class="c1"># concatenate female and male subjects who failed QC</span>
        <span class="n">fails</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">fail_females</span><span class="p">,</span> <span class="n">fail_males</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">fails</span></div>


<div class="viewcode-block" id="VariantQC.report_different_genotype_call_rate">
<a class="viewcode-back" href="../../VariantQC.html#ideal_genom_qc.VariantQC.VariantQC.report_different_genotype_call_rate">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">report_different_genotype_call_rate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">threshold</span><span class="p">:</span> <span class="nb">float</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Reports markers with different genotype call rates based on a given threshold.</span>
<span class="sd">        This function reads a .missing file, filters markers with a different genotype call rate</span>
<span class="sd">        below the specified threshold, and returns a DataFrame containing these markers.</span>

<span class="sd">        Parameters:</span>
<span class="sd">        -----------</span>
<span class="sd">            directory (str): The directory where the .missing file is located.</span>
<span class="sd">            filename (str): The name of the .missing file.</span>
<span class="sd">            threshold (float): The threshold for filtering markers based on the P-value.</span>
<span class="sd">        </span>
<span class="sd">        Returns:</span>
<span class="sd">        --------</span>
<span class="sd">            pd.DataFrame: A DataFrame containing markers with different genotype call rates</span>
<span class="sd">                          below the specified threshold. The DataFrame has two columns:</span>
<span class="sd">                          &#39;SNP&#39; and &#39;Failure&#39;, where &#39;Failure&#39; is set to &#39;Different genotype call rate&#39;.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># load .missing file</span>
        <span class="n">df_diffmiss</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
            <span class="n">filename</span><span class="p">,</span>
            <span class="n">sep</span><span class="o">=</span><span class="sa">r</span><span class="s2">&quot;\s+&quot;</span><span class="p">,</span>
            <span class="n">engine</span><span class="o">=</span><span class="s1">&#39;python&#39;</span>
        <span class="p">)</span>

        <span class="c1"># filter markers with different genotype call rate</span>
        <span class="n">fail_diffmiss</span> <span class="o">=</span> <span class="n">df_diffmiss</span><span class="p">[</span><span class="n">df_diffmiss</span><span class="p">[</span><span class="s1">&#39;P&#39;</span><span class="p">]</span><span class="o">&lt;</span><span class="n">threshold</span><span class="p">]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">fail_diffmiss</span> <span class="o">=</span> <span class="n">fail_diffmiss</span><span class="p">[[</span><span class="s1">&#39;SNP&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">fail_diffmiss</span><span class="p">[</span><span class="s1">&#39;Failure&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Different genotype call rate&#39;</span>

        <span class="k">return</span> <span class="n">fail_diffmiss</span></div>

    
<div class="viewcode-block" id="VariantQC.report_hwe">
<a class="viewcode-back" href="../../VariantQC.html#ideal_genom_qc.VariantQC.VariantQC.report_hwe">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">report_hwe</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">directory</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">filename</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">hwe_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">5e-8</span><span class="p">,</span> <span class="n">y_lim_cap</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Generate Hardy-Weinberg Equilibrium (HWE) test report and visualization.</span>

<span class="sd">        This method reads HWE test results from a file, identifies variants that fail HWE,</span>
<span class="sd">        creates a histogram of -log10(P) values, and returns failed variants.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        directory : Path</span>
<span class="sd">            Directory path where the HWE test results file is located</span>
<span class="sd">        filename : str</span>
<span class="sd">            Name of the file containing HWE test results</span>
<span class="sd">        hwe_threshold : float, optional</span>
<span class="sd">            P-value threshold for HWE test failure (default: 5e-8)</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        pd.DataFrame</span>
<span class="sd">            DataFrame containing variants that failed HWE test with columns:</span>
<span class="sd">            - SNP: variant identifier</span>
<span class="sd">            - Failure: reason for failure (always &#39;HWE&#39;)</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        The method creates a histogram plot saved as &#39;hwe-histogram&#39; showing the</span>
<span class="sd">        distribution of -log10(P) values from HWE tests.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">df_hwe</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
            <span class="n">directory</span> <span class="o">/</span> <span class="n">filename</span><span class="p">,</span>
            <span class="n">sep</span><span class="o">=</span><span class="sa">r</span><span class="s2">&quot;\s+&quot;</span><span class="p">,</span>
            <span class="n">engine</span><span class="o">=</span><span class="s1">&#39;python&#39;</span>
        <span class="p">)</span>

        <span class="n">fail_hwe</span> <span class="o">=</span> <span class="n">df_hwe</span><span class="p">[</span><span class="n">df_hwe</span><span class="p">[</span><span class="s1">&#39;P&#39;</span><span class="p">]</span><span class="o">&lt;</span><span class="n">hwe_threshold</span><span class="p">]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">fail_hwe</span> <span class="o">=</span> <span class="n">fail_hwe</span><span class="p">[[</span><span class="s1">&#39;SNP&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">fail_hwe</span><span class="p">[</span><span class="s1">&#39;Failure&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;HWE&#39;</span>

        <span class="n">df_all</span> <span class="o">=</span> <span class="n">df_hwe</span><span class="p">[</span><span class="n">df_hwe</span><span class="p">[</span><span class="s1">&#39;TEST&#39;</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;ALL&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">df_all</span><span class="p">[</span><span class="s1">&#39;P&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_all</span><span class="p">[</span><span class="s1">&#39;P&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">finfo</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span><span class="o">.</span><span class="n">tiny</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_make_histogram</span><span class="p">(</span>
            <span class="n">values</span><span class="o">=-</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">df_all</span><span class="p">[</span><span class="s1">&#39;P&#39;</span><span class="p">]),</span> <span class="c1"># type: ignore</span>
            <span class="n">output_name</span><span class="o">=</span><span class="s1">&#39;hwe-histogram&#39;</span><span class="p">,</span> 
            <span class="n">threshold</span><span class="o">=-</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">hwe_threshold</span><span class="p">),</span> 
            <span class="n">x_label</span><span class="o">=</span><span class="s1">&#39;-log10(P) of HWE test&#39;</span><span class="p">,</span> 
            <span class="n">title</span><span class="o">=</span><span class="s1">&#39;HWE test&#39;</span><span class="p">,</span>
            <span class="n">y_lim_cap</span><span class="o">=</span><span class="n">y_lim_cap</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">fail_hwe</span></div>

    
    <span class="k">def</span><span class="w"> </span><span class="nf">_make_histogram</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">values</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">,</span> <span class="n">output_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">threshold</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span> <span class="n">x_label</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">title</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">y_lim_cap</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Creates a histogram plot with a vertical threshold line and saves it to a PDF file.</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        values : pd.Series</span>
<span class="sd">            The data values to plot in the histogram.</span>
<span class="sd">        values: pd.Series</span>
<span class="sd">            Series containing the numeric values to plot.</span>
<span class="sd">        output_name : str</span>
<span class="sd">            Name of the output file (without extension).</span>
<span class="sd">        threshold : float</span>
<span class="sd">            Value where to draw the vertical threshold line.</span>
<span class="sd">        x_label : str</span>
<span class="sd">            Label for the x-axis.</span>
<span class="sd">        title : str</span>
<span class="sd">            Title of the plot.</span>
<span class="sd">        y_lim_cap : float, optional</span>
<span class="sd">            Upper limit for y-axis. If None, automatically determined. Defaults to None.</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None</span>
<span class="sd">            This function saves the plot to a file and displays it but does not return any value.</span>
<span class="sd">        </span>
<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        The plot is saved as a PDF file in the plots_dir directory defined in the class instance.</span>
<span class="sd">        The histogram uses 50 bins and a predefined color (#1B9E77).</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">plt</span><span class="o">.</span><span class="n">clf</span><span class="p">()</span>

        <span class="n">fig_path</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">plots_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">output_name</span><span class="si">}</span><span class="s2">.pdf&quot;</span>

        <span class="n">plt</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">values</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;#1B9E77&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="n">x_label</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Number of SNPs&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">y_lim_cap</span> <span class="k">if</span> <span class="n">y_lim_cap</span> <span class="k">else</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="n">title</span><span class="p">)</span>

        <span class="c1"># Draw the vertical line indicating the cut off threshold</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">threshold</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">)</span>

        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">fig_path</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">400</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">block</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

        <span class="k">return</span> <span class="kc">None</span></div>

</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, Luis Giraldo Gonzalez Ricardo, Amabel Tenghe, and Ashwin Ashok Kumar Sreelatha.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>