{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "# add parent directory to path\n",
    "library_path = os.path.abspath('..')\n",
    "if library_path not in sys.path:\n",
    "    sys.path.append(library_path)\n",
    "\n",
    "from cge_comrare_pipeline.AncestryQC import AncestryQC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import ipywidgets as widgets\n",
    "from IPython.display import display\n",
    "\n",
    "# Create interactive widgets for input\n",
    "input_path = widgets.Text(\n",
    "    value='/home/luis/data/rawdata-sexupdated/outputData/sample_qc_results/clean_files',\n",
    "    description='Path to input plink1.9 files:',\n",
    "    style={'description_width': 'initial'}\n",
    ")\n",
    "\n",
    "input_name = widgets.Text(\n",
    "    value='luxgiant_res-clean-samples',\n",
    "    description='Name of the plink1.9 files:',\n",
    "    style={'description_width': 'initial'}\n",
    ")\n",
    "\n",
    "dependables_path = widgets.Text(\n",
    "    value='/home/luis/data/rawdata-sexupdated/dependables/',\n",
    "    description='Path to dependable files:',\n",
    "    style={'description_width': 'initial'}\n",
    ")\n",
    "\n",
    "output_path = widgets.Text(\n",
    "    value='/home/luis/data/rawdata-sexupdated/outputData/',\n",
    "    description='Path to output files:',\n",
    "    style={'description_width': 'initial'}\n",
    ")\n",
    "output_name = widgets.Text(\n",
    "    value='luxgiant_res',\n",
    "    description='Name of the resulting files:',\n",
    "    style={'description_width': 'initial'}\n",
    ")\n",
    "# Display the widgets\n",
    "display(input_path, input_name, dependables_path, output_path, output_name)\n",
    "\n",
    "# Function to get the text parameter values\n",
    "def get_params():\n",
    "    return input_path.value, input_name.value, dependables_path.value, output_path.value, output_name.value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create interactive widgets for list input\n",
    "ind_par = widgets.Textarea(\n",
    "    value='50, 5, 0.2',\n",
    "    description='indep pairwise (comma-separated):',\n",
    "    style={'description_width': 'initial'}\n",
    ")\n",
    "\n",
    "maf = widgets.FloatText(\n",
    "    value=0.01,  # Default value\n",
    "    description='maf (float):',\n",
    "    style={'description_width': 'initial'}\n",
    ")\n",
    "\n",
    "pca = widgets.IntText(\n",
    "    value=10,  # Default value\n",
    "    description='mind (float):',\n",
    "    style={'description_width': 'initial'}\n",
    ")\n",
    "\n",
    "ref_threshold = widgets.FloatText(\n",
    "    value=4,  # Default value\n",
    "    description='threshold for the deviations from the mean on reference panel:',\n",
    "    style={'description_width': 'initial'}\n",
    ")\n",
    "\n",
    "stu_threshold = widgets.FloatText(\n",
    "    value=4,  # Default value\n",
    "    description='threshold for the deviations from the mean on study population:',\n",
    "    style={'description_width': 'initial'}\n",
    ")\n",
    "\n",
    "num_pcs = widgets.IntText(\n",
    "    value=9,  # Default value\n",
    "    description='Number of PCs to determine a sample outlier status:',\n",
    "    style={'description_width': 'initial'}\n",
    ")\n",
    "\n",
    "# display the widgets\n",
    "display(ind_par, maf, pca, stu_threshold, ref_threshold, num_pcs)\n",
    "\n",
    "def get_ancestry_qc_params():\n",
    "\n",
    "    ancestry_qc_params = dict()\n",
    "\n",
    "    indep = ind_par.value.split(',')\n",
    "\n",
    "    ancestry_qc_params['maf'] = maf.value\n",
    "    ancestry_qc_params['indep'] = [int(indep[0]), int(indep[1]), float(indep[2])]\n",
    "    ancestry_qc_params['pca'] = pca.value\n",
    "    ancestry_qc_params['ref_threshold'] = ref_threshold.value\n",
    "    ancestry_qc_params['stu_threshold'] = stu_threshold.value\n",
    "    ancestry_qc_params['num_pcs'] = num_pcs.value\n",
    "\n",
    "    return ancestry_qc_params"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ancestry_params = get_ancestry_qc_params()\n",
    "ancestry_params"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ancestry = AncestryQC(\n",
    "    input_path      =input_path.value,\n",
    "    input_name      =input_name.value,\n",
    "    output_path     =output_path.value,\n",
    "    output_name     =output_name.value,\n",
    "    dependables_path=dependables_path.value,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ancestry_qc_steps = {\n",
    "    #'filter_snps'              : (ancestry.execute_filter_prob_snps, ()),\n",
    "    #'LD_pruning'               : (ancestry.execute_ld_pruning, (ancestry_params['indep'],)),\n",
    "    #'reference_pruning'        : (ancestry.execute_ld_prune_ref_panel, ()),\n",
    "    #'chr_missmatch'            : (ancestry.execute_fix_chromosome_missmatch,()),\n",
    "    #'pos_missmatch_allele_flip': (ancestry.execute_fix_position_missmatch_allele_flip, ()),\n",
    "    #'remove_missmatch'         : (ancestry.execute_remove_missmatch, ()),\n",
    "    #'merging'                  : (ancestry.execute_merge_ref_study, ()),\n",
    "    #'pca_analysis'             : (ancestry.execute_pc_decomposition, (ancestry_params['pca'], ancestry_params['maf'],)),\n",
    "    'get_outliers'             : (ancestry.get_ancestry_outliers, (ancestry_params['ref_threshold'], ancestry_params['stu_threshold'], 'SAS', ancestry_params['num_pcs'],)),\n",
    "    'pca_plot'                 : (ancestry.pca_plot, ())\n",
    "}\n",
    "\n",
    "step_description = {\n",
    "    'filter_snps'              : 'filter problematic snps',\n",
    "    'LD_pruning'               : 'LD prune study_population',\n",
    "    'reference_pruning'        : 'LD prune reference panel',\n",
    "    'chr_missmatch'            : 'fix any chromosome missmatch',\n",
    "    'pos_missmatch_allele_flip': 'fix any issue with position missmatch and allele flip',\n",
    "    'remove_missmatch'         : 'remove missmatchesd SNPs',\n",
    "    'merging'                  : 'merge reference panel and study population',\n",
    "    'pca_analysis'             : 'execute PC decomposition',\n",
    "    'get_outliers'             : 'find samples with discordant ancestry',\n",
    "    'pca_plot'                 : 'generate PCA plot'\n",
    "}\n",
    "\n",
    "for name, (func, params) in ancestry_qc_steps.items():\n",
    "    print(f\"\\033[1m{step_description[name]}.\\033[0m\")\n",
    "    func(*params)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fail_path = os.path.join(ancestry.fails_dir, output_name.value+'.fail-ancestry-qc.txt')\n",
    "\n",
    "with open(fail_path, \"r\") as file:\n",
    "    num_fails= sum(1 for line in file)\n",
    "\n",
    "print('Number of samples that failed the ancestry QC:', num_fails) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ancestry.execute_drop_ancestry_outliers()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cge-comrare-pipeline-AOnVxG3n-py3.12",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
