{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Example notebook of a Ancestry QC analysis\n",
    "\n",
    "The present notebook serves as a guide of how use the `IDEAL-GENOM-QC` library to perform a ancestry quality control. We intend to show a possible use, because each user can adapt it to its particular needs. Up to this moment the library can only detects outliers from an homogenous population that overlaps with one of the `SuperPop` present in the **1000 Genomes** data.\n",
    "\n",
    "In this notebook the procedure to perform the ancestty quality control is more detailed so the user can get a deeper understanding of all the steps executed in this part of the pipeline.\n",
    "\n",
    "Let us import the required libraries."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "\n",
    "from pathlib import Path\n",
    "\n",
    "# add parent directory to path\n",
    "library_path = os.path.abspath('..')\n",
    "if library_path not in sys.path:\n",
    "    sys.path.append(library_path)\n",
    "\n",
    "library_path = Path(library_path)\n",
    "\n",
    "from ideal_genom_qc.AncestryQC import AncestryQC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the next cell the path variables associated with the project are set.\n",
    "\n",
    "Moreover, since each user can have a slightly different choices for the LD regions, the user can provide its own file. Nevertheless, we provide the functionality of automatically fetching high LD regions for builts **GRCh37** and **GRCH38**. \n",
    "\n",
    "Also, the user can provide the path to the reference genome files or let the library fetch the data automatically.\n",
    "\n",
    "When giving the path to the input files, the user should take into account that the input files of the ancestry check must the output of the sample QC."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_path = Path('/path/to/input/data/')\n",
    "input_name = 'input-prefix'\n",
    "output_path= Path('/path/to/output/data/')\n",
    "output_name= 'output-prefix'\n",
    "high_ld_file = Path('path/to/ld-regions/file') # if not available, set to Path()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the next cell we define a dictionary with the parameters to execute the ancestry QC.\n",
    "\n",
    "The explanation of the parameters is the following:\n",
    "\n",
    "1. `ind_pair`: parameter of **PLINK1.9** `--indep-pairwise`.\n",
    "2. `pca`: number of principal components to be computed, parameter `--pca` from **PLINK1.9**.\n",
    "3. `maf`: minor allele frequency, parameter `--maf` of **PLINK1.9**.\n",
    "4. `ref_threshold`: number of standard deviations from the mean of the reference panel `SuperPop` to a sample be considered a possible outlier.\n",
    "5. `stu_threshold`: number of standard deviations from the mean of the study population to flag a sample as possible outlier.\n",
    "6. `reference_pop`: Super population from the reference panel considered as reference for the study.\n",
    "7. `num_pcs`: number of principal components used to flag a sample as outlier."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ancestry_params ={\n",
    "    \"ind_pair\"     : [50, 5, 0.2],\n",
    "    \"pca\"          : 10,\n",
    "    \"maf\"          : 0.01,\n",
    "    \"ref_threshold\": 4,\n",
    "    \"stu_threshold\": 4,\n",
    "    \"reference_pop\": \"SAS\",\n",
    "    \"num_pcs\"      : 10,\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ancestry_qc = AncestryQC(\n",
    "    input_path = input_path, \n",
    "    input_name = input_name, \n",
    "    output_path= output_path, \n",
    "    output_name= output_name, \n",
    "    high_ld_file= high_ld_file,\n",
    "    recompute_merge=True, # if True, it will recompute the merge of the input files\n",
    "    built='38', # '38' it is the default value \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ancestry_qc_steps = {\n",
    "    'merge_study_reference'    : (ancestry_qc.merge_reference_study, {\"ind_pair\":ancestry_params['ind_pair']}),\n",
    "    #'delete_intermediate_files': (ancestry_qc._clean_merging_dir, {}),\n",
    "    'pca_analysis'             : (ancestry_qc.run_pca, \n",
    "        {\n",
    "            \"ref_population\": ancestry_params['reference_pop'],\n",
    "            \"pca\":ancestry_params['pca'],\n",
    "            \"maf\":ancestry_params['maf'],\n",
    "            \"num_pca\":ancestry_params['num_pcs'],\n",
    "            \"ref_threshold\":ancestry_params['ref_threshold'],\n",
    "            \"stu_threshold\":ancestry_params['stu_threshold'],\n",
    "        }\n",
    "    ),\n",
    "}\n",
    "\n",
    "step_description = {\n",
    "    'merge_study_reference'    : \"Merge reference genome with study genome\",\n",
    "    #'delete_intermediate_files': \"Delete intermediate files generated during merging\",\n",
    "    'pca_analysis'             : \"Run a PCA analysis to perfom ancestry QC\"\n",
    "}\n",
    "\n",
    "for name, (func, params) in ancestry_qc_steps.items():\n",
    "    print(f\"\\033[1m{step_description[name]}.\\033[0m\")\n",
    "    func(**params)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df_psam = pd.read_csv(\n",
    "    '/home/luis/CGE/ideal-genom-qc/data/1000genomes_built_38/1kG_phase3_GRCh38.psam',\n",
    "    sep='\\t',\n",
    "    header=0,\n",
    ")\n",
    "df_psam = df_psam[['#IID', 'SuperPop']].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_psam['FID'] = '0'\n",
    "df_psam = df_psam[['FID', '#IID', 'SuperPop']].copy()\n",
    "df_psam.to_csv(\n",
    "    '/media/luis/LaCie/data1/qc1000genomes/dependables/population_info.csv',\n",
    "    sep='\\t',\n",
    "    index=False,\n",
    "    header=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "FID",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "#IID",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "SuperPop",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "pheno",
         "rawType": "int64",
         "type": "integer"
        }
       ],
       "ref": "3641a9bf-000e-493f-89ea-cae7fd38841e",
       "rows": [
        [
         "0",
         "0",
         "HG00096",
         "EUR",
         "1"
        ],
        [
         "1",
         "0",
         "HG00097",
         "EUR",
         "2"
        ],
        [
         "2",
         "0",
         "HG00099",
         "EUR",
         "1"
        ],
        [
         "3",
         "0",
         "HG00100",
         "EUR",
         "1"
        ],
        [
         "4",
         "0",
         "HG00101",
         "EUR",
         "1"
        ]
       ],
       "shape": {
        "columns": 4,
        "rows": 5
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FID</th>\n",
       "      <th>#IID</th>\n",
       "      <th>SuperPop</th>\n",
       "      <th>pheno</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>HG00096</td>\n",
       "      <td>EUR</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>HG00097</td>\n",
       "      <td>EUR</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>HG00099</td>\n",
       "      <td>EUR</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>HG00100</td>\n",
       "      <td>EUR</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>HG00101</td>\n",
       "      <td>EUR</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  FID     #IID SuperPop  pheno\n",
       "0   0  HG00096      EUR      1\n",
       "1   0  HG00097      EUR      2\n",
       "2   0  HG00099      EUR      1\n",
       "3   0  HG00100      EUR      1\n",
       "4   0  HG00101      EUR      1"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "# Set seed for reproducibility\n",
    "np.random.seed(42)\n",
    "\n",
    "# Create a new column with random values 1 and 2\n",
    "df_psam['pheno'] = np.random.choice([1, 2], size=len(df_psam))\n",
    "\n",
    "# Display the first few rows to verify\n",
    "df_psam.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_psam[['FID', \"#IID\", 'pheno']].to_csv(\n",
    "    '/media/luis/LaCie/data1/qc1000genomes/dependables/pheno_info.csv',\n",
    "    sep='\\t',\n",
    "    index=False,\n",
    "    header=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "chr",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "snp",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "cm",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "pos",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "a1",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "a2",
         "rawType": "object",
         "type": "string"
        }
       ],
       "ref": "d2aaa66a-87f0-4d2a-b6d6-c2a31d79d0a7",
       "rows": [
        [
         "0",
         "1",
         "1:49554:A:G",
         "0",
         "49554",
         "G",
         "A"
        ],
        [
         "1",
         "1",
         "1:115746:C:T",
         "0",
         "115746",
         "T",
         "C"
        ],
        [
         "2",
         "1",
         "1:598858:A:C",
         "0",
         "598858",
         "C",
         "A"
        ],
        [
         "3",
         "1",
         "1:629488:C:T",
         "0",
         "629488",
         "T",
         "C"
        ],
        [
         "4",
         "1",
         "1:630053:C:T",
         "0",
         "630053",
         "T",
         "C"
        ],
        [
         "5",
         "1",
         "1:630096:C:G",
         "0",
         "630096",
         "G",
         "C"
        ],
        [
         "6",
         "1",
         "1:630760:G:A",
         "0",
         "630760",
         "A",
         "G"
        ],
        [
         "7",
         "1",
         "1:632199:T:C",
         "0",
         "632199",
         "C",
         "T"
        ],
        [
         "8",
         "1",
         "1:632287:C:T",
         "0",
         "632287",
         "T",
         "C"
        ],
        [
         "9",
         "1",
         "1:632828:T:C",
         "0",
         "632828",
         "C",
         "T"
        ],
        [
         "10",
         "1",
         "1:633147:G:A",
         "0",
         "633147",
         "A",
         "G"
        ],
        [
         "11",
         "1",
         "1:634038:A:G",
         "0",
         "634038",
         "G",
         "A"
        ],
        [
         "12",
         "1",
         "1:785744:A:G",
         "0",
         "785744",
         "G",
         "A"
        ],
        [
         "13",
         "1",
         "1:801883:G:A",
         "0",
         "801883",
         "A",
         "G"
        ],
        [
         "14",
         "1",
         "1:810809:A:G",
         "0",
         "810809",
         "G",
         "A"
        ],
        [
         "15",
         "1",
         "1:817341:G:A",
         "0",
         "817341",
         "A",
         "G"
        ],
        [
         "16",
         "1",
         "1:818025:A:C",
         "0",
         "818025",
         "C",
         "A"
        ],
        [
         "17",
         "1",
         "1:819049:T:G",
         "0",
         "819049",
         "G",
         "T"
        ],
        [
         "18",
         "1",
         "1:825811:C:T",
         "0",
         "825811",
         "T",
         "C"
        ],
        [
         "19",
         "1",
         "1:826384:G:A",
         "0",
         "826384",
         "A",
         "G"
        ],
        [
         "20",
         "1",
         "1:826578:C:T",
         "0",
         "826578",
         "T",
         "C"
        ],
        [
         "21",
         "1",
         "1:826807:C:T",
         "0",
         "826807",
         "T",
         "C"
        ],
        [
         "22",
         "1",
         "1:826940:C:T",
         "0",
         "826940",
         "T",
         "C"
        ],
        [
         "23",
         "1",
         "1:830507:A:C",
         "0",
         "830507",
         "C",
         "A"
        ],
        [
         "24",
         "1",
         "1:833068:G:A",
         "0",
         "833068",
         "A",
         "G"
        ],
        [
         "25",
         "1",
         "1:833439:C:T",
         "0",
         "833439",
         "T",
         "C"
        ],
        [
         "26",
         "1",
         "1:833994:G:A",
         "0",
         "833994",
         "A",
         "G"
        ],
        [
         "27",
         "1",
         "1:834997:A:T",
         "0",
         "834997",
         "T",
         "A"
        ],
        [
         "28",
         "1",
         "1:836324:A:G",
         "0",
         "836324",
         "G",
         "A"
        ],
        [
         "29",
         "1",
         "1:838618:C:T",
         "0",
         "838618",
         "T",
         "C"
        ],
        [
         "30",
         "1",
         "1:839745:C:T",
         "0",
         "839745",
         "T",
         "C"
        ],
        [
         "31",
         "1",
         "1:840046:A:G",
         "0",
         "840046",
         "G",
         "A"
        ],
        [
         "32",
         "1",
         "1:841014:G:A",
         "0",
         "841014",
         "A",
         "G"
        ],
        [
         "33",
         "1",
         "1:843906:C:A",
         "0",
         "843906",
         "A",
         "C"
        ],
        [
         "34",
         "1",
         "1:844364:T:G",
         "0",
         "844364",
         "G",
         "T"
        ],
        [
         "35",
         "1",
         "1:851882:C:G",
         "0",
         "851882",
         "G",
         "C"
        ],
        [
         "36",
         "1",
         "1:856473:G:A",
         "0",
         "856473",
         "A",
         "G"
        ],
        [
         "37",
         "1",
         "1:858390:G:A",
         "0",
         "858390",
         "A",
         "G"
        ],
        [
         "38",
         "1",
         "1:858952:G:A",
         "0",
         "858952",
         "A",
         "G"
        ],
        [
         "39",
         "1",
         "1:863579:G:A",
         "0",
         "863579",
         "A",
         "G"
        ],
        [
         "40",
         "1",
         "1:866563:C:T",
         "0",
         "866563",
         "T",
         "C"
        ],
        [
         "41",
         "1",
         "1:870176:T:A",
         "0",
         "870176",
         "A",
         "T"
        ],
        [
         "42",
         "1",
         "1:879486:T:G",
         "0",
         "879486",
         "G",
         "T"
        ],
        [
         "43",
         "1",
         "1:880823:C:T",
         "0",
         "880823",
         "T",
         "C"
        ],
        [
         "44",
         "1",
         "1:881049:T:G",
         "0",
         "881049",
         "G",
         "T"
        ],
        [
         "45",
         "1",
         "1:881832:C:A",
         "0",
         "881832",
         "A",
         "C"
        ],
        [
         "46",
         "1",
         "1:882166:G:A",
         "0",
         "882166",
         "A",
         "G"
        ],
        [
         "47",
         "1",
         "1:882457:G:A",
         "0",
         "882457",
         "A",
         "G"
        ],
        [
         "48",
         "1",
         "1:886546:C:G",
         "0",
         "886546",
         "G",
         "C"
        ],
        [
         "49",
         "1",
         "1:886724:C:A",
         "0",
         "886724",
         "A",
         "C"
        ]
       ],
       "shape": {
        "columns": 6,
        "rows": 1842636
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>snp</th>\n",
       "      <th>cm</th>\n",
       "      <th>pos</th>\n",
       "      <th>a1</th>\n",
       "      <th>a2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1:49554:A:G</td>\n",
       "      <td>0</td>\n",
       "      <td>49554</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1:115746:C:T</td>\n",
       "      <td>0</td>\n",
       "      <td>115746</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1:598858:A:C</td>\n",
       "      <td>0</td>\n",
       "      <td>598858</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1:629488:C:T</td>\n",
       "      <td>0</td>\n",
       "      <td>629488</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1:630053:C:T</td>\n",
       "      <td>0</td>\n",
       "      <td>630053</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1842631</th>\n",
       "      <td>24</td>\n",
       "      <td>Y:26671311:C:A</td>\n",
       "      <td>0</td>\n",
       "      <td>26671311</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1842632</th>\n",
       "      <td>24</td>\n",
       "      <td>Y:26671347:A:T</td>\n",
       "      <td>0</td>\n",
       "      <td>26671347</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1842633</th>\n",
       "      <td>24</td>\n",
       "      <td>Y:26671432:G:A</td>\n",
       "      <td>0</td>\n",
       "      <td>26671432</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1842634</th>\n",
       "      <td>24</td>\n",
       "      <td>Y:26671489:A:G</td>\n",
       "      <td>0</td>\n",
       "      <td>26671489</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1842635</th>\n",
       "      <td>24</td>\n",
       "      <td>Y:56747774:A:C</td>\n",
       "      <td>0</td>\n",
       "      <td>56747774</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1842636 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chr             snp  cm       pos a1 a2\n",
       "0          1     1:49554:A:G   0     49554  G  A\n",
       "1          1    1:115746:C:T   0    115746  T  C\n",
       "2          1    1:598858:A:C   0    598858  C  A\n",
       "3          1    1:629488:C:T   0    629488  T  C\n",
       "4          1    1:630053:C:T   0    630053  T  C\n",
       "...      ...             ...  ..       ... .. ..\n",
       "1842631   24  Y:26671311:C:A   0  26671311  A  C\n",
       "1842632   24  Y:26671347:A:T   0  26671347  T  A\n",
       "1842633   24  Y:26671432:G:A   0  26671432  A  G\n",
       "1842634   24  Y:26671489:A:G   0  26671489  G  A\n",
       "1842635   24  Y:56747774:A:C   0  56747774  C  A\n",
       "\n",
       "[1842636 rows x 6 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_bim = pd.read_csv(\n",
    "    '/media/luis/LaCie/rawdata-sexupdated/outputData/sample_qc_results/clean_files/qc-results-clean-samples.bim',\n",
    "    sep='\\t',\n",
    "    header=None,\n",
    "    names=['chr', 'snp', 'cm', 'pos', 'a1', 'a2']\n",
    ")\n",
    "df_bim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mKeyboardInterrupt\u001b[39m                         Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[5]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m df_gens = \u001b[43mpd\u001b[49m\u001b[43m.\u001b[49m\u001b[43mread_csv\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m      2\u001b[39m \u001b[43m    \u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43m/media/luis/LaCie/data1/qc1000genomes/inputData/all_hg38-filtered-renamed.bim\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m      3\u001b[39m \u001b[43m    \u001b[49m\u001b[43msep\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[38;5;130;43;01m\\t\u001b[39;49;00m\u001b[33;43m'\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m      4\u001b[39m \u001b[43m    \u001b[49m\u001b[43mheader\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[32m      5\u001b[39m \u001b[43m)\u001b[49m\n\u001b[32m      6\u001b[39m df_gens\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/.cache/pypoetry/virtualenvs/ideal-genom-qc-6U2l_yTg-py3.12/lib/python3.12/site-packages/pandas/io/parsers/readers.py:1026\u001b[39m, in \u001b[36mread_csv\u001b[39m\u001b[34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, date_format, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options, dtype_backend)\u001b[39m\n\u001b[32m   1013\u001b[39m kwds_defaults = _refine_defaults_read(\n\u001b[32m   1014\u001b[39m     dialect,\n\u001b[32m   1015\u001b[39m     delimiter,\n\u001b[32m   (...)\u001b[39m\u001b[32m   1022\u001b[39m     dtype_backend=dtype_backend,\n\u001b[32m   1023\u001b[39m )\n\u001b[32m   1024\u001b[39m kwds.update(kwds_defaults)\n\u001b[32m-> \u001b[39m\u001b[32m1026\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_read\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/.cache/pypoetry/virtualenvs/ideal-genom-qc-6U2l_yTg-py3.12/lib/python3.12/site-packages/pandas/io/parsers/readers.py:626\u001b[39m, in \u001b[36m_read\u001b[39m\u001b[34m(filepath_or_buffer, kwds)\u001b[39m\n\u001b[32m    623\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m parser\n\u001b[32m    625\u001b[39m \u001b[38;5;28;01mwith\u001b[39;00m parser:\n\u001b[32m--> \u001b[39m\u001b[32m626\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mparser\u001b[49m\u001b[43m.\u001b[49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnrows\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/.cache/pypoetry/virtualenvs/ideal-genom-qc-6U2l_yTg-py3.12/lib/python3.12/site-packages/pandas/io/parsers/readers.py:1923\u001b[39m, in \u001b[36mTextFileReader.read\u001b[39m\u001b[34m(self, nrows)\u001b[39m\n\u001b[32m   1916\u001b[39m nrows = validate_integer(\u001b[33m\"\u001b[39m\u001b[33mnrows\u001b[39m\u001b[33m\"\u001b[39m, nrows)\n\u001b[32m   1917\u001b[39m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[32m   1918\u001b[39m     \u001b[38;5;66;03m# error: \"ParserBase\" has no attribute \"read\"\u001b[39;00m\n\u001b[32m   1919\u001b[39m     (\n\u001b[32m   1920\u001b[39m         index,\n\u001b[32m   1921\u001b[39m         columns,\n\u001b[32m   1922\u001b[39m         col_dict,\n\u001b[32m-> \u001b[39m\u001b[32m1923\u001b[39m     ) = \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_engine\u001b[49m\u001b[43m.\u001b[49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43m  \u001b[49m\u001b[38;5;66;43;03m# type: ignore[attr-defined]\u001b[39;49;00m\n\u001b[32m   1924\u001b[39m \u001b[43m        \u001b[49m\u001b[43mnrows\u001b[49m\n\u001b[32m   1925\u001b[39m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   1926\u001b[39m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m:\n\u001b[32m   1927\u001b[39m     \u001b[38;5;28mself\u001b[39m.close()\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/.cache/pypoetry/virtualenvs/ideal-genom-qc-6U2l_yTg-py3.12/lib/python3.12/site-packages/pandas/io/parsers/c_parser_wrapper.py:234\u001b[39m, in \u001b[36mCParserWrapper.read\u001b[39m\u001b[34m(self, nrows)\u001b[39m\n\u001b[32m    232\u001b[39m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[32m    233\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m.low_memory:\n\u001b[32m--> \u001b[39m\u001b[32m234\u001b[39m         chunks = \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_reader\u001b[49m\u001b[43m.\u001b[49m\u001b[43mread_low_memory\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnrows\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    235\u001b[39m         \u001b[38;5;66;03m# destructive to chunks\u001b[39;00m\n\u001b[32m    236\u001b[39m         data = _concatenate_chunks(chunks)\n",
      "\u001b[36mFile \u001b[39m\u001b[32mparsers.pyx:838\u001b[39m, in \u001b[36mpandas._libs.parsers.TextReader.read_low_memory\u001b[39m\u001b[34m()\u001b[39m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mparsers.pyx:905\u001b[39m, in \u001b[36mpandas._libs.parsers.TextReader._read_rows\u001b[39m\u001b[34m()\u001b[39m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mparsers.pyx:874\u001b[39m, in \u001b[36mpandas._libs.parsers.TextReader._tokenize_rows\u001b[39m\u001b[34m()\u001b[39m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mparsers.pyx:891\u001b[39m, in \u001b[36mpandas._libs.parsers.TextReader._check_tokenize_status\u001b[39m\u001b[34m()\u001b[39m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mparsers.pyx:2053\u001b[39m, in \u001b[36mpandas._libs.parsers.raise_parser_error\u001b[39m\u001b[34m()\u001b[39m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m<frozen codecs>:319\u001b[39m, in \u001b[36mdecode\u001b[39m\u001b[34m(self, input, final)\u001b[39m\n",
      "\u001b[31mKeyboardInterrupt\u001b[39m: "
     ]
    }
   ],
   "source": [
    "# Define the path to the BIM file\n",
    "large_bim_path = '/media/luis/LaCie/data1/qc1000genomes/inputData/all_hg38-filtered-renamed.bim'\n",
    "\n",
    "# Define column names for BIM file\n",
    "bim_columns = ['chr', 'snp', 'cm', 'pos', 'a1', 'a2']\n",
    "\n",
    "# Initialize an empty list to store chunks\n",
    "chunks = []\n",
    "\n",
    "# Load the file in chunks of 500,000 rows\n",
    "chunk_size = 500000\n",
    "for chunk in pd.read_csv(large_bim_path, sep='\\t', header=None, names=bim_columns, \n",
    "                         chunksize=chunk_size):\n",
    "    # Process each chunk (e.g., print its shape)\n",
    "    print(f\"Loaded chunk with shape: {chunk.shape}\")\n",
    "    \n",
    "    # Optionally store the chunk for further processing\n",
    "    chunks.append(chunk)\n",
    "\n",
    "# If you need the entire dataframe after processing in chunks\n",
    "# Be careful with this if the file is very large\n",
    "# df_large_bim = pd.concat(chunks)\n",
    "# print(f\"Total shape of combined data: {df_large_bim.shape}\")\n",
    "\n",
    "# Alternatively, print summary statistics without loading the entire file\n",
    "print(\"\\nSample from first chunk:\")\n",
    "print(chunks[0].head())\n",
    "\n",
    "# Count total number of rows processed\n",
    "total_rows = sum(len(chunk) for chunk in chunks)\n",
    "print(f\"\\nTotal number of rows processed: {total_rows}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ideal-genom-qc-py3.12",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
