Search.setIndex({"alltitles": {"Dependencies:": [[0, "dependencies"], [0, "id5"]], "Documentation": [[6, "documentation"]], "Features": [[6, "features"]], "Installation": [[6, "installation"]], "Note:": [[5, "note"]], "Notes:": [[2, "notes"], [2, "id5"], [5, "notes"]], "Parameter:": [[2, "parameter"]], "Parameters:": [[4, "parameters"], [5, "parameters"]], "Raises:": [[2, "raises"], [2, "id2"], [2, "id4"], [5, "raises"]], "Requires:": [[0, "requires"]], "Returns:": [[0, "returns"], [0, "id1"], [0, "id3"], [0, "id6"], [0, "id8"], [2, "returns"], [2, "id1"], [2, "id3"], [4, "returns"], [4, "id1"], [4, "id3"], [5, "returns"], [5, "id1"]], "Side Effects:": [[0, "side-effects"], [0, "id2"], [0, "id4"], [0, "id9"]], "Side effects:": [[0, "id7"], [4, "side-effects"], [4, "id2"]], "ideal-genome-qc documentation": [[6, null]], "ideal_genom_qc.AncestryQC": [[0, null]], "ideal_genom_qc.Helpers": [[1, null]], "ideal_genom_qc.SampleQC": [[2, null]], "ideal_genom_qc.UMAPplot": [[3, null]], "ideal_genom_qc.VariantQC": [[4, null]], "ideal_genom_qc.get_references": [[5, null]]}, "docnames": ["AncestryQC", "Helpers", "SampleQC", "UMAPplot", "VariantQC", "get_references", "index"], "envversion": {"sphinx": 62, "sphinx.domains.c": 3, "sphinx.domains.changeset": 1, "sphinx.domains.citation": 1, "sphinx.domains.cpp": 9, "sphinx.domains.index": 1, "sphinx.domains.javascript": 3, "sphinx.domains.math": 2, "sphinx.domains.python": 4, "sphinx.domains.rst": 2, "sphinx.domains.std": 2}, "filenames": ["AncestryQC.rst", "Helpers.rst", "SampleQC.rst", "UMAPplot.rst", "VariantQC.rst", "get_references.rst", "index.rst"], "indexentries": {"ancestryqc (class in ideal_genom_qc.ancestryqc)": [[0, "ideal_genom_qc.AncestryQC.AncestryQC", false]], "arg_parser() (in module ideal_genom_qc.helpers)": [[1, "ideal_genom_qc.Helpers.arg_parser", false]], "clean_input_folder() (ideal_genom_qc.sampleqc.sampleqc method)": [[2, "ideal_genom_qc.SampleQC.SampleQC.clean_input_folder", false]], "clean_result_folder() (ideal_genom_qc.sampleqc.sampleqc method)": [[2, "ideal_genom_qc.SampleQC.SampleQC.clean_result_folder", false]], "compute_pcas() (ideal_genom_qc.umapplot.umapplot method)": [[3, "ideal_genom_qc.UMAPplot.UMAPplot.compute_pcas", false]], "delete_temp_files() (in module ideal_genom_qc.helpers)": [[1, "ideal_genom_qc.Helpers.delete_temp_files", false]], "draw_pca_plot() (ideal_genom_qc.ancestryqc.genomicoutlieranalyzer method)": [[0, "ideal_genom_qc.AncestryQC.GenomicOutlierAnalyzer.draw_pca_plot", false]], "execute_different_genotype_call_rate() (ideal_genom_qc.variantqc.variantqc method)": [[4, "ideal_genom_qc.VariantQC.VariantQC.execute_different_genotype_call_rate", false]], "execute_drop_ancestry_outliers() (ideal_genom_qc.ancestryqc.genomicoutlieranalyzer method)": [[0, "ideal_genom_qc.AncestryQC.GenomicOutlierAnalyzer.execute_drop_ancestry_outliers", false]], "execute_drop_samples() (ideal_genom_qc.sampleqc.sampleqc method)": [[2, "ideal_genom_qc.SampleQC.SampleQC.execute_drop_samples", false]], "execute_drop_variants() (ideal_genom_qc.variantqc.variantqc method)": [[4, "ideal_genom_qc.VariantQC.VariantQC.execute_drop_variants", false]], "execute_duplicate_relatedness() (ideal_genom_qc.sampleqc.sampleqc method)": [[2, "ideal_genom_qc.SampleQC.SampleQC.execute_duplicate_relatedness", false]], "execute_filter_prob_snps() (ideal_genom_qc.ancestryqc.referencegenomicmerger method)": [[0, "ideal_genom_qc.AncestryQC.ReferenceGenomicMerger.execute_filter_prob_snps", false]], "execute_fix_allele_flip() (ideal_genom_qc.ancestryqc.referencegenomicmerger method)": [[0, "ideal_genom_qc.AncestryQC.ReferenceGenomicMerger.execute_fix_allele_flip", false]], "execute_fix_chromosome_mismatch() (ideal_genom_qc.ancestryqc.referencegenomicmerger method)": [[0, "ideal_genom_qc.AncestryQC.ReferenceGenomicMerger.execute_fix_chromosome_mismatch", false]], "execute_fix_possition_mismatch() (ideal_genom_qc.ancestryqc.referencegenomicmerger method)": [[0, "ideal_genom_qc.AncestryQC.ReferenceGenomicMerger.execute_fix_possition_mismatch", false]], "execute_haploid_to_missing() (ideal_genom_qc.sampleqc.sampleqc method)": [[2, "ideal_genom_qc.SampleQC.SampleQC.execute_haploid_to_missing", false]], "execute_heterozygosity_rate() (ideal_genom_qc.sampleqc.sampleqc method)": [[2, "ideal_genom_qc.SampleQC.SampleQC.execute_heterozygosity_rate", false]], "execute_hwe_test() (ideal_genom_qc.variantqc.variantqc method)": [[4, "ideal_genom_qc.VariantQC.VariantQC.execute_hwe_test", false]], "execute_ibd() (ideal_genom_qc.sampleqc.sampleqc method)": [[2, "ideal_genom_qc.SampleQC.SampleQC.execute_ibd", false]], "execute_kingship() (ideal_genom_qc.sampleqc.sampleqc method)": [[2, "ideal_genom_qc.SampleQC.SampleQC.execute_kingship", false]], "execute_ld_pruning() (ideal_genom_qc.ancestryqc.referencegenomicmerger method)": [[0, "ideal_genom_qc.AncestryQC.ReferenceGenomicMerger.execute_ld_pruning", false]], "execute_ld_pruning() (ideal_genom_qc.sampleqc.sampleqc method)": [[2, "ideal_genom_qc.SampleQC.SampleQC.execute_ld_pruning", false]], "execute_merge_data() (ideal_genom_qc.ancestryqc.referencegenomicmerger method)": [[0, "ideal_genom_qc.AncestryQC.ReferenceGenomicMerger.execute_merge_data", false]], "execute_miss_genotype() (ideal_genom_qc.sampleqc.sampleqc method)": [[2, "ideal_genom_qc.SampleQC.SampleQC.execute_miss_genotype", false]], "execute_missing_data_rate() (ideal_genom_qc.variantqc.variantqc method)": [[4, "ideal_genom_qc.VariantQC.VariantQC.execute_missing_data_rate", false]], "execute_pca() (ideal_genom_qc.ancestryqc.genomicoutlieranalyzer method)": [[0, "ideal_genom_qc.AncestryQC.GenomicOutlierAnalyzer.execute_pca", false]], "execute_remove_mismatches() (ideal_genom_qc.ancestryqc.referencegenomicmerger method)": [[0, "ideal_genom_qc.AncestryQC.ReferenceGenomicMerger.execute_remove_mismatches", false]], "execute_rename_snpid() (ideal_genom_qc.sampleqc.sampleqc method)": [[2, "ideal_genom_qc.SampleQC.SampleQC.execute_rename_snpid", false]], "execute_sex_check() (ideal_genom_qc.sampleqc.sampleqc method)": [[2, "ideal_genom_qc.SampleQC.SampleQC.execute_sex_check", false]], "fetcher1000genome (class in ideal_genom_qc.get_references)": [[5, "ideal_genom_qc.get_references.Fetcher1000Genome", false]], "fetcherldregions (class in ideal_genom_qc.get_references)": [[5, "ideal_genom_qc.get_references.FetcherLDRegions", false]], "find_ancestry_outliers() (ideal_genom_qc.ancestryqc.genomicoutlieranalyzer method)": [[0, "ideal_genom_qc.AncestryQC.GenomicOutlierAnalyzer.find_ancestry_outliers", false]], "generate_plots() (ideal_genom_qc.umapplot.umapplot method)": [[3, "ideal_genom_qc.UMAPplot.UMAPplot.generate_plots", false]], "genomicoutlieranalyzer (class in ideal_genom_qc.ancestryqc)": [[0, "ideal_genom_qc.AncestryQC.GenomicOutlierAnalyzer", false]], "get_1000genomes() (ideal_genom_qc.get_references.fetcher1000genome method)": [[5, "ideal_genom_qc.get_references.Fetcher1000Genome.get_1000genomes", false]], "get_1000genomes_binaries() (ideal_genom_qc.get_references.fetcher1000genome method)": [[5, "ideal_genom_qc.get_references.Fetcher1000Genome.get_1000genomes_binaries", false]], "get_fail_samples() (ideal_genom_qc.sampleqc.sampleqc method)": [[2, "ideal_genom_qc.SampleQC.SampleQC.get_fail_samples", false]], "get_fail_variants() (ideal_genom_qc.variantqc.variantqc method)": [[4, "ideal_genom_qc.VariantQC.VariantQC.get_fail_variants", false]], "get_ld_regions() (ideal_genom_qc.get_references.fetcherldregions method)": [[5, "ideal_genom_qc.get_references.FetcherLDRegions.get_ld_regions", false]], "ideal_genom_qc.ancestryqc": [[0, "module-ideal_genom_qc.AncestryQC", false]], "ideal_genom_qc.get_references": [[5, "module-ideal_genom_qc.get_references", false]], "ideal_genom_qc.helpers": [[1, "module-ideal_genom_qc.Helpers", false]], "ideal_genom_qc.sampleqc": [[2, "module-ideal_genom_qc.SampleQC", false]], "ideal_genom_qc.umapplot": [[3, "module-ideal_genom_qc.UMAPplot", false]], "ideal_genom_qc.variantqc": [[4, "module-ideal_genom_qc.VariantQC", false]], "ld_pruning() (ideal_genom_qc.umapplot.umapplot method)": [[3, "ideal_genom_qc.UMAPplot.UMAPplot.ld_pruning", false]], "merge_reference_study() (ideal_genom_qc.ancestryqc.ancestryqc method)": [[0, "ideal_genom_qc.AncestryQC.AncestryQC.merge_reference_study", false]], "module": [[0, "module-ideal_genom_qc.AncestryQC", false], [1, "module-ideal_genom_qc.Helpers", false], [2, "module-ideal_genom_qc.SampleQC", false], [3, "module-ideal_genom_qc.UMAPplot", false], [4, "module-ideal_genom_qc.VariantQC", false], [5, "module-ideal_genom_qc.get_references", false]], "referencegenomicmerger (class in ideal_genom_qc.ancestryqc)": [[0, "ideal_genom_qc.AncestryQC.ReferenceGenomicMerger", false]], "report_call_rate() (ideal_genom_qc.sampleqc.sampleqc method)": [[2, "ideal_genom_qc.SampleQC.SampleQC.report_call_rate", false]], "report_different_genotype_call_rate() (ideal_genom_qc.variantqc.variantqc method)": [[4, "ideal_genom_qc.VariantQC.VariantQC.report_different_genotype_call_rate", false]], "report_heterozygosity_rate() (ideal_genom_qc.sampleqc.sampleqc method)": [[2, "ideal_genom_qc.SampleQC.SampleQC.report_heterozygosity_rate", false]], "report_hwe() (ideal_genom_qc.variantqc.variantqc method)": [[4, "ideal_genom_qc.VariantQC.VariantQC.report_hwe", false]], "report_ibd_analysis() (ideal_genom_qc.sampleqc.sampleqc method)": [[2, "ideal_genom_qc.SampleQC.SampleQC.report_ibd_analysis", false]], "report_missing_data() (ideal_genom_qc.variantqc.variantqc method)": [[4, "ideal_genom_qc.VariantQC.VariantQC.report_missing_data", false]], "report_sex_check() (ideal_genom_qc.sampleqc.sampleqc method)": [[2, "ideal_genom_qc.SampleQC.SampleQC.report_sex_check", false]], "run_pca() (ideal_genom_qc.ancestryqc.ancestryqc method)": [[0, "ideal_genom_qc.AncestryQC.AncestryQC.run_pca", false]], "sampleqc (class in ideal_genom_qc.sampleqc)": [[2, "ideal_genom_qc.SampleQC.SampleQC", false]], "shell_do() (in module ideal_genom_qc.helpers)": [[1, "ideal_genom_qc.Helpers.shell_do", false]], "umapplot (class in ideal_genom_qc.umapplot)": [[3, "ideal_genom_qc.UMAPplot.UMAPplot", false]], "variantqc (class in ideal_genom_qc.variantqc)": [[4, "ideal_genom_qc.VariantQC.VariantQC", false]]}, "objects": {"ideal_genom_qc": [[0, 0, 0, "-", "AncestryQC"], [1, 0, 0, "-", "Helpers"], [2, 0, 0, "-", "SampleQC"], [3, 0, 0, "-", "UMAPplot"], [4, 0, 0, "-", "VariantQC"], [5, 0, 0, "-", "get_references"]], "ideal_genom_qc.AncestryQC": [[0, 1, 1, "", "AncestryQC"], [0, 1, 1, "", "GenomicOutlierAnalyzer"], [0, 1, 1, "", "ReferenceGenomicMerger"]], "ideal_genom_qc.AncestryQC.AncestryQC": [[0, 2, 1, "", "merge_reference_study"], [0, 2, 1, "", "run_pca"]], "ideal_genom_qc.AncestryQC.GenomicOutlierAnalyzer": [[0, 2, 1, "", "draw_pca_plot"], [0, 2, 1, "", "execute_drop_ancestry_outliers"], [0, 2, 1, "", "execute_pca"], [0, 2, 1, "", "find_ancestry_outliers"]], "ideal_genom_qc.AncestryQC.ReferenceGenomicMerger": [[0, 2, 1, "", "execute_filter_prob_snps"], [0, 2, 1, "", "execute_fix_allele_flip"], [0, 2, 1, "", "execute_fix_chromosome_mismatch"], [0, 2, 1, "", "execute_fix_possition_mismatch"], [0, 2, 1, "", "execute_ld_pruning"], [0, 2, 1, "", "execute_merge_data"], [0, 2, 1, "", "execute_remove_mismatches"]], "ideal_genom_qc.Helpers": [[1, 3, 1, "", "arg_parser"], [1, 3, 1, "", "delete_temp_files"], [1, 3, 1, "", "shell_do"]], "ideal_genom_qc.SampleQC": [[2, 1, 1, "", "SampleQC"]], "ideal_genom_qc.SampleQC.SampleQC": [[2, 2, 1, "", "clean_input_folder"], [2, 2, 1, "", "clean_result_folder"], [2, 2, 1, "", "execute_drop_samples"], [2, 2, 1, "", "execute_duplicate_relatedness"], [2, 2, 1, "", "execute_haploid_to_missing"], [2, 2, 1, "", "execute_heterozygosity_rate"], [2, 2, 1, "", "execute_ibd"], [2, 2, 1, "", "execute_kingship"], [2, 2, 1, "", "execute_ld_pruning"], [2, 2, 1, "", "execute_miss_genotype"], [2, 2, 1, "", "execute_rename_snpid"], [2, 2, 1, "", "execute_sex_check"], [2, 2, 1, "", "get_fail_samples"], [2, 2, 1, "", "report_call_rate"], [2, 2, 1, "", "report_heterozygosity_rate"], [2, 2, 1, "", "report_ibd_analysis"], [2, 2, 1, "", "report_sex_check"]], "ideal_genom_qc.UMAPplot": [[3, 1, 1, "", "UMAPplot"]], "ideal_genom_qc.UMAPplot.UMAPplot": [[3, 2, 1, "", "compute_pcas"], [3, 2, 1, "", "generate_plots"], [3, 2, 1, "", "ld_pruning"]], "ideal_genom_qc.VariantQC": [[4, 1, 1, "", "VariantQC"]], "ideal_genom_qc.VariantQC.VariantQC": [[4, 2, 1, "", "execute_different_genotype_call_rate"], [4, 2, 1, "", "execute_drop_variants"], [4, 2, 1, "", "execute_hwe_test"], [4, 2, 1, "", "execute_missing_data_rate"], [4, 2, 1, "", "get_fail_variants"], [4, 2, 1, "", "report_different_genotype_call_rate"], [4, 2, 1, "", "report_hwe"], [4, 2, 1, "", "report_missing_data"]], "ideal_genom_qc.get_references": [[5, 1, 1, "", "Fetcher1000Genome"], [5, 1, 1, "", "FetcherLDRegions"]], "ideal_genom_qc.get_references.Fetcher1000Genome": [[5, 2, 1, "", "get_1000genomes"], [5, 2, 1, "", "get_1000genomes_binaries"]], "ideal_genom_qc.get_references.FetcherLDRegions": [[5, 2, 1, "", "get_ld_regions"]]}, "objnames": {"0": ["py", "module", "Python module"], "1": ["py", "class", "Python class"], "2": ["py", "method", "Python method"], "3": ["py", "function", "Python function"]}, "objtypes": {"0": "py:module", "1": "py:class", "2": "py:method", "3": "py:function"}, "terms": {"": [0, 2, 3, 4], "0": [0, 2, 3, 4], "001": 3, "01": [0, 2], "02": [2, 3], "05": 4, "08": [3, 4], "1": [0, 2, 3, 4, 5], "10": [0, 2, 3, 4], "1000": 5, "100000": 2, "185": 2, "1b9e77": 2, "1e": 4, "2": [0, 2, 3, 4, 5], "22": 5, "24": 4, "26": 4, "2d": 0, "3": [0, 2, 3, 4], "354": 2, "37": 5, "38": [0, 2, 3, 5], "3d": 0, "4": [0, 2], "5": [0, 2, 3, 4], "50": [0, 2, 3], "5e": [3, 4], "6": 0, "7": 0, "8": [2, 3, 4], "80": 2, "98": 2, "A": [0, 2, 4], "By": 2, "For": [2, 4, 5], "If": [0, 2, 3, 4, 5], "It": [0, 2, 3, 6], "Or": 6, "The": [0, 2, 3, 4, 5], "_histogram": 2, "_scatterplot": 2, "a1": 2, "a2": 2, "abov": [2, 3, 4], "access": [0, 5], "across": 4, "actual": [0, 5], "addit": [3, 6], "against": 0, "aid": 2, "al": 1, "algorithm": 2, "all": [0, 2, 3, 4], "allel": [0, 2, 3, 4], "allele1": 2, "allele2": 2, "alloc": [0, 4], "allow": 4, "alreadi": [0, 5], "also": [2, 4], "alwai": [2, 4], "an": [0, 3, 4], "analysi": [0, 2, 3, 4, 6], "analyz": [0, 2, 4], "ancestri": [0, 6], "ancestry_fail": 0, "ancestryqc": 6, "ani": [0, 2], "api": 6, "appli": 0, "approach": 0, "appropri": 2, "approxim": [3, 4], "ar": [0, 1, 2, 3, 4, 5], "arg_pars": [1, 6], "argument": 3, "assess": 4, "assign": 0, "attribut": [0, 2, 4], "autom": 6, "automat": [0, 2], "autosom": 2, "autosomal_filenam": 2, "avail": [0, 2, 4], "axi": [2, 4], "back": 2, "base": [0, 2, 3, 4, 5], "becaus": 6, "bed": [0, 2, 4, 5], "been": 2, "being": 0, "below": [2, 3, 4], "between": [0, 2, 3, 4], "bim": [0, 2, 4, 5], "binari": [0, 2, 3, 4, 5], "bioinformat": 6, "blue": 2, "bmerg": 0, "bool": [0, 2, 3], "boolean": 2, "both": [0, 2], "bound": 2, "boundari": 2, "build": 5, "built": [0, 2, 3, 5], "c": 0, "calcul": [0, 2, 4], "call": [1, 2, 4], "call_rate_": 2, "call_rate_thr": 2, "can": [3, 6], "cannot": 2, "cap": 2, "capabl": 0, "case": [3, 4], "case_control_mark": 3, "case_control_miss": 4, "case_controls_thr": 4, "categori": 4, "cge": 6, "chang": 6, "check": [2, 4], "chr": 2, "chr_y": 4, "chromosom": [0, 2, 4, 5], "chunk_siz": 2, "circl": 2, "class": [0, 2, 3, 4, 5], "clean": [0, 2, 4, 6], "clean_input_fold": 2, "clean_result_fold": 2, "cleanup": 5, "clone": 6, "code": 3, "coeffici": 2, "color": [0, 2, 3], "color_hue_fil": 3, "column": [0, 2, 3, 4], "com": 6, "combin": [3, 4], "command": [0, 1, 2, 3, 4], "compar": 0, "comparison": [0, 2], "complet": 0, "complex": 2, "compon": [0, 3], "comput": [2, 3, 4], "compute_pca": 3, "condit": 4, "consid": 2, "consolid": 4, "constant": 4, "construct": 0, "constructor": 3, "contain": [0, 2, 3, 4, 6], "control": [0, 2, 3, 4, 6], "convers": [2, 5], "convert": [2, 5], "coordin": 5, "core": [0, 2], "correct": [0, 3], "correl": [0, 2, 3], "correspond": [2, 5], "could": 6, "count": [0, 2, 3, 4], "cpu": [0, 2, 3], "creat": [0, 1, 2, 3, 4, 5], "criteria": 4, "criterion": 4, "csv": [0, 3], "current": 0, "custom": 5, "d95f02": 2, "data": [0, 2, 3, 4, 5, 6], "datafram": [2, 4], "dataset": [0, 2, 3, 4, 5], "decompress": 5, "default": [0, 2, 3, 4, 5], "defin": 2, "degre": 2, "delet": 1, "delete_temp_fil": [1, 6], "descent": 2, "descript": 2, "design": 6, "destin": 5, "detail": 2, "detect": 0, "determin": [0, 2], "deviant": 2, "deviat": 2, "dict": [0, 1, 2, 3], "dictionari": [2, 3], "differ": [2, 3, 4, 5], "differenti": 4, "dimens": 3, "direct": 2, "directori": [0, 1, 2, 3, 4, 5], "disclaim": 6, "discrep": 2, "disequilibrium": [0, 2, 3, 5], "disk": 3, "distanc": [0, 3], "distribut": [2, 4], "doe": 2, "doesn": [0, 2, 5], "don": 2, "done": [0, 5], "dot": 2, "download": 5, "draw": 3, "draw_pca_plot": 0, "drop": 0, "duplic": [2, 4], "dure": [0, 1, 2, 3], "each": [0, 2, 3, 4], "eigenv": 0, "eigenvalu": [0, 3], "eigenvec": 0, "eigenvector": [0, 3], "einvector": 0, "either": [2, 4], "element": [0, 1, 2], "empti": [0, 2, 3, 5], "equilibrium": [3, 4], "equival": 2, "err": 1, "error": 6, "estim": 2, "euclidean": 3, "exact": 3, "exactli": 2, "except": [0, 5], "exclud": [0, 2, 4], "execut": [0, 1, 2, 4], "execute_different_genotype_call_r": 4, "execute_drop_ancestry_outli": 0, "execute_drop_sampl": 2, "execute_drop_vari": 4, "execute_duplicate_related": 2, "execute_filter_prob_snp": 0, "execute_fix_allele_flip": 0, "execute_fix_chromosome_mismatch": 0, "execute_fix_possition_mismatch": 0, "execute_haploid_to_miss": 2, "execute_heterozygosity_r": 2, "execute_hwe_test": 4, "execute_ibd": 2, "execute_kingship": 2, "execute_ld_prun": [0, 2], "execute_merge_data": 0, "execute_miss_genotyp": 2, "execute_missing_data_r": 4, "execute_pca": 0, "execute_remove_mismatch": 0, "execute_rename_snpid": 2, "execute_sex_check": 2, "exist": [0, 5], "expect": [0, 2], "explicitli": 5, "extens": [5, 6], "extract": 2, "f": 2, "f_miss": 2, "fail": [0, 2, 4, 5], "fail_mark": 4, "fail_sampl": 2, "fails_dir": 0, "failur": [2, 4], "fall": 2, "fals": [0, 1, 2, 3], "fam": [0, 2, 3, 4, 5], "famili": 2, "femal": [2, 4], "females_missing_data": 4, "fetcher1000genom": [5, 6], "fetcherldregion": [5, 6], "fid": 2, "file": [0, 1, 2, 3, 4, 5], "filenam": [0, 2, 4], "filename_femal": 4, "filename_mal": 4, "filenotfounderror": [2, 3], "files_to_keep": 1, "fill": 2, "filter": [0, 2, 3, 4, 6], "final": [0, 2, 3, 4, 5], "find_ancestry_outli": 0, "first": [0, 2, 3], "fix": 0, "flag": 2, "flip": 0, "float": [0, 2, 3, 4], "focu": 6, "folder": [1, 2], "follow": [0, 2, 4], "format": [2, 4, 5], "found": [2, 3], "free": 2, "freiburg": 5, "frequenc": [0, 2, 3, 4], "from": [0, 1, 2, 3, 4, 5, 6], "full": [1, 6], "function": [0, 1, 2, 4], "g": 0, "genepi": 5, "gener": [0, 2, 3, 4, 5, 6], "generate_plot": 3, "genet": [0, 2, 3], "geno": [3, 4], "genom": [2, 5], "genomicoutlieranalyz": [0, 6], "genotool": 1, "genotyp": [2, 4], "get": 2, "get_1000genom": 5, "get_1000genomes_binari": 5, "get_fail_sampl": 2, "get_fail_vari": 4, "get_ld_region": 5, "get_refer": 6, "git": 6, "github": [5, 6], "given": 4, "green": 2, "grid": 3, "group": [2, 4], "gwa": 5, "ha": 4, "half": [0, 2], "handl": [0, 2, 5], "haploid": 2, "hardcod": 5, "hardi": [3, 4], "have": [2, 3, 6], "help": 2, "helper": 6, "heterozygos": 2, "hh": 2, "hh_to_miss": 2, "high": [0, 2, 4, 5], "high_ld_fil": [0, 2, 3], "high_ld_region": 0, "higher": 2, "highli": [0, 3], "histogram": [2, 4], "hollow": 2, "http": 6, "hue": 3, "hwe": [3, 4], "hwe_result": 4, "hwe_threshold": 4, "hwe_y_cap": 4, "i": [0, 2, 3, 4, 5, 6], "ibd": 2, "ibd_miss": 2, "ibd_threshold": 2, "id": [2, 5], "id1": [0, 3], "id2": [0, 3], "ideal_genom_qc": 6, "ident": 2, "identifi": [0, 2, 4], "iid": 2, "imag": 2, "imiss": [2, 4], "includ": [0, 3, 4, 5], "incorpor": 3, "ind_pair": [0, 2, 3], "indep": [0, 2], "indic": [2, 4], "individu": [2, 3], "inform": [0, 2, 3, 4], "initi": [0, 3], "input": [2, 3, 4], "input_nam": [0, 2, 3, 4], "input_path": [0, 2, 3, 4], "instal": [0, 5], "instanc": [0, 2], "insuffici": 3, "int": [0, 2, 3, 4], "integ": [0, 2, 3, 4], "integr": 6, "intermedi": [2, 3], "intern": 0, "invalid": 3, "its": [1, 2], "jpeg": [0, 2, 3], "keep": 2, "kept": 1, "keyword": 3, "king": 2, "kingship": 2, "kinship": 2, "kinship_miss": 2, "latest": 6, "ld": [0, 2, 3, 5], "ld_prune": 3, "ldprune": [2, 3], "ldregionexclud": 2, "least": 3, "leav": 2, "level": 6, "librari": 6, "limit": 4, "line": [0, 2], "line_color": 2, "linkag": [0, 2, 3, 5], "list": [0, 1, 2, 3], "lmiss": 4, "load": [0, 2], "local": 6, "locat": [1, 4, 5], "log": [0, 1, 2], "log10": 4, "log_fil": 1, "logic": 2, "lower": 2, "maf": [0, 2, 3, 4], "maf_het": 2, "mai": [3, 5], "main": 2, "maintain": [0, 2], "male": [2, 4], "male_female_y_cap": 4, "males_missing_data": 4, "mani": 6, "manifold": 3, "map": 2, "mark": 2, "marker": [3, 4], "marker_call_rate_thr": 4, "match": 0, "matrix": 2, "max": 2, "maximum": [0, 2, 3, 4], "meaning": 3, "memori": [0, 2, 4], "merg": [0, 2], "merge_reference_studi": 0, "merged_fil": 0, "metadata": 3, "method": [0, 2, 3, 4, 5], "metric": 3, "min_dist": 3, "mind": [2, 3], "minor": [0, 2, 3, 4], "mismatch": 0, "miss": [2, 3, 4], "missing": [2, 4], "missing_data_femal": 4, "missing_data_mal": 4, "modern": 6, "modifi": 0, "modul": [2, 3, 4, 6], "moreov": 1, "mt": 5, "multi": 0, "multipl": [2, 4], "multithread": 3, "must": [0, 2, 3, 4], "mz": 2, "n_neighbor": 3, "name": [0, 1, 2, 4, 5], "neg": 3, "new": [0, 2], "non": [3, 4], "none": [0, 1, 2, 3, 4, 5], "note": [0, 3, 4], "num_pc": 0, "num_pca": 0, "number": [0, 2, 3, 4], "o": 5, "object": [0, 2, 3, 4, 5], "often": 2, "onc": [2, 4], "onli": [2, 4, 5], "oper": [0, 2, 4], "optim": [0, 2, 3], "option": [0, 2, 3, 4, 5], "order": [0, 2], "organ": 2, "origin": [0, 1], "out": 3, "outlier": 0, "output": [0, 2, 3, 4, 5], "output_dir": 0, "output_nam": [0, 2, 3, 4], "output_path": [0, 2, 3, 4], "outsid": [2, 3], "p": [3, 4], "pair": 2, "pairwis": [0, 2, 3], "panel": 0, "paramet": [0, 1, 3], "particularli": 2, "pass": 3, "path": [0, 1, 2, 3, 4, 5], "path_to_fold": 1, "pattern": 4, "pc": 0, "pc1": 0, "pc2": 0, "pc3": 0, "pca": [0, 3, 6], "pca_plot": 0, "pd": [2, 4], "ped": 2, "pedsex": 2, "per": [3, 4], "perform": [0, 2, 3, 4, 6], "permiss": 0, "pfile": 5, "pgen": 5, "pi_hat": 2, "pip": 6, "pipelin": [1, 5, 6], "plink": [0, 2, 3, 4, 5], "plink2": [2, 5], "plot": [0, 2, 3, 4, 6], "plot_dir": 0, "plot_nam": 0, "plots_dir": 2, "plots_paramet": 3, "po": 2, "point": 5, "popul": [0, 3, 4, 6], "population_tag": 0, "posit": [0, 2, 3], "posixpath": 0, "possibl": [0, 3], "potenti": [0, 2], "predefin": 5, "prefix": [0, 3, 5], "preserv": 2, "previou": [0, 2], "previous": [0, 2], "princip": [0, 3], "print_cmd": 1, "problemat": [0, 2], "process": [0, 2, 3, 4, 5], "progress": 0, "project": 3, "proper": [0, 5], "proport": 2, "provid": [0, 3], "prun": 2, "prune": [0, 2, 3], "pruned_fil": 2, "pruned_refer": 0, "pruned_studi": 0, "psam": 5, "pvar": 5, "pypi": 6, "python": [4, 6], "qc": [2, 4], "qualiti": [0, 2, 3, 4, 6], "r2": 0, "rais": [0, 3, 4], "random": 3, "random_st": 3, "rang": [2, 3], "rate": [2, 3, 4], "read": [0, 2, 3, 4], "reason": [2, 4], "reassign": 0, "recommend": [2, 3], "recompute_merg": 0, "recompute_pca": 3, "red": 2, "reduct": 3, "ref_popul": 0, "ref_threshold": 0, "refer": [0, 5], "reference_ac_gt_filt": 0, "reference_bim_stem": 0, "reference_fil": 0, "reference_fixed_po": 0, "reference_flip": 0, "reference_pop": 0, "reference_tag": 0, "referencegenomicmerg": [0, 6], "region": [0, 2, 5], "regions_grch": 5, "rel": 2, "relat": [2, 5], "related": 2, "remain": [0, 4], "remov": [0, 2, 3, 4], "renam": [0, 2, 5], "rename_snp": 0, "renamed_snp": [0, 2], "report": [2, 4, 6], "report_call_r": 2, "report_different_genotype_call_r": 4, "report_heterozygosity_r": 2, "report_hw": 4, "report_ibd_analysi": 2, "report_missing_data": 4, "report_sex_check": 2, "repositori": [5, 6], "reproduc": [3, 6], "requir": [2, 5], "reserv": 0, "resourc": [0, 2], "result": [0, 2, 3, 4, 6], "results_dir": 2, "retriev": 5, "return": 3, "return_log": 1, "row": 2, "run": [0, 2, 4], "run_pca": 0, "r\u00b2": [0, 2, 3], "same": 5, "sampl": [0, 2, 3, 6], "sampleqc": 6, "save": [0, 1, 2, 3, 4], "scatter": [0, 2], "scatterplot": 2, "scenario": 0, "score": 2, "second": 2, "seed": 3, "self": [0, 2, 4], "separ": [0, 2, 3, 4], "seri": [0, 2], "set": [0, 2, 4], "sex": [2, 4], "sex_check": 2, "sex_check_filenam": 2, "sexcheck": 2, "shell": 0, "shell_do": [1, 6], "shift": [0, 2], "should": [0, 1, 2, 3], "show": [0, 2, 4], "significantli": 4, "size": [0, 2, 3], "skip": [0, 2, 3, 5], "slide": 0, "snp": [0, 2, 3, 4, 5], "some": 6, "space": 0, "specif": [2, 4], "specifi": [0, 2, 3, 4, 5], "split": 2, "standard": [2, 5], "statist": [0, 2], "std_deviation_het": 2, "step": [0, 2, 3, 4, 5], "store": [0, 2, 4], "str": [0, 1, 2, 3, 4, 5], "string": [0, 1], "structur": [0, 3, 6], "stu_threshold": 0, "studi": 0, "study_ac_gt_filt": 0, "subject": 4, "suffix": [0, 2, 3, 4], "sum": 2, "summari": [2, 4], "summary_ped_filenam": 2, "super": 0, "superpop": 0, "support": 6, "system": [0, 2, 3, 4, 5], "t": [0, 2, 5], "tab": [0, 3, 4], "tag": 0, "temporari": [1, 2], "term": 2, "test": [2, 3, 4], "thei": [2, 3], "them": [2, 5], "thi": [0, 2, 3, 4, 5, 6], "third": [0, 2, 3], "those": 0, "thread": [0, 2, 3], "three": [0, 2, 4], "threshold": [0, 2, 3, 4], "tidi": 2, "time": 2, "toflip": 0, "tool": [0, 6], "toremov": 0, "total": [0, 2, 4], "tradit": 2, "trigger": 3, "true": [0, 2, 3], "tubingen": 6, "twin": 2, "two": [0, 2, 3, 4, 5], "txt": [2, 4, 5], "type": [0, 2, 3, 4, 5], "typeerror": [0, 2, 3, 4], "typic": 2, "umap": [3, 6], "umap_kwarg": 3, "umapplot": 6, "uncap": 2, "under": 4, "unexpect": 2, "uniform": 3, "uniqu": 4, "updat": [0, 2, 5], "updatechr": 0, "upper": [2, 4], "url": 5, "url_pgen": 5, "url_psam": 5, "url_pvar": 5, "us": [0, 2, 3, 4, 5, 6], "usag": 2, "use_k": 2, "userwarn": 3, "util": 0, "v": [0, 2], "valid": [0, 2, 3], "valu": [2, 3, 4], "valueerror": [0, 2, 3, 4], "var": 2, "variant": [0, 2, 3, 4, 5, 6], "variantqc": 6, "variou": 3, "version": [5, 6], "vertic": 2, "via": 0, "view": 2, "visual": [0, 2, 3, 4, 6], "warn": 3, "weinberg": [3, 4], "were": 1, "when": [0, 2], "where": [0, 1, 2, 3, 4], "whether": [2, 3, 4], "which": 2, "while": [0, 2], "wide": 2, "window": [0, 2, 3], "without": 5, "workflow": 0, "workspac": 2, "write": 0, "written": [3, 4], "x": [2, 5], "xchr": 2, "xchr_imiss_filenam": 2, "y": [2, 4, 5], "y_axis_cap": [2, 4], "y_lim_cap": 4, "you": 6}, "titles": ["ideal_genom_qc.AncestryQC", "ideal_genom_qc.Helpers", "ideal_genom_qc.SampleQC", "ideal_genom_qc.UMAPplot", "ideal_genom_qc.VariantQC", "ideal_genom_qc.get_references", "ideal-genome-qc documentation"], "titleterms": {"ancestryqc": 0, "depend": 0, "document": 6, "effect": [0, 4], "featur": 6, "genom": 6, "get_refer": 5, "helper": 1, "ideal": 6, "ideal_genom_qc": [0, 1, 2, 3, 4, 5], "instal": 6, "note": [2, 5], "paramet": [2, 4, 5], "qc": 6, "rais": [2, 5], "requir": 0, "return": [0, 2, 4, 5], "sampleqc": 2, "side": [0, 4], "umapplot": 3, "variantqc": 4}})